

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Bioinformatic analysis &#8212; eDNA workshop HKU</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'bioinformaticanalysis';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Taxonomy assignment" href="taxonomyassignment.html" />
    <link rel="prev" title="Setting up the environment" href="setup.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo3.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo3.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    <no title>
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Bioinformatic analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxonomyassignment.html">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datapreprocessing.html">Data pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogenetictree.html">Phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="statisticalanalysis.html">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fbioinformaticanalysis.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button"
   title="Open an issue"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/bioinformaticanalysis.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Bioinformatic analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preparing-data-for-processing">1. Preparing data for processing</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unzipping-data">1.1 Unzipping data</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#counting-files-pipe">1.1.1 Counting files (pipe <code class="docutils literal notranslate"><span class="pre">|</span></code>)</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fastq-and-fasta-file-structure">1.2 Fastq and Fasta file structure</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#changing-file-names">1.3 Changing file names</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-control-of-raw-data">1.4 Quality control of raw data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merging-forward-and-reverse-reads">2. Merging forward and reverse reads</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-single-sample-example">2.1 A single sample example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-merging-reads">2.2 Batch merging reads</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarizing-the-output">2.3 Summarizing the output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-primer-sequences">3. Removing primer sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">3.1 A single sample example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-trimming">3.2 batch trimming</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarizing-cutadapt-output">3.3 Summarizing cutadapt output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-filtering">4. Quality filtering</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-processing">4.1 Batch processing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-output">4.2 Summary output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#check-filtering-step">4.3 Check filtering step</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dereplication">5. Dereplication</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#combining-data-into-one-file">5.1 Combining data into one file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#finding-unique-sequences">5.2 Finding unique sequences</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#denoising">6. Denoising</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chimera-removal">7. Chimera removal</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#frequency-table">8. Frequency table</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarise-sequence-counts">8.1 Summarise sequence counts</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="bioinformatic-analysis">
<h1>Bioinformatic analysis<a class="headerlink" href="#bioinformatic-analysis" title="Permalink to this heading">#</a></h1>
<section id="preparing-data-for-processing">
<h2>1. Preparing data for processing<a class="headerlink" href="#preparing-data-for-processing" title="Permalink to this heading">#</a></h2>
<section id="unzipping-data">
<h3>1.1 Unzipping data<a class="headerlink" href="#unzipping-data" title="Permalink to this heading">#</a></h3>
<p>The sequencing data for this project came as a single zipped file (<strong>HKeDNAworkshop2023.zip</strong>) which we moved to the <strong>sequenceData/2-raw</strong> subdirectory. To check if our sequence data is in the appropriate subfolder, let us use the <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-ltr</span></code> command again.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-ltr<span class="w"> </span>sequenceData/2-raw
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span> <span class="mi">568074346</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">HKeDNAworkshop2023</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The sequencing data has the extension <code class="docutils literal notranslate"><span class="pre">.zip</span></code>. This is not the same as the extension <code class="docutils literal notranslate"><span class="pre">.gz</span></code>, which is a g-zipped file structure frequently provided by the sequencing service during file transfer. The following set of commands that deal with the <code class="docutils literal notranslate"><span class="pre">.zip</span></code> extension are specific for this data type, a result due to the alteration of files after subsetting from the original data. It is unlikely that these steps will be necessary for your project data if your data was received from a sequencing service.</p>
</div>
<p>Since the data comes zipped, we need to use the <code class="docutils literal notranslate"><span class="pre">unzip</span></code> command to get access to the sequencing files. We can use the <code class="docutils literal notranslate"><span class="pre">d</span></code> parameter to tell <code class="docutils literal notranslate"><span class="pre">unzip</span></code> where we would like to place the unzipped files. In our tutorial, we will place the files in a subdirectory of <strong>sequenceData/2-raw/</strong> called <strong>sequenceData/2-raw/unzipped/</strong>. Note that we do not need to create this subdirectory first. The <code class="docutils literal notranslate"><span class="pre">unzip</span></code> command will generate this subdirectory automatically if it not yet exists.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>unzip<span class="w"> </span>sequenceData/2-raw/HKeDNAworkshop2023.zip<span class="w"> </span>-d<span class="w"> </span>sequenceData/2-raw/unzipped/
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Archive</span><span class="p">:</span>  <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">HKeDNAworkshop2023</span><span class="o">.</span><span class="n">zip</span>
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK37_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK37_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK38_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK38_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK39_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK39_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK40_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK40_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK41_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK41_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK42_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK42_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK49_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK49_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK50_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK50_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK51_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK51_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK52_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK52_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK53_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK53_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK54_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_100_HK54_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK37_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK37_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK38_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK38_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK39_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK39_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK40_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK40_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK41_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK41_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK42_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK42_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK49_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK49_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK50_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK50_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK51_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK51_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK52_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK52_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK53_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK53_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK54_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">COI_500_HK54_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">NTC_1_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">NTC_1_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">NTC_2_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">NTC_2_2</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">NTC_3_1</span><span class="o">.</span><span class="n">fastq</span>  
  <span class="n">inflating</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">unzipped</span><span class="o">/</span><span class="n">NTC_3_2</span><span class="o">.</span><span class="n">fastq</span>  
</pre></div>
</div>
</div>
<p>Unzipping the <strong>HKeDNAworkshop2023.zip</strong> file has generated a bunch of files for all of the samples in our experiment, each represented with a forward (<strong>_1.fastq</strong>) and reverse (<strong>_2.fastq</strong>) fastq file.</p>
<section id="counting-files-pipe">
<h4>1.1.1 Counting files (pipe <code class="docutils literal notranslate"><span class="pre">|</span></code>)<a class="headerlink" href="#counting-files-pipe" title="Permalink to this heading">#</a></h4>
<p>Now that we are getting more familiar with the Terminal and bash commands, let’s introduce something that is called a <code class="docutils literal notranslate"><span class="pre">pipe</span></code>, which is represented by the <code class="docutils literal notranslate"><span class="pre">|</span></code> symbol. A pipe or placing the <code class="docutils literal notranslate"><span class="pre">|</span></code> in your terminal allows you to use the output from one command as the input of a second command. Hence, the terminology “pipe”, as <code class="docutils literal notranslate"><span class="pre">|</span></code> acts as a connection between two different commands. To show you the power of piping commands together, <strong>let’s look at the exercise of how you might count the number of samples you have in our data set.</strong></p>
<p><strong>STEP 1:</strong> To solve the issue of counting the number of samples, we can use a step-by-step approach. First, we can use a command you already know, the <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-1</span></code> command to list all the files in the directory. The <code class="docutils literal notranslate"><span class="pre">-1</span></code> parameter will print the files into a single column. Let’s see the output of this command first before we continue.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-1<span class="w"> </span>sequenceData/2-raw/unzipped/
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">COI_100_HK37_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK37_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK38_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK38_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK39_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK39_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK40_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK40_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK41_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK41_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK42_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK42_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK49_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK49_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK50_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK50_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK51_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK51_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK52_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK52_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK53_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK53_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK54_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_100_HK54_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK37_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK37_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK38_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK38_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK39_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK39_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK40_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK40_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK41_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK41_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK42_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK42_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK49_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK49_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK50_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK50_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK51_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK51_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK52_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK52_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK53_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK53_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK54_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">COI_500_HK54_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">NTC_1_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">NTC_1_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">NTC_2_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">NTC_2_2</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">NTC_3_1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">NTC_3_2</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
</div>
<p><strong>STEP 2:</strong> Once we have a list, we can use a count function such as <code class="docutils literal notranslate"><span class="pre">wc</span></code> which stands for <em>word count</em>. However, we want to count lines, not words, so we need to add the <code class="docutils literal notranslate"><span class="pre">-l</span></code> parameter to specify this. When we pipe these two commands together, the Terminal window will output the number of files in our folder. Let’s try!</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-1<span class="w"> </span>sequenceData/2-raw/unzipped/<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>      <span class="mi">54</span>
</pre></div>
</div>
</div>
<p><strong>STEP 3:</strong> Remember that we wanted to know the number of samples for which we have data files and that each sample was comprised of one forward (<strong>_1.fastq</strong>) and one reverse (<strong>_2.fastq</strong>) file. So, for the last step, we need to divide the number of files by 2. The easiest way to accomplis this step is to pipe the number of files in our directory to an <strong>AWK</strong> command (a separate language in the Terminal) that takes the number and divides it by 2. For this last command, we specify that we would like to <code class="docutils literal notranslate"><span class="pre">print</span></code> the first item (<code class="docutils literal notranslate"><span class="pre">$1</span></code>), which in our case is the number of files, and divide it by two (<code class="docutils literal notranslate"><span class="pre">/2</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-1<span class="w"> </span>sequenceData/2-raw/unzipped<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{print $1/2}&#39;</span><span class="w">  </span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">27</span>
</pre></div>
</div>
</div>
<p>By using one line of code and two pipes, we have shown that we have 27 samples in our data set. In our tutorial data set, these 27 samples are comprised of 3 negative controls, and 2 size fractions * 3 replicates * 4 sites.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><strong>AWK</strong>, <strong>sed</strong>, and <strong>grep</strong> are very powerful computer languages to process text. Throughout this tutorial, we will be using snippets of code from these three computer languages. Unfortunately, the syntax of these languages is quite complex in my opinion. I recommend you to investigate these languages on your own time if bioinformatics is of interest to you. However, to keep this tutorial beginner friendly, we will not expand on the syntax further.</p>
</div>
</section>
</section>
<section id="fastq-and-fasta-file-structure">
<h3>1.2 Fastq and Fasta file structure<a class="headerlink" href="#fastq-and-fasta-file-structure" title="Permalink to this heading">#</a></h3>
<p>When unzipping the <strong>HKeDNAworkshop2023.zip</strong> file, we have revealed our sequence data files. Those sequence data files have the extension <strong>.fastq</strong>. Sequence data is most frequently represented as either <strong>.fastq</strong> or <strong>.fasta</strong> files. Both are simple text files that are structured in a particular way.</p>
<p>For example, within <strong>.fastq</strong> files each sequence record is represented by 4 lines. The first line contains the header information and the line starts with the symbol <code class="docutils literal notranslate"><span class="pre">&#64;</span></code>. The second line contains the actual sequence. The third line can contain metadata, but usually is restricted to the <code class="docutils literal notranslate"><span class="pre">+</span></code> symbol. The fourth line provides the quality of the base call and should, therefore, have the same length as the second line containing the sequence. We can inspect the file structure of one of our sequence files we just unzipped using the <code class="docutils literal notranslate"><span class="pre">head</span></code> or <code class="docutils literal notranslate"><span class="pre">tail</span></code> command. Both commands print the first or last N number of lines in a document, respectively. By using the <code class="docutils literal notranslate"><span class="pre">-n</span></code> parameter, we can specify the number of lines to be printed.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">4</span><span class="w"> </span>sequenceData/2-raw/unzipped/COI_100_HK37_1.fastq
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>@M04617:136:000000000-J6LL9:1:1101:14901:1936 1:N:0:31
GGTACTGGATGAACAGTATATCCCCCCCTAAGCTCCAATATTGCCCACGCCGGGGCGTCTGTTGACCTTGCTATCTTTAGGCTACACTTGGCTGGGGTTTCTTCTCTACTCGGGGCTGTAAACTTTATTAGAACTATCGCTAACCTGCGAGCTTTAGGGCTAATTCTTGACCGTATAACACTATTCACATGATCAGTTCTTATCACCGCCATCCTTCTCCTTCTTTCTCTACCTGTTCTCGCAGGGGCTAT
+
3AA@AFFFCFFFBG5GGBGBFGHCFE?AEA233FFE33D3FGBFFGBA0AEE???E?EEEAGHFB@BGGBEGDFCGGG43F3GBGBFFGFBGFECCCFDGDGHFHFHB&gt;GFA///AD?G1&lt;1&lt;FGHBGH1DB1&gt;FBGD..C../&lt;GCF---;AC0000.:C:0CBBFFFBBB-C.C/0B0:F/FFB099C9BCF00;C0FF/BB/B..---;/.BFBFBBFFFFFFFFFF/BFFBFFFFF.9-.---9@.B
</pre></div>
</div>
</div>
<p><strong>.fasta</strong> files, on the other hand, are simple text files structured in a slightly different way, whereby each sequence record is represented by 2 lines. The first line contains the header information and the line starts with the symbol <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>. The second line contains the actual sequence. The third and fourth line that are in the .fastq files are missing in the .fasta files, as this file structure does not incorporate the quality of a sequence. We can inspect the file structure of a .fasta file by using the <code class="docutils literal notranslate"><span class="pre">head</span></code> or <code class="docutils literal notranslate"><span class="pre">tail</span></code> command on the reference database.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tail<span class="w"> </span>-n<span class="w"> </span><span class="m">2</span><span class="w"> </span>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">AJ488637</span><span class="p">;</span><span class="n">tax</span><span class="o">=</span><span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Mollusca</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Gastropoda</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Littorinimorpha</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Littorinidae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Littoraria</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">Littoraria_scabra</span>
<span class="n">TCTTGCAGGCAACCTGGCTCACGCCGGGGGCTCTGTAGATCTAGCAATTTTTTCACTCCATCTAGCCGGTGTGTCTTCTATTTTAGGGGCTGTAAATTTCATTACAACCATCATTAATATGCGATGACGAGGTATGCAGTTTGAACGTCTACCTCTCTTTGTTTGATCAGTAAAGATTACAGCTATTCTTCTTCTTTTATCTCTCCCAGTTTTAGCTGGTGCAATTACCATACTCTTAACGGATCGAAACTTCAATACTGCCTTCTTTGACCCTGCCGGAGGAGGAGATCC</span>
</pre></div>
</div>
</div>
<div class="hint admonition">
<p class="admonition-title">Exercise 2</p>
<p>Knowing the file structure of the .fastq and .fasta sequence files. How would you calculate the number of sequences incorporated in the files <strong>sequenceData/2-raw/unzipped/COI_100_HK37_1.fastq</strong> and <strong>sequenceData/7-refdb/leray_COI_sintax.fasta</strong>?</p>
<div class="title dropdown admonition">
<p class="admonition-title">Answer 2</p>
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-0">
bash Option 1</label><div class="sd-tab-content docutils">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;^@M04617&quot;</span><span class="w"> </span>sequenceData/2-raw/unzipped/COI_100_HK37_1.fastq
grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;^&gt;&quot;</span><span class="w"> </span>sequenceData/7-refdb/leray_COI_sintax.fasta
</pre></div>
</div>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-1">
bash Option 2</label><div class="sd-tab-content docutils">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wc<span class="w"> </span>-l<span class="w"> </span>sequenceData/2-raw/unzipped/COI_100_HK37_1.fastq
wc<span class="w"> </span>-l<span class="w"> </span>sequenceData/7-refdb/leray_COI_sintax.fasta
</pre></div>
</div>
</div>
<input id="sd-tab-item-2" name="sd-tab-set-0" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-2">
python</label><div class="sd-tab-content docutils">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;sequenceData/2-raw/unzipped/COI_100_HK37_1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
  <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">infile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span>
  <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Total lines:&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="changing-file-names">
<h3>1.3 Changing file names<a class="headerlink" href="#changing-file-names" title="Permalink to this heading">#</a></h3>
<p>When we inspect the file names of our .fastq files by using the <code class="docutils literal notranslate"><span class="pre">ls</span></code> command, we can see that they contain the <code class="docutils literal notranslate"><span class="pre">_</span></code> symbols.</p>
<figure class="align-default" id="list-of-file-names">
<img alt="_images/filenamelist3.png" src="_images/filenamelist3.png" />
<figcaption>
<p><span class="caption-number">Fig. 4 </span><span class="caption-text">: A list of the fastq file names</span><a class="headerlink" href="#list-of-file-names" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>While <code class="docutils literal notranslate"><span class="pre">_</span></code> symbols will not cause downstream incompatability issues, some symbols in file names can cause problems downstream with certain programs. One frequently observed symbol that affects one of the programs we will use during our bioinformatic pipeline is the dash <code class="docutils literal notranslate"><span class="pre">-</span></code> symbol. While not pertinent for our tutorial file names, it is important to know how to remove or change such symbols in file names in the Terminal without having do to this manually. As an exercise, let’s replace all the underscores <code class="docutils literal notranslate"><span class="pre">_</span></code> to dashes <code class="docutils literal notranslate"><span class="pre">-</span></code> and back to show you how the code would work if you need it in the future.</p>
<p>Luckily there is a simple perl one-liner to batch replace characters in file names using the <code class="docutils literal notranslate"><span class="pre">rename</span></code> command. We can tell <code class="docutils literal notranslate"><span class="pre">rename</span></code> that we want to <em>substitute</em> a pattern by providing the <code class="docutils literal notranslate"><span class="pre">s</span></code> parameter. The <code class="docutils literal notranslate"><span class="pre">/_/-/</span></code> parameter tells <code class="docutils literal notranslate"><span class="pre">rename</span></code> the pattern we want to substitute, i.e., replace <code class="docutils literal notranslate"><span class="pre">_</span></code> with <code class="docutils literal notranslate"><span class="pre">-</span></code>. Finally, we tell <code class="docutils literal notranslate"><span class="pre">rename</span></code> that we want to replace all instances using the <code class="docutils literal notranslate"><span class="pre">g</span></code> or <em>global</em> parameter. Without this parameter, <code class="docutils literal notranslate"><span class="pre">rename</span></code> would only substitute the first instance. The <code class="docutils literal notranslate"><span class="pre">*</span></code> at the end of the code tells the computer to go over all the files in the <strong>unzipped/</strong> directory.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>sequenceData/2-raw/
rename<span class="w"> </span><span class="s1">&#39;s/_/-/g&#39;</span><span class="w"> </span>unzipped/*
</pre></div>
</div>
<figure class="align-default" id="list-4-of-file-names">
<img alt="_images/filenamelist4.png" src="_images/filenamelist4.png" />
<figcaption>
<p><span class="caption-number">Fig. 5 </span><span class="caption-text">: A list of fastq file names whereby the underscores are replaced by dashes</span><a class="headerlink" href="#list-4-of-file-names" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>rename<span class="w"> </span><span class="s1">&#39;s/-/_/g&#39;</span><span class="w"> </span>unzipped/*
<span class="nb">cd</span><span class="w"> </span>../../
</pre></div>
</div>
<figure class="align-default" id="list-5-of-file-names">
<img alt="_images/filenamelist5.png" src="_images/filenamelist5.png" />
<figcaption>
<p><span class="caption-number">Fig. 6 </span><span class="caption-text">: A list of fastq file names whereby the dashes are replaced back by underscores</span><a class="headerlink" href="#list-5-of-file-names" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="quality-control-of-raw-data">
<h3>1.4 Quality control of raw data<a class="headerlink" href="#quality-control-of-raw-data" title="Permalink to this heading">#</a></h3>
<p>Up to this point, we have unzipped our sequence data, checked the file structure, and batch renamed our files to exclude problematic symbols in the file names. Before processing the contents of each file, it is best to check the quality of the raw sequence data to ensure no problems were faced during sequencing. The programs we will be using to check the quality of all of our files are <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a> and <a class="reference external" href="https://multiqc.info">MULTIQC</a>. First, we will generate reports for all of the files using FastQC. We can specify the output folder using the <code class="docutils literal notranslate"><span class="pre">-o</span></code> parameter and the number of threads or cores using the <code class="docutils literal notranslate"><span class="pre">-t</span></code> parameter.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><strong>Make sure to change the value of the <code class="docutils literal notranslate"><span class="pre">-t</span></code> parameter to the number of cores available on your system!</strong></p>
</div>
<p>To run <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a> on the supercomputer, we can modify the template script <strong>sequenceData/1-scripts/example-script.sh</strong>. We can open text documents with CLI text editors, such as <strong>nano</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<p>When opening the script, we can add the following line of code to execute the <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>fastqc<span class="w"> </span>sequenceData/2-raw/unzipped/*<span class="w"> </span>-o<span class="w"> </span>sequenceData/3-fastqc/<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>We can exit out of the editor by pressing <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code>. We will change the name of the script to <strong>fastqc_raw.sh</strong> and press <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name. After completing these steps, you should end back up in the normal Terminal window. To check if we created this new script, let’s use the <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-ltr</span></code> command again.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-ltr<span class="w"> </span>sequenceData/1-scripts
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span> <span class="mi">776</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">example</span><span class="o">-</span><span class="n">script</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span> <span class="mi">169</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">module_load</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span> <span class="mi">833</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">11</span><span class="p">:</span><span class="mi">33</span> <span class="n">fastqc_raw</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
</div>
<p>To run the command on the supercomputer, we can use the <code class="docutils literal notranslate"><span class="pre">sbatch</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/fastqc_raw.sh
</pre></div>
</div>
<p>Once this command is executed, it will be placed in the queue to be run on the cluster. We can use the <code class="docutils literal notranslate"><span class="pre">squeue</span></code> command to check if our code is running or if the job is finished. It is best to specify your jobs specifically using the <code class="docutils literal notranslate"><span class="pre">-u</span></code> or user parameter, followed by your user name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>squeue<span class="w"> </span>-u<span class="w"> </span>ednaw01
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><strong>Make sure to change the <code class="docutils literal notranslate"><span class="pre">-u</span></code> parameter to your user name!</strong></p>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>             <span class="n">JOBID</span> <span class="n">PARTITION</span>     <span class="n">NAME</span>     <span class="n">USER</span> <span class="n">ST</span>       <span class="n">TIME</span>  <span class="n">NODES</span> <span class="n">NODELIST</span><span class="p">(</span><span class="n">REASON</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Once finished, FastQC will generate a .html report for every single file in the subdirectory <strong>sequenceData/2-raw/unzipped/</strong>. Let’s open one report to see what it contains. The .html reports are stored in the subdirectory <strong>sequenceData/3-fastqc/</strong>. From the FastQC .html reports, we are particularly interested in the <strong>summary tab</strong>, as well as the <strong>per base sequence quality</strong> and <strong>sequence length distribution figures</strong>.</p>
<p>Since we are working on the supercomputer, we need to download the .html report to our computer before we can open it. To do this, let’s open a new Terminal window that is not connected to the supercomputer, move to the <strong>Desktop</strong> directory using the <code class="docutils literal notranslate"><span class="pre">cd</span></code> command and download the file using the <code class="docutils literal notranslate"><span class="pre">scp</span></code> command, which stands for secure copy. While not essential in this case, you can add the <code class="docutils literal notranslate"><span class="pre">-r</span></code> parameter to allow the download of directories. We also need to specify the server location of the file to download and where we would like to download the file to on our system.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>Desktop
scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/sequenceData/3-fastqc/COI_100_HK37_1_fastqc.html<span class="w"> </span>./
</pre></div>
</div>
<figure class="align-default" id="fastqc-raw-report">
<img alt="_images/fastqcraw.png" src="_images/fastqcraw.png" />
<figcaption>
<p><span class="caption-number">Fig. 7 </span><span class="caption-text">: A FastQC report of the raw sequencing data.</span><a class="headerlink" href="#fastqc-raw-report" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The quality of file <strong>COI_100_HK37_1.fastq</strong> is looking very good and the number of sequences reported for this sample (160,732 reads) is a good sequencing depth as well!</p>
<p>Opening each report for the 54 files separately, however, is quite tedious and makes it difficult to compare differences between files. Luckily, multiQC was developed to collate these reports into a single .html report. We can use the <code class="docutils literal notranslate"><span class="pre">.</span></code> symbol in combination with the path to specify a minimal output to the Terminal window. The <code class="docutils literal notranslate"><span class="pre">-o</span></code> parameter let’s us set the output directory.</p>
<p>Again, to run this on the supercomputer, we need to modify the initial template script and add in the <code class="docutils literal notranslate"><span class="pre">multiqc</span></code> code.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>multiqc<span class="w"> </span>sequenceData/3-fastqc/.<span class="w"> </span>-o<span class="w"> </span>sequenceData/3-fastqc/
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>multiqc_raw.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/multiqc_raw.sh
</pre></div>
</div>
<figure class="align-default" id="multiqc-terminal-output">
<img alt="_images/multiqcterminal2.png" src="_images/multiqcterminal2.png" />
<figcaption>
<p><span class="caption-number">Fig. 8 </span><span class="caption-text">: Terminal output for multiQC</span><a class="headerlink" href="#multiqc-terminal-output" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The multiQC program will combine all 54 FastQC reports into a single .html document. Let’s download and open this file to see how our raw sequence data is looking like.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/sequenceData/3-fastqc/multiqc_report.html<span class="w"> </span>./
</pre></div>
</div>
<figure class="align-default" id="multiqc-report">
<img alt="_images/multiqcreport2.png" src="_images/multiqcreport2.png" />
<figcaption>
<p><span class="caption-number">Fig. 9 </span><span class="caption-text">: The .html multiQC report</span><a class="headerlink" href="#multiqc-report" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The sequence quality, average read length, and number of sequences are all very comparable between the different files representing the samples. This is an excellent starting point! Interestingly, note the difference in length, read depth, and quality with the control samples. As those are our negative controls, these differences aren’t worrysome. If, on the other hand, some samples were looking like this, you might want to consider resequencing those samples to achieve a similar sequencing depth.</p>
</section>
</section>
<section id="merging-forward-and-reverse-reads">
<h2>2. Merging forward and reverse reads<a class="headerlink" href="#merging-forward-and-reverse-reads" title="Permalink to this heading">#</a></h2>
<section id="a-single-sample-example">
<h3>2.1 A single sample example<a class="headerlink" href="#a-single-sample-example" title="Permalink to this heading">#</a></h3>
<p>Once we have assessed the quality of the raw sequence data and did not observe any issues, we can move ahead with merging the forward and reverse reads of each sample. As a reminder, we have 54 .fastq files, indicating that we have 27 samples in our experiment, one forward and one reverse file for each sample.</p>
<p>For this experiment and particular library, we have amplified a ~313 bp fragment of the COI gene (excluding primer-binding regions). With the sequencing run specifications of a MiSeq 2x250 paired-end V2 sequencing kit, we have a partial overlap between the forward and reverse reads in the middle of the amplicon region. This overlap is essential for successful merging. Today, we will be using the <code class="docutils literal notranslate"><span class="pre">--fastq_mergepairs</span></code> command in <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a> to merge reads between the forward and reverse .fastq sequence files. We can specify the reverse sequence file through the <code class="docutils literal notranslate"><span class="pre">--reverse</span></code> parameter and the output file through the <code class="docutils literal notranslate"><span class="pre">--fastqout</span></code> parameter. To check the options available within the program, you can always read the documentation and pull up the help vignette by executing <code class="docutils literal notranslate"><span class="pre">vsearch</span> <span class="pre">--help</span></code> in the Terminal window. Before merging reads immediately on all samples, it is best to play around with the parameters on a single sample.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--fastq_mergepairs<span class="w"> </span>sequenceData/2-raw/unzipped/COI_100_HK37_1.fastq<span class="w"> </span>--reverse<span class="w"> </span>sequenceData/2-raw/unzipped/COI_100_HK37_2.fastq<span class="w"> </span>--fastqout<span class="w"> </span>sequenceData/2-raw/COI_100_HK37merged.fastq
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>merge_single.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/merge_single.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">160732</span>  <span class="n">Pairs</span>
    <span class="mi">146040</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">90.9</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">14692</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">9.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">144</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">5751</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">8584</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
       <span class="mi">213</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">249.48</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">359.43</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">27.01</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.32</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.53</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.22</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.56</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.77</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>
</pre></div>
</div>
</div>
<p>Using the default settings, we managed to merge 90.9% of reads. For this tutorial, we’ll take that as sufficient. However, for your own project, you might want to explore the various options within the <code class="docutils literal notranslate"><span class="pre">fastq_mergepairs</span></code> command further to try and increase the percentage of reads merged. From the <code class="docutils literal notranslate"><span class="pre">vsearch</span> <span class="pre">--fastq_mergepairs</span></code> output that was printed to the Terminal window, we can see that VSEARCH reports the number of observed pairs in the raw data files, the number of reads successfully merged, and the number of reads unable to be merged. Additionally, information on why reads couldn’t be merged and statistics about merged reads are provided.</p>
</section>
<section id="batch-merging-reads">
<h3>2.2 Batch merging reads<a class="headerlink" href="#batch-merging-reads" title="Permalink to this heading">#</a></h3>
<p>Since we need to merge sequences for 27 samples, we will introduce a new coding concept, the <strong>for loop</strong>. A for loop allows us to move through a list and execute a command on each item. While we will not go into too much detail for this tutorial, the basic syntax of for loops is as follows:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="k">for</span><span class="w"> </span>file<span class="w"> </span><span class="k">in</span><span class="w"> </span>list
<span class="k">do</span>
<span class="w">  </span>execute<span class="w"> </span><span class="nb">command</span>
<span class="k">done</span>
</pre></div>
</div>
<p>Basically, the first line tells us that the program will go through a list and <code class="docutils literal notranslate"><span class="pre">for</span></code> each item in the list, <code class="docutils literal notranslate"><span class="pre">do</span></code> something (line 2), which in this case is <code class="docutils literal notranslate"><span class="pre">execute</span> <span class="pre">command</span></code> (line 3). Once the program has executed the command in all items within the list, we need to close the loop by specifying <code class="docutils literal notranslate"><span class="pre">done</span></code> (line 4). So, with a for loop, we can tell the computer to merge reads for our 27 samples automatically, saving us the time to have to execute the code to merge reads 27 times by ourselves.</p>
<p>Within our for loop to merge reads for all samples, we will generate a list of all forward sequencing files (<code class="docutils literal notranslate"><span class="pre">*_1.fastq</span></code>). Before writing the code to merge reads in our for loop, we will first print which sample is being merged, as the VSEARCH output does not specify this (<code class="docutils literal notranslate"><span class="pre">echo</span> <span class="pre">&quot;Merging</span> <span class="pre">reads</span> <span class="pre">for:</span> <span class="pre">${R1/_1.fastq/}&quot;</span></code>). To print to the Terminal, we can use the <code class="docutils literal notranslate"><span class="pre">echo</span></code> command. With the last parameter for the <code class="docutils literal notranslate"><span class="pre">echo</span></code> command (<code class="docutils literal notranslate"><span class="pre">${R1/_1.fastq/}&quot;</span></code>), we specify that from our forward sequencing file name <code class="docutils literal notranslate"><span class="pre">$R1</span></code>, substitute <code class="docutils literal notranslate"><span class="pre">_1.fastq</span></code> with nothing. After the <code class="docutils literal notranslate"><span class="pre">echo</span></code> command, we execute the <code class="docutils literal notranslate"><span class="pre">vsearch</span> <span class="pre">--fastq_mergepairs</span></code> command as we did above, but we need to use the for loop syntax to tell the program what our forward, reverse, and output file names are. Remember that those names change when the program iterates over the list, so we cannot hard code them into the command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>rm<span class="w"> </span>sequenceData/2-raw/COI_100_HK37merged.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>sequenceData/2-raw/unzipped/
<span class="k">for</span><span class="w"> </span>R1<span class="w"> </span><span class="k">in</span><span class="w"> </span>*_1.fastq
<span class="k">do</span>

<span class="w">  </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;\n\nMerging reads for: </span><span class="si">${</span><span class="nv">R1</span><span class="p">/_1.fastq/</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="w">  </span>vsearch<span class="w"> </span>--fastq_mergepairs<span class="w"> </span><span class="si">${</span><span class="nv">R1</span><span class="si">}</span><span class="w"> </span>--reverse<span class="w"> </span><span class="si">${</span><span class="nv">R1</span><span class="p">/_1.fastq/_2.fastq</span><span class="si">}</span><span class="w"> </span>--fastqout<span class="w"> </span>../<span class="si">${</span><span class="nv">R1</span><span class="p">/1.fastq/merged.fastq</span><span class="si">}</span>

<span class="k">done</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>merge_batch.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/merge_batch.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK37</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">160732</span>  <span class="n">Pairs</span>
    <span class="mi">146040</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">90.9</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">14692</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">9.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">144</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">5751</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">8584</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
       <span class="mi">213</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">249.48</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">359.43</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">27.01</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.32</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.53</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.22</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.56</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.77</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK38</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">129473</span>  <span class="n">Pairs</span>
    <span class="mi">118917</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">91.8</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">10556</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">8.2</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">116</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">4920</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">5434</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
        <span class="mi">86</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">249.92</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">361.90</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">20.07</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.30</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.54</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.25</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.21</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.58</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.79</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK39</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">181198</span>  <span class="n">Pairs</span>
    <span class="mi">162326</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">89.6</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">18872</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">10.4</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">137</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
         <span class="mi">2</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
      <span class="mi">6778</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
     <span class="mi">11862</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">1</span>  <span class="n">overlap</span> <span class="n">too</span> <span class="n">short</span>
        <span class="mi">92</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.08</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">362.18</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">18.32</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.33</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.48</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.50</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.73</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK40</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">130928</span>  <span class="n">Pairs</span>
    <span class="mi">115722</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">88.4</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">15206</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">11.6</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">144</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">4604</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
     <span class="mi">10378</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
        <span class="mi">80</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.12</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">362.11</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">19.02</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.30</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.52</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.75</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK41</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">115679</span>  <span class="n">Pairs</span>
    <span class="mi">102320</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">88.5</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">13359</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">11.5</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">108</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
         <span class="mi">2</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
      <span class="mi">4047</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">9067</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
       <span class="mi">135</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.07</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">361.47</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">20.81</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.28</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.54</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.20</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.59</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.79</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK42</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">164258</span>  <span class="n">Pairs</span>
    <span class="mi">148463</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">90.4</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">15795</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">9.6</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">112</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
         <span class="mi">1</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
      <span class="mi">5409</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
     <span class="mi">10244</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
        <span class="mi">29</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.39</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">363.83</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">11.18</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.29</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.48</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.21</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.51</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.72</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK49</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span>  
     <span class="mi">97006</span>  <span class="n">Pairs</span>
     <span class="mi">91171</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">94.0</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">5835</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">6.0</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">51</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2739</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3039</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">6</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.31</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.45</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">10.51</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.18</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.54</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.72</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK50</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">85963</span>  <span class="n">Pairs</span>
     <span class="mi">80114</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">93.2</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">5849</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">6.8</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">48</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2669</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3127</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">5</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.46</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.77</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">6.27</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.50</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.18</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.54</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.72</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK51</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">89943</span>  <span class="n">Pairs</span>
     <span class="mi">83923</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">93.3</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">6020</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">6.7</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">41</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2547</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3423</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">9</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.36</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.57</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">9.74</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.26</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.52</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.25</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.19</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.57</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.76</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK52</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">80397</span>  <span class="n">Pairs</span>
     <span class="mi">74255</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">92.4</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">6142</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">7.6</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">31</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2511</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3597</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">3</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.38</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.59</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">8.22</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.27</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.44</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.22</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.46</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.69</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK53</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">82962</span>  <span class="n">Pairs</span>
     <span class="mi">76890</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">92.7</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">6072</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">7.3</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">34</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2217</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3815</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">6</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.37</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.52</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">8.68</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.26</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.43</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.22</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.19</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.46</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.65</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK54</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">81013</span>  <span class="n">Pairs</span>
     <span class="mi">75723</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">93.5</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">5290</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">6.5</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">38</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2445</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">2802</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">5</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.41</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.70</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">7.32</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.46</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.18</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.50</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.68</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK37</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">120585</span>  <span class="n">Pairs</span>
    <span class="mi">109424</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">90.7</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">11161</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">9.3</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">111</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">4412</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">6635</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">3</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.44</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.34</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">4.63</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.33</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.47</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.25</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.25</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.74</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK38</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">93778</span>  <span class="n">Pairs</span>
     <span class="mi">86827</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">92.6</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">6951</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">7.4</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">88</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2664</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">4191</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">8</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.52</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.40</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">4.26</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.30</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.48</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.21</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.71</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK39</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">146550</span>  <span class="n">Pairs</span>
    <span class="mi">134723</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">91.9</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">11827</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">8.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">156</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
         <span class="mi">1</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
      <span class="mi">4664</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">6996</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
        <span class="mi">10</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.45</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.09</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">8.67</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.32</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.45</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.47</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.71</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK40</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">119622</span>  <span class="n">Pairs</span>
    <span class="mi">109897</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">91.9</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">9725</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">8.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">128</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">3864</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">5728</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">5</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.48</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.28</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">5.43</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.31</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.50</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.73</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK41</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">128094</span>  <span class="n">Pairs</span>
    <span class="mi">117560</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">91.8</span><span class="o">%</span><span class="p">)</span>
     <span class="mi">10534</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">8.2</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">179</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">4244</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">6104</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">7</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.51</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.37</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">6.25</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.27</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.48</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.19</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.54</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.72</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK42</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">84467</span>  <span class="n">Pairs</span>
     <span class="mi">77020</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">91.2</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">7447</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">8.8</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">69</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
         <span class="mi">1</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
      <span class="mi">3141</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">4233</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">3</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.61</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.63</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">6.61</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.53</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.17</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.64</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.80</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK49</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">74265</span>  <span class="n">Pairs</span>
     <span class="mi">67718</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">91.2</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">6547</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">8.8</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">81</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2989</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3473</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">4</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.47</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.75</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">6.74</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.56</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.26</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.17</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.62</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.79</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK50</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">81354</span>  <span class="n">Pairs</span>
     <span class="mi">75588</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">92.9</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">5766</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">7.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">81</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2482</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">3199</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">4</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.52</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.66</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">7.57</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.27</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.47</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.20</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.70</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK51</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">113932</span>  <span class="n">Pairs</span>
    <span class="mi">105503</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">92.6</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">8429</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">7.4</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">110</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">3809</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">4505</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">5</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.54</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.66</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">7.14</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.26</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.52</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.24</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.20</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.54</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.74</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK52</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">121246</span>  <span class="n">Pairs</span>
    <span class="mi">113930</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">94.0</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">7316</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">6.0</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">91</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2602</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">4619</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">4</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.43</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.74</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">6.49</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.21</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.44</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.20</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.15</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.46</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.61</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK53</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">108046</span>  <span class="n">Pairs</span>
    <span class="mi">100604</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">93.1</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">7442</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">6.9</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">104</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2833</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">4502</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">3</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.43</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.62</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">6.46</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.43</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.22</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.18</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.47</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.64</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK54</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
    <span class="mi">104984</span>  <span class="n">Pairs</span>
     <span class="mi">95343</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">90.8</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">9641</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">9.2</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">74</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
      <span class="mi">2427</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
      <span class="mi">7134</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">6</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">250.33</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">364.39</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">10.29</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.23</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.44</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.21</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.17</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.46</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.63</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_1</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
       <span class="mi">960</span>  <span class="n">Pairs</span>
       <span class="mi">306</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">31.9</span><span class="o">%</span><span class="p">)</span>
       <span class="mi">654</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">68.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
        <span class="mi">13</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
       <span class="mi">134</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
       <span class="mi">501</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
         <span class="mi">6</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">233.35</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
    <span class="mf">287.81</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
    <span class="mf">132.89</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.49</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">1.03</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.63</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.56</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">1.49</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">2.05</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_2</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
     <span class="mi">10408</span>  <span class="n">Pairs</span>
      <span class="mi">7842</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">75.3</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">2566</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">24.7</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
      <span class="mi">1590</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
        <span class="mi">15</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
       <span class="mi">109</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
       <span class="mi">574</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
       <span class="mi">278</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
     <span class="mf">80.31</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
     <span class="mf">64.72</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">42.73</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.09</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.11</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.05</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.18</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.25</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.43</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>


<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_3</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Merging</span> <span class="n">reads</span> <span class="mi">100</span><span class="o">%</span> 
      <span class="mi">7779</span>  <span class="n">Pairs</span>
      <span class="mi">6114</span>  <span class="n">Merged</span> <span class="p">(</span><span class="mf">78.6</span><span class="o">%</span><span class="p">)</span>
      <span class="mi">1665</span>  <span class="n">Not</span> <span class="n">merged</span> <span class="p">(</span><span class="mf">21.4</span><span class="o">%</span><span class="p">)</span>

<span class="n">Pairs</span> <span class="n">that</span> <span class="n">failed</span> <span class="n">merging</span> <span class="n">due</span> <span class="n">to</span> <span class="n">various</span> <span class="n">reasons</span><span class="p">:</span>
       <span class="mi">413</span>  <span class="n">too</span> <span class="n">few</span> <span class="n">kmers</span> <span class="n">found</span> <span class="n">on</span> <span class="n">same</span> <span class="n">diagonal</span>
        <span class="mi">10</span>  <span class="n">multiple</span> <span class="n">potential</span> <span class="n">alignments</span>
       <span class="mi">162</span>  <span class="n">too</span> <span class="n">many</span> <span class="n">differences</span>
       <span class="mi">819</span>  <span class="n">alignment</span> <span class="n">score</span> <span class="n">too</span> <span class="n">low</span><span class="p">,</span> <span class="ow">or</span> <span class="n">score</span> <span class="n">drop</span> <span class="n">too</span> <span class="n">high</span>
       <span class="mi">261</span>  <span class="n">staggered</span> <span class="n">read</span> <span class="n">pairs</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">reads</span><span class="p">:</span>
     <span class="mf">93.17</span>  <span class="n">Mean</span> <span class="n">read</span> <span class="n">length</span>

<span class="n">Statistics</span> <span class="n">of</span> <span class="n">merged</span> <span class="n">reads</span><span class="p">:</span>
     <span class="mf">72.01</span>  <span class="n">Mean</span> <span class="n">fragment</span> <span class="n">length</span>
     <span class="mf">52.72</span>  <span class="n">Standard</span> <span class="n">deviation</span> <span class="n">of</span> <span class="n">fragment</span> <span class="n">length</span>
      <span class="mf">0.12</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.14</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.09</span>  <span class="n">Mean</span> <span class="n">expected</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">sequences</span>
      <span class="mf">0.28</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">forward</span> <span class="n">sequences</span>
      <span class="mf">0.30</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span> <span class="n">of</span> <span class="n">reverse</span> <span class="n">sequences</span>
      <span class="mf">0.58</span>  <span class="n">Mean</span> <span class="n">observed</span> <span class="n">errors</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">region</span>
</pre></div>
</div>
</div>
<p>When executing this for loop, VSEARCH will attempt to merge all reads from the forward and reverse sequencing file for each sample separately. The output files containing the merged reads will be placed in the subdirectory <strong>sequenceData/2-raw/</strong> and will have the extension <strong>_merged.fastq</strong> rather than <strong>_1.fastq</strong> or <strong>_2.fastq</strong>. Within the Terminal window, all the statistics are reported for each sample separately. A quick glance shows us that we managed to merge roughly 90% of reads for each sample. However, due to the number of samples, it is difficult to get an overview of the read merging success rate and compare different samples.</p>
</section>
<section id="summarizing-the-output">
<h3>2.3 Summarizing the output<a class="headerlink" href="#summarizing-the-output" title="Permalink to this heading">#</a></h3>
<p>To get a better overview, we can use the following python script to generate a bar plot with the raw and merged read statistics. The python code takes in two arguments, the first is the location where the merged sequence files are stored, the second is the location of the raw sequence files. Since this is python code, we cannot directly copy-paste the code into the Terminal as we did before. Rather, we have to create our first script. To do this, we will use a text editor that is implemented in the Terminal called <code class="docutils literal notranslate"><span class="pre">nano</span></code>. Running <code class="docutils literal notranslate"><span class="pre">nano</span></code> will open a new text window where we can copy-paste our python code. For this tutorial, it is not important or essential to understand python, though it is an extremely powerful and useful coding language for bioinforamticians!. To provide some context, I’ll give a brief explanation during the workshop when we’re executing this code.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/rawMergedStatistics.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="c1">## import modules</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1">## user arguments</span>
<span class="n">mergedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">rawPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="c1">## first, create a sample name list</span>
<span class="n">mergedFileList</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">mergedPath</span><span class="p">)</span>
<span class="n">sampleNameList</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">mergedFile</span> <span class="ow">in</span> <span class="n">mergedFileList</span><span class="p">:</span>
    <span class="n">sampleName</span> <span class="o">=</span> <span class="n">mergedFile</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_merged.fastq&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;COI_&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;NTC_&#39;</span><span class="p">):</span>
        <span class="n">sampleNameList</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sampleName</span><span class="p">)</span>

<span class="c1">## count number of raw and merged sequences for each sample in sampleNameList</span>
<span class="n">rawSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">mergedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sampleNameList</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">mergedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_merged.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">mergedFile</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mergedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">mergedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">rawPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">rawFile</span><span class="p">:</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rawFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">rawSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>

<span class="c1">## create a dataframe from the dictionaries</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Sample&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;Raw&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Merged&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">mergedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">())})</span>

<span class="c1">## sort the dataframe by raw reads in descending order</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">## calculate the percentage of merged/raw and format it with 2 decimal places and the &#39;%&#39; symbol</span>
<span class="n">df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Merged&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;%&#39;</span>

<span class="c1">## create a horizontal bar plot using seaborn</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Adjust the figure size as needed</span>

<span class="c1">## use seaborn&#39;s barplot</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#BBC6C8&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#469597&#39;</span><span class="p">)</span>

<span class="c1">## add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Number of sequences&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Samples&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Horizontal bar graph of raw and merged reads (Sorted by Total in Reverse)&#39;</span><span class="p">)</span>

<span class="c1">## add a legend</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Add raw read count next to the bars</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">50</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

<span class="c1">## save the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;raw_and_merged_bargraph.png&#39;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
<p>Once we have copy-pasted the code, we can press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code> to save the file. After doing so, we’re back in the normal Terminal window. Before we can run or execute our first script, we need to make it executable.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Remember the permissions for each file that we discussed before? By running the <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-ltr</span></code> command, we can see that for the script we have just created, we only have read (<code class="docutils literal notranslate"><span class="pre">r</span></code>) and write (<code class="docutils literal notranslate"><span class="pre">w</span></code>) access, but no execution (<code class="docutils literal notranslate"><span class="pre">x</span></code>) permission. This is essential in this case, as we won’t be running the script using the <code class="docutils literal notranslate"><span class="pre">sbatch</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-ltr<span class="w"> </span>sequenceData/1-scripts/
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">776</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">example</span><span class="o">-</span><span class="n">script</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">169</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">module_load</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">833</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">11</span><span class="p">:</span><span class="mi">33</span> <span class="n">fastqc_raw</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">822</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">03</span> <span class="n">multiqc_raw</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">955</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">10</span> <span class="n">merge_single</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">985</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">21</span> <span class="n">merge_batch</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span> <span class="mi">2146</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">26</span> <span class="n">rawMergedStatistics</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
</div>
<p>To change the permissions or modifiers of a file, we can use the <code class="docutils literal notranslate"><span class="pre">chmod</span></code> command, which stands for <em>change modifier</em>. Since we want to make our file executable, we can specify the parameter <code class="docutils literal notranslate"><span class="pre">+x</span></code>.</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/rawMergedStatistics.py
</pre></div>
</div>
<p>Rerunning the <code class="docutils literal notranslate"><span class="pre">ls</span> <span class="pre">-ltr</span></code> command shows that we have changed the permissions and that we can execute our script.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-ltr<span class="w"> </span>sequenceData/1-scripts/
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">776</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">example</span><span class="o">-</span><span class="n">script</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">169</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">10</span><span class="p">:</span><span class="mi">59</span> <span class="n">module_load</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">833</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">11</span><span class="p">:</span><span class="mi">33</span> <span class="n">fastqc_raw</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">822</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">03</span> <span class="n">multiqc_raw</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">955</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">10</span> <span class="n">merge_single</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span>  <span class="mi">985</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">21</span> <span class="n">merge_batch</span><span class="o">.</span><span class="n">sh</span>
<span class="o">-</span><span class="n">rwxr</span><span class="o">-</span><span class="n">xr</span><span class="o">-</span><span class="n">x</span> <span class="mi">1</span> <span class="n">ednaw01</span> <span class="n">others</span> <span class="mi">2146</span> <span class="n">Oct</span> <span class="mi">21</span> <span class="mi">12</span><span class="p">:</span><span class="mi">26</span> <span class="n">rawMergedStatistics</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
</div>
<p>To execute the script, we can use the <code class="docutils literal notranslate"><span class="pre">./</span></code> command followed by the python script and our two user parameters where our (<em>parameter 1</em>) merged and (<em>parameter 2</em>) raw files are located.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/rawMergedStatistics.py<span class="w"> </span>sequenceData/2-raw/<span class="w"> </span>sequenceData/2-raw/unzipped/
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>check_merging.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/check_merging.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/raw_and_merged_bargraph.png<span class="w"> </span>./
</pre></div>
</div>
<figure class="align-default" id="raw-and-merged-read-count">
<img alt="_images/raw_and_merged_bargraph.png" src="_images/raw_and_merged_bargraph.png" />
<figcaption>
<p><span class="caption-number">Fig. 10 </span><span class="caption-text">: Raw and merged read count</span><a class="headerlink" href="#raw-and-merged-read-count" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The figure we generated using the python script shows a horizontal bar graph with the merged reads in green and the raw sequence counts in grey. We can see that the percentage of merged reads is similar across all the samples, roughly 90%. We can also see that there is some abundance differences between the samples, which is to be expected, as pooling samples equimolarly is hard to do in the lab, something we covered during the first week of this eDNA workshop. Also, note the big difference between the samples and the negative controls, something we observed earlier in the <a class="reference external" href="https://multiqc.info">MULTIQC</a> report as well.</p>
<div class="tip admonition">
<p class="admonition-title">Amplicon length exceeding cycle number</p>
<p>To merge forward and reverse reads, it is essential to have at least a partial overlap in the amplicon region with the forward and reverse reads. Even better would be a full overlap, as it increases the base call quality score. Sometimes, however, the amplicon length is too long to achieve partial overlap with paired-end reads on an Illumina platform, since the Illumina sequencing technology is restricted by sequence length. While such a scenario should be avoided at all cost, if this was a situation you would find yourself in, you can concatenate paired-end reads using the following line of code, rather than merge them.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--fastq_join<span class="w"> </span>input_R1.fastq<span class="w"> </span>--reverse<span class="w"> </span>input_R2.fastq<span class="w"> </span>--fastqout<span class="w"> </span>output_joined.fastq
</pre></div>
</div>
<p>Keep in mind 3rd generation sequencing technologies, which we discussed during the first week of this workshop, if your experiment would benefit from using longer amplicon lengths than can be covered by Illumina sequencing technology.</p>
</div>
</section>
</section>
<section id="removing-primer-sequences">
<h2>3. Removing primer sequences<a class="headerlink" href="#removing-primer-sequences" title="Permalink to this heading">#</a></h2>
<section id="id1">
<h3>3.1 A single sample example<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h3>
<p>At this point, our merged reads still contain the primer sequences. Since these regions are artefacts from the PCR amplification and not biological, they will need to be removed from the reads before continuing with the bioinformatic pipeline. For this library and experiment, we have used the mlCOIintF/jgHCO2198 primer set (Leray et al., 2013). The forward primer corresponds to 5’-GGWACWGGWTGAACWGTWTAYCCYCC-3’ and the reverse primer sequence is 5’-TAIACYTCIGGRTGICCRAARAAYCA-3’. Before batch processing every sample, let’s test our code on a single sample again to start with. For primer or adapter removal, we can use the program <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a>. To specify the primers to be removed, we can use the <code class="docutils literal notranslate"><span class="pre">-a</span></code> parameter. Since we’re removing both the forward and reverse primer, we can link them together using <code class="docutils literal notranslate"><span class="pre">...</span></code>. Remember to use the reverse complement of the reverse primer, as this would be the direction the reverse primer is found in our sequence data after merging. The minimum and maximum length of the amplicon can be specified with the <code class="docutils literal notranslate"><span class="pre">-m</span></code> and <code class="docutils literal notranslate"><span class="pre">-M</span></code> parameters, respectively. To only keep reads for which both primers were found and removed, we need to specify the <code class="docutils literal notranslate"><span class="pre">--discard-untrimmed</span></code> option. The <code class="docutils literal notranslate"><span class="pre">--no-indels</span></code> and <code class="docutils literal notranslate"><span class="pre">-e</span> <span class="pre">2</span></code> parameters allow us to tell the program to not include insertions and deletions in the search and allow a maximum of 2 errors in the primer sequence. We can specify the <code class="docutils literal notranslate"><span class="pre">--revcomp</span></code> parameter to search for the primer sequences in both directions. Finally, we can use the <code class="docutils literal notranslate"><span class="pre">--cores=0</span></code> parameter to automatically detect the number of available cores.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cutadapt<span class="w"> </span>sequenceData/2-raw/COI_100_HK37_merged.fastq<span class="w"> </span>-a<span class="w"> </span>GGWACWGGWTGAACWGTWTAYCCYCC...TGRTTYTTYGGHCAYCCHGARGTHTA<span class="w"> </span>-m<span class="w"> </span><span class="m">300</span><span class="w"> </span>-M<span class="w"> </span><span class="m">330</span><span class="w"> </span>--discard-untrimmed<span class="w"> </span>-o<span class="w"> </span>sequenceData/4-demux/COI_100_HK37_trimmed.fastq<span class="w"> </span>--no-indels<span class="w"> </span>-e<span class="w"> </span><span class="m">2</span><span class="w"> </span>--revcomp<span class="w"> </span>--cores<span class="o">=</span><span class="m">0</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>adapter_removal_single.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/adapter_removal_single.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">This</span> <span class="ow">is</span> <span class="n">cutadapt</span> <span class="mf">4.4</span> <span class="k">with</span> <span class="n">Python</span> <span class="mf">3.11.5</span>
<span class="n">Command</span> <span class="n">line</span> <span class="n">parameters</span><span class="p">:</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">2</span><span class="o">-</span><span class="n">raw</span><span class="o">/</span><span class="n">COI_100_HK37_merged</span><span class="o">.</span><span class="n">fastq</span> <span class="o">-</span><span class="n">a</span> <span class="n">GGWACWGGWTGAACWGTWTAYCCYCC</span><span class="o">...</span><span class="n">TGRTTYTTYGGHCAYCCHGARGTHTA</span> <span class="o">-</span><span class="n">m</span> <span class="mi">300</span> <span class="o">-</span><span class="n">M</span> <span class="mi">330</span> <span class="o">--</span><span class="n">discard</span><span class="o">-</span><span class="n">untrimmed</span> <span class="o">-</span><span class="n">o</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">4</span><span class="o">-</span><span class="n">demux</span><span class="o">/</span><span class="n">COI_100_HK37_trimmed</span><span class="o">.</span><span class="n">fastq</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indels</span> <span class="o">-</span><span class="n">e</span> <span class="mi">2</span> <span class="o">--</span><span class="n">revcomp</span> <span class="o">--</span><span class="n">cores</span><span class="o">=</span><span class="mi">0</span>
<span class="n">Processing</span> <span class="n">single</span><span class="o">-</span><span class="n">end</span> <span class="n">reads</span> <span class="n">on</span> <span class="mi">8</span> <span class="n">cores</span> <span class="o">...</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">146</span><span class="p">,</span><span class="mi">040</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.1</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.69</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">Finished</span> <span class="ow">in</span> <span class="mf">0.598</span> <span class="n">s</span> <span class="p">(</span><span class="mf">4.093</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span> <span class="mf">14.66</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span><span class="p">)</span><span class="o">.</span>

<span class="o">===</span> <span class="n">Summary</span> <span class="o">===</span>

<span class="n">Total</span> <span class="n">reads</span> <span class="n">processed</span><span class="p">:</span>                 <span class="mi">146</span><span class="p">,</span><span class="mi">040</span>
<span class="n">Reads</span> <span class="k">with</span> <span class="n">adapters</span><span class="p">:</span>                   <span class="mi">145</span><span class="p">,</span><span class="mi">999</span> <span class="p">(</span><span class="mf">100.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reverse</span><span class="o">-</span><span class="n">complemented</span><span class="p">:</span>                        <span class="mi">4</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>

<span class="o">==</span> <span class="n">Read</span> <span class="n">fate</span> <span class="n">breakdown</span> <span class="o">==</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">short</span><span class="p">:</span>               <span class="mi">4</span><span class="p">,</span><span class="mi">877</span> <span class="p">(</span><span class="mf">3.3</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">long</span><span class="p">:</span>                <span class="mi">4</span><span class="p">,</span><span class="mi">754</span> <span class="p">(</span><span class="mf">3.3</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">discarded</span> <span class="k">as</span> <span class="n">untrimmed</span><span class="p">:</span>                <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">written</span> <span class="p">(</span><span class="n">passing</span> <span class="n">filters</span><span class="p">):</span>       <span class="mi">136</span><span class="p">,</span><span class="mi">409</span> <span class="p">(</span><span class="mf">93.4</span><span class="o">%</span><span class="p">)</span>

<span class="n">Total</span> <span class="n">basepairs</span> <span class="n">processed</span><span class="p">:</span>    <span class="mi">52</span><span class="p">,</span><span class="mi">491</span><span class="p">,</span><span class="mi">397</span> <span class="n">bp</span>
<span class="n">Total</span> <span class="n">written</span> <span class="p">(</span><span class="n">filtered</span><span class="p">):</span>     <span class="mi">42</span><span class="p">,</span><span class="mi">623</span><span class="p">,</span><span class="mi">291</span> <span class="n">bp</span> <span class="p">(</span><span class="mf">81.2</span><span class="o">%</span><span class="p">)</span>

<span class="o">===</span> <span class="n">Adapter</span> <span class="mi">1</span> <span class="o">===</span>

<span class="n">Sequence</span><span class="p">:</span> <span class="n">GGWACWGGWTGAACWGTWTAYCCYCC</span><span class="o">...</span><span class="n">TGRTTYTTYGGHCAYCCHGARGTHTA</span><span class="p">;</span> <span class="n">Type</span><span class="p">:</span> <span class="n">linked</span><span class="p">;</span> <span class="n">Length</span><span class="p">:</span> <span class="mi">26</span><span class="o">+</span><span class="mi">26</span><span class="p">;</span> <span class="mi">5</span><span class="s1">&#39; trimmed: 143311 times; 3&#39;</span> <span class="n">trimmed</span><span class="p">:</span> <span class="mi">143808</span> <span class="n">times</span><span class="p">;</span> <span class="n">Reverse</span><span class="o">-</span><span class="n">complemented</span><span class="p">:</span> <span class="mi">4</span> <span class="n">times</span>

<span class="n">Minimum</span> <span class="n">overlap</span><span class="p">:</span> <span class="mi">3</span><span class="o">+</span><span class="mi">3</span>
<span class="n">No</span><span class="o">.</span> <span class="n">of</span> <span class="n">allowed</span> <span class="n">errors</span><span class="p">:</span>
<span class="mi">1</span><span class="o">-</span><span class="mi">12</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">0</span><span class="p">;</span> <span class="mi">13</span><span class="o">-</span><span class="mi">25</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span> <span class="mi">26</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">2</span>

<span class="n">No</span><span class="o">.</span> <span class="n">of</span> <span class="n">allowed</span> <span class="n">errors</span><span class="p">:</span>
<span class="mi">1</span><span class="o">-</span><span class="mi">12</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">0</span><span class="p">;</span> <span class="mi">13</span><span class="o">-</span><span class="mi">25</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span> <span class="mi">26</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">2</span>

<span class="n">Overview</span> <span class="n">of</span> <span class="n">removed</span> <span class="n">sequences</span> <span class="n">at</span> <span class="mi">5</span><span class="s1">&#39; end</span>
<span class="n">length</span> <span class="n">count</span> <span class="n">expect</span> <span class="nb">max</span><span class="o">.</span><span class="n">err</span> <span class="n">error</span> <span class="n">counts</span>
<span class="mi">4</span> <span class="mi">2</span> <span class="mf">570.5</span> <span class="mi">0</span> <span class="mi">2</span>
<span class="mi">5</span> <span class="mi">1</span> <span class="mf">142.6</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">11</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">13</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">1</span>
<span class="mi">14</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">15</span> <span class="mi">6</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">6</span>
<span class="mi">16</span> <span class="mi">5</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">4</span> <span class="mi">1</span>
<span class="mi">17</span> <span class="mi">10</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">8</span>
<span class="mi">18</span> <span class="mi">5</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">4</span> <span class="mi">1</span>
<span class="mi">19</span> <span class="mi">20</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">14</span> <span class="mi">6</span>
<span class="mi">20</span> <span class="mi">13</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">11</span> <span class="mi">2</span>
<span class="mi">21</span> <span class="mi">15</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">6</span> <span class="mi">9</span>
<span class="mi">22</span> <span class="mi">12</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">10</span> <span class="mi">2</span>
<span class="mi">23</span> <span class="mi">10</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">9</span> <span class="mi">1</span>
<span class="mi">24</span> <span class="mi">31</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">14</span> <span class="mi">17</span>
<span class="mi">25</span> <span class="mi">851</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">552</span> <span class="mi">299</span>
<span class="mi">26</span> <span class="mi">142152</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">140479</span> <span class="mi">1461</span> <span class="mi">212</span>
<span class="mi">27</span> <span class="mi">175</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">142</span> <span class="mi">27</span> <span class="mi">6</span>

<span class="n">Overview</span> <span class="n">of</span> <span class="n">removed</span> <span class="n">sequences</span> <span class="n">at</span> <span class="mi">3</span><span class="s1">&#39; end</span>
<span class="n">length</span> <span class="n">count</span> <span class="n">expect</span> <span class="nb">max</span><span class="o">.</span><span class="n">err</span> <span class="n">error</span> <span class="n">counts</span>
<span class="mi">13</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">14</span> <span class="mi">2</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">2</span>
<span class="mi">15</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">1</span>
<span class="mi">16</span> <span class="mi">5</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">2</span> <span class="mi">3</span>
<span class="mi">17</span> <span class="mi">3</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">3</span>
<span class="mi">18</span> <span class="mi">10</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">4</span> <span class="mi">6</span>
<span class="mi">19</span> <span class="mi">3</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">0</span> <span class="mi">3</span>
<span class="mi">20</span> <span class="mi">6</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">4</span> <span class="mi">2</span>
<span class="mi">21</span> <span class="mi">14</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">9</span> <span class="mi">5</span>
<span class="mi">22</span> <span class="mi">19</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">15</span> <span class="mi">4</span>
<span class="mi">23</span> <span class="mi">65</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">48</span> <span class="mi">17</span>
<span class="mi">24</span> <span class="mi">77</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">62</span> <span class="mi">15</span>
<span class="mi">25</span> <span class="mi">174</span> <span class="mf">0.0</span> <span class="mi">1</span> <span class="mi">45</span> <span class="mi">129</span>
<span class="mi">26</span> <span class="mi">142067</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">140032</span> <span class="mi">1686</span> <span class="mi">349</span>
<span class="mi">27</span> <span class="mi">1351</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">749</span> <span class="mi">370</span> <span class="mi">232</span>
<span class="mi">28</span> <span class="mi">8</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">3</span> <span class="mi">2</span> <span class="mi">3</span>
<span class="mi">63</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">70</span> <span class="mi">1</span> <span class="mf">0.0</span> <span class="mi">2</span> <span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<p>The cutadapt output provides information on how many reads were analysed, how many reads were trimmed on the plus and minus strand, plus a detailed overview of where the adapters were cut in the sequence. For our sample <strong>sequenceData/2-raw/COI_100_HK37_merged.fastq</strong>, 146,040 reads were processed, 145,999 reads were found to contain both primers, and 4 reads were found where the primers were present as the reverse complement of the read. As these options that we specified in our command managed to remove the primer sequences from nearly all reads, we will use these settings in our for loop to batch process all samples.</p>
</section>
<section id="batch-trimming">
<h3>3.2 batch trimming<a class="headerlink" href="#batch-trimming" title="Permalink to this heading">#</a></h3>
<p>Similarly to the merging of reads, we need to process 27 samples. Hence, we will write another for loop to accomplish this goal, rather than executing the code to trim primer sequences 27 times by ourselves. Because the cutadapt output that is written to the Terminal window is very elaborate, we will write it to a file instead. In the next section (<em>Summarizing cutadapt output</em>), we’ll parse the output to produce some graphs in order to get a better overview of how cutadapt performed across the 27 samples. First, we will remove the file we just generated and move to the <strong>sequenceData/2-raw/</strong> directory. As we’ve introduced the for loop syntax before and explained the cutadapt code snippet, I won’t be going into too much detail on each aspect.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>rm<span class="w"> </span>sequenceData/4-demux/COI_100_HK37_trimmed.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>sequenceData/2-raw/
<span class="k">for</span><span class="w"> </span>fq<span class="w"> </span><span class="k">in</span><span class="w"> </span>*merged.fastq
<span class="k">do</span>

<span class="w">  </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;trimming primer seqs for: </span><span class="si">${</span><span class="nv">fq</span><span class="p">/_merged.fastq/</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="w">  </span>cutadapt<span class="w"> </span><span class="si">${</span><span class="nv">fq</span><span class="si">}</span><span class="w"> </span>-a<span class="w"> </span>GGWACWGGWTGAACWGTWTAYCCYCC...TGRTTYTTYGGHCAYCCHGARGTHTA<span class="w"> </span>-m<span class="w"> </span><span class="m">300</span><span class="w"> </span>-M<span class="w"> </span><span class="m">330</span><span class="w"> </span>--discard-untrimmed<span class="w"> </span>-o<span class="w"> </span>../4-demux/<span class="si">${</span><span class="nv">fq</span><span class="p">/merged.fastq/trimmed.fastq</span><span class="si">}</span><span class="w"> </span>--no-indels<span class="w"> </span>-e<span class="w"> </span><span class="m">2</span><span class="w"> </span>--revcomp<span class="w"> </span>--cores<span class="o">=</span><span class="m">0</span><span class="w"> </span>&gt;&gt;<span class="w"> </span>../0-metadata/cutadapt_primer_trimming.txt

<span class="k">done</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>adapter_removal_batch.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/adapter_removal_batch.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK37</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">146</span><span class="p">,</span><span class="mi">040</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.4</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">17.58</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK38</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">118</span><span class="p">,</span><span class="mi">917</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.7</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">16.14</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK39</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">162</span><span class="p">,</span><span class="mi">326</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.5</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">17.39</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK40</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">115</span><span class="p">,</span><span class="mi">722</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.8</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">15.94</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK41</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">102</span><span class="p">,</span><span class="mi">320</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.8</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">15.74</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK42</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">148</span><span class="p">,</span><span class="mi">463</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.6</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">16.78</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK49</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">91</span><span class="p">,</span><span class="mi">171</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.1</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.62</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK50</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">80</span><span class="p">,</span><span class="mi">114</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.11</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK51</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">83</span><span class="p">,</span><span class="mi">923</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.7</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">12.89</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK52</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">74</span><span class="p">,</span><span class="mi">255</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.2</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.24</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK53</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">76</span><span class="p">,</span><span class="mi">890</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.04</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK54</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">75</span><span class="p">,</span><span class="mi">723</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.11</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK37</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">109</span><span class="p">,</span><span class="mi">424</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.7</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">16.40</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK38</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">86</span><span class="p">,</span><span class="mi">827</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">13.98</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK39</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">134</span><span class="p">,</span><span class="mi">723</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">3.9</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">15.53</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK40</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">109</span><span class="p">,</span><span class="mi">897</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.0</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">15.04</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK41</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">117</span><span class="p">,</span><span class="mi">560</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.0</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.88</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK42</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">77</span><span class="p">,</span><span class="mi">020</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">5.1</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">11.87</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK49</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">67</span><span class="p">,</span><span class="mi">718</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">6.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>   <span class="mf">9.52</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK50</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">75</span><span class="p">,</span><span class="mi">588</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">5.2</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">11.48</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK51</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">105</span><span class="p">,</span><span class="mi">503</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.4</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">13.75</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK52</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">113</span><span class="p">,</span><span class="mi">930</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.0</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">15.13</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK53</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>       <span class="mi">100</span><span class="p">,</span><span class="mi">604</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">14.01</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK54</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>        <span class="mi">95</span><span class="p">,</span><span class="mi">343</span> <span class="n">reads</span> <span class="o">@</span>   <span class="mf">4.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>  <span class="mf">13.86</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_1</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>           <span class="mi">306</span> <span class="n">reads</span> <span class="o">@</span> <span class="mf">435.8</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>   <span class="mf">0.14</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_2</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>         <span class="mi">7</span><span class="p">,</span><span class="mi">842</span> <span class="n">reads</span> <span class="o">@</span>  <span class="mf">21.1</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>   <span class="mf">2.85</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
<span class="n">trimming</span> <span class="n">primer</span> <span class="n">seqs</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_3</span>
<span class="n">Done</span>           <span class="mi">00</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span>         <span class="mi">6</span><span class="p">,</span><span class="mi">114</span> <span class="n">reads</span> <span class="o">@</span>  <span class="mf">24.3</span> <span class="n">µs</span><span class="o">/</span><span class="n">read</span><span class="p">;</span>   <span class="mf">2.47</span> <span class="n">M</span> <span class="n">reads</span><span class="o">/</span><span class="n">minute</span>
</pre></div>
</div>
</div>
<p>When executing this for loop, cutadapt will attempt to remove the primer sequences from all reads in both directions (parameter <code class="docutils literal notranslate"><span class="pre">--revcomp</span></code>). The output files containing the reads where the primers were successfully removed are placed in the subdirectory <strong>sequenceData/4-demux/</strong> and will have the extension <strong>_trimmed.fastq</strong> rather than <strong>_merged.fastq</strong>. As mentioned above, the cutadapt output was written to <strong>sequenceData/0-metadata/cutadapt_primer_trimming.txt</strong>, rather than the Terminal window. We can open this text file using a text editor, such as Sublime Text. However, due to the number of samples, it is difficult to get an overview of the success rate and compare the different samples, a similar issue we encountered when merging forward and reverse reads.</p>
<figure class="align-default" id="cutadapt-statistics">
<img alt="_images/cutadaptstatistics2.png" src="_images/cutadaptstatistics2.png" />
<figcaption>
<p><span class="caption-number">Fig. 11 </span><span class="caption-text">: Output of the cutadapt code, providing information about adapter removal success rate.</span><a class="headerlink" href="#cutadapt-statistics" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="summarizing-cutadapt-output">
<h3>3.3 Summarizing cutadapt output<a class="headerlink" href="#summarizing-cutadapt-output" title="Permalink to this heading">#</a></h3>
<p>To get a better overview of the read statistics after primer trimming and the success rate of cutadapt to locate and trim the primer-binding regions, we can alter the python code producing the bar graph to incorporate the sequence counts after primer trimming. This new version of the code takes in three arguments, the first is the location where the merged sequence files are stored, the second is the location of the raw sequence files, and the third is the location of the trimmed sequence files. Rather than trying to alter the code in the already-existing script <strong>sequenceData/1-scripts/rawMergedStatistics.py</strong>, we will generate a new script where we can simply cut and paste the code below. Generating, saving, and making the script executable follows the same process as discussed previously.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/rawMergedTrimmedStatistics.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="c1">## import modules</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1">## user arguments</span>
<span class="n">mergedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">rawPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">trimmedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

<span class="c1">## first, create a sample name list</span>
<span class="n">mergedFileList</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">mergedPath</span><span class="p">)</span>
<span class="n">sampleNameList</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">mergedFile</span> <span class="ow">in</span> <span class="n">mergedFileList</span><span class="p">:</span>
    <span class="n">sampleName</span> <span class="o">=</span> <span class="n">mergedFile</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_merged.fastq&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;COI_&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;NTC_&#39;</span><span class="p">):</span>
        <span class="n">sampleNameList</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sampleName</span><span class="p">)</span>

<span class="c1">## count number of raw and merged sequences for each sample in sampleNameList</span>
<span class="n">rawSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">mergedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">trimmedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sampleNameList</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">mergedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_merged.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">mergedFile</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mergedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">mergedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">rawPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">rawFile</span><span class="p">:</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rawFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">rawSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">trimmedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_trimmed.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">trimmedFile</span><span class="p">:</span>
        <span class="n">z</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">trimmedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">trimmedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>

<span class="c1">## create a dataframe from the dictionaries</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Sample&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;Raw&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Merged&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">mergedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Trimmed&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">trimmedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">())})</span>

<span class="c1">## sort the dataframe by raw reads in descending order</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">## calculate the percentage of merged/raw and format it with 2 decimal places and the &#39;%&#39; symbol</span>
<span class="n">df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;%&#39;</span>

<span class="c1">## create a horizontal bar plot using seaborn</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Adjust the figure size as needed</span>

<span class="c1">## use seaborn&#39;s barplot</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#BBC6C8&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#469597&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#DDBEAA&#39;</span><span class="p">)</span>

<span class="c1">## add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Number of sequences&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Samples&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Horizontal bar graph of raw, merged, and trimmed reads (Sorted by Total in Reverse)&#39;</span><span class="p">)</span>

<span class="c1">## add a legend</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Add raw read count next to the bars</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">50</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

<span class="c1">## save the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;raw_and_merged_and_trimmed_bargraph.png&#39;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/rawMergedTrimmedStatistics.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/rawMergedTrimmedStatistics.py<span class="w"> </span>sequenceData/2-raw/<span class="w"> </span>sequenceData/2-raw/unzipped/<span class="w"> </span>sequenceData/4-demux/
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>check_adapter_trimming.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/check_adapter_trimming.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/raw_and_merged_and_trimmed_bargraph.png<span class="w"> </span>./
</pre></div>
</div>
<figure class="align-default" id="raw-and-merged-and-trimmed-read-count">
<img alt="_images/raw_and_merged_and_trimmed_bargraph.png" src="_images/raw_and_merged_and_trimmed_bargraph.png" />
<figcaption>
<p><span class="caption-number">Fig. 12 </span><span class="caption-text">: Read count of raw, merged, and trimmed files.</span><a class="headerlink" href="#raw-and-merged-and-trimmed-read-count" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The figure shows a similar success rate for cutadapt to locate and trim primers across all samples, with roughly 85% of raw sequences still included after merging and primer trimming. Interestingly, for the negative control samples (<strong>NTC_1, NTC_2, and NTC_3</strong>) hardly any reads are returned after primer trimming. To see what happened with the negative controls, let’s run this little python code below to print out the cutadapt summary from the file we generated during the for loop. The python code takes two user arguments, which are (1) the file name of the cutadapt summary file to parse and (2) a list of sample names to print the summary statistics for. This list should be separated by <code class="docutils literal notranslate"><span class="pre">+</span></code> symbols.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/cutadaptParsing.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="c1">## import modules</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="c1">## user arguments</span>
<span class="n">cutadaptResults</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sampleNames</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="c1">## create sample list</span>
<span class="n">sampleList</span> <span class="o">=</span> <span class="n">sampleNames</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">)</span>

<span class="c1">## parse cutadapt file</span>
<span class="n">printLine</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">startPrint</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">cutadaptResults</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;=== Adapter 1 ===&#39;</span><span class="p">):</span>
      <span class="n">startPrint</span> <span class="o">=</span> <span class="mi">0</span>
      <span class="n">printLine</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">startPrint</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
      <span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">printLine</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
      <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;=== Summary ===&#39;</span><span class="p">):</span>
        <span class="n">startPrint</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Command line parameters:&#39;</span><span class="p">):</span>
      <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sampleList</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
          <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s1"> summary statistics:&#39;</span><span class="p">)</span>
          <span class="n">printLine</span> <span class="o">=</span> <span class="mi">1</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/cutadaptParsing.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/cutadaptParsing.py<span class="w"> </span>sequenceData/0-metadata/cutadapt_primer_trimming.txt<span class="w"> </span>NTC_1_merged.fastq+NTC_2_merged.fastq+NTC_3_merged.fastq
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>check_neg_trimming.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/check_neg_trimming.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">NTC_1_merged</span><span class="o">.</span><span class="n">fastq</span> <span class="n">summary</span> <span class="n">statistics</span><span class="p">:</span>
<span class="o">===</span> <span class="n">Summary</span> <span class="o">===</span>

<span class="n">Total</span> <span class="n">reads</span> <span class="n">processed</span><span class="p">:</span>                     <span class="mi">306</span>
<span class="n">Reads</span> <span class="k">with</span> <span class="n">adapters</span><span class="p">:</span>                       <span class="mi">306</span> <span class="p">(</span><span class="mf">100.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reverse</span><span class="o">-</span><span class="n">complemented</span><span class="p">:</span>                        <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>

<span class="o">==</span> <span class="n">Read</span> <span class="n">fate</span> <span class="n">breakdown</span> <span class="o">==</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">short</span><span class="p">:</span>                  <span class="mi">77</span> <span class="p">(</span><span class="mf">25.2</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">long</span><span class="p">:</span>                    <span class="mi">3</span> <span class="p">(</span><span class="mf">1.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">discarded</span> <span class="k">as</span> <span class="n">untrimmed</span><span class="p">:</span>                <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">written</span> <span class="p">(</span><span class="n">passing</span> <span class="n">filters</span><span class="p">):</span>           <span class="mi">226</span> <span class="p">(</span><span class="mf">73.9</span><span class="o">%</span><span class="p">)</span>

<span class="n">Total</span> <span class="n">basepairs</span> <span class="n">processed</span><span class="p">:</span>        <span class="mi">88</span><span class="p">,</span><span class="mi">071</span> <span class="n">bp</span>
<span class="n">Total</span> <span class="n">written</span> <span class="p">(</span><span class="n">filtered</span><span class="p">):</span>         <span class="mi">70</span><span class="p">,</span><span class="mi">668</span> <span class="n">bp</span> <span class="p">(</span><span class="mf">80.2</span><span class="o">%</span><span class="p">)</span>

<span class="n">NTC_2_merged</span><span class="o">.</span><span class="n">fastq</span> <span class="n">summary</span> <span class="n">statistics</span><span class="p">:</span>
<span class="o">===</span> <span class="n">Summary</span> <span class="o">===</span>

<span class="n">Total</span> <span class="n">reads</span> <span class="n">processed</span><span class="p">:</span>                   <span class="mi">7</span><span class="p">,</span><span class="mi">842</span>
<span class="n">Reads</span> <span class="k">with</span> <span class="n">adapters</span><span class="p">:</span>                     <span class="mi">7</span><span class="p">,</span><span class="mi">839</span> <span class="p">(</span><span class="mf">100.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reverse</span><span class="o">-</span><span class="n">complemented</span><span class="p">:</span>                        <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>

<span class="o">==</span> <span class="n">Read</span> <span class="n">fate</span> <span class="n">breakdown</span> <span class="o">==</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">short</span><span class="p">:</span>               <span class="mi">7</span><span class="p">,</span><span class="mi">691</span> <span class="p">(</span><span class="mf">98.1</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">long</span><span class="p">:</span>                    <span class="mi">6</span> <span class="p">(</span><span class="mf">0.1</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">discarded</span> <span class="k">as</span> <span class="n">untrimmed</span><span class="p">:</span>                <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">written</span> <span class="p">(</span><span class="n">passing</span> <span class="n">filters</span><span class="p">):</span>           <span class="mi">145</span> <span class="p">(</span><span class="mf">1.8</span><span class="o">%</span><span class="p">)</span>

<span class="n">Total</span> <span class="n">basepairs</span> <span class="n">processed</span><span class="p">:</span>       <span class="mi">507</span><span class="p">,</span><span class="mi">546</span> <span class="n">bp</span>
<span class="n">Total</span> <span class="n">written</span> <span class="p">(</span><span class="n">filtered</span><span class="p">):</span>         <span class="mi">45</span><span class="p">,</span><span class="mi">343</span> <span class="n">bp</span> <span class="p">(</span><span class="mf">8.9</span><span class="o">%</span><span class="p">)</span>

<span class="n">NTC_3_merged</span><span class="o">.</span><span class="n">fastq</span> <span class="n">summary</span> <span class="n">statistics</span><span class="p">:</span>
<span class="o">===</span> <span class="n">Summary</span> <span class="o">===</span>

<span class="n">Total</span> <span class="n">reads</span> <span class="n">processed</span><span class="p">:</span>                   <span class="mi">6</span><span class="p">,</span><span class="mi">114</span>
<span class="n">Reads</span> <span class="k">with</span> <span class="n">adapters</span><span class="p">:</span>                     <span class="mi">6</span><span class="p">,</span><span class="mi">105</span> <span class="p">(</span><span class="mf">99.9</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reverse</span><span class="o">-</span><span class="n">complemented</span><span class="p">:</span>                        <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>

<span class="o">==</span> <span class="n">Read</span> <span class="n">fate</span> <span class="n">breakdown</span> <span class="o">==</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">short</span><span class="p">:</span>               <span class="mi">5</span><span class="p">,</span><span class="mi">928</span> <span class="p">(</span><span class="mf">97.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">that</span> <span class="n">were</span> <span class="n">too</span> <span class="n">long</span><span class="p">:</span>                    <span class="mi">7</span> <span class="p">(</span><span class="mf">0.1</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">discarded</span> <span class="k">as</span> <span class="n">untrimmed</span><span class="p">:</span>                <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span>
<span class="n">Reads</span> <span class="n">written</span> <span class="p">(</span><span class="n">passing</span> <span class="n">filters</span><span class="p">):</span>           <span class="mi">179</span> <span class="p">(</span><span class="mf">2.9</span><span class="o">%</span><span class="p">)</span>

<span class="n">Total</span> <span class="n">basepairs</span> <span class="n">processed</span><span class="p">:</span>       <span class="mi">440</span><span class="p">,</span><span class="mi">273</span> <span class="n">bp</span>
<span class="n">Total</span> <span class="n">written</span> <span class="p">(</span><span class="n">filtered</span><span class="p">):</span>         <span class="mi">55</span><span class="p">,</span><span class="mi">985</span> <span class="n">bp</span> <span class="p">(</span><span class="mf">12.7</span><span class="o">%</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The cutadapt summary statistics for the negative control samples shows that a large proportion of the reads were discarded as they were too short (NTC_2: 98.1%; NTC_3: 97.0%). This probably meant that some primer-dimer sequences seeped through during the size selection step of the library preparation protocol and got sequenced. While primer-dimers should be removed during library preparation, as Illumina favours sequencing shorter fragments, such a low number as observed in this tutorial data set is a normal occurrence and nothing to be worried about.</p>
</section>
</section>
<section id="quality-filtering">
<h2>4. Quality filtering<a class="headerlink" href="#quality-filtering" title="Permalink to this heading">#</a></h2>
<p>During the bioinformatic pipeline, it is critical to only retain high-quality reads to reduce the abundance and impact of spurious sequences. There is an intrinsic error rate to all polymerases used during PCR amplification, as well as sequencing technologies. For example, the most frequently used polymerase during PCR is <em>Taq</em>, though lacks 3’ to 5’ exonuclease proofreading activity, resulting in relatively low replication fidelity. These errors will generate some portion of sequences that vary from their biological origin sequence. Such reads can substantially inflate metrics such as alpha diversity, especially in a denoising approach (more on this later). While it is near-impossible to remove all of these sequences bioinformatically, especially PCR errors, we will attempt to remove erroneous reads by filtering on base calling quality scores (the fourth line of a sequence record in a .fastq file).</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>When processing your own metabarcoding data, I suggest you follow the same structure as we did with merging and trimming reads, i.e., check the code on one sample, play around with the parameter settings, and only batch process your samples when everything works fine. However, to save time during this tutorial, we will move straight to batch processing the samples for the quality filtering steps with the correct parameters.</p>
</div>
<section id="batch-processing">
<h3>4.1 Batch processing<a class="headerlink" href="#batch-processing" title="Permalink to this heading">#</a></h3>
<p>For quality filtering, we will be using the <code class="docutils literal notranslate"><span class="pre">--fastq_filter</span></code> command in <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a>. During this step, we will discard all sequences that do not adhere to a specific set of rules. Quality filtering parameters are not standardized, but rather specific for each library. For our tutorial data, we will filter out all sequences on a much stricter size range compared to the settings we used during primer trimming (<code class="docutils literal notranslate"><span class="pre">--fastq_minlen</span> <span class="pre">310</span></code> and <code class="docutils literal notranslate"><span class="pre">--fastq_maxlen</span> <span class="pre">316</span></code>). Additionally, we will remove sequences that have ambiguous base calls, bases denotes as N, rather than A, C, G, or T (<code class="docutils literal notranslate"><span class="pre">--fastq_maxns</span> <span class="pre">0</span></code>). Ambiguous base calls, however, are not a real issue with Illumina sequencing data. The last parameter we will use to filter our sequences is a maximum expected error rate (<code class="docutils literal notranslate"><span class="pre">--fastq_maxee</span> <span class="pre">1.0</span></code>). As this is the last step in our bioinformatic pipeline where quality scores are essential, we will export our output files both in .fastq and .fasta format. Additionally, we will be merging all files together after this step in the bioinformatic pipeline. Before merging all files, however, we need to change the sequence headers to contain the information to which sample the sequence belongs to, which is currently stored in the file name. We can rename sequence header based on the file name using the <code class="docutils literal notranslate"><span class="pre">--relabel</span></code> parameter.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>sequenceData/4-demux/
<span class="k">for</span><span class="w"> </span>fq<span class="w"> </span><span class="k">in</span><span class="w"> </span>*_trimmed.fastq
<span class="k">do</span>

<span class="w">  </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;Merging reads for: </span><span class="si">${</span><span class="nv">fq</span><span class="p">/_trimmed.fastq/</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="w">  </span>vsearch<span class="w"> </span>--fastq_filter<span class="w"> </span><span class="si">${</span><span class="nv">fq</span><span class="si">}</span><span class="w"> </span>--fastq_maxee<span class="w"> </span><span class="m">1</span>.0<span class="w"> </span>--fastq_maxlen<span class="w"> </span><span class="m">316</span><span class="w"> </span>--fastq_minlen<span class="w"> </span><span class="m">310</span><span class="w"> </span>--fastq_maxns<span class="w"> </span><span class="m">0</span><span class="w"> </span>--fastqout<span class="w"> </span>../5-filter/<span class="si">${</span><span class="nv">fq</span><span class="p">/_trimmed.fastq/_filtered.fastq</span><span class="si">}</span><span class="w"> </span>--fastaout<span class="w"> </span>../5-filter/<span class="si">${</span><span class="nv">fq</span><span class="p">/_trimmed.fastq/_filtered.fasta</span><span class="si">}</span><span class="w"> </span>--relabel<span class="w"> </span><span class="si">${</span><span class="nv">fq</span><span class="p">/_trimmed.fastq/</span><span class="si">}</span>.

<span class="k">done</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>quality_filter.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/quality_filter.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK37</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">128093</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">8316</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK38</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">105809</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">6969</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK39</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">145327</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">9152</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK40</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">103665</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">6297</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK41</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">91076</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">5477</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK42</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">134371</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">8278</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK49</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">82338</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">5241</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK50</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">72895</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4298</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK51</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">75547</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">5111</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK52</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">67509</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">3961</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK53</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">70090</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4008</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_100_HK54</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">68948</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">3952</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK37</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">98678</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">6897</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK38</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">78856</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4859</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK39</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">122200</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">7674</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK40</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">99565</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">6561</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK41</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">106658</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">6567</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK42</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">70167</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4090</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK49</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">61051</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4086</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK50</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">68680</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4195</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK51</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">95654</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">6049</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK52</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">104924</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4927</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK53</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">91320</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">5605</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">COI_500_HK54</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">87233</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">4500</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_1</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">161</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">65</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_2</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">86</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">59</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Merging</span> <span class="n">reads</span> <span class="k">for</span><span class="p">:</span> <span class="n">NTC_3</span>
<span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="nb">input</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">92</span> <span class="n">sequences</span> <span class="n">kept</span> <span class="p">(</span><span class="n">of</span> <span class="n">which</span> <span class="mi">0</span> <span class="n">truncated</span><span class="p">),</span> <span class="mi">87</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
</pre></div>
</div>
</div>
<div class="hint admonition">
<p class="admonition-title">Exercise 3</p>
<p>How would you check if the sequence headers of the .fastq and .fasta output files are relabeled and contain the sample information?</p>
<div class="title dropdown admonition">
<p class="admonition-title">Answer 3</p>
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-3" name="sd-tab-set-1" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-3">
bash: head</label><div class="sd-tab-content docutils">
<p>Using the <code class="docutils literal notranslate"><span class="pre">head</span></code> command, we can print the first N lines of a document.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>sequenceData/4-demux/COI_100_HK37_trimmed.fastq
head<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>sequenceData/5-filter/COI_100_HK37_filtered.fastq
head<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>sequenceData/5-filter/COI_100_HK37_filtered.fasta
</pre></div>
</div>
</div>
<input id="sd-tab-item-4" name="sd-tab-set-1" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-4">
bash: tail</label><div class="sd-tab-content docutils">
<p>Using the <code class="docutils literal notranslate"><span class="pre">tail</span></code> command we can print the last N lines of a document. Note the different number of lines needed to be printed for the .fasta file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tail<span class="w"> </span>-n<span class="w"> </span><span class="m">4</span><span class="w"> </span>sequenceData/4-demux/COI_100_HK37_trimmed.fastq
tail<span class="w"> </span>-n<span class="w"> </span><span class="m">4</span><span class="w"> </span>sequenceData/5-filter/COI_100_HK37_filtered.fastq
tail<span class="w"> </span>-n<span class="w"> </span><span class="m">5</span><span class="w"> </span>sequenceData/5-filter/COI_100_HK37_filtered.fasta
</pre></div>
</div>
</div>
<input id="sd-tab-item-5" name="sd-tab-set-1" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-5">
bash: grep</label><div class="sd-tab-content docutils">
<p>Using the <code class="docutils literal notranslate"><span class="pre">grep</span></code> command on the sample name within the file, we can verify its presence in the file and also count the number of lines (sequences) that contain the sample information. Note the different search term for the .fasta file, due to the different sequence header structure. The <code class="docutils literal notranslate"><span class="pre">grep</span></code> output for the filtered .fastq and .fasta files should be identical.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;^@COI_100_HK37&quot;</span><span class="w"> </span>sequenceData/4-demux/COI_100_HK37_trimmed.fastq
grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;^@COI_100_HK37&quot;</span><span class="w"> </span>sequenceData/5-filter/COI_100_HK37_filtered.fastq
grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;^&gt;COI_100_HK37&quot;</span><span class="w"> </span>sequenceData/5-filter/COI_100_HK37_filtered.fasta
</pre></div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="summary-output">
<h3>4.2 Summary output<a class="headerlink" href="#summary-output" title="Permalink to this heading">#</a></h3>
<p>While VSEARCH provides the numbers for each sample before and after quality filtering, we will update the python script that generates the bar graph again, now to also include the filtered reads. This new python script takes in four user arguments, the same three as before, plus the path to where the filtered sequence files are stored. This will allow us to easily identify how the quality filtering performed and compare differences between samples.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/rawMergedTrimmedFilteredStatistics.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="c1">## import modules</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1">## user arguments</span>
<span class="n">mergedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">rawPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">trimmedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">filteredPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>


<span class="c1">## first, create a sample name list</span>
<span class="n">mergedFileList</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">mergedPath</span><span class="p">)</span>
<span class="n">sampleNameList</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">mergedFile</span> <span class="ow">in</span> <span class="n">mergedFileList</span><span class="p">:</span>
    <span class="n">sampleName</span> <span class="o">=</span> <span class="n">mergedFile</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_merged.fastq&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;COI_&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;NTC_&#39;</span><span class="p">):</span>
        <span class="n">sampleNameList</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sampleName</span><span class="p">)</span>

<span class="c1">## count number of raw and merged sequences for each sample in sampleNameList</span>
<span class="n">rawSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">mergedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">trimmedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">filteredSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sampleNameList</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">mergedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_merged.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">mergedFile</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mergedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">mergedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">rawPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">rawFile</span><span class="p">:</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rawFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">rawSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">trimmedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_trimmed.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">trimmedFile</span><span class="p">:</span>
        <span class="n">z</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">trimmedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">trimmedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">filteredPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_filtered.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">filteredFile</span><span class="p">:</span>
        <span class="n">a</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">filteredFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">filteredSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>


<span class="c1">## create a dataframe from the dictionaries</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Sample&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;Raw&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Merged&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">mergedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Trimmed&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">trimmedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Filtered&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">filteredSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">())})</span>

<span class="c1">## sort the dataframe by raw reads in descending order</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">## calculate the percentage of merged/raw and format it with 2 decimal places and the &#39;%&#39; symbol</span>
<span class="n">df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Filtered&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;%&#39;</span>


<span class="c1">## create a horizontal bar plot using seaborn</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Adjust the figure size as needed</span>

<span class="c1">## use seaborn&#39;s barplot</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#BBC6C8&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#469597&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#DDBEAA&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Filtered&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Filtered&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#806491&#39;</span><span class="p">)</span>

<span class="c1">## add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Number of sequences&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Samples&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Horizontal bar graph of raw, merged, trimmed, and filtered reads (Sorted by Total in Reverse)&#39;</span><span class="p">)</span>

<span class="c1">## add a legend</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Add raw read count next to the bars</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">50</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

<span class="c1">## save the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;raw_and_merged_and_trimmed_and_filtered_bargraph.png&#39;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/rawMergedTrimmedFilteredStatistics.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/rawMergedTrimmedFilteredStatistics.py<span class="w"> </span>sequenceData/2-raw/<span class="w"> </span>sequenceData/2-raw/unzipped/<span class="w"> </span>sequenceData/4-demux/<span class="w"> </span>sequenceData/5-filter/
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>check_quality_filter.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/check_quality_filter.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/raw_and_merged_and_trimmed_and_filtered_bargraph.png<span class="w"> </span>./
</pre></div>
</div>
<figure class="align-default" id="raw-and-merged-and-trimmed-and-filtered-read-count">
<img alt="_images/raw_and_merged_and_trimmed_and_filtered_bargraph.png" src="_images/raw_and_merged_and_trimmed_and_filtered_bargraph.png" />
<figcaption>
<p><span class="caption-number">Fig. 13 </span><span class="caption-text">: Read count of raw, merged, trimmed, and filtered files.</span><a class="headerlink" href="#raw-and-merged-and-trimmed-and-filtered-read-count" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>This bar graph shows that after quality filtering, we still keep around 80% of the raw reads, which is very good and due to the high quality data we started with (remember the multiQC output from before!). Again, we see a big difference between the actual samples and negative controls is observed.</p>
</section>
<section id="check-filtering-step">
<h3>4.3 Check filtering step<a class="headerlink" href="#check-filtering-step" title="Permalink to this heading">#</a></h3>
<p>Before continuing with the bioinformatic pipeline, it is best to check if the quality filtering step retained only high-quality sequences that are of length ~313 bp (estimated length of the amplicon). If quality filtering was successful, we can move on to the next steps in our bioinformatic pipeline. The check the quality and length of our filtered sequence files, we will use <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a> and <a class="reference external" href="https://multiqc.info">MULTIQC</a>, same as before.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>fastqc<span class="w"> </span>sequenceData/5-filter/*.fastq<span class="w"> </span>-o<span class="w"> </span>sequenceData/3-fastqc/<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>fastqc_quality_filter.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/fastqc_quality_filter.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Started</span> <span class="n">analysis</span> <span class="n">of</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">5</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">10</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">15</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">20</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">25</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Started</span> <span class="n">analysis</span> <span class="n">of</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">30</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">5</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">35</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">10</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">40</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">15</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">45</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">20</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">50</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">25</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">30</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">55</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">35</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">60</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">40</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">65</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">45</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">70</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK37_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">Approx</span> <span class="mi">50</span><span class="o">%</span> <span class="n">complete</span> <span class="k">for</span> <span class="n">COI_100_HK38_filtered</span><span class="o">.</span><span class="n">fastq</span>
<span class="o">...</span>
</pre></div>
</div>
</div>
<p>FastQC will generate a .html report for every single .fastq file in the subdirectory <strong>sequenceData/5-filter/</strong>. Since we are working with 27 files, it will be easier to compare the reports by collating them, something we can do using multiQC.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>multiqc<span class="w"> </span>sequenceData/3-fastqc/*filtered*<span class="w"> </span>-o<span class="w"> </span>sequenceData/3-fastqc/
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>multiqc_quality_filter.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/multiqc_quality_filter.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/sequenceData/3-fastqc/multiqc_report.html<span class="w"> </span>./
</pre></div>
</div>
<p>The multiQC program will combine all 27 FastQC reports into a single .html document. Let’s open this to see how our filtered sequence data is looking like. Since we have written both multiQC reports (raw and filtered) to the same <strong>sequenceData/3-fastqc/</strong> directory, make sure to open the latest report. The file name will most likely end in <strong>_1.html</strong>.</p>
<figure class="align-default" id="multiqc-report-filtered">
<img alt="_images/multiqcreportfiltered2.png" src="_images/multiqcreportfiltered2.png" />
<figcaption>
<p><span class="caption-number">Fig. 14 </span><span class="caption-text">: The .html multiQC report for the quality filtered reads</span><a class="headerlink" href="#multiqc-report-filtered" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>From this multiQC report, we can see that we only retained high quality reads that are of length 313 bp. Exactly what we need to move forward with our bioinformatic pipeline!</p>
</section>
</section>
<section id="dereplication">
<h2>5. Dereplication<a class="headerlink" href="#dereplication" title="Permalink to this heading">#</a></h2>
<section id="combining-data-into-one-file">
<h3>5.1 Combining data into one file<a class="headerlink" href="#combining-data-into-one-file" title="Permalink to this heading">#</a></h3>
<p>Once we have verified that the reads passing quality filtering are of high quality and of a length similar to the expected amplicon size, we can move forward with our bioinformatic pipeline. For the next final steps of the bioinformatic pipeline, it makes more sense to combine all the files into a single file. So, let’s do this first using the <code class="docutils literal notranslate"><span class="pre">cat</span></code> command, which stands for <em>concatenate</em>. Since base calling quality scores are not essential for our analysis from this point onwards, we can combine all the .fasta files to reduce the file size and computational cost. To concatenate all files, we can use the <code class="docutils literal notranslate"><span class="pre">*</span></code> symbol. Make sure to specify the output file using <code class="docutils literal notranslate"><span class="pre">&gt;</span></code>, otherwise <code class="docutils literal notranslate"><span class="pre">cat</span></code> will print the combined file to the Terminal window instead.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>sequenceData/5-filter/*.fasta<span class="w"> </span>&gt;<span class="w"> </span>sequenceData/6-quality/combined.fasta
</pre></div>
</div>
<div class="hint admonition">
<p class="admonition-title">Exercise 4</p>
<p>How would you determine if no sequences were lost during the <code class="docutils literal notranslate"><span class="pre">cat</span></code> command to merge all .fasta files?</p>
<div class="title dropdown admonition">
<p class="admonition-title">Answer 4</p>
<p>First, let’s start with the easiest problem, i.e., determine the number of sequences in the merged file. Here we can simply pipe the output of <code class="docutils literal notranslate"><span class="pre">wc</span> <span class="pre">-l</span></code> to the <code class="docutils literal notranslate"><span class="pre">awk</span></code> command and divide the number of lines by 5 (= one sequence record).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wc<span class="w"> </span>-l<span class="w"> </span>sequenceData/6-quality/combined.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{print $1/5}&#39;</span>
</pre></div>
</div>
<div class="note admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">2230993</span>
</pre></div>
</div>
</div>
<p>To count the total number of sequences in the separate files before concatenation, we will need to write a for loop to count all the lines in each file, pipe that number to <code class="docutils literal notranslate"><span class="pre">awk</span></code> to divide it by 5, and sum the number of sequences for each file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">total_lines</span><span class="o">=</span><span class="m">0</span>
<span class="k">for</span><span class="w"> </span>item<span class="w"> </span><span class="k">in</span><span class="w"> </span>sequenceData/5-filter/*.fasta<span class="w"> </span>
<span class="k">do</span>
<span class="w">    </span><span class="nv">lines</span><span class="o">=</span><span class="k">$(</span>wc<span class="w"> </span>-l<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$item</span><span class="s2">&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{print $1/5}&#39;</span><span class="k">)</span>
<span class="w">    </span><span class="nv">total_lines</span><span class="o">=</span><span class="k">$((</span><span class="nv">total_lines</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nv">lines</span><span class="k">))</span>
<span class="k">done</span>

<span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$total_lines</span><span class="s2">&quot;</span>
</pre></div>
</div>
<div class="note admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">2230993</span>
</pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="finding-unique-sequences">
<h3>5.2 Finding unique sequences<a class="headerlink" href="#finding-unique-sequences" title="Permalink to this heading">#</a></h3>
<p>With the <code class="docutils literal notranslate"><span class="pre">cat</span></code> command, we have created a single file. The next step in our bioinformatic pipeline is to dereplicate our data, i.e., find unique sequences. Since metabarcoding data is based on a PCR amplification method, the same DNA molecule will have been copied thousands or millions of times and, therefore, sequenced multiple times. In order to reduce the file size and computational cost, it is convenient to work with unique sequences and keep track of how many times each unique sequence was found in our data set. This dereplication step is also the reason why we combined all samples into a single file, as the same barcode (or species if you will) can be found across different samples.</p>
<p>We can dereplicate our data by using the <code class="docutils literal notranslate"><span class="pre">--derep_fulllength</span></code> command in <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a>. The <code class="docutils literal notranslate"><span class="pre">--sizeout</span></code> parameter keeps a tally of the number of times each unique sequence was observed across all samples and places this information in the sequence header. Since we are looking at unique sequences across samples, it wouldn’t make much sense to keep the current sequence header information, as it currently refers to which sample a sequence belongs to. We will, therefore, rename the sequence headers by using the <code class="docutils literal notranslate"><span class="pre">--relabel</span></code> parameter.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--derep_fulllength<span class="w"> </span>sequenceData/6-quality/combined.fasta<span class="w"> </span>--sizeout<span class="w"> </span>--relabel<span class="w"> </span>uniq.<span class="w"> </span>--output<span class="w"> </span>sequenceData/6-quality/uniques.fasta
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong><a class="reference external" href="http://dereplication.sh">dereplication.sh</a></strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/dereplication.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Dereplicating</span> <span class="n">file</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">6</span><span class="o">-</span><span class="n">quality</span><span class="o">/</span><span class="n">combined</span><span class="o">.</span><span class="n">fasta</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">697630765</span> <span class="n">nt</span> <span class="ow">in</span> <span class="mi">2230993</span> <span class="n">seqs</span><span class="p">,</span> <span class="nb">min</span> <span class="mi">310</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">316</span><span class="p">,</span> <span class="n">avg</span> <span class="mi">313</span>
<span class="n">Sorting</span> <span class="mi">100</span><span class="o">%</span>
<span class="mi">244379</span> <span class="n">unique</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">avg</span> <span class="n">cluster</span> <span class="mf">9.1</span><span class="p">,</span> <span class="n">median</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">68022</span>
<span class="n">Writing</span> <span class="n">FASTA</span> <span class="n">output</span> <span class="n">file</span> <span class="mi">100</span><span class="o">%</span> 
</pre></div>
</div>
</div>
<p>We can see from the VSEARCH output that 244,379 out of 2,230,993 sequences were unique. The average number a unique sequence was observed was 9.1 times, while the median is 1 and the maximum is 68,022 times. As mentioned before, the <code class="docutils literal notranslate"><span class="pre">--sizeout</span></code> parameter will have added the abundance information for each unique sequence to the header. Let’s use the <code class="docutils literal notranslate"><span class="pre">head</span></code> command to investigate the first couple of sequences.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>sequenceData/6-quality/uniques.fasta
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.1</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">68022</span>
<span class="n">CCTAAGCTCCAATATTGCCCACGCCGGGGCGTCTGTTGACCTTGCTATCTTTAGGCTACACTTGGCTGGGGTTTCTTCTC</span>
<span class="n">TACTCGGGGCTGTAAACTTTATTAGAACTATCGCTAACCTGCGAGCTTTAGGGCTAATTCTTGACCGTATAACACTATTC</span>
<span class="n">ACATGATCAGTTCTTATCACCGCCATCCTTCTCCTTCTTTCTCTACCTGTTCTCGCAGGGGCTATTACGATACTCCTTAC</span>
<span class="n">CGACCGAAATCTAAATACCTCTTTTTATGACCCTAGAGGAGGGGGAGACCCTATTCTCTACCAACACCTGTTT</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.2</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">59045</span>
<span class="n">CCTCTCCGCTAGCGGAGCCCATGGATCTGCCTCAGTAGATTTAAGGATCTTTTCCCTCCATCTAGCAGGAGTATCTTCTA</span>
<span class="n">TCCTCGGATCCATCAACTTTATCACCACAATTATTAATATGCGAACCCCAAACCTCACATGAGACAAAATCTCTCTCTTT</span>
<span class="n">ACTTGATCCATCCTTATTACTACTATCCTACTCCTCCTCTCTCTCCCTGTCCTAGCCGGAGCTTTAACTATACTTCTAAC</span>
<span class="n">AGACCGTAACTTCAACACAACATTTTTTGACCCTAGAGGAGGAGGCGACCCTATCCTTTACCAGCACCTTTTC</span>
</pre></div>
</div>
</div>
<p>The output of the <code class="docutils literal notranslate"><span class="pre">head</span></code> command shows that the unique sequences are already sorted by abundance, with the most abundant unique sequence occurring 68,022 times and the second most abundant sequence occurring 59,045 times.</p>
<div class="hint admonition">
<p class="admonition-title">Exercise 5</p>
<p>How would you determine the number of times the 10 most and 10 least abundant unique sequences were observed?</p>
<div class="title dropdown admonition">
<p class="admonition-title">Answer 5</p>
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-6" name="sd-tab-set-2" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-6">
Two-liner</label><div class="sd-tab-content docutils">
<p>The simplest way to determine the abundance of the 10 most and 10 least abundant unique sequences, based on the code we’ve learned during this tutorial thus far, would be to write two lines of code, one for the 10 most abundant and one for the 10 least abundant. We can solve this problem by taking out all the lines that start with <code class="docutils literal notranslate"><span class="pre">^&gt;uniq</span></code> using <code class="docutils literal notranslate"><span class="pre">grep</span></code> and pipe that list to either the <code class="docutils literal notranslate"><span class="pre">head</span> <span class="pre">-n</span> <span class="pre">10</span></code> or <code class="docutils literal notranslate"><span class="pre">tail</span> <span class="pre">-n</span> <span class="pre">10</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep<span class="w"> </span><span class="s1">&#39;^&gt;uniq&#39;</span><span class="w"> </span>sequenceData/6-quality/uniques.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span>
grep<span class="w"> </span><span class="s1">&#39;^&gt;uniq&#39;</span><span class="w"> </span>sequenceData/6-quality/uniques.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span>
</pre></div>
</div>
<div class="note admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.1</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">68022</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.2</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">59045</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.3</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">57634</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.4</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">42670</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.5</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">42507</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.6</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">40947</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.7</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">38849</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.8</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">38764</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.9</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">37125</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.10</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">34995</span>

<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244370</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244371</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244372</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244373</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244374</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244375</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244376</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244377</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244378</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244379</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
<input id="sd-tab-item-7" name="sd-tab-set-2" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-7">
One-liner</label><div class="sd-tab-content docutils">
<p>To solve this problem with one line of code, we need to use <code class="docutils literal notranslate"><span class="pre">awk</span></code>. As mentioned before, the syntax of this language is quite complex. To provide a bit of background to the solution:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/^&gt;uniq/</span></code>: is the pattern we are looking for (similar to <code class="docutils literal notranslate"><span class="pre">grep</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{a[i++]=$0}</span></code>: when a line matches the pattern, it is added to an array <code class="docutils literal notranslate"><span class="pre">a</span></code>, and the index <code class="docutils literal notranslate"><span class="pre">i</span></code> is incremented. <code class="docutils literal notranslate"><span class="pre">$0</span></code> represents the entire line.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">END</span></code>: the next section of the code is only executed after processing all lines in the input file.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">for</span> <span class="pre">(j=i-10;</span> <span class="pre">j&lt;i;</span> <span class="pre">j++)</span> <span class="pre">print</span> <span class="pre">a[j]</span></code>: this loop prints the last 10 lines that matched the pattern. It starts from <code class="docutils literal notranslate"><span class="pre">i</span> <span class="pre">-</span> <span class="pre">10</span></code> to <code class="docutils literal notranslate"><span class="pre">i</span></code>, where <code class="docutils literal notranslate"><span class="pre">i</span></code> is the number of lines that matched the pattern, and prints those lines from the array <code class="docutils literal notranslate"><span class="pre">a</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">for</span> <span class="pre">(k=0;</span> <span class="pre">k&lt;10;</span> <span class="pre">k++)</span> <span class="pre">print</span> <span class="pre">a[k]</span></code>: this loop prints the first 10 lines that matched the pattern, iterating from <code class="docutils literal notranslate"><span class="pre">0</span></code> to <code class="docutils literal notranslate"><span class="pre">9</span></code> (first 10 elements) and prints those lines from the array <code class="docutils literal notranslate"><span class="pre">a</span></code>.</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>awk<span class="w"> </span><span class="s1">&#39;/^&gt;uniq/ {a[i++]=$0} END {for (j=i-10; j&lt;i; j++) print a[j]; for (k=0; k&lt;10; k++) print a[k]}&#39;</span><span class="w"> </span>sequenceData/6-quality/uniques.fasta
</pre></div>
</div>
<div class="note admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244370</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244371</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244372</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244373</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244374</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244375</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244376</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244377</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244378</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.244379</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">1</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.1</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">68022</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.2</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">59045</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.3</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">57634</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.4</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">42670</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.5</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">42507</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.6</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">40947</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.7</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">38849</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.8</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">38764</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.9</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">37125</span>
<span class="o">&gt;</span><span class="n">uniq</span><span class="mf">.10</span><span class="p">;</span><span class="n">size</span><span class="o">=</span><span class="mi">34995</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="denoising">
<h2>6. Denoising<a class="headerlink" href="#denoising" title="Permalink to this heading">#</a></h2>
<p>Now that our data set is filtered and unique sequences have been retrieved, we are ready for the next step in the bioinformatic pipeline, i.e., looking for <strong>biologically meaningful</strong> or <strong>biologically correct</strong> sequences. Two approaches to achieve this goal exist, including <strong>denoising</strong> and <strong>clustering</strong>. There is still an ongoing debate on what the best approach is to obtain these <strong>biologicall meaningful</strong> sequences. For more information, these are two good papers to start with: <a class="reference external" href="https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13398">Brandt et al., 2021</a> and <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04115-6">Antich et al., 2021</a>. For this tutorial we won’t be discussing this topic in much detail, but this is the basic idea…</p>
<p>When <strong>clustering</strong> the dataset, OTUs (Operational Taxonomic Units) will be generated by combining sequences that are similar to a set percentage level (traditionally 97%), with the most abundant sequence identified as the true sequence. When clustering at 97%, this means that if a sequence is more than 3% different than the generated OTU, a second OTU will be generated. The concept of OTU clustering was introduced in the 1960s and has been debated since. Clustering the dataset is usually used for identifying species in metabarcoding data.</p>
<p><strong>Denoising</strong>, on the other hand, attempts to identify all correct biological sequences through an algorithm. In short, denoising will cluster the data with a 100% threshold and tries to identify errors based on abundance differences. The retained sequences are called ZOTU (Zero-radius Operation Taxonomic Unit) or ASVs (Amplicon Sequence Variants). Denoising the dataset is usually used for identifying intraspecific variation in metabarcoding data. A schematic of both approaches can be found below.</p>
<figure class="align-default" id="schematic-of-denoising-and-clustering">
<img alt="_images/denoisingandclustering.png" src="_images/denoisingandclustering.png" />
<figcaption>
<p><span class="caption-number">Fig. 15 </span><span class="caption-text">: A schematic representation of denoising and clustering. Copyright: <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04115-6">Antich et al., 2021</a></span><a class="headerlink" href="#schematic-of-denoising-and-clustering" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>This difference in approach may seem small but has a very big impact on your final dataset!</p>
<p>When you denoise the dataset, it is expected that one species may have more than one ZOTU. This means that when the reference database is incomplete, or you plan to work with ZOTUs instead of taxonomic assignments, your diversity estimates will be highly inflated. When clustering the dataset, on the other hand, it is expected that an OTU may have more than one species assigned to it, meaning that you may lose some correct biological sequences that are present in your data by merging species with barcodes more similar than 97%. In other words, you will miss out on differentiating closely related species and intraspecific variation.</p>
<p>For this tutorial, we will use the <strong>denoising</strong> approach, as it is favoured in recent years. However, clustering is still a valid option. So, feel free to explore this approach for your own data set if you prefer.</p>
<p>For denoising, we will use the <a class="reference external" href="https://drive5.com/usearch/manual/cmd_unoise3.html">unoise3 algorithm</a> as implemented in the <code class="docutils literal notranslate"><span class="pre">--cluster_unoise</span></code> command in <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a>. Since denoising is based on read abundance of the unique sequences, we can specify the <code class="docutils literal notranslate"><span class="pre">--sizein</span></code> parameter. The minimum abundance threshold for a true denoised read is defaulted to 8 reads, as specified by the unoise3 algorithm developer. However, more recent research by <a class="reference external" href="https://www.nature.com/articles/nmeth.2276">Bokulich et al., 2013</a>, identified a minimum abundance threshold to be more appropriate. Hence, we will set the <code class="docutils literal notranslate"><span class="pre">--minsize</span></code> parameter to 0.001%, which in our case is ~22 reads. As we will be merging different reads with varying abundances, we need to recalculate the new count for each denoised read using the <code class="docutils literal notranslate"><span class="pre">--sizeout</span></code> parameter. Relabeling the sequence headers can be done through the <code class="docutils literal notranslate"><span class="pre">--relabel</span></code> parameter and the output file is specified using <code class="docutils literal notranslate"><span class="pre">--centroids</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--cluster_unoise<span class="w"> </span>sequenceData/6-quality/uniques.fasta<span class="w"> </span>--sizein<span class="w"> </span>--minsize<span class="w"> </span><span class="m">22</span><span class="w"> </span>--sizeout<span class="w"> </span>--relabel<span class="w"> </span>denoised.<span class="w"> </span>--centroids<span class="w"> </span>sequenceData/6-quality/denoised.fasta
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong><a class="reference external" href="http://denoising.sh">denoising.sh</a></strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/denoising.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="n">file</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">6</span><span class="o">-</span><span class="n">quality</span><span class="o">/</span><span class="n">uniques</span><span class="o">.</span><span class="n">fasta</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">1519235</span> <span class="n">nt</span> <span class="ow">in</span> <span class="mi">4859</span> <span class="n">seqs</span><span class="p">,</span> <span class="nb">min</span> <span class="mi">310</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">316</span><span class="p">,</span> <span class="n">avg</span> <span class="mi">313</span>
<span class="n">minsize</span> <span class="mi">22</span><span class="p">:</span> <span class="mi">239520</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Masking</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Sorting</span> <span class="n">by</span> <span class="n">abundance</span> <span class="mi">100</span><span class="o">%</span>
<span class="n">Counting</span> <span class="n">k</span><span class="o">-</span><span class="n">mers</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Clustering</span> <span class="mi">100</span><span class="o">%</span>  
<span class="n">Sorting</span> <span class="n">clusters</span> <span class="mi">100</span><span class="o">%</span>
<span class="n">Writing</span> <span class="n">clusters</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Clusters</span><span class="p">:</span> <span class="mi">1372</span> <span class="n">Size</span> <span class="nb">min</span> <span class="mi">22</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">187320</span><span class="p">,</span> <span class="n">avg</span> <span class="mf">3.5</span>
<span class="n">Singletons</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.0</span><span class="o">%</span> <span class="n">of</span> <span class="n">seqs</span><span class="p">,</span> <span class="mf">0.0</span><span class="o">%</span> <span class="n">of</span> <span class="n">clusters</span>
</pre></div>
</div>
</div>
<p>From the output, we can see that we have generated 1,372 denoised reads and discarded 239,520 unique sequences, as they did not achieve an abundance of at least 22 reads.</p>
</section>
<section id="chimera-removal">
<h2>7. Chimera removal<a class="headerlink" href="#chimera-removal" title="Permalink to this heading">#</a></h2>
<p>The second to last step in our bioinformatic pipeline is to remove chimeric sequences. Amplicon sequencing has the potential to generate chimeric reads, which can cause spurious inference of biological variation. Chimeric amplicons form when an incomplete DNA strand anneals to a different template and primes synthesis of a new template derived from two different biological sequences, or in other words chimeras are artefact sequences formed by two or more biological sequences incorrectly joined together. More information can be found on this <a class="reference external" href="https://www.biorxiv.org/content/biorxiv/early/2016/09/09/074252.full.pdf">website</a> and a simple illustration can be found below.</p>
<figure class="align-default" id="schematic-of-chimera-formation">
<img alt="_images/chimeras.jpg" src="_images/chimeras.jpg" />
<figcaption>
<p><span class="caption-number">Fig. 16 </span><span class="caption-text">: A schematic representation of chimera formation. Copyright: <a class="reference external" href="https://genome.cshlp.org/content/21/3/494/F1.expansion.html">Genome Research</a></span><a class="headerlink" href="#schematic-of-chimera-formation" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>We will use the <code class="docutils literal notranslate"><span class="pre">--uchime3_denovo</span></code> algorithm as implemented in <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a> for removing chimeric sequences from our denoised reads. This method is also based on sequence abundance, hence, we need to provide the <code class="docutils literal notranslate"><span class="pre">--sizein</span></code> parameter. As the output (parameter <code class="docutils literal notranslate"><span class="pre">--nonchimeras</span></code>) file will contain our <strong>biologically relevant</strong> sequences, we will relabel our sequence headers using <code class="docutils literal notranslate"><span class="pre">--relabel</span> <span class="pre">zotu.</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--uchime3_denovo<span class="w"> </span>sequenceData/6-quality/denoised.fasta<span class="w"> </span>--sizein<span class="w"> </span>--nonchimeras<span class="w"> </span>sequenceData/8-final/zotus.fasta<span class="w"> </span>--relabel<span class="w"> </span>zotu.
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong><a class="reference external" href="http://chimeras.sh">chimeras.sh</a></strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/chimeras.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="n">file</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">6</span><span class="o">-</span><span class="n">quality</span><span class="o">/</span><span class="n">denoised</span><span class="o">.</span><span class="n">fasta</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">429255</span> <span class="n">nt</span> <span class="ow">in</span> <span class="mi">1372</span> <span class="n">seqs</span><span class="p">,</span> <span class="nb">min</span> <span class="mi">310</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">316</span><span class="p">,</span> <span class="n">avg</span> <span class="mi">313</span>
<span class="n">Masking</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Sorting</span> <span class="n">by</span> <span class="n">abundance</span> <span class="mi">100</span><span class="o">%</span>
<span class="n">Counting</span> <span class="n">k</span><span class="o">-</span><span class="n">mers</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Detecting</span> <span class="n">chimeras</span> <span class="mi">100</span><span class="o">%</span>  
<span class="n">Found</span> <span class="mi">27</span> <span class="p">(</span><span class="mf">2.0</span><span class="o">%</span><span class="p">)</span> <span class="n">chimeras</span><span class="p">,</span> <span class="mi">1345</span> <span class="p">(</span><span class="mf">98.0</span><span class="o">%</span><span class="p">)</span> <span class="n">non</span><span class="o">-</span><span class="n">chimeras</span><span class="p">,</span>
<span class="ow">and</span> <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span> <span class="n">borderline</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="mi">1372</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Taking</span> <span class="n">abundance</span> <span class="n">information</span> <span class="n">into</span> <span class="n">account</span><span class="p">,</span> <span class="n">this</span> <span class="n">corresponds</span> <span class="n">to</span>
<span class="mi">1085</span> <span class="p">(</span><span class="mf">0.1</span><span class="o">%</span><span class="p">)</span> <span class="n">chimeras</span><span class="p">,</span> <span class="mi">1785492</span> <span class="p">(</span><span class="mf">99.9</span><span class="o">%</span><span class="p">)</span> <span class="n">non</span><span class="o">-</span><span class="n">chimeras</span><span class="p">,</span>
<span class="ow">and</span> <span class="mi">0</span> <span class="p">(</span><span class="mf">0.0</span><span class="o">%</span><span class="p">)</span> <span class="n">borderline</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="mi">1786577</span> <span class="n">total</span> <span class="n">sequences</span><span class="o">.</span>
</pre></div>
</div>
</div>
<p>The output shows that 98% of our denoised reads were kept and 27 sequences were identified as chimeric. Taking into account abundance information, this would result into 0.1% of reads identified as chimeras.</p>
<p><strong>The <em>sequenceData/8-final/zotus.fasta</em> file is the first output file we have created that we need for our statistical analysis!</strong></p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If you use the unoise3 algorithm as implemented in USEARCH, chimera removal is built into the denoising step and will not have to be conducted separately!</p>
</div>
</section>
<section id="frequency-table">
<h2>8. Frequency table<a class="headerlink" href="#frequency-table" title="Permalink to this heading">#</a></h2>
<p>Now that we have created our list of <strong>biologically relevant</strong> sequences or ZOTUs or “species”, we are ready to generate a frequency table, also known as a count table. This will be the last step in our bioinformatic pipeline. A frequency table is something you might have encountered before during some traditional ecological surveys you have conducted, whereby a table was created with site names as column headers, a species list as rows, and the number in each cell representing the number of individuals observed for a specific species at a specific site.</p>
<p>In metabarcoding studies, a frequency table is analogous, where it tells us how many times each sequence has appeared in each sample. It is the end-product of all the bioinformatic processing steps we have conducted today. Now that we have identified what we believe to be true biological sequences, we are going to generate our frequency table by matching the merged sequences to our ZOTU sequence list. Remember that the sequences within the <strong>sequenceData/6-quality/combined.fasta</strong> file have the information of the sample they belong to present in the sequence header, which is how the <code class="docutils literal notranslate"><span class="pre">--usearch_global</span></code> command in <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a> can generate the frequency table. The <code class="docutils literal notranslate"><span class="pre">--db</span></code> parameter allows us to set the ZOTU sequence list (<strong>sequenceData/8-final/zotus.fasta</strong>) as the database to search against, while we can specify the <code class="docutils literal notranslate"><span class="pre">--strand</span></code> parameter as <em>plus</em>, since all sequences are in the same direction after primer trimming. Finally, we need to incorporate a 97% identity threshold for this function through the <code class="docutils literal notranslate"><span class="pre">--id</span></code> parameter. This might seem counter-intuitive, since we employed a denoising approach. However, providing some leniency on which sequences can map to our ZOTU will allow us to incorporate a larger percentage of the data set. As some ZOTUs might be more similar to each other than 97%, the algorithm will sort out the best match and add the sequence to the correct ZOTU sequence. If you’d like to be more conservative, you can set this threshold to 99%, though this is not recommended by the authors.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--usearch_global<span class="w"> </span>sequenceData/6-quality/combined.fasta<span class="w"> </span>--db<span class="w"> </span>sequenceData/8-final/zotus.fasta<span class="w"> </span>--strand<span class="w"> </span>plus<span class="w"> </span>--id<span class="w"> </span><span class="m">0</span>.97<span class="w"> </span>--otutabout<span class="w"> </span>sequenceData/8-final/zotutable.txt
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>frequency_table.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/frequency_table.sh
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="n">file</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">8</span><span class="o">-</span><span class="n">final</span><span class="o">/</span><span class="n">zotus</span><span class="o">.</span><span class="n">fasta</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">420804</span> <span class="n">nt</span> <span class="ow">in</span> <span class="mi">1345</span> <span class="n">seqs</span><span class="p">,</span> <span class="nb">min</span> <span class="mi">310</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">316</span><span class="p">,</span> <span class="n">avg</span> <span class="mi">313</span>
<span class="n">Masking</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Counting</span> <span class="n">k</span><span class="o">-</span><span class="n">mers</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Creating</span> <span class="n">k</span><span class="o">-</span><span class="n">mer</span> <span class="n">index</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Searching</span> <span class="mi">100</span><span class="o">%</span>  
<span class="n">Matching</span> <span class="n">unique</span> <span class="n">query</span> <span class="n">sequences</span><span class="p">:</span> <span class="mi">2195816</span> <span class="n">of</span> <span class="mi">2230993</span> <span class="p">(</span><span class="mf">98.42</span><span class="o">%</span><span class="p">)</span>
<span class="n">Writing</span> <span class="n">OTU</span> <span class="n">table</span> <span class="p">(</span><span class="n">classic</span><span class="p">)</span> <span class="mi">100</span><span class="o">%</span>  
</pre></div>
</div>
</div>
<p>The output printed to the Terminal window shows we managed to map 2,195,816 (98.42%) sequences to the ZOTU list, which were incorporated in our frequency table. The output file is a simple text file, which we can open in Excel. Below is a screenshot where I coloured the samples (column headers) blue and the sequence list (row names) yellow.</p>
<figure class="align-default" id="frequency-table-output">
<img alt="_images/freqtable.png" src="_images/freqtable.png" />
<figcaption>
<p><span class="caption-number">Fig. 17 </span><span class="caption-text">: The frequency table output file in Excel, where the samples are coloured blue and the sequences are coloured yellow.</span><a class="headerlink" href="#frequency-table-output" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<section id="summarise-sequence-counts">
<h3>8.1 Summarise sequence counts<a class="headerlink" href="#summarise-sequence-counts" title="Permalink to this heading">#</a></h3>
<p>As a final step, let’s update the python script that produces a bar graph again, now including the final read count incorporated for each sample. This new python script takes in five user arguments, the same four as before, plus the frequency table file name. This final bar graph will allow us to get a great overview of what we have accomplished today.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/rawMergedTrimmedFilteredFinalStatistics.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="c1">## import modules</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="c1">## user arguments</span>
<span class="n">mergedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">rawPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">trimmedPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">filteredPath</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
<span class="n">freqTable</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>


<span class="c1">## first, create a sample name list</span>
<span class="n">mergedFileList</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">mergedPath</span><span class="p">)</span>
<span class="n">sampleNameList</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">mergedFile</span> <span class="ow">in</span> <span class="n">mergedFileList</span><span class="p">:</span>
    <span class="n">sampleName</span> <span class="o">=</span> <span class="n">mergedFile</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_merged.fastq&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;COI_&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">sampleName</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;NTC_&#39;</span><span class="p">):</span>
        <span class="n">sampleNameList</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sampleName</span><span class="p">)</span>

<span class="c1">## count number of raw and merged sequences for each sample in sampleNameList</span>
<span class="n">rawSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">mergedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">trimmedSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">filteredSeqCount</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sampleNameList</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">mergedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_merged.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">mergedFile</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mergedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">mergedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">rawPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">rawFile</span><span class="p">:</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rawFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">rawSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">trimmedPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_trimmed.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">trimmedFile</span><span class="p">:</span>
        <span class="n">z</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">trimmedFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">trimmedSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">z</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">filteredPath</span><span class="si">}{</span><span class="n">sample</span><span class="si">}</span><span class="s1">_filtered.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">filteredFile</span><span class="p">:</span>
        <span class="n">a</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">filteredFile</span><span class="o">.</span><span class="n">readlines</span><span class="p">())</span> <span class="o">/</span> <span class="mi">4</span>
        <span class="n">filteredSeqCount</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>

<span class="n">freqTableDF</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">freqTable</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="n">freqTableDF</span><span class="p">[</span><span class="s1">&#39;Final&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">freqTableDF</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">columnsToDrop</span> <span class="o">=</span> <span class="n">freqTableDF</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">difference</span><span class="p">([</span><span class="s1">&#39;Final&#39;</span><span class="p">])</span>
<span class="n">freqTableDF</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">columnsToDrop</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

<span class="c1">## create a dataframe from the dictionaries</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Sample&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;Raw&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">rawSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Merged&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">mergedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Trimmed&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">trimmedSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="s1">&#39;Filtered&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">filteredSeqCount</span><span class="o">.</span><span class="n">values</span><span class="p">())})</span>

<span class="c1">## merge both data frames</span>
<span class="n">merged_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">freqTableDF</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Sample&#39;</span><span class="p">],</span> <span class="n">right_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;inner&#39;</span><span class="p">)</span>

<span class="c1">## sort the dataframe by raw reads in descending order</span>
<span class="n">merged_df</span> <span class="o">=</span> <span class="n">merged_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">## calculate the percentage of merged/raw and format it with 2 decimal places and the &#39;%&#39; symbol</span>
<span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Final&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;%&#39;</span>


<span class="c1">## create a horizontal bar plot using seaborn</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>  <span class="c1"># Adjust the figure size as needed</span>

<span class="c1">## use seaborn&#39;s barplot</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Raw&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#BBC6C8&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Merged&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#469597&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Trimmed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#DDBEAA&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Filtered&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Filtered&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#806491&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;Final&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Sample&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Final&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#2F70AF&#39;</span><span class="p">)</span>

<span class="c1">## add labels and title</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Number of sequences&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Samples&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Horizontal bar graph of raw, merged, trimmed, and filtered reads (Sorted by Total in Reverse)&#39;</span><span class="p">)</span>

<span class="c1">## add a legend</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Add raw read count next to the bars</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;Raw&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">50</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>

<span class="c1">## save the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;raw_and_merged_and_trimmed_and_filtered_and_final_bargraph.png&#39;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/rawMergedTrimmedFilteredFinalStatistics.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/example-script.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/rawMergedTrimmedFilteredFinalStatistics.py<span class="w"> </span>sequenceData/2-raw/<span class="w"> </span>sequenceData/2-raw/unzipped/<span class="w"> </span>sequenceData/4-demux/<span class="w"> </span>sequenceData/5-filter/<span class="w"> </span>sequenceData/8-final/zotutable.txt
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save the file and change the name to <strong>check_table.sh</strong>. Press <code class="docutils literal notranslate"><span class="pre">return</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> to save under a different name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch<span class="w"> </span>sequenceData/1-scripts/check_table.sh
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>scp<span class="w"> </span>-r<span class="w"> </span>ednaw01@hpc2021-io1.hku.hk:~/ednaw01/raw_and_merged_and_trimmed_and_filtered_and_final_bargraph.png<span class="w"> </span>./
</pre></div>
</div>
<figure class="align-default" id="raw-and-merged-and-trimmed-and-filtered-and-final-read-count">
<img alt="_images/raw_and_merged_and_trimmed_and_filtered_and_final_bargraph.png" src="_images/raw_and_merged_and_trimmed_and_filtered_and_final_bargraph.png" />
<figcaption>
<p><span class="caption-number">Fig. 18 </span><span class="caption-text">: Read count of raw, merged, trimmed, filtered, and final frequency table for each sample.</span><a class="headerlink" href="#raw-and-merged-and-trimmed-and-filtered-and-final-read-count" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>This bar graph shows that for the final frequency table, we have kept around 80% of the raw reads for each sample, which is very good and due to the high quality data we started with (remember the multiQC output from before!). Again, we see a big difference between the actual samples and negative controls is observed.</p>
<p><strong>That’s it for today, see you tomorrow when we will assign a taxonomic ID to our ZOTU sequence list!</strong></p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="setup.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Setting up the environment</p>
      </div>
    </a>
    <a class="right-next"
       href="taxonomyassignment.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Taxonomy assignment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preparing-data-for-processing">1. Preparing data for processing</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unzipping-data">1.1 Unzipping data</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#counting-files-pipe">1.1.1 Counting files (pipe <code class="docutils literal notranslate"><span class="pre">|</span></code>)</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fastq-and-fasta-file-structure">1.2 Fastq and Fasta file structure</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#changing-file-names">1.3 Changing file names</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-control-of-raw-data">1.4 Quality control of raw data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merging-forward-and-reverse-reads">2. Merging forward and reverse reads</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-single-sample-example">2.1 A single sample example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-merging-reads">2.2 Batch merging reads</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarizing-the-output">2.3 Summarizing the output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-primer-sequences">3. Removing primer sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">3.1 A single sample example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-trimming">3.2 batch trimming</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarizing-cutadapt-output">3.3 Summarizing cutadapt output</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-filtering">4. Quality filtering</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-processing">4.1 Batch processing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-output">4.2 Summary output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#check-filtering-step">4.3 Check filtering step</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#dereplication">5. Dereplication</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#combining-data-into-one-file">5.1 Combining data into one file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#finding-unique-sequences">5.2 Finding unique sequences</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#denoising">6. Denoising</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chimera-removal">7. Chimera removal</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#frequency-table">8. Frequency table</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarise-sequence-counts">8.1 Summarise sequence counts</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Dr Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>