{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d9f90285",
   "metadata": {},
   "source": [
    "# Phylogenetic tree\n",
    "\n",
    "Before conducting the statistical analysis, there is one last output file we can create to help us explore our data, a phylogenetic tree of the ZOTU sequences. Phylogenetic relatedness is commonly used to inform downstream analyses, especially the calculation of phylogeny-aware distances. We can create the phylogenetic tree in R. Therefore, let's open a new R script and save it as **sequenceData/1-scripts/phylotree.R**. We can copy-paste the R code below into this script to run all the code.\n",
    "\n",
    "First, let's set up the R environment by setting the working directory and loading all the necessary R packages.\n",
    "\n",
    "```{code-block} R\n",
    "#########################\n",
    "# PREPARE R ENVIRONMENT #\n",
    "#########################\n",
    "library(msa)\n",
    "library(ape)\n",
    "library(ade4)\n",
    "library(scales)\n",
    "library(basetheme)\n",
    "library(Biostrings)\n",
    "\n",
    "# set working directory\n",
    "setwd('/Users/gjeunen/Documents/work/lectures/HKU2023/tutorial/sequenceData/8-final')\n",
    "```\n",
    "\n",
    "## 1. Reading/Parsing data\n",
    "\n",
    "To generate the phylogenetic tree file, we need to import the ZOTU sequence file (**sequenceData/8-final/zotusFiltered.fasta**) and the taxonomy assignment file (**sequenceData/8-final/taxonomyFiltered.txt**) into R. Once the files are read into R, we can explore the sequence file and make sure the names between both files match.\n",
    "\n",
    "```{code-block} R\n",
    "##################\n",
    "# READ DATA IN R #\n",
    "##################\n",
    "dna <- readDNAStringSet('zotusFiltered.fasta')\n",
    "taxonomy <- read.table('taxonomyFiltered.txt', header = TRUE, sep = '\\t', fill = TRUE, na.strings = c('', 'NA'), row.names = 1)\n",
    "\n",
    "# explore data sets\n",
    "dna\n",
    "all(taxonomy$ID == rownames(dna))\n",
    "table(taxonomy$phylum)\n",
    "table(taxonomy$class)\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: tip\n",
    "```\n",
    "> all(taxonomy$ID == rownames(dna))\n",
    "[1] TRUE\n",
    "\n",
    "> table(taxonomy$phylum)\n",
    "\n",
    "       Annelida      Arthropoda Bacillariophyta         Bryozoa     Chlorophyta        Chordata        Cnidaria \n",
    "            198             427               7              27               1              19              33 \n",
    "       Discosea   Echinodermata      Entoprocta   Eukaryota_nan    Gastrotricha        Mollusca        Nematoda \n",
    "              1              15               2               9               6              71               6 \n",
    "       Nemertea Platyhelminthes        Porifera      Rhodophyta      Tardigrada \n",
    "              7               9              11              15               1 \n",
    "\n",
    "> table(taxonomy$class)\n",
    "\n",
    "        Actinopteri            Amphibia        Annelida_nan            Anthozoa           Arachnida \n",
    "                 11                   1                   6                  10                  30 \n",
    "     Arthropoda_nan          Ascidiacea                Aves   Bacillariophyceae            Bivalvia \n",
    "                  2                   1                   1                   2                  16 \n",
    "       Branchiopoda         Bryozoa_nan         Cephalopoda           Chilopoda         Chromadorea \n",
    "                  2                   2                   1                   4                   2 \n",
    "         Clitellata          Collembola Coscinodiscophyceae        Demospongiae           Diplopoda \n",
    "                 11                   6                   4                  11                   8 \n",
    "       Discosea_nan              Enopla             Enoplea      Entoprocta_nan   Eukaryota_nan_nan \n",
    "                  1                   4                   4                   2                   5 \n",
    "    Florideophyceae          Gastropoda    Gastrotricha_nan        Gymnolaemata    Heterotardigrada \n",
    "                 15                  53                   6                  25                   1 \n",
    "        Hexanauplia            Hydrozoa             Insecta        Malacostraca            Mammalia \n",
    "                101                  23                 157                 103                   5 \n",
    "       Mediophyceae        Nemertea_nan         Ophiuroidea           Ostracoda      Palaeonemertea \n",
    "                  1                   1                  15                   1                   1 \n",
    "       Phaeophyceae        Pilidiophora          Polychaeta      Polyplacophora Pyramimonadophyceae \n",
    "                  4                   1                 176                   1                   1 \n",
    "      Rhabditophora           Sipuncula         Thecostraca \n",
    "                  9                   5                  13 \n",
    "```\n",
    "````\n",
    "\n",
    "```{figure} dnastringsetphylogeny.png\n",
    ":name: DNA string set phylogeny\n",
    "\n",
    ": The R output of the DNA string set\n",
    "```\n",
    "\n",
    "## 2. Calculating MSA\n",
    "\n",
    "Once the data is read into memory, we can calculate the multiple sequence alignment using `ClustalW` from the [msa R package](https://bioconductor.org/packages/release/bioc/html/msa.html).\n",
    "\n",
    "```{important}\n",
    "Generating a multiple sequence alignment from a small DNA fragment (~313 bp) of such a large data set containing a very broad range of taxonomic groups is incredibly complex and near-impossible. Using standard settings, the alignment will most likely end up nonsensical with a length of thousands of basepairs. To circumvent this issue, we will subset the data into taxonomic groups of interest. For our tutorial data, let's generate a simple phylogenetic tree for the phylum Cnidaria.\n",
    "```\n",
    "\n",
    "```{code-block} R\n",
    "#######\n",
    "# MSA #\n",
    "#######\n",
    "# 1. subset the data for the phylum Cnidaria\n",
    "taxonomyCnidaria <- subset(taxonomy, phylum == 'Cnidaria')\n",
    "dnaCnidaria <- dna[rownames(taxonomyCnidaria)]\n",
    "\n",
    "# 1. run MSA for Cnidaria\n",
    "multCnidaria <- msa(dnaCnidaria, method = 'ClustalW', type = 'dna', order = 'input')\n",
    "```\n",
    "\n",
    "## 3. Distance-based phylogenies\n",
    "\n",
    "Distance-based phylogenetic reconstruction consists of (i) computing pairwise genetic distances between sequences, (ii) representing these distances using a tree, and (iii) evaluating the relevance of this representation.\n",
    "\n",
    "### 3.1 Computing distances\n",
    "\n",
    "We first compute genetic distances using the `dist.dna` function in the [ape R package](https://cran.r-project.org/web/packages/ape/index.html). This package contains more than 15 different genetic distance models (see `?dist.dna` for details). Here, we use Tamura and Nei 1993's model, which allows for different rates of transitions and transversions, heterogeneous base frequencies, and between-site variation of the substitution rate.\n",
    "\n",
    "```{code-block} R\n",
    "#########################\n",
    "# CALCULATING DISTANCES #\n",
    "#########################\n",
    "# 1. distance calculation for Cnidaria\n",
    "multConvertCnidaria <- msaConvert(multCnidaria, 'ape::DNAbin')\n",
    "distanceCnidaria <- dist.dna(multConvertCnidaria, model = 'TN93')\n",
    "class(distanceCnidaria)\n",
    "length(distanceCnidaria)\n",
    "temp <- as.data.frame(as.matrix(distanceCnidaria))\n",
    "table.paint(temp, cleg = 0, clabel.row = 0.5, clabel.col = 0.5)\n",
    "```\n",
    "\n",
    "```{figure} distancecnidaria.png\n",
    ":name: Genetic distance visualisation for Cnidaria\n",
    "\n",
    ": Genetic distance visualisation for the phylum Cnidaria.\n",
    "```\n",
    "\n",
    "### 3.2 Building trees\n",
    "\n",
    "We use trees to get a better representation of the genetic distances between sequences. It is important, however, to bear in mind that the obtained trees are not necessarily an efficient representation of the original distances, and information can - and likely will - be lost in the process. A wide array of algorithms for constructing trees from a distance matrix are available. In this tutorial, we will use the classical Neighbour-joining algorithm.\n",
    "\n",
    "```{code-block} R\n",
    "##################\n",
    "# BUILDING TREES #\n",
    "##################\n",
    "# 1. for Cnidaria\n",
    "treeCnidaria <- nj(distanceCnidaria)\n",
    "class(treeCnidaria)\n",
    "treeCnidaria\n",
    "plot(treeCnidaria, cex = 0.6)\n",
    "title('A simple NJ tree of the phylum Cnidaria')\n",
    "```\n",
    "\n",
    "```{figure} simplecnidarianjtree.png\n",
    ":name: Simple NJ tree of the phylum Cnidaria\n",
    "\n",
    ": A simple NJ tree of the phylum Cnidaria\n",
    "```\n",
    "\n",
    "We can see that the tree is plotted with the ZOTU labels. This makes it difficult to check if the tree makes any sense, since the ZOTU labels do not provide any information about the taxonomic ID. We can change this by changing the tip labels with a taxonomic ID name. Let's use the class IDs for the phylum Mollusca and the order IDs for the class Malacostraca.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. plot tree with colored labels for Cnidaria\n",
    "plot(treeCnidaria, show.tip = FALSE)\n",
    "title('Unrooted NJ tree for the phylum Cnidaria')\n",
    "tiplabels(taxonomyCnidaria$class, bg = lab2col(taxonomyCnidaria$class), cex = 0.5)\n",
    "```\n",
    "\n",
    "```{figure} simplecolorednjtreecnidaria.png\n",
    ":name: Simple NJ tree of the phylum Cnidaria with coloured tip labels\n",
    "\n",
    ": A simple NJ tree of the phylum Cnidaria with coloured tip labels based on the taxonomic ID at the class level.\n",
    "```\n",
    "\n",
    "The results are a bit messed up in our trees it seems, with classes mixed together in the tree. However, we need to keep in mind that we currently have created an unrooted tree. The default plot view, though, is rooted. We can check if our tree is rooted or not by using the function `is.rooted`.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. check if tree is rooted\n",
    "is.rooted(treeCnidaria)\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: tip\n",
    "```\n",
    "# 1. check if tree is rooted\n",
    "is.rooted(treeCnidaria)\n",
    "[1] FALSE\n",
    "```\n",
    "````\n",
    "\n",
    "So, a better representation of our tree would be to specify `type = 'unrooted'` in our `plot` command.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. plot unrooted tree\n",
    "plot(treeCnidaria, type = 'unrooted', show.tip = FALSE)\n",
    "title('Unrooted NJ tree of phylum Cnidaria')\n",
    "tiplabels(taxonomyCnidaria$class, bg = lab2col(taxonomyCnidaria$class), cex = 0.5)\n",
    "```\n",
    "\n",
    "```{figure} unrootedCnidariaColoured.png\n",
    ":name: Unrooted NJ tree of the phylum Cnidaria with coloured tip labels based on the taxonomic ID at the class level\n",
    "\n",
    ": Unrooted NJ tree of the phylum Cnidaria with coloured tip labels based on the taxonomic ID at the class level.\n",
    "```\n",
    "\n",
    "To explore our data set through a phylogenetic tree, it is easiest to work on a rooted tree however. To root the tree, we can use a simple \"hack\", where we set one of the `Anthozoa` sequences as outgroup. Check the taxnoomy data frame at which row number an `Anthozoa` sequence is located. In our case, ZOTU 1117 in row 1 is assigned an anthazoan taxonomic ID.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. set root of tree as Anthozoa ZOTU 1117 (row 1)\n",
    "treeCnidaria2 <- root(treeCnidaria, out = 1)\n",
    "plot(treeCnidaria2, show.tip = FALSE, edge.width = 2)\n",
    "title('Rooted NJ tree for phylum Cnidaria')\n",
    "tiplabels(taxonomyCnidaria$class, bg = lab2col(taxonomyCnidaria$class), cex = 0.5, fg = 'transparent')\n",
    "axisPhylo()\n",
    "```\n",
    "\n",
    "```{figure} rootedCnidariaColoured.png\n",
    ":name: Rooted NJ tree of the phylum Cnidaria with coloured tip labels based on the taxonomic ID at the class level\n",
    "\n",
    ": Rooted NJ tree of the phylum Cnidaria with coloured tip labels based on the taxonomic ID at the class level.\n",
    "```\n",
    "\n",
    "This is looking much better already! However, there is this one group that isn't as resolved as it should be between the `Anthozoa` and the `Hydrozoa`. We can change the labels to plot the quality of the BLAST hit and keep the colour of the taxonomic class ID to determine if mis-classification might have occurred. So, let's plot the tree again with these new labels.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. combine the percent identity and query coverage columns, 2. set root of tree as Anthozoa ZOTU 1117 (row 1)\n",
    "cols <- c('pident', 'qcov')\n",
    "taxonomyCnidaria$taxqual <- apply(taxonomyCnidaria[, cols], 1, paste, collapse = '__')\n",
    "treeCnidaria3 <- root(treeCnidaria, out = 1)\n",
    "plot(treeCnidaria3, show.tip = FALSE, edge.width = 2)\n",
    "title('Rooted NJ tree for phylum Cnidaria')\n",
    "tiplabels(taxonomyCnidaria$taxqual, bg = lab2col(taxonomyCnidaria$class), cex = 0.5, fg = 'transparent')\n",
    "axisPhylo()\n",
    "```\n",
    "\n",
    "```{figure} rootedCnidariaColouredPident.png\n",
    ":name: Rooted NJ tree of the phylum Cnidaria with coloured tip labels based on the taxonomic ID at the class level and labels as taxonomic quality\n",
    "\n",
    ": Rooted NJ tree of the phylum Cnidaria with coloured tips based on the taxonomic ID at the class level and labels based on the quality of the BLAST hit.\n",
    "```\n",
    "\n",
    "From the output, we can see that the badly resolved section of the tree, based on the BLAST taxonomic ID, all contain low scores for the BLAST hits, indicating the potential for mis-classification. For your project, you might want to consider removing such signals when using a taxonomic-dependent analysis or, instead, use a taxonomic-independent analysis approach. For this tutorial, we'll keep the tree as it is and move forward to the next section.\n",
    "\n",
    "### 3.3 Tree quality\n",
    "\n",
    "Many genetic distances and hierarchical clustering algorithms can be used to build trees; not all of them are appropriate for a given dataset. Genetic distances rely on hypotheses about the evolution of DNA sequences, which should be taken into account. Once you have chosen an appropriate genetic distance and built a tree using this distance, an essential but yet most often overlooked question is whether this tree is actually a good representation of the original distance matrix. This is easily investigated using simple biplots and correlation indices. The function `cophenetic` is used to compete distances between the tips of the tree. Note that more distances are availabe in the `adephylo` R pacakge and `distTips` function.\n",
    "\n",
    "```{code-block} R\n",
    "################\n",
    "# TREE QUALITY #\n",
    "################\n",
    "x <- as.vector(distanceCnidaria)\n",
    "y <- as.vector(as.dist(cophenetic(treeCnidaria2)))\n",
    "plot(x, y, xlab = 'original distance', ylab = 'distance in the tree',\n",
    "     main = 'Is NJ appropriate?', pch = 20, col = alpha('black', 0.4), cex = 0.3)\n",
    "abline(lm(y ~ x), col = 'red')\n",
    "cor(x, y)^2\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: tip\n",
    "```\n",
    "> cor(x, y)^2\n",
    "[1] 0.9697193\n",
    "```\n",
    "````\n",
    "\n",
    "```{figure} treequalityCnidaria.png\n",
    ":name: Biplot and correlation analysis of the quality of the tree\n",
    "\n",
    ": Biplot and correlation analysis of the quality of the tree with the original distance matrix values on the x-axis and the distance on the tree on the y-axis.\n",
    "```\n",
    "\n",
    "In our case, we can see that this simple Neighbour Joining tree we created for the phylum Cnidaria in our data set is a good representation of the initial distance matrix, with an R2-value of 0.97.\n",
    "\n",
    "## 4. Exporting trees\n",
    "\n",
    "Once we have created the phylogenetic tree, we can export the tree in Newick format to save it using the R code below.\n",
    "\n",
    "```{code-block} R\n",
    "###############\n",
    "# EXPORT TREE #\n",
    "###############\n",
    "ape::write.tree(treeCnidaria2, file = 'njTreeCnidaria.txt')\n",
    "```\n",
    "\n",
    "## 5. Intro to data visualization\n",
    "\n",
    "With this tree we just created, we can start exploring our data set. There are many elaborate ways of exploring your data and creating stunning visualisations. However, most of this will be outside the scope of this tutorial. Today, we will show you how to add other data types to the phylogenetic tree, whereby we will look at differences in detection of the Cnidaria sequences between highly and low-impacted sites. The R package we will be using to generate such figures is `ggtree`, which was developed by Prof. Guangchaung Yu at the University of Hong Kong. For a lot more information to create more elaborate figures and details on the `ggtree` package, I recommend starting with [this book](https://yulab-smu.top/treedata-book/author.html).\n",
    "\n",
    "### 5.1 Reading data into R\n",
    "\n",
    "The first thing we need to do is read our files into the R environment, as well as some R packages. Next, we need to subset our data to only include Cnidarian sequences, as this is what our phylogenetic tree is based on. Last, we can import our data frames into one phyloseq object.\n",
    "\n",
    "```{code-block} R\n",
    "#########################\n",
    "# PREPARE R ENVIRONMENT #\n",
    "#########################\n",
    "library(ggtree)\n",
    "library(phyloseq)\n",
    "library(Biostrings)\n",
    "library(speedyseq)\n",
    "library(microbiome)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(egg)\n",
    "\n",
    "# set working directory\n",
    "setwd('/Users/gjeunen/Documents/work/lectures/HKU2023/tutorial/sequenceData/8-final')\n",
    "\n",
    "##################\n",
    "# READ DATA IN R #\n",
    "##################\n",
    "metaData <- read.table('../0-metadata/sampleMetadataFiltered.txt', header = TRUE, sep = '\\t', row.names = 1, check.names = FALSE, comment.char = '')\n",
    "freqTable <- read.table('zotutableFiltered.txt', header = TRUE, sep = '\\t', row.names = 1, check.names = FALSE, comment.char = '')\n",
    "sequenceTable <- readDNAStringSet('zotusFiltered.fasta')\n",
    "phyloTree <- read.tree('njTreeCnidaria.txt')\n",
    "taxonomyTable <- read.table('taxonomyFiltered.txt', header = TRUE, sep = '\\t', row.names = 1, check.names = FALSE, comment.char = '')\n",
    "\n",
    "###############\n",
    "# SUBSET DATA #\n",
    "###############\n",
    "taxonomyCnidaria <- subset(taxonomyTable, phylum == 'Cnidaria')\n",
    "sequenceCnidaria <- sequenceTable[rownames(taxonomyCnidaria)]\n",
    "freqTableCnidaria <- freqTable[rownames(taxonomyCnidaria), ]\n",
    "\n",
    "# 1. import dataframes into phyloseq\n",
    "OTU = otu_table(freqTableCnidaria, taxa_are_rows = TRUE)\n",
    "TAX = tax_table(as.matrix(taxonomyCnidaria))\n",
    "META = sample_data(metaData)\n",
    "physeq = merge_phyloseq(OTU, TAX, META, phyloTree, sequenceTable)\n",
    "physeq\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: tip\n",
    "```\n",
    "> physeq\n",
    "phyloseq-class experiment-level object\n",
    "otu_table()   OTU Table:         [ 33 taxa and 24 samples ]\n",
    "sample_data() Sample Data:       [ 24 samples by 32 sample variables ]\n",
    "tax_table()   Taxonomy Table:    [ 33 taxa by 9 taxonomic ranks ]\n",
    "phy_tree()    Phylogenetic Tree: [ 33 tips and 31 internal nodes ]\n",
    "refseq()      DNAStringSet:      [ 33 reference sequences ]\n",
    "```\n",
    "````\n",
    "\n",
    "### 5.2 Visualise tree\n",
    "\n",
    "Now that we have all our data imported into R and phyloseq objects, let's visualise our phylogenetic tree using `ggtree`.\n",
    "\n",
    "```{code-block} R\n",
    "#####################\n",
    "# PHYLOGENETIC TREE #\n",
    "#####################\n",
    "# 1. create phylogenetic tree\n",
    "ptree <- ggtree(phyloTree) +\n",
    "  geom_tiplab(size = 2, align = TRUE, linesize = 0.5, ladderize = FALSE) +\n",
    "  theme_tree2()\n",
    "ptree\n",
    "```\n",
    "\n",
    "```{figure} ggtreebasic.png\n",
    ":name: Basic plot of the phylogenetic tree using the ggtree R package\n",
    "\n",
    ": Basic plot of the phylogenetic tree of the phylum Cnidaria using the `ggtree` R package.\n",
    "```\n",
    "\n",
    "### 5.3 Data massaging\n",
    "\n",
    "For the plot we want to create, i.e., a multiple bar graph showing differences in Cnidaria detection between high and low impacted sites, we do not need to differentiate between size fractions. Therefore, let's merge the 100 µm and 500 µm samples originating from the same ARMS. Afterwards, we can transform the data to presence-absence, as we would like to determine detection frequency differences between the sites.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. merge different size fractions of the same ARMS, 2. transform data to presence-absence\n",
    "physeq.arms <- merge_samples2(physeq, 'ARMS_ID')\n",
    "physeq.arms.pa <- microbiome::transform(physeq.arms, 'pa')\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: tip\n",
    "```\n",
    "> physeq.arms.pa\n",
    "phyloseq-class experiment-level object\n",
    "otu_table()   OTU Table:          [ 33 taxa and 12 samples ]:\n",
    "sample_data() Sample Data:        [ 12 samples by 32 sample variables ]:\n",
    "tax_table()   Taxonomy Table:     [ 33 taxa by 9 taxonomic ranks ]:\n",
    "phy_tree()    Phylogenetic Tree:  [ 33 tips and 31 internal nodes ]:\n",
    "refseq()      DNAStringSet:       [ 33 reference sequences ]\n",
    "taxa are rows\n",
    "```\n",
    "````\n",
    "\n",
    "### 5.4 Bar plots\n",
    "\n",
    "Next, we can create separate bar plots for the low-impacted and the high-impacted sites. We need to make sure that the y-axis, which displays the ZOTU sequences, is in the same order as the tree. We can accomplish this through the `get_taxa_name` function in `ggtree`. Note that we have to reverse this order to make it match with the bar plot.\n",
    "\n",
    "```{code-block} R\n",
    "# 1. generate bar graph for low-impacted sites\n",
    "lowColor = '#528B8B'\n",
    "orderedTipLabels <- get_taxa_name(ptree)\n",
    "physeq.arms.pa.low <- subset_samples(physeq.arms.pa, impact == 'low')\n",
    "bar.freq.low <- physeq.arms.pa.low %>% psmelt() %>% as_tibble()\n",
    "barfreq.low <- ggplot(data = bar.freq.low) + \n",
    "  geom_bar(aes(y = OTU, x = Abundance, fill = impact), stat = 'identity', show.legend = FALSE) +\n",
    "  scale_fill_manual(values = lowColor) +\n",
    "  scale_y_discrete(limits = rev(orderedTipLabels)) +\n",
    "  scale_x_continuous(limits = c(0, 6), expand = c(0, 0), breaks = c(0, 3, 6)) +\n",
    "  theme_classic() + \n",
    "  theme(axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.line.y = element_blank())\n",
    "barfreq.low\n",
    "```\n",
    "\n",
    "```{figure} barplotlow.png\n",
    ":name: bar plot of the low impacted sites\n",
    "\n",
    ": Bar plot of the low impacted sites.\n",
    "```\n",
    "\n",
    "```{code-block} R\n",
    "# 1. generate bar graph for high-impacted sites\n",
    "highColor = '#CBDCDC'\n",
    "orderedTipLabels <- get_taxa_name(ptree)\n",
    "physeq.arms.pa.high <- subset_samples(physeq.arms.pa, impact == 'high')\n",
    "bar.freq.high <- physeq.arms.pa.high %>% psmelt() %>% as_tibble()\n",
    "barfreq.high <- ggplot(data = bar.freq.high) + \n",
    "  geom_bar(aes(y = OTU, x = Abundance, fill = impact), stat = 'identity', show.legend = FALSE) +\n",
    "  scale_fill_manual(values = highColor) +\n",
    "  scale_y_discrete(limits = rev(orderedTipLabels)) +\n",
    "  scale_x_continuous(limits = c(0, 6), expand = c(0, 0), breaks = c(0, 3, 6)) +\n",
    "  theme_classic() + \n",
    "  theme(axis.title.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.line.y = element_blank())\n",
    "barfreq.high\n",
    "```\n",
    "\n",
    "```{figure} barplothigh.png\n",
    ":name: bar plot of the high impacted sites\n",
    "\n",
    ": Bar plot of the high impacted sites.\n",
    "```\n",
    "\n",
    "### 5.5 Combine all plots\n",
    "\n",
    "One final aspect is to combine all plots. We can easily do this using the `egg` R package, which allows us to specify the number of rows we want to plot (`nrow`) and the width of each plot (`widths`).\n",
    "\n",
    "```{code-block} R\n",
    "# 1. combine using egg\n",
    "ggarrange(ptree, barfreq.low, barfreq.high, nrow = 1, widths = c(8, 1, 1))\n",
    "```\n",
    "\n",
    "```{figure} combinedphylotree.png\n",
    ":name: combined phylo tree\n",
    "\n",
    ": A combined plot showing the Cnidarian phylogenetic tree, as well as the detection frequency of each observed Cnidarian sequence for the low and high impacted sites.\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "md:myst",
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.13,
    "jupytext_version": "1.11.5"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "source_map": [
   13
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}