{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a81de7af",
   "metadata": {},
   "source": [
    "# Setting up the environment\n",
    "\n",
    "## 1. Folder structure\n",
    "\n",
    "First, we will have a look at the folder structure we will use during the workshop. For any project, it is important to organise your various files into subfolders. Through the course you will be navigating between folders to run your analyses. It may seem confusing or tedious at first, especially as there are only a few files we will be dealing with during this workshop. However, remember that for many projects you can easily generate hundreds of files, so it is best to start with good practices from the beginning.\n",
    "\n",
    "We will be using the following folder structure for this experiment:\n",
    "\n",
    "```none\n",
    "└── sequenceData\n",
    "    ├── 0-metadata   \n",
    "    ├── 1-scripts      \n",
    "    ├── 2-raw    \n",
    "    ├── 3-fastqc\n",
    "    ├── 4-demux\n",
    "    ├── 5-filter\n",
    "    ├── 6-quality\n",
    "    ├── 7-refdb\n",
    "    └── 8-final\n",
    "```\n",
    "\n",
    "To set up this folder structure, let us navigate to the starting directory of this workshop. We can do this using the `cd` command, which stands for *change directory*.\n",
    "\n",
    "```{warning}\n",
    "The code below is the path on my system. Make sure to alter the path that specifies the location of the starting folder on your system.\n",
    "```\n",
    "\n",
    "```{code-block} bash\n",
    "cd /Users/gjeunen/Documents/work/lectures/HKU2023/tutorial/\n",
    "```\n",
    "\n",
    "It is always a good idea to check if the code executed as expected. In this case, we can verify we are in the correct working directory by using the `pwd` command, which stands for *print working directory*.\n",
    "\n",
    "```{code-block} bash\n",
    "pwd\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: note, dropdown\n",
    "```\n",
    "/Users/gjeunen/Documents/work/lectures/HKU2023/tutorial\n",
    "```\n",
    "````\n",
    "\n",
    "We can also list all the documents and subfolders in our working directory using the `ls` command. Additionally, we will specify the `-ltr` parameters to the `ls` command. The `l` parameter is providing a list with long format that includes the permissions, which will come in handy when we are creating our scripts (more on that later). The `t` parameter is sorting the output by time and date, while the `r` parameter is printing the list in reverse order, i.e, the newest files will be printed as the latest in the list.\n",
    "\n",
    "```{code-block} bash\n",
    "ls -ltr\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: note, dropdown\n",
    "```\n",
    "-rwxrwxrwx@ 1 gjeunen  staff  240602642  7 Feb  2023 newDBcrabsCOIsintax.fasta\n",
    "-rw-r--r--@ 1 gjeunen  staff      11958  6 Oct 13:50 metadata-COI-selected-updated.txt\n",
    "-rw-r--r--  1 gjeunen  staff  568074346  6 Oct 13:54 HKeDNAworkshop2023.zip\n",
    "```\n",
    "````\n",
    "\n",
    "The output of the `ls` command provides us with a list of the documents in our starting folder, which are the documents we need to complete the tutorial. More on this in the next section.\n",
    "\n",
    "Before we go more in detail about the starting files, let us first set up the abovementioned folder structure. While you could complete this task within the folder by generating \"new folder\", we can use the `mkdir` command in the terminal to set up all the folders and subfolders in one go. Additionally, we will use the `-p` parameter to automatically make any necessary parent directories that might not yet exist.\n",
    "\n",
    "```{code-block} bash\n",
    "mkdir -p sequenceData/0-metadata sequenceData/1-scripts sequenceData/2-raw sequenceData/3-fastqc sequenceData/4-demux sequenceData/5-filter sequenceData/6-quality sequenceData/7-refdb sequenceData/8-final\n",
    "```\n",
    "\n",
    "When we now use the `ls -ltr` command again, we see that the directory **sequenceData** has been added to the output list.\n",
    "\n",
    "```{code-block} bash\n",
    "ls -ltr\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: note, dropdown\n",
    "```\n",
    "-rwxrwxrwx@  1 gjeunen  staff  240602642  7 Feb  2023 newDBcrabsCOIsintax.fasta\n",
    "-rw-r--r--@  1 gjeunen  staff      11958  6 Oct 13:50 metadata-COI-selected-updated.txt\n",
    "-rw-r--r--   1 gjeunen  staff  568074346  6 Oct 13:54 HKeDNAworkshop2023.zip\n",
    "drwxr-xr-x  11 gjeunen  staff        352  6 Oct 20:11 sequenceData\n",
    "```\n",
    "````\n",
    "\n",
    "Note that the `sequenceData` directory has appeared last in the printed list due to the `r` parameter we passed to the `ls` command, as it was the most recently created. When we'll be going through our bioinformatic pipeline and create multiple files at various stages, the `r` parameter will make it easy to see which files were recently created, as they will appear at the bottom and will avoid us having to scroll through the list of files.\n",
    "\n",
    "Also note the `d` at the beginning of the line of `sequenceData`. These are the permissions that the `l` parameter that we passed to the `ls` command has given us access to. Several other letters are shown for `sequenceData` and other files as well. Below is a list of what they mean:\n",
    "\n",
    "1. `d`: directory - indicate that this is in fact a folder or directory, rather than a file.\n",
    "2. `r`: read - specifies that we have access to read the file.\n",
    "3. `w`: write - specifies that we can write to the file.\n",
    "4. `x`: execute - specifies that we have permission to execute the file.\n",
    "5. `@`: extended - a novel symbol for MacOS indicating that the file has extended attributes.\n",
    "\n",
    "## 2. Starting files\n",
    "\n",
    "For the bioinformatic and statistical analysis, we need several starting files, including a zipped file containing all the sequencing data (**HKeDNAworkshop2023.zip**), a sample metadata file for the statistical analysis (**metadata-COI-selected-updated.txt**), and our reference database for the taxonomy assignment (**newDBcrabsCOIintax.fasta**).\n",
    "\n",
    "```{admonition} Starting files\n",
    ":class: important\n",
    "1. HKeDNAworkshop2023.zip\n",
    "2. metadata-COI-selected-updated.txt\n",
    "3. newDBcrabsCOIsintax.fasta\n",
    "```\n",
    "\n",
    "With the folder structure set up, let's move the starting files to their respective subfolders using the `cp` command.\n",
    "\n",
    "```{code-block} bash\n",
    "cp HKeDNAworkshop2023.zip sequenceData/2-raw\n",
    "cp metadata-COI-selected-updated.txt sequenceData/0-metadata\n",
    "cp newDBcrabsCOIsintax.fasta sequenceData/7-refdb\n",
    "```\n",
    "\n",
    "Again, we can check if the command executed as expected by listing the files within the subfolder **2-raw** using the `ls -ltr` command. Note that we do not need to first use the `cd` command to move to the directory for which we would like to list all of the files, but that we can specify which directory we would like to list by referring to it after the `ls -ltr` command.\n",
    "\n",
    "```{code-block} bash\n",
    "ls -ltr sequenceData/2-raw\n",
    "```\n",
    "\n",
    "````{admonition} Output\n",
    ":class: note, dropdown\n",
    "```\n",
    "-rw-r--r--  1 gjeunen  staff  568074346  6 Oct 20:28 HKeDNAworkshop2023.zip\n",
    "```\n",
    "````\n",
    "\n",
    "````{admonition} Exercise 1\n",
    ":class: hint\n",
    "\n",
    "Moving files around using the Terminal can also be accomplished through the `mv` command. What is the difference between `mv` and `cp`? Why is it better to use `cp`?\n",
    "\n",
    "```{admonition} Answer 1\n",
    ":class: title, dropdown\n",
    "The `mv` command will move the files from one folder to another. Hence, the files will be removed from the initial directory. If something goes wrong during the process or during the bioinformatic pipeline, such actions cannot be reverted back and data will be lost. \n",
    "\n",
    "With the current setup, if anything will go wrong during our pipeline, we can simply remove everything within the **sequenceData** directory using the `rm -r sequenceData*` command and start over, since we still have all the information and files within our initial starting folder.\n",
    "```\n",
    "````\n",
    "\n",
    "**Now that we have everything set up, we can get started with the bioinformatic processing of our data!**"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "md:myst",
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": 0.13,
    "jupytext_version": "1.11.5"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "source_map": [
   13
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}