

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Statistical analysis &#8212; eDNA workshop HKU</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'statisticalanalysis';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Phylogenetic tree" href="phylogenetictree.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo3.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo3.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    <no title>
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinformaticanalysis.html">Bioinformatic analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxonomyassignment.html">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datapreprocessing.html">Data pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogenetictree.html">Phylogenetic tree</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fstatisticalanalysis.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button"
   title="Open an issue"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/statisticalanalysis.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Statistical analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-parsing-data">2. Reading/Parsing data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-stats">3. Summary stats</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-exploration">4. Data exploration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#phyloseq-functions">4.1 Phyloseq functions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#bar-plots">4.1.1 Bar plots</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#heat-maps">4.1.2 Heat maps</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#richness">4.1.3 Richness</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-diversity">5. Alpha diversity</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#species-accumulation-curves">5.1 Species accumulation curves</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-diversity-comparison">5.2 Alpha diversity comparison</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#estimate-replication">5.3 Estimate replication</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#phylodiversity">5.4 Phylodiversity</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beta-diversity">6. Beta diversity</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ordination">6.1 Ordination</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#nmds">6.1.1 NMDS</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pcoa">6.1.2 PCoA</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#permanova">6.2 PERMANOVA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#permdisp">6.3 PERMDISP</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#indicator-species">6.4 Indicator species</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="statistical-analysis">
<h1>Statistical analysis<a class="headerlink" href="#statistical-analysis" title="Permalink to this heading">#</a></h1>
<section id="introduction">
<h2>1. Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">#</a></h2>
<p>During this section, we will cover some common statistical analyses that are conducted in eDNA metabarcoding research. Such methods mainly focus on alpha and beta diversity comparisons between sites/treatments.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Before we start with the statistical analysis, it is essential to keep an overview of the experimental design. Basically, we want to determine if we can observe differences in the biological community between sites that are highly impacted by anthropogenic stressors and sites that are less impacted by anthropogenic stressors. To test this, 3 ARMS (Autonomous Reef Monitoring Structures) were placed at each site, including 2 sites identified as highly impacted and two sites identified with low impact. After deployment, the biological community was subsetted by size, including 100 µm and 500 µm.</p>
<p>All of this means that we have 2 DNA samples per ARMS, 3 ARMS per site, and 4 sites = 24 samples.</p>
<p>For the statistical comparison, we could test differences between high and low impacted sites for each size fraction of the biological community separately, as well as a combined approach of both size fractions. During this tutorial, we will only cover the analysis of the separate size fractions to keep things simple.</p>
</div>
<figure class="align-default" id="experimental-design-2">
<img alt="_images/experimental_design.png" src="_images/experimental_design.png" />
<figcaption>
<p><span class="caption-number">Fig. 51 </span><span class="caption-text">: The experimental design of the data we will be analysing during the next few days. Note that this is the information of the full experiment. Please read the section above to provide an overview of what subset will be analysed during this workshop.</span><a class="headerlink" href="#experimental-design-2" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="reading-parsing-data">
<h2>2. Reading/Parsing data<a class="headerlink" href="#reading-parsing-data" title="Permalink to this heading">#</a></h2>
<p>Similarly to the previous section, most code will be run in R, unless specified otherwise. Therefore, let us open a new R script and save it as <strong>sequenceData/1-scripts/statisticalAnalysis.R</strong>. We can copy-paste the R code below into this script to run all the code. We can set up the R environment by specifying the working directory and loading all the necessary libraries.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">#########################</span>
<span class="c1"># PREPARE R ENVIRONMENT #</span>
<span class="c1">#########################</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Biostrings</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">phyloseq</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">vegan</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">iNEXT.3D</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">microbiome</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">scales</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">readxl</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ampvis2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">car</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">FSA</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">agricolae</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">speedyseq</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">BiodiversityR</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">indicspecies</span><span class="p">)</span>

<span class="c1"># set working directory</span>
<span class="nf">setwd</span><span class="p">(</span><span class="s">&#39;/Users/gjeunen/Documents/work/lectures/HKU2023/tutorial/sequenceData/8-final&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>We begin by reading the files into R, including the filtered frequency table <strong>sequenceData/8-fianl/zotutableFiltered.txt</strong>, the updated metadata file <strong>sequenceData/0-metadata/sampleMetadataFiltered.txt</strong>, the filtered ZOTU sequence file <strong>sequenceData/8-final/zotusFiltered.fasta</strong>, and the updated taxonomy table <strong>sequenceData/8-final/taxonomyFiltered.txt</strong>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">##################</span>
<span class="c1"># READ DATA IN R #</span>
<span class="c1">##################</span>
<span class="n">metaData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="s">&#39;../0-metadata/sampleMetadataFiltered.txt&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">freqTable</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="s">&#39;zotutableFiltered.txt&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">taxonomyTable</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="s">&#39;taxonomyFiltered.txt&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">sequenceTable</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readDNAStringSet</span><span class="p">(</span><span class="s">&#39;zotusFiltered.fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Once the data is read into R, we can import the dataframes into phyloseq as a single phyloseq object.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. import dataframes into phyloseq</span>
<span class="n">OTU</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">otu_table</span><span class="p">(</span><span class="n">freqTable</span><span class="p">,</span><span class="w"> </span><span class="n">taxa_are_rows</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">TAX</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">tax_table</span><span class="p">(</span><span class="nf">as.matrix</span><span class="p">(</span><span class="n">taxonomyTable</span><span class="p">))</span>
<span class="n">META</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sample_data</span><span class="p">(</span><span class="n">metaData</span><span class="p">)</span>
<span class="n">physeq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">merge_phyloseq</span><span class="p">(</span><span class="n">OTU</span><span class="p">,</span><span class="w"> </span><span class="n">TAX</span><span class="p">,</span><span class="w"> </span><span class="n">META</span><span class="p">,</span><span class="w"> </span><span class="n">sequenceTable</span><span class="p">)</span>
<span class="n">physeq</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">physeq</span>
<span class="n">phyloseq</span><span class="o">-</span><span class="k">class</span> <span class="nc">experiment</span><span class="o">-</span><span class="n">level</span> <span class="nb">object</span>
<span class="n">otu_table</span><span class="p">()</span>   <span class="n">OTU</span> <span class="n">Table</span><span class="p">:</span>         <span class="p">[</span> <span class="mi">865</span> <span class="n">taxa</span> <span class="ow">and</span> <span class="mi">24</span> <span class="n">samples</span> <span class="p">]</span>
<span class="n">sample_data</span><span class="p">()</span> <span class="n">Sample</span> <span class="n">Data</span><span class="p">:</span>       <span class="p">[</span> <span class="mi">24</span> <span class="n">samples</span> <span class="n">by</span> <span class="mi">32</span> <span class="n">sample</span> <span class="n">variables</span> <span class="p">]</span>
<span class="n">tax_table</span><span class="p">()</span>   <span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>    <span class="p">[</span> <span class="mi">865</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span> <span class="p">]</span>
<span class="n">refseq</span><span class="p">()</span>      <span class="n">DNAStringSet</span><span class="p">:</span>      <span class="p">[</span> <span class="mi">865</span> <span class="n">reference</span> <span class="n">sequences</span> <span class="p">]</span>
</pre></div>
</div>
</div>
</section>
<section id="summary-stats">
<h2>3. Summary stats<a class="headerlink" href="#summary-stats" title="Permalink to this heading">#</a></h2>
<p>One of the first things we will look at are some summary statistics on read number prior and post quality filtering, as well as average read number per sample, number of taxa, most abundant and most frequently detected taxa, read count in negative controls, taxa observed in negative controls, sample drop out, etc. These numbers are usually reported in the first paragraph of the results section in eDNA metabarcoding publications.</p>
<p>Most of these numbers have already been reported on during the bioinformatic and data processing session. For example, we know that after the bioinformatic processing of the raw sequencing data, we ended up with <code class="docutils literal notranslate"><span class="pre">2,195,816</span> <span class="pre">reads</span></code> across <code class="docutils literal notranslate"><span class="pre">1,345</span> <span class="pre">ZOTU</span></code> sequences. Additionally, we know that during data pre-processing, we filtered the <code class="docutils literal notranslate"><span class="pre">total</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">ZOTU</span> <span class="pre">sequences</span> <span class="pre">to</span> <span class="pre">865</span></code> and after contaminant and abundance filtering, we ended up with a <code class="docutils literal notranslate"><span class="pre">total</span> <span class="pre">of</span> <span class="pre">1,768,932</span> <span class="pre">reads</span></code>.</p>
<p>Information about the negative controls has also been provided during the data pre-processing session. For example, we know that there were 99 ZOTUs with a positive detection in the negative controls, including the top 9 most-abundant ZOTU sequences in our data set. Of those 99 ZOTUs with a positive detection in the negative controls, <code class="docutils literal notranslate"><span class="pre">total</span> <span class="pre">read</span> <span class="pre">count</span> <span class="pre">reached</span> <span class="pre">287</span> <span class="pre">(0.01%)</span></code> and most abundant read count for a single detection in the negative controls was 12 reads and 23 when summed across all negative control samples. This highest number of reads was observed for ZOTU 1, which is assigned to the class <code class="docutils literal notranslate"><span class="pre">Hexanauplia</span></code>.</p>
<p>Another aspect that is frequently mentioned is information about the taxonomic assignment. For the tutorial data, 58 ZOTUs did not achieve a taxonomic ID through BLAST (the taxonomic ID we used during the data pre-processing steps). Removal of these signals resulted in a reduction of 14,735 reads from the final data set.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>During the tutorial, we have not gone over the quality of the taxonomic assignment. Some studies will filter out sequences with low-quality taxonomic ID scores or set arbitrary thresholds at which level the taxonomic ID should be set. Please keep in mind that this is something you can and should do for your data set as well if you would like to focus on a taxonomic-dependent analysis.</p>
</div>
<p>Finally, we can report that we used the LULU algorithm to identify artefact sequences. Through this approach, 369 child or artefact sequences were merged with their parent.</p>
<p>To get some further summary statistics on the final frequency table (such as minimum and maximum read count, average read count, ZOTUs summing to 0, etc.) we can make use of multiple functions within the <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> and <code class="docutils literal notranslate"><span class="pre">microbiome</span></code> R packages.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">#################</span>
<span class="c1"># SUMMARY STATS #</span>
<span class="c1">#################</span>
<span class="nf">summarize_phyloseq</span><span class="p">(</span><span class="n">physeq</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="nf">readcount</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span>
<span class="nf">sd</span><span class="p">(</span><span class="nf">readcount</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span>
<span class="n">se</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sd</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
<span class="nf">se</span><span class="p">(</span><span class="nf">readcount</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[[</span><span class="mi">1</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;1] Min. number of reads = 35142&quot;</span>

<span class="p">[[</span><span class="mi">2</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;2] Max. number of reads = 126301&quot;</span>

<span class="p">[[</span><span class="mi">3</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;3] Total number of reads = 1768932&quot;</span>

<span class="p">[[</span><span class="mi">4</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;4] Average number of reads = 73705.5&quot;</span>

<span class="p">[[</span><span class="mi">5</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;5] Median number of reads = 69433.5&quot;</span>

<span class="p">[[</span><span class="mi">6</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;7] Sparsity = 0.785452793834297&quot;</span>

<span class="p">[[</span><span class="mi">7</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;6] Any OTU sum to 1 or less? NO&quot;</span>

<span class="p">[[</span><span class="mi">8</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;8] Number of singletons = 0&quot;</span>

<span class="p">[[</span><span class="mi">9</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;9] Percent of OTUs that are singletons </span><span class="se">\n</span><span class="s2">        (i.e. exactly one read detected across all samples)0&quot;</span>

<span class="p">[[</span><span class="mi">10</span><span class="p">]]</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;10] Number of sample variables are: 32&quot;</span>

<span class="p">[[</span><span class="mi">11</span><span class="p">]]</span>
 <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;sample_name&quot;</span>                     <span class="s2">&quot;ARMS_ID&quot;</span>                        
 <span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="s2">&quot;SAMPLE&quot;</span>                          <span class="s2">&quot;sampleSite&quot;</span>                     
 <span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="s2">&quot;TargetLoci&quot;</span>                      <span class="s2">&quot;impact&quot;</span>                         
 <span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="s2">&quot;replicateLabel&quot;</span>                  <span class="s2">&quot;island&quot;</span>                         
 <span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="s2">&quot;siteDetails&quot;</span>                     <span class="s2">&quot;StationID&quot;</span>                      
<span class="p">[</span><span class="mi">11</span><span class="p">]</span> <span class="s2">&quot;habitat&quot;</span>                         <span class="s2">&quot;PercentCoralCover&quot;</span>              
<span class="p">[</span><span class="mi">13</span><span class="p">]</span> <span class="s2">&quot;locality&quot;</span>                        <span class="s2">&quot;decimalLongitude&quot;</span>               
<span class="p">[</span><span class="mi">15</span><span class="p">]</span> <span class="s2">&quot;decimalLatitude&quot;</span>                 <span class="s2">&quot;actualDeploymentDate&quot;</span>           
<span class="p">[</span><span class="mi">17</span><span class="p">]</span> <span class="s2">&quot;actualRecoveryDate&quot;</span>              <span class="s2">&quot;totalmonth&quot;</span>                     
<span class="p">[</span><span class="mi">19</span><span class="p">]</span> <span class="s2">&quot;Month_Ret&quot;</span>                       <span class="s2">&quot;Ret_ID&quot;</span>                         
<span class="p">[</span><span class="mi">21</span><span class="p">]</span> <span class="s2">&quot;Temperature...C.&quot;</span>                <span class="s2">&quot;Salinity..psu.&quot;</span>                 
<span class="p">[</span><span class="mi">23</span><span class="p">]</span> <span class="s2">&quot;Dissolved.Oxygen..mg.L.&quot;</span>         <span class="s2">&quot;Turbidity..NTU.&quot;</span>                
<span class="p">[</span><span class="mi">25</span><span class="p">]</span> <span class="s2">&quot;Total.Inorganic.Nitrogen..mg.L.&quot;</span> <span class="s2">&quot;Total.Nitrogen..mg.L.&quot;</span>          
<span class="p">[</span><span class="mi">27</span><span class="p">]</span> <span class="s2">&quot;Chlorophyll.a..μg.L.&quot;</span>            <span class="s2">&quot;Faecal.Coliforms..cfu.100mL.&quot;</span>   
<span class="p">[</span><span class="mi">29</span><span class="p">]</span> <span class="s2">&quot;station.gov&quot;</span>                     <span class="s2">&quot;EPD_date_from&quot;</span>                  
<span class="p">[</span><span class="mi">31</span><span class="p">]</span> <span class="s2">&quot;EPD_date_to&quot;</span>                     <span class="s2">&quot;comments&quot;</span>                       
</pre></div>
</div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">summary</span><span class="p">(</span><span class="n">readcount</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span>
   <span class="n">Min</span><span class="o">.</span> <span class="mi">1</span><span class="n">st</span> <span class="n">Qu</span><span class="o">.</span>  <span class="n">Median</span>    <span class="n">Mean</span> <span class="mi">3</span><span class="n">rd</span> <span class="n">Qu</span><span class="o">.</span>    <span class="n">Max</span><span class="o">.</span> 
  <span class="mi">35142</span>   <span class="mi">54186</span>   <span class="mi">69434</span>   <span class="mi">73706</span>   <span class="mi">89468</span>  <span class="mi">126301</span> 

<span class="o">&gt;</span> <span class="n">sd</span><span class="p">(</span><span class="n">readcount</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">24364.47</span>

<span class="o">&gt;</span> <span class="n">se</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">sd</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="n">sqrt</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
<span class="o">&gt;</span> <span class="n">se</span><span class="p">(</span><span class="n">readcount</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">4973.378</span>                   
</pre></div>
</div>
</div>
<p>Besides read count per sample, we can also investigate the most abundant (with regards to read count) and most frequently detected (occurrence count) taxa through some simple <code class="docutils literal notranslate"><span class="pre">for</span> <span class="pre">loops</span></code> in R.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. list 10 most abundant (read count) taxa</span>
<span class="nf">for </span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">top_taxa</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">taxa</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tax_table</span><span class="p">(</span><span class="n">physeq</span><span class="p">)[</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="w">  </span><span class="n">count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">physeq</span><span class="p">)[</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="p">])</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">count</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">physeq</span><span class="p">))</span><span class="o">*</span><span class="m">100</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># 1. list 10 most frequently detected (occurrences) taxa</span>
<span class="nf">for </span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">top_taxa</span><span class="p">(</span><span class="n">microbiome</span><span class="o">::</span><span class="nf">transform</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;pa&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">taxa</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tax_table</span><span class="p">(</span><span class="n">physeq</span><span class="p">)[</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="w">  </span><span class="n">count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">microbiome</span><span class="o">::</span><span class="nf">transform</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;pa&#39;</span><span class="p">))[</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="p">])</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">count</span><span class="o">/</span><span class="nf">ncol</span><span class="p">(</span><span class="n">freqTable</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>         <span class="nc">order</span>        <span class="n">family</span>       <span class="n">genus</span>            <span class="n">species</span>          <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.1</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Hexanauplia&quot;</span> <span class="s2">&quot;Cyclopoida&quot;</span> <span class="s2">&quot;Cyclopidae&quot;</span> <span class="s2">&quot;Cyclopidae_nan&quot;</span> <span class="s2">&quot;Cyclopidae_sp.&quot;</span> <span class="s2">&quot; 75.421&quot;</span> <span class="s2">&quot; 95&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">222411</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">12.57318</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>       <span class="nc">order</span>       <span class="n">family</span>           <span class="n">genus</span>       <span class="n">species</span>                 <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.2</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Arachnida&quot;</span> <span class="s2">&quot;Opiliones&quot;</span> <span class="s2">&quot;Nemastomatidae&quot;</span> <span class="s2">&quot;Mitostoma&quot;</span> <span class="s2">&quot;Mitostoma_chrysomelas&quot;</span> <span class="s2">&quot; 76.596&quot;</span> <span class="s2">&quot; 89&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">168289</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">9.513594</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>          <span class="nc">order</span>      <span class="n">family</span>      <span class="n">genus</span>     <span class="n">species</span>  <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.3</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Malacostraca&quot;</span> <span class="s2">&quot;Decapoda&quot;</span> <span class="s2">&quot;Alpheidae&quot;</span> <span class="s2">&quot;Athanas&quot;</span> <span class="s2">&quot;parvus&quot;</span> <span class="s2">&quot; 99.681&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">137326</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">7.763215</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>          <span class="n">family</span>       <span class="n">genus</span>         <span class="n">species</span>                 <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.7</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Phyllodocida&quot;</span> <span class="s2">&quot;Nereididae&quot;</span> <span class="s2">&quot;Platynereis&quot;</span> <span class="s2">&quot;Platynereis_dumerilii&quot;</span> <span class="s2">&quot; 84.158&quot;</span> <span class="s2">&quot; 96&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">65860</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">3.72315</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>          <span class="n">family</span>       <span class="n">genus</span>         <span class="n">species</span> <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.8</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Phyllodocida&quot;</span> <span class="s2">&quot;Nereididae&quot;</span> <span class="s2">&quot;Platynereis&quot;</span> <span class="s2">&quot;HK01&quot;</span>  <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">57249</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">3.23636</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>          <span class="nc">order</span>      <span class="n">family</span>             <span class="n">genus</span>            <span class="n">species</span>  <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.10</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Malacostraca&quot;</span> <span class="s2">&quot;Decapoda&quot;</span> <span class="s2">&quot;Rhynchocinetidae&quot;</span> <span class="s2">&quot;Rhynchocinetes&quot;</span> <span class="s2">&quot;brucei&quot;</span> <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">47045</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">2.659514</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>          <span class="nc">order</span>      <span class="n">family</span>       <span class="n">genus</span>            <span class="n">species</span> <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.11</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Malacostraca&quot;</span> <span class="s2">&quot;Decapoda&quot;</span> <span class="s2">&quot;Pilumnidae&quot;</span> <span class="s2">&quot;Pilumnidae_nan&quot;</span> <span class="s2">&quot;HK03&quot;</span>  <span class="s2">&quot; 79.769&quot;</span> <span class="s2">&quot; 55&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">45386</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">2.565729</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>          <span class="n">family</span>       <span class="n">genus</span>         <span class="n">species</span> <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.9</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Phyllodocida&quot;</span> <span class="s2">&quot;Nereididae&quot;</span> <span class="s2">&quot;Platynereis&quot;</span> <span class="s2">&quot;HK01&quot;</span>  <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">45334</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">2.562789</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>         <span class="nc">order</span>         <span class="n">family</span>     <span class="n">genus</span>         <span class="n">species</span>                    <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.16</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Chordata&quot;</span> <span class="s2">&quot;Actinopteri&quot;</span> <span class="s2">&quot;Gobiiformes&quot;</span> <span class="s2">&quot;Gobiidae&quot;</span> <span class="s2">&quot;Tridentiger&quot;</span> <span class="s2">&quot;Tridentiger_nudicervicus&quot;</span> <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot; 97&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">32197</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">1.820138</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>     <span class="nc">order</span>     <span class="n">family</span>     <span class="n">genus</span>     <span class="n">species</span>             <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.13</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Insecta&quot;</span> <span class="s2">&quot;Diptera&quot;</span> <span class="s2">&quot;Muscidae&quot;</span> <span class="s2">&quot;Muscina&quot;</span> <span class="s2">&quot;Muscina_pascuorum&quot;</span> <span class="s2">&quot; 80.412&quot;</span> <span class="s2">&quot; 93&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">30681</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">1.734436</span>                 
</pre></div>
</div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>      <span class="n">family</span>      <span class="n">genus</span>          <span class="n">species</span>               <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.57</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Spionida&quot;</span> <span class="s2">&quot;Spionidae&quot;</span> <span class="s2">&quot;Boccardiella&quot;</span> <span class="s2">&quot;Boccardiella_hamata&quot;</span> <span class="s2">&quot; 85.761&quot;</span> <span class="s2">&quot; 99&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">24</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">100</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
         <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>      <span class="n">family</span>      <span class="n">genus</span>          <span class="n">species</span>               <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.122</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Spionida&quot;</span> <span class="s2">&quot;Spionidae&quot;</span> <span class="s2">&quot;Boccardiella&quot;</span> <span class="s2">&quot;Boccardiella_hamata&quot;</span> <span class="s2">&quot; 84.091&quot;</span> <span class="s2">&quot; 98&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">23</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">95.83333</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>          <span class="nc">order</span>      <span class="n">family</span>      <span class="n">genus</span>       <span class="n">species</span> <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.14</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Malacostraca&quot;</span> <span class="s2">&quot;Decapoda&quot;</span> <span class="s2">&quot;Xanthidae&quot;</span> <span class="s2">&quot;Leptodius&quot;</span> <span class="s2">&quot;HK01&quot;</span>  <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">19</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">79.16667</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
         <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>         <span class="nc">order</span>          <span class="n">family</span>      <span class="n">genus</span>     <span class="n">species</span>            <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.155</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Thecostraca&quot;</span> <span class="s2">&quot;Balanomorpha&quot;</span> <span class="s2">&quot;Balanidae&quot;</span> <span class="s2">&quot;Balanus&quot;</span> <span class="s2">&quot;Balanus_trigonus&quot;</span> <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">19</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">79.16667</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
       <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>          <span class="n">family</span>       <span class="n">genus</span>         <span class="n">species</span> <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.8</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Phyllodocida&quot;</span> <span class="s2">&quot;Nereididae&quot;</span> <span class="s2">&quot;Platynereis&quot;</span> <span class="s2">&quot;HK01&quot;</span>  <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">19</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">79.16667</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>       <span class="nc">order</span>         <span class="n">family</span>          <span class="n">genus</span>       <span class="n">species</span>              <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.21</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Diplopoda&quot;</span> <span class="s2">&quot;Polydesmida&quot;</span> <span class="s2">&quot;Xystodesmidae&quot;</span> <span class="s2">&quot;Brachoria&quot;</span> <span class="s2">&quot;Brachoria_laminata&quot;</span> <span class="s2">&quot; 76.654&quot;</span> <span class="s2">&quot; 81&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">18</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">75</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>          <span class="n">family</span>       <span class="n">genus</span>      <span class="n">species</span>        <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.79</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Phyllodocida&quot;</span> <span class="s2">&quot;Hesionidae&quot;</span> <span class="s2">&quot;Syllidia&quot;</span> <span class="s2">&quot;Syllidia_sp.&quot;</span> <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">18</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">75</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
         <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>         <span class="nc">order</span>          <span class="n">family</span>      <span class="n">genus</span>     <span class="n">species</span>            <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.221</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Thecostraca&quot;</span> <span class="s2">&quot;Balanomorpha&quot;</span> <span class="s2">&quot;Balanidae&quot;</span> <span class="s2">&quot;Balanus&quot;</span> <span class="s2">&quot;Balanus_trigonus&quot;</span> <span class="s2">&quot; 99.681&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">17</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">70.83333</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
         <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>        <span class="nc">order</span>      <span class="n">family</span>      <span class="n">genus</span>      <span class="n">species</span>             <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.388</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Annelida&quot;</span> <span class="s2">&quot;Polychaeta&quot;</span> <span class="s2">&quot;Spionida&quot;</span> <span class="s2">&quot;Spionidae&quot;</span> <span class="s2">&quot;Polydora&quot;</span> <span class="s2">&quot;Polydora_haswelli&quot;</span> <span class="s2">&quot; 97.444&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">17</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">70.83333</span>
<span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>     <span class="p">[</span><span class="mi">1</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span><span class="p">]:</span>
        <span class="n">kingdom</span>     <span class="n">phylum</span>       <span class="k">class</span>         <span class="nc">order</span>          <span class="n">family</span>      <span class="n">genus</span>          <span class="n">species</span>                    <span class="n">pident</span>    <span class="n">qcov</span> 
<span class="n">zotu</span><span class="mf">.52</span> <span class="s2">&quot;Eukaryota&quot;</span> <span class="s2">&quot;Arthropoda&quot;</span> <span class="s2">&quot;Thecostraca&quot;</span> <span class="s2">&quot;Balanomorpha&quot;</span> <span class="s2">&quot;Balanidae&quot;</span> <span class="s2">&quot;Amphibalanus&quot;</span> <span class="s2">&quot;Amphibalanus_reticulatus&quot;</span> <span class="s2">&quot;100.000&quot;</span> <span class="s2">&quot;100&quot;</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">17</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mf">70.83333</span>                
</pre></div>
</div>
</div>
</section>
<section id="data-exploration">
<h2>4. Data exploration<a class="headerlink" href="#data-exploration" title="Permalink to this heading">#</a></h2>
<p>We already covered several data exploration techniques in the previous sections. For example, we created a phylogenetic tree to investigate differences in detection frequency of Cnidarian ZOTUs between low and high impact sites in the <strong>Phylogenetic tree</strong> section. Additionally, we investigated if sufficient sequencing depth was achieved for samples through rarefaction curves in the <strong>Data pre-processing</strong> section.</p>
<section id="phyloseq-functions">
<h3>4.1 Phyloseq functions<a class="headerlink" href="#phyloseq-functions" title="Permalink to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> R package itself also provides some basic graphs to explore your metabarcoding data even further, including bar graphs (<code class="docutils literal notranslate"><span class="pre">plot_bar</span></code>), heat maps (<code class="docutils literal notranslate"><span class="pre">plot_heatmap</span></code>), richness (<code class="docutils literal notranslate"><span class="pre">plot_richness</span></code>), basic tree figures (<code class="docutils literal notranslate"><span class="pre">plot_tree</span></code>), and ordination graphs (<code class="docutils literal notranslate"><span class="pre">plot_ordination</span></code>; more on this in section <strong>6. Beta diversity</strong>). Below are some examples of these plotting functions. Note that you can expand on this by subsetting the data in a meaningful manner. More information about the <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> R package and some excellent tutorials can be found on the <a class="reference external" href="https://joey711.github.io/phyloseq/index.html">phyloseq webpage</a>.</p>
<section id="bar-plots">
<h4>4.1.1 Bar plots<a class="headerlink" href="#bar-plots" title="Permalink to this heading">#</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. create bar plot of samples coloured by phylum, 2. group by impact, 3. facet by location, 4. use ggplot to remove ZOTU lines</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;phylum&quot;</span><span class="p">)</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;impact&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;phylum&#39;</span><span class="p">)</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;SAMPLE&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;phylum&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">facet_grid</span><span class="o">=~</span><span class="n">impact</span><span class="p">)</span>
<span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plot_bar</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;SAMPLE&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;phylum&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">facet_grid</span><span class="o">=~</span><span class="n">impact</span><span class="p">)</span>
<span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">phylum</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="o">=</span><span class="n">phylum</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="o">=</span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="o">=</span><span class="s">&quot;stack&quot;</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="simple-phyloseq-bar-plot">
<img alt="_images/simplebarplot1.png" src="_images/simplebarplot1.png" />
<figcaption>
<p><span class="caption-number">Fig. 52 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> bar plot of samples whereby abundance of ZOTU sequences are coloured by phylum ID.</span><a class="headerlink" href="#simple-phyloseq-bar-plot" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="simple-phyloseq-bar-plot-2">
<img alt="_images/simplebarplot2.png" src="_images/simplebarplot2.png" />
<figcaption>
<p><span class="caption-number">Fig. 53 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> bar plot of samples whereby abundance of ZOTU sequences are coloured by phylum ID and samples are merged based on impact metadata information.</span><a class="headerlink" href="#simple-phyloseq-bar-plot-2" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="simple-phyloseq-bar-plot-3">
<img alt="_images/simplebarplot3.png" src="_images/simplebarplot3.png" />
<figcaption>
<p><span class="caption-number">Fig. 54 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> bar plot of samples whereby abundance of ZOTU sequences are coloured by phylum ID, samples are merged based on size fraction and facetted by impact metadata information.</span><a class="headerlink" href="#simple-phyloseq-bar-plot-3" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="simple-phyloseq-bar-plot-4">
<img alt="_images/simplebarplot4.png" src="_images/simplebarplot4.png" />
<figcaption>
<p><span class="caption-number">Fig. 55 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> bar plot as above, but whereby <code class="docutils literal notranslate"><span class="pre">ggplot</span></code> is used to remove the black ZOTU lines.</span><a class="headerlink" href="#simple-phyloseq-bar-plot-4" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="heat-maps">
<h4>4.1.2 Heat maps<a class="headerlink" href="#heat-maps" title="Permalink to this heading">#</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. plot simple heatmap, 2. plot heatmap with relabeled x-axis labels and different colours</span>
<span class="nf">plot_heatmap</span><span class="p">(</span><span class="n">physeq</span><span class="p">)</span>
<span class="nf">plot_heatmap</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">sample.label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;sampleSite&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">low</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;#66CCFF&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">high</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;#000033&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;white&#39;</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="simple-phyloseq-heat-map-1">
<img alt="_images/simpleheatmap1.png" src="_images/simpleheatmap1.png" />
<figcaption>
<p><span class="caption-number">Fig. 56 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> heat map.</span><a class="headerlink" href="#simple-phyloseq-heat-map-1" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="simple-phyloseq-heat-map-2">
<img alt="_images/simpleheatmap2.png" src="_images/simpleheatmap2.png" />
<figcaption>
<p><span class="caption-number">Fig. 57 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> heat map whereby x-axis labels are altered and heat map colours are changed.</span><a class="headerlink" href="#simple-phyloseq-heat-map-2" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="richness">
<h4>4.1.3 Richness<a class="headerlink" href="#richness" title="Permalink to this heading">#</a></h4>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. plot richness</span>
<span class="nf">plot_richness</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="s">&quot;impact&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="o">=</span><span class="s">&quot;SAMPLE&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">measures</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;Observed&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Chao1&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Shannon&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Simpson&#39;</span><span class="p">))</span>
</pre></div>
</div>
<figure class="align-default" id="simple-phyloseq-richness">
<img alt="_images/simplerichness1.png" src="_images/simplerichness1.png" />
<figcaption>
<p><span class="caption-number">Fig. 58 </span><span class="caption-text">: Simple <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> multiple richness measurements grouped by low and high impact and coloured by size fraction.</span><a class="headerlink" href="#simple-phyloseq-richness" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
</section>
</section>
<section id="alpha-diversity">
<h2>5. Alpha diversity<a class="headerlink" href="#alpha-diversity" title="Permalink to this heading">#</a></h2>
<p>In short, alpha diversity is the characterisation of diversity within a site. Researchers use this characterisation to quantify how diverse different systems are. Over the last century, a myriad of approaches and tools have been developed to perform such operations, each embedded within a statistical background. One important aspect, and where these various approaches can differ, is the differential weighing of abundant and rare species. The simplest approach, and most-frequently used in metabarcoding research, is <code class="docutils literal notranslate"><span class="pre">species</span> <span class="pre">richness</span></code>, which only considers the presence or absence of a species. However, OTUs in a system are rarely distributed evenly; thus using richness is rarely the best approach with which to reflect the diversity of a system. Other popular approaches that do account for the evenness of a system are the <code class="docutils literal notranslate"><span class="pre">Shannon</span> <span class="pre">index</span></code> and the <code class="docutils literal notranslate"><span class="pre">Simpson</span> <span class="pre">index</span></code>. However, it is important to note that both are not acutal measures of diversity. <code class="docutils literal notranslate"><span class="pre">Shannon</span></code> measures entropy and therefore yields the uncertainty in species identity of a randomly chosen species in the system, while <code class="docutils literal notranslate"><span class="pre">Simpson</span></code> provides the probability that two randomly chosen species actually belong to different species. Hence, both <code class="docutils literal notranslate"><span class="pre">Shannon</span></code> and <code class="docutils literal notranslate"><span class="pre">Simpson</span></code> are difficult to interpret and are not intuitive. For example, a doubling of the species in a system (10 compared to 20 species for example), does not yield double the value of Shannon or Simpson. Fortunately, these indices belong to a single statistical framework. Hence, we can rework these formulas and calculate the actual diversities using so-called Hill numbers (qD), which is a simple statistical framework allowing us to calculate various diversity metrics using a single system. Within the Hill number framework, q = 0 represents species richness, q = 1 represents the exponential of the Shannon index, and q = 2 represents the inverse of the Simpson index. For more information, we can have a look at the figure below. I also strongly suggest reading the excellent paper by <a class="reference external" href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.13014">Alberdi &amp; Gilbert, 2019</a>, which covers all of this in much more detail and in context of a metabarcoding approach. They also provide all the essential citations of the original statistical framework development.</p>
<figure class="align-default" id="hill-numbers-intro">
<img alt="_images/hillnumbersintro.png" src="_images/hillnumbersintro.png" />
<figcaption>
<p><span class="caption-number">Fig. 59 </span><span class="caption-text">: Introduction to the Hill number statistical framework. Copyright: <a class="reference external" href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.13014">Alberdi &amp; Gilbert, 2019</a>.</span><a class="headerlink" href="#hill-numbers-intro" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>For the alpha diversity analysis in this tutorial, we will be using the Hill number framework. However, since metabarcoding approaches do not hold true abundance information, or at least the correlation between eDNA signal strength and species abundance is very weak, we will need to approximate relative abundance measurements from our data set. We can do so by using a so-called <code class="docutils literal notranslate"><span class="pre">incidence-based</span></code> approach. This approach works by transforming the data set to presence-absence and combining multiple samples from the same site/treatment to obtain the relative abundance information.</p>
<figure class="align-default" id="incidence-based-data">
<img alt="_images/incidencebased.png" src="_images/incidencebased.png" />
<figcaption>
<p><span class="caption-number">Fig. 60 </span><span class="caption-text">: Introduction to incidence-based data transformation. Copyright: <a class="reference external" href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.13014">Alberdi &amp; Gilbert, 2019</a>.</span><a class="headerlink" href="#incidence-based-data" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<section id="species-accumulation-curves">
<h3>5.1 Species accumulation curves<a class="headerlink" href="#species-accumulation-curves" title="Permalink to this heading">#</a></h3>
<p>To expand upon rarefaction curves, whereby the plateauing of the curves indicate sufficient sequencing coverage was achieved for samples, we can take advantage of the fact that we have multiple ARMS (3) at each site. Transforming our frequency table to presence-absence data allows us to transform our data set to an incidence-raw format. When merging samples based on site location, we can form an incidence-frequency table to approximate a semi-quantitative data set. Through inter- and extrapolation calculations, we can draw and investigate alpha diversity accumulation curves (similar to species accumulation curves). Such an approach is implemented in the <code class="docutils literal notranslate"><span class="pre">iNEXT.3D</span></code> <a class="reference external" href="https://github.com/KaiHsiangHu/iNEXT.3D">R package</a>. These graphs will help us determine if sufficient sampling was conducted at each site, based on the plateauing of the curves, similar to the rarefaction curves we discussed in section <strong>Data pre-processing</strong>. Unlike the rarefaction curves we generated earlier, iNEXT.3D can extrapolate alpha diversity numbers beyond the number of samples collected. By default, the extrapolation value is set to double the number of samples. We will leave this default value (in our case 6 samples; double of the 3 ARMS per site), as increasing this value increases the error profile. We can also specify the different Hill number we want to investigate through the <code class="docutils literal notranslate"><span class="pre">q</span></code> parameter.</p>
<p>First, we need to subset our phyloseq object into two different objects, one containing the information about the 100 µm size fraction and one containing the information about the 500 µm size fraction.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">############</span>
<span class="c1"># iNEXT.3D #</span>
<span class="c1">############</span>
<span class="c1"># 1. subset data based on size fractions</span>
<span class="n">physeq.100</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">subset_samples</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">SAMPLE</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;COI_100&quot;</span><span class="p">)</span>
<span class="n">physeq.500</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">subset_samples</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="n">SAMPLE</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;COI_500&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, we need to convert our data format to the iNEXT.3D specific format of <code class="docutils literal notranslate"><span class="pre">incidence-frequency</span></code>. We can use the code below to achieve this reformatting of the dataframe. We need to run this twice, once for each subsetted dataframe.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. reformat data to iNEXT.3D standard for 100 size fraction</span>
<span class="n">categories.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="nf">sample_data</span><span class="p">(</span><span class="n">physeq.100</span><span class="p">)</span><span class="o">$</span><span class="n">sampleSite</span><span class="p">)</span>
<span class="n">split_physeq_list_100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">category</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">categories.100</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">sub_physeq.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset_samples</span><span class="p">(</span><span class="n">microbiome</span><span class="o">::</span><span class="nf">transform</span><span class="p">(</span><span class="n">physeq.100</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;pa&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">sampleSite</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">category</span><span class="p">)</span>
<span class="w">  </span><span class="n">split_physeq_list_100</span><span class="p">[[</span><span class="n">category</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">otu_table</span><span class="p">(</span><span class="n">sub_physeq.100</span><span class="p">)</span>
<span class="p">}</span><span class="w">  </span>
<span class="n">matrix_list.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="n">split_physeq_list_100</span><span class="p">,</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">otu_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;matrix&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">otu_table</span><span class="p">)</span>
<span class="p">})</span>
<span class="c1"># Create a list named &quot;data&quot; to store matrices</span>
<span class="n">matrix_list.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">())</span>
<span class="c1"># Convert each split phyloseq object into a matrix and store it under &quot;data&quot;</span>
<span class="nf">for </span><span class="p">(</span><span class="n">category</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">categories.100</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">otu_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">split_physeq_list_100</span><span class="p">[[</span><span class="n">category</span><span class="p">]]),</span><span class="w"> </span><span class="s">&quot;matrix&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">matrix_list.100</span><span class="p">[[</span><span class="s">&quot;data&quot;</span><span class="p">]][[</span><span class="n">category</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">otu_table</span>
<span class="p">}</span>

<span class="c1"># 1. reformat data to iNEXT.3D standard for 500 size fraction</span>
<span class="n">categories.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="nf">sample_data</span><span class="p">(</span><span class="n">physeq.500</span><span class="p">)</span><span class="o">$</span><span class="n">sampleSite</span><span class="p">)</span>
<span class="n">split_physeq_list_500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">category</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">categories.500</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">sub_physeq.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset_samples</span><span class="p">(</span><span class="n">microbiome</span><span class="o">::</span><span class="nf">transform</span><span class="p">(</span><span class="n">physeq.500</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;pa&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">sampleSite</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">category</span><span class="p">)</span>
<span class="w">  </span><span class="n">split_physeq_list_500</span><span class="p">[[</span><span class="n">category</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">otu_table</span><span class="p">(</span><span class="n">sub_physeq.500</span><span class="p">)</span>
<span class="p">}</span><span class="w">  </span>
<span class="n">matrix_list.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="n">split_physeq_list_500</span><span class="p">,</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">otu_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;matrix&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">otu_table</span><span class="p">)</span>
<span class="p">})</span>
<span class="c1"># Create a list named &quot;data&quot; to store matrices</span>
<span class="n">matrix_list.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">())</span>
<span class="c1"># Convert each split phyloseq object into a matrix and store it under &quot;data&quot;</span>
<span class="nf">for </span><span class="p">(</span><span class="n">category</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">categories.500</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">otu_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">split_physeq_list_500</span><span class="p">[[</span><span class="n">category</span><span class="p">]]),</span><span class="w"> </span><span class="s">&quot;matrix&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">matrix_list.500</span><span class="p">[[</span><span class="s">&quot;data&quot;</span><span class="p">]][[</span><span class="n">category</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">otu_table</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Once the reformatted dataframe according to iNEXT.3D specifications is created, we can run the <code class="docutils literal notranslate"><span class="pre">iNEXT3D</span></code> function to calculate the inter- and extrapolation values. Next, we can run the iNEXT.3D wrapper around ggplot <code class="docutils literal notranslate"><span class="pre">ggiNEXT3D</span></code> to plot the alpha diversity accumulation curves. Note the measurements are provided up to a value of 6, double the number of samples. Again, we need to run this twice, once for each subsetted data frame.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. run iNEXT3D for the 100 size fraction</span>
<span class="n">out.raw.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">iNEXT3D</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">matrix_list.100</span><span class="o">$</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;TD&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;incidence_raw&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">nboot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
<span class="nf">ggiNEXT3D</span><span class="p">(</span><span class="n">out.raw.100</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">facet.var</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Assemblage&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="n">Assemblage</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>

<span class="c1"># 1. run iNEXT 3D for the 500 size fraction</span>
<span class="n">out.raw.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">iNEXT3D</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">matrix_list.500</span><span class="o">$</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;TD&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;incidence_raw&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">nboot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
<span class="nf">ggiNEXT3D</span><span class="p">(</span><span class="n">out.raw.500</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">facet.var</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Assemblage&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="n">Assemblage</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="species-accumulation-curves-100">
<img alt="_images/speciesaccumulationcurves100.png" src="_images/speciesaccumulationcurves100.png" />
<figcaption>
<p><span class="caption-number">Fig. 61 </span><span class="caption-text">: Species accumulation curves per site for the 100 µm size fraction through inter- and extrapolation from the <code class="docutils literal notranslate"><span class="pre">iNEXT.3D</span></code> R package.</span><a class="headerlink" href="#species-accumulation-curves-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="species-accumulation-curves-500">
<img alt="_images/speciesaccumulationcurves500.png" src="_images/speciesaccumulationcurves500.png" />
<figcaption>
<p><span class="caption-number">Fig. 62 </span><span class="caption-text">: Species accumulation curves per site for the 500 µm size fraction through inter- and extrapolation from the <code class="docutils literal notranslate"><span class="pre">iNEXT.3D</span></code> R package.</span><a class="headerlink" href="#species-accumulation-curves-500" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="alpha-diversity-comparison">
<h3>5.2 Alpha diversity comparison<a class="headerlink" href="#alpha-diversity-comparison" title="Permalink to this heading">#</a></h3>
<p>For the alpha diversity comparison through the Hill number framework, we can use the same data format as the species accumulation curves and use the <code class="docutils literal notranslate"><span class="pre">AO3D</span></code> function in the <code class="docutils literal notranslate"><span class="pre">iNEXT.3D</span></code> R package.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Since we combined our 3 ARMS samples per site to obtain an incidence-frequency data table, we are comparing two low impacted sites against two high impacted sites. Such a comparison is not statistically sound and would require data from more low and high impacted sites. Hence, this analysis should only be seen as an introduction to the methodology, not as the outcome or endorsement of this analysis on such a data set.</p>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. run function AO3D</span>
<span class="n">out.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AO3D</span><span class="p">(</span><span class="n">matrix_list.100</span><span class="o">$</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;TD&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;incidence_raw&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Observed&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">nboot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="n">conf</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.95</span><span class="p">)</span>
<span class="n">out.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AO3D</span><span class="p">(</span><span class="n">matrix_list.500</span><span class="o">$</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;TD&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;incidence_raw&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;Observed&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">nboot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="n">conf</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.95</span><span class="p">)</span>
</pre></div>
</div>
<p>To determine differences between low and high impacted sites (and when you would have more than 2 values for each “treatment”), you can draw box plots for each Hill order and run ANOVAs to determine significant differences between the treatments for each Hill number analysed. It is important to test for the assumptions of parametric tests, such as ANOVA, to determine the need to use a non-parametric alternative instead. As mentioned above, this approach is not valid with two data points for each treatment, but below is the code to show how this would work in theory.</p>
<p>First, we need to subset the dataframes to contain the essential information about the Hill numbers.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">################</span>
<span class="c1"># HILL NUMBERS #</span>
<span class="c1">################</span>
<span class="c1"># 1. subset data to include essential hill numbers, 2. add a new column to the data frame containing group info</span>
<span class="n">qOrders</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.0</span><span class="p">,</span><span class="w"> </span><span class="m">1.0</span><span class="p">,</span><span class="w"> </span><span class="m">2.0</span><span class="p">)</span>
<span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">)</span>
<span class="n">out.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">out.100</span><span class="p">[</span><span class="n">out.100</span><span class="o">$</span><span class="n">Order.q</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">qOrders</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="n">out.100</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">group</span>
<span class="n">out.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">out.500</span><span class="p">[</span><span class="n">out.500</span><span class="o">$</span><span class="n">Order.q</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">qOrders</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="n">out.500</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">group</span>
</pre></div>
</div>
<p>Next, we can draw the box plots.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. draw boxplot for 100 size fraction</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">out.100</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">Order.q</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">qD</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">interaction</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">Order.q</span><span class="p">))))</span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_boxplot</span><span class="p">(</span><span class="n">outlier.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_point</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_jitterdodge</span><span class="p">(</span><span class="n">jitter.width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">),</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_bw</span><span class="p">()</span>

<span class="c1"># 1. draw boxplot for 500 size fraction</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">out.500</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">Order.q</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">qD</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">interaction</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">Order.q</span><span class="p">))))</span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_boxplot</span><span class="p">(</span><span class="n">outlier.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_point</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_jitterdodge</span><span class="p">(</span><span class="n">jitter.width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">),</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_bw</span><span class="p">()</span>
</pre></div>
</div>
<figure class="align-default" id="boxplot-100">
<img alt="_images/boxplot100.png" src="_images/boxplot100.png" />
<figcaption>
<p><span class="caption-number">Fig. 63 </span><span class="caption-text">: Box plot of the 100 µm size fraction of three orders of Hill numbers to compare alpha diversity measures between high and low impacted sites.</span><a class="headerlink" href="#boxplot-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="boxplot-500">
<img alt="_images/boxplot500.png" src="_images/boxplot500.png" />
<figcaption>
<p><span class="caption-number">Fig. 64 </span><span class="caption-text">: Box plot of the 500 µm size fraction of three orders of Hill numbers to compare alpha diversity measures between high and low impacted sites</span><a class="headerlink" href="#boxplot-500" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>To test for significant differences between the treatments, we can run an ANOVA analysis. We will also test for the assumptions of this parametric test. Since we’re working with two groups, this ANOVA will default to a simple Student’s t-test. Note the warning before about the number of samples in the tutorial data set!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The code below shows the analysis for q = 0. To test the other two orders (q = 1; q = 2), the values in the initial subsetting can be changed.</p>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. subset out.100 and out.500 to only keep q0 values</span>
<span class="n">q0.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">out.100</span><span class="p">[</span><span class="n">out.100</span><span class="o">$</span><span class="n">Order.q</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="m">0.0</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="n">q0.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">out.500</span><span class="p">[</span><span class="n">out.500</span><span class="o">$</span><span class="n">Order.q</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="m">0.0</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>

<span class="c1"># 1. run ANOVA for q0.100</span>
<span class="n">anovaHill100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">q0.100</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">group</span><span class="p">)))</span>
<span class="nf">Summarize</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q0.100</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
<span class="n">modelq0.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lm</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q0.100</span><span class="p">)</span>
<span class="nf">Anova</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;II&#39;</span><span class="p">)</span>
<span class="nf">anova</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">)</span>

<span class="c1"># 1. check assumptions for ANOVA q0.100</span>
<span class="nf">hist</span><span class="p">(</span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">),</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;darkgray&#39;</span><span class="p">)</span>
<span class="nf">shapiro.test</span><span class="p">(</span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="nf">fitted</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">),</span><span class="w"> </span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelq0.100</span><span class="p">))</span>
<span class="nf">leveneTest</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anovaHill100</span><span class="p">)</span>
<span class="nf">bartlett.test</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anovaHill100</span><span class="p">)</span>

<span class="c1"># 1. run ANOVA for q0.500</span>
<span class="n">anovaHill500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">q0.500</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">group</span><span class="p">)))</span>
<span class="nf">Summarize</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q0.500</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
<span class="n">modelq0.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lm</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q0.500</span><span class="p">)</span>
<span class="nf">Anova</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;II&#39;</span><span class="p">)</span>
<span class="nf">anova</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">)</span>

<span class="c1"># 1. check assumptions for ANOVA q0.500</span>
<span class="nf">hist</span><span class="p">(</span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">),</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;darkgray&#39;</span><span class="p">)</span>
<span class="nf">shapiro.test</span><span class="p">(</span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="nf">fitted</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">),</span><span class="w"> </span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelq0.500</span><span class="p">))</span>
<span class="nf">leveneTest</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anovaHill500</span><span class="p">)</span>
<span class="nf">bartlett.test</span><span class="p">(</span><span class="n">qD</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anovaHill500</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="sd-tab-set docutils">
<input checked="checked" id="sd-tab-item-0" name="sd-tab-set-0" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-0">
100 µm size fraction</label><div class="sd-tab-content docutils">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">Summarize</span><span class="p">(</span><span class="n">qD</span> <span class="o">~</span> <span class="n">group</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">q0</span><span class="mf">.100</span><span class="p">,</span> <span class="n">digits</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
  <span class="n">group</span> <span class="n">n</span>  <span class="n">mean</span>     <span class="n">sd</span> <span class="nb">min</span>     <span class="n">Q1</span> <span class="n">median</span>     <span class="n">Q3</span> <span class="nb">max</span>
<span class="mi">1</span>  <span class="n">high</span> <span class="mi">2</span> <span class="mf">291.5</span> <span class="mf">78.489</span> <span class="mi">236</span> <span class="mf">263.75</span>  <span class="mf">291.5</span> <span class="mf">319.25</span> <span class="mi">347</span>
<span class="mi">2</span>   <span class="n">low</span> <span class="mi">2</span> <span class="mf">399.0</span>  <span class="mf">0.000</span> <span class="mi">399</span> <span class="mf">399.00</span>  <span class="mf">399.0</span> <span class="mf">399.00</span> <span class="mi">399</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">Anova</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.100</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;II&#39;</span><span class="p">)</span>
<span class="n">Anova</span> <span class="n">Table</span> <span class="p">(</span><span class="n">Type</span> <span class="n">II</span> <span class="n">tests</span><span class="p">)</span>

<span class="n">Response</span><span class="p">:</span> <span class="n">qD</span>
           <span class="n">Sum</span> <span class="n">Sq</span> <span class="n">Df</span> <span class="n">F</span> <span class="n">value</span> <span class="n">Pr</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>
<span class="n">group</span>     <span class="mf">11556.3</span>  <span class="mi">1</span>  <span class="mf">3.7517</span> <span class="mf">0.1924</span>
<span class="n">Residuals</span>  <span class="mf">6160.5</span>  <span class="mi">2</span>  
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">anova</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.100</span><span class="p">)</span>
<span class="n">Analysis</span> <span class="n">of</span> <span class="n">Variance</span> <span class="n">Table</span>

<span class="n">Response</span><span class="p">:</span> <span class="n">qD</span>
          <span class="n">Df</span>  <span class="n">Sum</span> <span class="n">Sq</span> <span class="n">Mean</span> <span class="n">Sq</span> <span class="n">F</span> <span class="n">value</span> <span class="n">Pr</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>
<span class="n">group</span>      <span class="mi">1</span> <span class="mf">11556.3</span> <span class="mf">11556.3</span>  <span class="mf">3.7517</span> <span class="mf">0.1924</span>
<span class="n">Residuals</span>  <span class="mi">2</span>  <span class="mf">6160.5</span>  <span class="mf">3080.2</span>   
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; summary(modelq0.100)

Call:
lm(formula = qD ~ group, data = q0.100)

Residuals:
         1         12         23         34 
 2.132e-14 -1.926e-14  5.550e+01 -5.550e+01 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)  
(Intercept)   291.50      39.24   7.428   0.0176 *
grouplow      107.50      55.50   1.937   0.1924  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 55.5 on 2 degrees of freedom
Multiple R-squared:  0.6523, Adjusted R-squared:  0.4784 
F-statistic: 3.752 on 1 and 2 DF,  p-value: 0.1924
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">shapiro</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">residuals</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.100</span><span class="p">))</span>

 <span class="n">Shapiro</span><span class="o">-</span><span class="n">Wilk</span> <span class="n">normality</span> <span class="n">test</span>

<span class="n">data</span><span class="p">:</span>  <span class="n">residuals</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.100</span><span class="p">)</span>
<span class="n">W</span> <span class="o">=</span> <span class="mf">0.94466</span><span class="p">,</span> <span class="n">p</span><span class="o">-</span><span class="n">value</span> <span class="o">=</span> <span class="mf">0.683</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; leveneTest(qD ~ group, data = anovaHill100)
Levene&#39;s Test for Homogeneity of Variance (center = median)
      Df    F value    Pr(&gt;F)    
group  1 1.8254e+32 &lt; 2.2e-16 ***
       2                         
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
Warning message:
In anova.lm(lm(resp ~ group)) :
  ANOVA F-tests on an essentially perfect fit are unreliable
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">bartlett</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">qD</span> <span class="o">~</span> <span class="n">group</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">anovaHill100</span><span class="p">)</span>

 <span class="n">Bartlett</span> <span class="n">test</span> <span class="n">of</span> <span class="n">homogeneity</span> <span class="n">of</span> <span class="n">variances</span>

<span class="n">data</span><span class="p">:</span>  <span class="n">qD</span> <span class="n">by</span> <span class="n">group</span>
<span class="n">Bartlett</span><span class="s1">&#39;s K-squared = Inf, df = 1, p-value &lt; 2.2e-16</span>
</pre></div>
</div>
<figure class="align-default" id="normality-100">
<img alt="_images/normality100.png" src="_images/normality100.png" />
<figcaption>
<p><span class="caption-number">Fig. 65 </span><span class="caption-text">: Histogram to determine normal distribution of variances for the 100 µm size fraction.</span><a class="headerlink" href="#normality-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="residuals-plot-100">
<img alt="_images/residualsplot100.png" src="_images/residualsplot100.png" />
<figcaption>
<p><span class="caption-number">Fig. 66 </span><span class="caption-text">: Plot of the residuals for the 100 µm size fraction.</span><a class="headerlink" href="#residuals-plot-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</div>
<input id="sd-tab-item-1" name="sd-tab-set-0" type="radio">
</input><label class="sd-tab-label" for="sd-tab-item-1">
500 µm size fraction</label><div class="sd-tab-content docutils">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">Summarize</span><span class="p">(</span><span class="n">qD</span> <span class="o">~</span> <span class="n">group</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">q0</span><span class="mf">.500</span><span class="p">,</span> <span class="n">digits</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
  <span class="n">group</span> <span class="n">n</span>  <span class="n">mean</span>     <span class="n">sd</span> <span class="nb">min</span>     <span class="n">Q1</span> <span class="n">median</span>     <span class="n">Q3</span> <span class="nb">max</span>
<span class="mi">1</span>  <span class="n">high</span> <span class="mi">2</span> <span class="mf">224.5</span> <span class="mf">27.577</span> <span class="mi">205</span> <span class="mf">214.75</span>  <span class="mf">224.5</span> <span class="mf">234.25</span> <span class="mi">244</span>
<span class="mi">2</span>   <span class="n">low</span> <span class="mi">2</span> <span class="mf">229.0</span>  <span class="mf">7.071</span> <span class="mi">224</span> <span class="mf">226.50</span>  <span class="mf">229.0</span> <span class="mf">231.50</span> <span class="mi">234</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">Anova</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.500</span><span class="p">,</span> <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;II&#39;</span><span class="p">)</span>
<span class="n">Anova</span> <span class="n">Table</span> <span class="p">(</span><span class="n">Type</span> <span class="n">II</span> <span class="n">tests</span><span class="p">)</span>

<span class="n">Response</span><span class="p">:</span> <span class="n">qD</span>
          <span class="n">Sum</span> <span class="n">Sq</span> <span class="n">Df</span> <span class="n">F</span> <span class="n">value</span> <span class="n">Pr</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>
<span class="n">group</span>      <span class="mf">20.25</span>  <span class="mi">1</span>    <span class="mf">0.05</span> <span class="mf">0.8439</span>
<span class="n">Residuals</span> <span class="mf">810.50</span>  <span class="mi">2</span>   
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">anova</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.500</span><span class="p">)</span>
<span class="n">Analysis</span> <span class="n">of</span> <span class="n">Variance</span> <span class="n">Table</span>

<span class="n">Response</span><span class="p">:</span> <span class="n">qD</span>
          <span class="n">Df</span> <span class="n">Sum</span> <span class="n">Sq</span> <span class="n">Mean</span> <span class="n">Sq</span> <span class="n">F</span> <span class="n">value</span> <span class="n">Pr</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>
<span class="n">group</span>      <span class="mi">1</span>  <span class="mf">20.25</span>   <span class="mf">20.25</span>    <span class="mf">0.05</span> <span class="mf">0.8439</span>
<span class="n">Residuals</span>  <span class="mi">2</span> <span class="mf">810.50</span>  <span class="mf">405.25</span>   
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; summary(modelq0.500)

Call:
lm(formula = qD ~ group, data = q0.500)

Residuals:
    1    12    23    34 
  5.0  -5.0  19.5 -19.5 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)   
(Intercept)   224.50      14.23  15.771    0.004 **
grouplow        4.50      20.13   0.224    0.844   
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 20.13 on 2 degrees of freedom
Multiple R-squared:  0.02438, Adjusted R-squared:  -0.4634 
F-statistic: 0.04997 on 1 and 2 DF,  p-value: 0.8439
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">shapiro</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">residuals</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.500</span><span class="p">))</span>

 <span class="n">Shapiro</span><span class="o">-</span><span class="n">Wilk</span> <span class="n">normality</span> <span class="n">test</span>

<span class="n">data</span><span class="p">:</span>  <span class="n">residuals</span><span class="p">(</span><span class="n">modelq0</span><span class="mf">.500</span><span class="p">)</span>
<span class="n">W</span> <span class="o">=</span> <span class="mf">0.99982</span><span class="p">,</span> <span class="n">p</span><span class="o">-</span><span class="n">value</span> <span class="o">=</span> <span class="mf">0.9995</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; leveneTest(qD ~ group, data = anovaHill500)
Levene&#39;s Test for Homogeneity of Variance (center = median)
      Df    F value    Pr(&gt;F)    
group  1 2.5893e+31 &lt; 2.2e-16 ***
       2                         
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
Warning message:
In anova.lm(lm(resp ~ group)) :
  ANOVA F-tests on an essentially perfect fit are unreliable
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">bartlett</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">qD</span> <span class="o">~</span> <span class="n">group</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">anovaHill500</span><span class="p">)</span>

 <span class="n">Bartlett</span> <span class="n">test</span> <span class="n">of</span> <span class="n">homogeneity</span> <span class="n">of</span> <span class="n">variances</span>

<span class="n">data</span><span class="p">:</span>  <span class="n">qD</span> <span class="n">by</span> <span class="n">group</span>
<span class="n">Bartlett</span><span class="s1">&#39;s K-squared = 0.97534, df = 1, p-value = 0.3234</span>
</pre></div>
</div>
<figure class="align-default" id="normality-500">
<img alt="_images/normality500.png" src="_images/normality500.png" />
<figcaption>
<p><span class="caption-number">Fig. 67 </span><span class="caption-text">: Histogram to determine normal distribution of variances for the 500 µm size fraction.</span><a class="headerlink" href="#normality-500" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="residuals-plot-500">
<img alt="_images/residualsplot500.png" src="_images/residualsplot500.png" />
<figcaption>
<p><span class="caption-number">Fig. 68 </span><span class="caption-text">: Plot of the residuals for the 500 µm size fraction.</span><a class="headerlink" href="#residuals-plot-500" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="estimate-replication">
<h3>5.3 Estimate replication<a class="headerlink" href="#estimate-replication" title="Permalink to this heading">#</a></h3>
<p>Through the inter- and extrapolation of alpha diversity measures calculated by the <code class="docutils literal notranslate"><span class="pre">iNEXT.3D</span></code> R package, we can also estimate the needed replication or sampling within a site to recover a set percentage of the diversity contained within that site. For this, we need to run the <code class="docutils literal notranslate"><span class="pre">estimate3D</span></code> function. We can then run a similar analysis as section <strong>5.2 Alpha diversity comparison</strong>, whereby we can draw box plots and run ANOVAs to determine differences between treatments. The same warning applies here as mentioned in the previous section about the analysis!</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">########################</span>
<span class="c1"># ESTIMATE REPLICATION #</span>
<span class="c1">########################</span>
<span class="c1"># 1. run the estimate3D function for the 100 and 500 size fraction dataframes for q = 0</span>
<span class="n">out.est.100</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">estimate3D</span><span class="p">(</span><span class="n">matrix_list.100</span><span class="o">$</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;TD&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;incidence_raw&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;coverage&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.9</span><span class="p">)</span>
<span class="n">out.est.500</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">estimate3D</span><span class="p">(</span><span class="n">matrix_list.500</span><span class="o">$</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;TD&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">datatype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;incidence_raw&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;coverage&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.9</span><span class="p">)</span>

<span class="c1"># 1. add a new column to the data frames containing group info</span>
<span class="n">group.est</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;low&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;high&#39;</span><span class="p">)</span>
<span class="n">out.est.100</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">group.est</span>
<span class="n">out.est.500</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">group.est</span>

<span class="c1"># 1. draw boxplot for size fraction 100</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">out.est.100</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nt</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_boxplot</span><span class="p">(</span><span class="n">outlier.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_jitter</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_bw</span><span class="p">()</span>

<span class="c1"># 1. draw boxplot for size fraction 500</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">out.est.500</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nt</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_boxplot</span><span class="p">(</span><span class="n">outlier.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_jitter</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_bw</span><span class="p">()</span>
</pre></div>
</div>
<figure class="align-default" id="estimate-100">
<img alt="_images/estimate100.png" src="_images/estimate100.png" />
<figcaption>
<p><span class="caption-number">Fig. 69 </span><span class="caption-text">: Box plot of the 100 µm size fraction of q = 0 to estimate number of samples to be collected to detect 90% of the diversity.</span><a class="headerlink" href="#estimate-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="estimate-500">
<img alt="_images/estimate500.png" src="_images/estimate500.png" />
<figcaption>
<p><span class="caption-number">Fig. 70 </span><span class="caption-text">: Box plot of the 500 µm size fraction of q = 0 to estimate number of samples to be collected to detect 90% of the diversity.</span><a class="headerlink" href="#estimate-500" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. run ANOVA for estimateD for 100</span>
<span class="n">anovaEstimateD</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">out.est.100</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">group</span><span class="p">)))</span>
<span class="nf">Summarize</span><span class="p">(</span><span class="n">nt</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">out.est.100</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span>
<span class="n">modelEstimateD</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lm</span><span class="p">(</span><span class="n">nt</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">out.est.100</span><span class="p">)</span>
<span class="nf">Anova</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;II&#39;</span><span class="p">)</span>
<span class="nf">anova</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">)</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">)</span>

<span class="c1"># 1. check assumptions for ANOVA estimateD for 100</span>
<span class="nf">hist</span><span class="p">(</span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">),</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;darkgray&#39;</span><span class="p">)</span>
<span class="nf">shapiro.test</span><span class="p">(</span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="nf">fitted</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">),</span><span class="w"> </span><span class="nf">residuals</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">))</span>
<span class="nf">leveneTest</span><span class="p">(</span><span class="n">nt</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anovaEstimateD</span><span class="p">)</span>
<span class="nf">bartlett.test</span><span class="p">(</span><span class="n">nt</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">anovaEstimateD</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">Summarize</span><span class="p">(</span><span class="n">nt</span> <span class="o">~</span> <span class="n">group</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">out</span><span class="o">.</span><span class="n">est</span><span class="mf">.100</span><span class="p">,</span> <span class="n">digits</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
  <span class="n">group</span> <span class="n">n</span>  <span class="n">mean</span>   <span class="n">sd</span>   <span class="nb">min</span>    <span class="n">Q1</span> <span class="n">median</span>    <span class="n">Q3</span>   <span class="nb">max</span>
<span class="mi">1</span>  <span class="n">high</span> <span class="mi">2</span> <span class="mf">1.146</span> <span class="mf">0.00</span> <span class="mf">1.146</span> <span class="mf">1.146</span>  <span class="mf">1.146</span> <span class="mf">1.146</span> <span class="mf">1.146</span>
<span class="mi">2</span>   <span class="n">low</span> <span class="mi">2</span> <span class="mf">3.247</span> <span class="mf">0.04</span> <span class="mf">3.218</span> <span class="mf">3.232</span>  <span class="mf">3.247</span> <span class="mf">3.261</span> <span class="mf">3.275</span>                
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; Anova(modelEstimateD, type = &#39;II&#39;)
Anova Table (Type II tests)

Response: nt
          Sum Sq Df F value    Pr(&gt;F)    
group     4.4129  1  5387.7 0.0001856 ***
Residuals 0.0016  2                      
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; anova(modelEstimateD)
Analysis of Variance Table

Response: nt
          Df Sum Sq Mean Sq F value    Pr(&gt;F)    
group      1 4.4129  4.4129  5387.7 0.0001856 ***
Residuals  2 0.0016  0.0008                      
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; summary(modelEstimateD)

Call:
lm(formula = nt ~ group, data = out.est.100)

Residuals:
         1          2          3          4 
-2.862e-02  2.862e-02  3.469e-18  3.469e-18 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  1.14596    0.02024   56.63 0.000312 ***
grouplow     2.10068    0.02862   73.40 0.000186 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.02862 on 2 degrees of freedom
Multiple R-squared:  0.9996, Adjusted R-squared:  0.9994 
F-statistic:  5388 on 1 and 2 DF,  p-value: 0.0001856
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">shapiro</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">residuals</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">))</span>

 <span class="n">Shapiro</span><span class="o">-</span><span class="n">Wilk</span> <span class="n">normality</span> <span class="n">test</span>

<span class="n">data</span><span class="p">:</span>  <span class="n">residuals</span><span class="p">(</span><span class="n">modelEstimateD</span><span class="p">)</span>
<span class="n">W</span> <span class="o">=</span> <span class="mf">0.94466</span><span class="p">,</span> <span class="n">p</span><span class="o">-</span><span class="n">value</span> <span class="o">=</span> <span class="mf">0.683</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; leveneTest(nt ~ group, data = anovaEstimateD)
Levene&#39;s Test for Homogeneity of Variance (center = median)
      Df  F value    Pr(&gt;F)    
group  1 1.63e+28 &lt; 2.2e-16 ***
       2                       
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
Warning message:
In anova.lm(lm(resp ~ group)) :
  ANOVA F-tests on an essentially perfect fit are unreliable
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">bartlett</span><span class="o">.</span><span class="n">test</span><span class="p">(</span><span class="n">nt</span> <span class="o">~</span> <span class="n">group</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">anovaEstimateD</span><span class="p">)</span>

 <span class="n">Bartlett</span> <span class="n">test</span> <span class="n">of</span> <span class="n">homogeneity</span> <span class="n">of</span> <span class="n">variances</span>

<span class="n">data</span><span class="p">:</span>  <span class="n">nt</span> <span class="n">by</span> <span class="n">group</span>
<span class="n">Bartlett</span><span class="s1">&#39;s K-squared = Inf, df = 1, p-value &lt; 2.2e-16</span>
</pre></div>
</div>
<figure class="align-default" id="hist-estimate-100">
<img alt="_images/histestimate100.png" src="_images/histestimate100.png" />
<figcaption>
<p><span class="caption-number">Fig. 71 </span><span class="caption-text">: Histogram of the 100 µm size fraction of q = 0 to estimate number of samples to be collected to detect 90% of the diversity.</span><a class="headerlink" href="#hist-estimate-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="residuals-estimate-100">
<img alt="_images/residualsestimate100.png" src="_images/residualsestimate100.png" />
<figcaption>
<p><span class="caption-number">Fig. 72 </span><span class="caption-text">: Residual plot of the 100 µm size fraction of q = 0 to estimate number of samples to be collected to detect 90% of the diversity.</span><a class="headerlink" href="#residuals-estimate-100" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</div>
</section>
<section id="phylodiversity">
<h3>5.4 Phylodiversity<a class="headerlink" href="#phylodiversity" title="Permalink to this heading">#</a></h3>
<p>A quick note on phylodiversity measurements: There is a push to incorporate phylogenetic relatedness in alpha diversity measurements and comparisons. Intuitively, it makes much more sense that a community comprised of 20 birds within the same family is less diverse than 20 birds comprised of different families. Such statistical frameworks that incorporate phylogenetic relatedness already exist and make use of the phylogenetic tree we generated in section <strong>Phylogenetic tree</strong>. Phylogenetic diversity measures usually take on three facets, including divergence, diversification, and survival time. <a class="reference external" href="https://royalsocietypublishing.org/doi/epdf/10.1098/rstb.2017.0397">Kling et al., 2018</a> and <a class="reference external" href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.13014">Alberdi &amp; Gilbert, 2019</a> are two excellent publications providing additional information about this topic.</p>
<figure class="align-default" id="id1">
<img alt="_images/phylodiversity.png" src="_images/phylodiversity.png" />
<figcaption>
<p><span class="caption-number">Fig. 73 </span><span class="caption-text">: An introduction to the facets of phylogenetic diversity. Copyright: <a class="reference external" href="https://royalsocietypublishing.org/doi/epdf/10.1098/rstb.2017.0397">Kling et al., 2018</a>.</span><a class="headerlink" href="#id1" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
</section>
<section id="beta-diversity">
<h2>6. Beta diversity<a class="headerlink" href="#beta-diversity" title="Permalink to this heading">#</a></h2>
<p>Besides the number of species within a site (alpha diversity), species composition or beta diversity is another important aspect of analysis within metabarcoding research. To analyse species composition between sites, we can use ordination and cluster visualisations to depict sample/site/treatment similarity or dissimilarity. PERMANOVA and PERMDISP, on the other hand, are the statistical tests to investigate significant differences between treatments based on species composition between treatments. Lastly, Indicator Species Analysis (ISA) allows us to identify which signals are driving the difference between groups.</p>
<section id="ordination">
<h3>6.1 Ordination<a class="headerlink" href="#ordination" title="Permalink to this heading">#</a></h3>
<p>Ordination is the multivariate analysis which aims to find gradients or changes in species composition across groups. Multivariate species composition can be imagined as samples in multidimensional space, where every species represents its own axis. Because multidimensional space is not easy to display, describe, or even imagine, it is worth reducing it into a few main dimensions while preserving the maximum information. In ordination space, objects that are closer together are more similar, and dissimilar objects are more separated from each other.</p>
<p>There are multiple ordination methods available. Some of the most common ones are Principal Components Analysis (PCA), Principle Coordinates Analysis (PCoA), and Non-metric MultiDimensional Scaling (NMDS). Although many other ordination methods exist as well, including RDA (Redundancy Analysis), CAP (Canonical Analysis of Principle coordinates), and dbRDA (Distance Cased Redundancy Analysis) The choice of ordination methods depends on (1) the type of data and (2) the similarity distance matrix you intend to use. Some methods, such as PCA are restricted to using Euclidean distance matrices, while NMDS and PCoA are able to use any distance matrix. The choice of distance matrix, is usually dependent on the type of data. For metabarcoding data, which contains lots of null values, Bray-Curtis (abundance or relative abundance) and Jaccard (presence-absence) are frequently chosen. Both these options are non-euclidean distances, which is why most metabarcoding studies opt for PCoA and NMDS ordination methods. For more information about ordination, please visit <a class="reference external" href="https://www.davidzeleny.net/anadat-r/doku.php/en:ordination">this blog</a> or this <a class="reference external" href="https://onlinelibrary.wiley.com/doi/full/10.1111/mec.13536">review article</a>.</p>
<p>Since metabarcoding studies usually opt for NMDS and PCoA ordination, this tutorial will cover these two options.</p>
<section id="nmds">
<h4>6.1.1 NMDS<a class="headerlink" href="#nmds" title="Permalink to this heading">#</a></h4>
<p>The first option we will cover is NMDS. NMDS is a technique that attempts to represent the pairwise dissimilarities or distances between samples in a lower-dimensional space. The goal is to position the samples in such a way that their relative distances in the ordination plot reflect the original dissimilarity structure as close as possible.</p>
<p>NMDS is a non-metric technique, meaning that it does not assume a linear relationship between the original distances and the distances in the ordination space. Instead, NMDS focuses on preserving the ranking or order of dissimilarities rather than their precise values.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">##############</span>
<span class="c1"># ORDINATION #</span>
<span class="c1">##############</span>
<span class="c1"># 1. generate frequency-occurrence data frame</span>
<span class="n">physeq.pa</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">microbiome</span><span class="o">::</span><span class="nf">transform</span><span class="p">(</span><span class="n">physeq</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;pa&#39;</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="n">physeq.pa</span><span class="p">)</span>

<span class="c1"># 1. run NMDS</span>
<span class="n">physeq.pa.nmds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ordinate</span><span class="p">(</span><span class="n">physeq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">physeq.pa</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;NMDS&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">distance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;jaccard&#39;</span><span class="p">)</span>

<span class="c1"># 1. plot NMDS</span>
<span class="n">sample_colors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;high&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;low&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">)</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">physeq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">physeq.pa</span><span class="p">,</span><span class="w"> </span><span class="n">ordination</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">physeq.pa.nmds</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">  </span><span class="nf">geom_point</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SAMPLE</span><span class="p">,</span><span class="w"> </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">island</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">stat_ellipse</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">NMDS1</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">NMDS2</span><span class="p">,</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">impact</span><span class="p">),</span><span class="w"> </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_equal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_shape_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">21</span><span class="p">,</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="m">23</span><span class="p">,</span><span class="w"> </span><span class="m">24</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample_colors</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_classic</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="w">                             </span>
<span class="w">    </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span><span class="w">                               </span><span class="c1">#changes legend size</span>
<span class="w">    </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span><span class="w">                                      </span><span class="c1">#removes legend title</span>
<span class="w">    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">))</span><span class="o">+</span><span class="w">  </span><span class="c1">#adds black boarder around legend</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y.left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">        </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">        </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">        </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="o">+</span>
<span class="w">  </span><span class="nf">guides</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">guide_legend</span><span class="p">(</span><span class="n">override.aes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">21</span><span class="p">)))</span><span class="w"> </span>

<span class="c1"># 1. output stress value</span>
<span class="n">physeq.pa.nmds</span><span class="o">$</span><span class="n">stress</span>

<span class="c1"># 1. plot stress values in shepard&#39;s plot</span>
<span class="nf">stressplot</span><span class="p">(</span><span class="n">physeq.pa.nmds</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="nb">print</span><span class="p">(</span><span class="n">physeq</span><span class="o">.</span><span class="n">pa</span><span class="p">)</span>
<span class="n">phyloseq</span><span class="o">-</span><span class="k">class</span> <span class="nc">experiment</span><span class="o">-</span><span class="n">level</span> <span class="nb">object</span>
<span class="n">otu_table</span><span class="p">()</span>   <span class="n">OTU</span> <span class="n">Table</span><span class="p">:</span>         <span class="p">[</span> <span class="mi">865</span> <span class="n">taxa</span> <span class="ow">and</span> <span class="mi">24</span> <span class="n">samples</span> <span class="p">]</span>
<span class="n">sample_data</span><span class="p">()</span> <span class="n">Sample</span> <span class="n">Data</span><span class="p">:</span>       <span class="p">[</span> <span class="mi">24</span> <span class="n">samples</span> <span class="n">by</span> <span class="mi">32</span> <span class="n">sample</span> <span class="n">variables</span> <span class="p">]</span>
<span class="n">tax_table</span><span class="p">()</span>   <span class="n">Taxonomy</span> <span class="n">Table</span><span class="p">:</span>    <span class="p">[</span> <span class="mi">865</span> <span class="n">taxa</span> <span class="n">by</span> <span class="mi">9</span> <span class="n">taxonomic</span> <span class="n">ranks</span> <span class="p">]</span>
<span class="n">refseq</span><span class="p">()</span>      <span class="n">DNAStringSet</span><span class="p">:</span>      <span class="p">[</span> <span class="mi">865</span> <span class="n">reference</span> <span class="n">sequences</span> <span class="p">]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; physeq.pa.nmds$stress
[1] 0.08059097
</pre></div>
</div>
</div>
<figure class="align-default" id="nmds-plot">
<img alt="_images/nmdsplot.png" src="_images/nmdsplot.png" />
<figcaption>
<p><span class="caption-number">Fig. 74 </span><span class="caption-text">: The NMDS plot whereby shapes provide information about the location, colour of shape indicate size fraction, and 95% confidence ellipses are drawn around low and high impact samples.</span><a class="headerlink" href="#nmds-plot" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="shepards-plot">
<img alt="_images/shepardsplot.png" src="_images/shepardsplot.png" />
<figcaption>
<p><span class="caption-number">Fig. 75 </span><span class="caption-text">: The Shepard plot of the NMDS indicating a strong correlation between observed dissimilarity and ordination distance (R2 = 0.977), highlighting an accurate representation of the NMDS.</span><a class="headerlink" href="#shepards-plot" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>NMDS will provide a stress level to identify how well the NMDS plot is representing the data. As a rule of thumb, the following cut-off values are agreed upon:</p>
<ul class="simple">
<li><p>stress &gt; 0.2: poor representation; risk of false interpretation of results</p></li>
<li><p>0.1 &lt; stress &lt; 0.2: fair representation; some distances can be misleading for interpretation</p></li>
<li><p>0.05 &lt; stress &lt; 0.1: good representation; confident about the results</p></li>
<li><p>stress &lt; 0.05: excellent representation.</p></li>
</ul>
</section>
<section id="pcoa">
<h4>6.1.2 PCoA<a class="headerlink" href="#pcoa" title="Permalink to this heading">#</a></h4>
<p>Principal Coordinates analysis, also known as metric Multi-Dimensional Scaling (MDS), aims to preserve the original pairwise distances as closely as possible while transforming the data into a low-dimensional space. PCoA is a metric technique, thereby assuming a linear relationship between the original distances and the distances in the ordination space. If the distance measure used is Euclidean, PCoA is identical to PCA.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># 1. run PCoA</span>
<span class="n">physeq.pa.pcoa</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ordinate</span><span class="p">(</span><span class="n">physeq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">physeq.pa</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;PCoA&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">distance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;jaccard&#39;</span><span class="p">)</span>

<span class="c1"># 1. plot PCoA</span>
<span class="n">sample_colors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;high&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;lightgoldenrod&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;low&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;steelblue&quot;</span><span class="p">)</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">physeq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">physeq.pa</span><span class="p">,</span><span class="w"> </span><span class="n">ordination</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">physeq.pa.pcoa</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">  </span><span class="nf">geom_point</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SAMPLE</span><span class="p">,</span><span class="w"> </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">island</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">stat_ellipse</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Axis.1</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Axis.2</span><span class="p">,</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">impact</span><span class="p">),</span><span class="w"> </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_equal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_shape_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">21</span><span class="p">,</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="m">23</span><span class="p">,</span><span class="w"> </span><span class="m">24</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample_colors</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_classic</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="w">                             </span>
<span class="w">    </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span><span class="w">                               </span><span class="c1">#changes legend size</span>
<span class="w">    </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span><span class="w">                                      </span><span class="c1">#removes legend title</span>
<span class="w">    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">))</span><span class="o">+</span><span class="w">  </span><span class="c1">#adds black boarder around legend</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.y.left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">        </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">        </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">        </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="o">+</span>
<span class="w">  </span><span class="nf">guides</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">guide_legend</span><span class="p">(</span><span class="n">override.aes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">21</span><span class="p">)))</span><span class="w"> </span>

<span class="c1"># 1. scree plot</span>
<span class="nf">plot_scree</span><span class="p">(</span><span class="n">physeq.pa.pcoa</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="pcoa-plot">
<img alt="_images/pcoaplot.png" src="_images/pcoaplot.png" />
<figcaption>
<p><span class="caption-number">Fig. 76 </span><span class="caption-text">: The PCoA plot whereby shapes provide information about the location, colour of shape indicate size fraction, and 95% confidence ellipses are drawn around low and high impact samples.</span><a class="headerlink" href="#pcoa-plot" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="scree-plot">
<img alt="_images/screeplot.png" src="_images/screeplot.png" />
<figcaption>
<p><span class="caption-number">Fig. 77 </span><span class="caption-text">: The Scree plot providing information about the percentage of variation explained on each axis.</span><a class="headerlink" href="#scree-plot" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>In PCoA, the eigenvalues represent the amount of variation explained by each axis or dimension in the ordination plot. The larger the eigenvalue, the more important that axis is in capturing the variability in the original data. We can plot a Scree plot to visualise the importance of each axis in our data set.</p>
</section>
</section>
<section id="permanova">
<h3>6.2 PERMANOVA<a class="headerlink" href="#permanova" title="Permalink to this heading">#</a></h3>
<p>While ordination is a great way to visualise differences in species community composition between samples, it does not provide a p-value to determine if those differences are significant. To check if the differences between groups are significant, we can run a PERMANOVA or PERmutational Multivariate Analysis of VAriance. The PERMANOVA procedure works by using permutation testing. Here’s a basic outline of how it works:</p>
<ol class="arabic simple">
<li><p>Data Setup: You have a dataset with multiple variables (multivariate) and multiple groups or conditions.</p></li>
<li><p>Dissimilarity Matrix: A dissimilarity matrix is created based on the multivariate data. This matrix quantifies the dissimilarity or distance between individual observations within the dataset (here you could pick what matrix you would like to use, in our example we are using a Jaccard matrix as we are using Presence-Absence data).</p></li>
<li><p>Permutation Testing: PERMANOVA works by shuffling the group labels (permuting) and then recalculating the dissimilarity matrix and relevant statistical tests. This process is repeated many times to create a distribution of test statistics under the assumption that the group labels have no effect on the data.</p></li>
<li><p>Comparison to Observed Data: The observed test statistic (usually a sum of squares or sums of squared distances) is then compared to the distribution of test statistics generated through permutation testing. This comparison helps determine the likelihood of observing the observed test statistic if the group labels have no effect.</p></li>
<li><p>P-Value Calculation: The p-value is calculated as the proportion of permuted test statistics that are more extreme than the observed test statistic. A low p-value indicates that the observed differences between groups are unlikely to occur by random chance alone.</p></li>
<li><p>Interpretation: If the p-value is below a pre-defined significance level (e.g., 0.05), you can then conclude that there are statistically significant differences between the groups based on the multivariate data.</p></li>
</ol>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">#############</span>
<span class="c1"># PERMANOVA #</span>
<span class="c1">#############</span>
<span class="c1"># 1. extract frequency table and metadata from phyloseq object</span>
<span class="n">permanova.metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as</span><span class="p">(</span><span class="nf">sample_data</span><span class="p">(</span><span class="n">physeq.pa</span><span class="p">),</span><span class="w"> </span><span class="s">&#39;data.frame&#39;</span><span class="p">)</span>
<span class="n">permanova.freqtable</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">as</span><span class="p">(</span><span class="nf">otu_table</span><span class="p">(</span><span class="n">physeq.pa</span><span class="p">),</span><span class="w"> </span><span class="s">&#39;matrix&#39;</span><span class="p">))</span>

<span class="c1"># 1. create distance matrix</span>
<span class="n">permanova.dist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">vegdist</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">permanova.freqtable</span><span class="p">),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;jaccard&#39;</span><span class="p">)</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">123</span><span class="p">)</span>
<span class="n">permanova.results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">adonis2</span><span class="p">(</span><span class="n">permanova.dist</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">impact</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">permanova.metadata</span><span class="p">,</span><span class="w"> </span><span class="n">permutations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10000</span><span class="p">)</span>
<span class="n">permanova.results</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; permanova.results
Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 10000

adonis2(formula = permanova.dist ~ impact, data = permanova.metadata, permutations = 10000)
         Df SumOfSqs      R2      F    Pr(&gt;F)    
impact    1   2.2516 0.27691 8.4251 9.999e-05 ***
Residual 22   5.8795 0.72309                     
Total    23   8.1311 1.00000                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
</pre></div>
</div>
</div>
</section>
<section id="permdisp">
<h3>6.3 PERMDISP<a class="headerlink" href="#permdisp" title="Permalink to this heading">#</a></h3>
<p>A PERMANOVA analysis can provide significant p-values if (i) the centroids of the groups are placed differently in the ordination space, (ii) the dispersion of the points between groups is different, or (iii) both. While we can verify why PERMANOVA is significant for our tutorial data, it is useful to run one additional test, PERMDISP, which provides a significant p-value only when points are dispersed differently between groups within the ordination space.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">##############</span>
<span class="c1"># BETADISPER #</span>
<span class="c1">##############</span>
<span class="n">mod</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">betadisper</span><span class="p">(</span><span class="n">permanova.dist</span><span class="p">,</span><span class="w"> </span><span class="n">permanova.metadata</span><span class="o">$</span><span class="n">impact</span><span class="p">)</span>
<span class="n">mod</span>
<span class="nf">boxplot</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
<span class="nf">permutest</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">permutest</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>

<span class="n">Permutation</span> <span class="n">test</span> <span class="k">for</span> <span class="n">homogeneity</span> <span class="n">of</span> <span class="n">multivariate</span> <span class="n">dispersions</span>
<span class="n">Permutation</span><span class="p">:</span> <span class="n">free</span>
<span class="n">Number</span> <span class="n">of</span> <span class="n">permutations</span><span class="p">:</span> <span class="mi">999</span>

<span class="n">Response</span><span class="p">:</span> <span class="n">Distances</span>
          <span class="n">Df</span>   <span class="n">Sum</span> <span class="n">Sq</span>   <span class="n">Mean</span> <span class="n">Sq</span>      <span class="n">F</span> <span class="n">N</span><span class="o">.</span><span class="n">Perm</span> <span class="n">Pr</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>
<span class="n">Groups</span>     <span class="mi">1</span> <span class="mf">0.000128</span> <span class="mf">0.0001278</span> <span class="mf">0.0308</span>    <span class="mi">999</span>  <span class="mf">0.863</span>
<span class="n">Residuals</span> <span class="mi">22</span> <span class="mf">0.091226</span> <span class="mf">0.0041466</span>  
</pre></div>
</div>
</div>
<figure class="align-default" id="id2">
<img alt="_images/permdisp.png" src="_images/permdisp.png" />
<figcaption>
<p><span class="caption-number">Fig. 78 </span><span class="caption-text">: Box plot showin the distance to the centroid for each point grouped per impact variable.</span><a class="headerlink" href="#id2" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="indicator-species">
<h3>6.4 Indicator species<a class="headerlink" href="#indicator-species" title="Permalink to this heading">#</a></h3>
<p>To determine which signals are contributing to the difference observed in our ordination and PERMANOVA analysis, we can run an Indicator Species Analysis (ISA) using the <code class="docutils literal notranslate"><span class="pre">indicspecies</span></code> R package. The Indicator Value (IndVal) is a simple calculation based on the specificity (occurring only in one group or multiple) and fidelity (how frequently is it detected in each group). We can specify <code class="docutils literal notranslate"><span class="pre">indvalcomp</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> in the summary to provide the information about the specificity (component <code class="docutils literal notranslate"><span class="pre">A</span></code>) and fidelity (component <code class="docutils literal notranslate"><span class="pre">B</span></code>). For the tutorial data, we will restrict the indicator species as the ones with perfect specificity (<code class="docutils literal notranslate"><span class="pre">At=1.0</span></code>) and fidelity (<code class="docutils literal notranslate"><span class="pre">Bt=1.0</span></code>).</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">#####################</span>
<span class="c1"># INDICATOR SPECIES #</span>
<span class="c1">#####################</span>
<span class="c1"># 1. run indicator species analysis</span>
<span class="n">indval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">multipatt</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">permanova.freqtable</span><span class="p">),</span><span class="w"> </span><span class="n">permanova.metadata</span><span class="o">$</span><span class="n">impact</span><span class="p">,</span><span class="w"> </span><span class="n">control</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">how</span><span class="p">(</span><span class="n">nperm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">9999</span><span class="p">))</span>
<span class="nf">summary</span><span class="p">(</span><span class="n">indval</span><span class="p">,</span><span class="w"> </span><span class="n">indvalcomp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">At</span><span class="o">=</span><span class="m">1.0</span><span class="p">,</span><span class="w"> </span><span class="n">Bt</span><span class="o">=</span><span class="m">1.0</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; summary(indval, indvalcomp = TRUE, At=1.0, Bt=1.0)

 Multilevel pattern analysis
 ---------------------------

 Association function: IndVal.g
 Significance level (alpha): 0.05
 Minimum positive predictive value (At): 1
 Minimum sensitivity (Bt): 1

 Total number of species: 865
 Selected number of species: 9 
 Number of species associated to 1 group: 9 

 List of species associated to each combination: 

 Group high  #sps.  4 
         A B stat p.value    
zotu.128 1 1    1   1e-04 ***
zotu.346 1 1    1   1e-04 ***
zotu.631 1 1    1   1e-04 ***
zotu.673 1 1    1   1e-04 ***

 Group low  #sps.  5 
         A B stat p.value    
zotu.117 1 1    1   1e-04 ***
zotu.145 1 1    1   1e-04 ***
zotu.2   1 1    1   1e-04 ***
zotu.260 1 1    1   1e-04 ***
zotu.436 1 1    1   1e-04 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1 
</pre></div>
</div>
</div>
<p>Finally, looping over the indicator species taxonomic IDs will provide information on which signals were deemed as indicative of low impact and high impact sites.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">lowindval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;zotu.117&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.145&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.2&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.260&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.436&#39;</span><span class="p">)</span>
<span class="n">highindval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;zotu.128&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.346&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.631&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;zotu.673&#39;</span><span class="p">)</span>
<span class="nf">for </span><span class="p">(</span><span class="n">item</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">lowindval</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">taxonomyTable</span><span class="p">[</span><span class="n">item</span><span class="p">,</span><span class="w"> </span><span class="p">])</span>
<span class="p">}</span>
<span class="nf">for </span><span class="p">(</span><span class="n">item</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">highindval</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">taxonomyTable</span><span class="p">[</span><span class="n">item</span><span class="p">,</span><span class="w"> </span><span class="p">])</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>           <span class="n">kingdom</span>     <span class="n">phylum</span>           <span class="k">class</span>           <span class="nc">order</span>           <span class="n">family</span>        <span class="n">genus</span>          <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.117</span> <span class="n">Eukaryota</span> <span class="n">Rhodophyta</span> <span class="n">Florideophyceae</span> <span class="n">Peyssonneliales</span> <span class="n">Peyssonneliaceae</span> <span class="n">Peyssonnelia</span> <span class="n">Peyssonnelia_sp</span><span class="o">.</span>  <span class="mf">93.59</span>   <span class="mi">99</span>
           <span class="n">kingdom</span>     <span class="n">phylum</span>   <span class="k">class</span>     <span class="nc">order</span>     <span class="n">family</span>     <span class="n">genus</span>       <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.145</span> <span class="n">Eukaryota</span> <span class="n">Arthropoda</span> <span class="n">Insecta</span> <span class="n">Blattodea</span> <span class="n">Termitidae</span> <span class="n">Amitermes</span> <span class="n">Amitermes_sp</span><span class="o">.</span>  <span class="mf">75.08</span>   <span class="mi">98</span>
         <span class="n">kingdom</span>     <span class="n">phylum</span>     <span class="k">class</span>     <span class="nc">order</span>         <span class="n">family</span>     <span class="n">genus</span>               <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">Eukaryota</span> <span class="n">Arthropoda</span> <span class="n">Arachnida</span> <span class="n">Opiliones</span> <span class="n">Nemastomatidae</span> <span class="n">Mitostoma</span> <span class="n">Mitostoma_chrysomelas</span> <span class="mf">76.596</span>   <span class="mi">89</span>
           <span class="n">kingdom</span>     <span class="n">phylum</span>           <span class="k">class</span>           <span class="nc">order</span>           <span class="n">family</span>      <span class="n">genus</span>        <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.260</span> <span class="n">Eukaryota</span> <span class="n">Rhodophyta</span> <span class="n">Florideophyceae</span> <span class="n">Peyssonneliales</span> <span class="n">Peyssonneliaceae</span> <span class="n">Polystrata</span> <span class="n">Polystrata_sp</span><span class="o">.</span> <span class="mf">91.054</span>  <span class="mi">100</span>
           <span class="n">kingdom</span>     <span class="n">phylum</span>        <span class="k">class</span>    <span class="nc">order</span>      <span class="n">family</span>    <span class="n">genus</span> <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.436</span> <span class="n">Eukaryota</span> <span class="n">Arthropoda</span> <span class="n">Malacostraca</span> <span class="n">Decapoda</span> <span class="n">Galatheidae</span> <span class="n">Galathea</span>   <span class="n">senta</span>    <span class="mi">100</span>  <span class="mi">100</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>           <span class="n">kingdom</span>          <span class="n">phylum</span>               <span class="k">class</span>           <span class="nc">order</span>           <span class="n">family</span>         <span class="n">genus</span>              <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.128</span> <span class="n">Eukaryota</span> <span class="n">Bacillariophyta</span> <span class="n">Coscinodiscophyceae</span> <span class="n">Coscinodiscales</span> <span class="n">Coscinodiscaceae</span> <span class="n">Coscinodiscus</span> <span class="n">Coscinodiscus_granii</span> <span class="mf">88.103</span>   <span class="mi">99</span>
           <span class="n">kingdom</span>   <span class="n">phylum</span>      <span class="k">class</span>    <span class="nc">order</span>    <span class="n">family</span>    <span class="n">genus</span>       <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.346</span> <span class="n">Eukaryota</span> <span class="n">Annelida</span> <span class="n">Polychaeta</span> <span class="n">Spionida</span> <span class="n">Spionidae</span> <span class="n">Polydora</span> <span class="n">Polydora_aura</span> <span class="mf">81.788</span>   <span class="mi">96</span>
           <span class="n">kingdom</span>   <span class="n">phylum</span>    <span class="k">class</span>        <span class="nc">order</span>    <span class="n">family</span>  <span class="n">genus</span>          <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.631</span> <span class="n">Eukaryota</span> <span class="n">Cnidaria</span> <span class="n">Hydrozoa</span> <span class="n">Leptothecata</span> <span class="n">Obeliidae</span> <span class="n">Obelia</span> <span class="n">Obelia_bidentata</span> <span class="mf">87.622</span>   <span class="mi">98</span>
           <span class="n">kingdom</span>   <span class="n">phylum</span>      <span class="k">class</span>    <span class="nc">order</span>    <span class="n">family</span>    <span class="n">genus</span>                <span class="n">species</span> <span class="n">pident</span> <span class="n">qcov</span>
<span class="n">zotu</span><span class="mf">.673</span> <span class="n">Eukaryota</span> <span class="n">Annelida</span> <span class="n">Polychaeta</span> <span class="n">Spionida</span> <span class="n">Spionidae</span> <span class="n">Polydora</span> <span class="n">Polydora_lingshuiensis</span>    <span class="mi">100</span>  <span class="mi">100</span>
</pre></div>
</div>
</div>
<p><strong>That is it for the introduction to the statistical analysis of metabarcoding data, Congratulations on making it this far in the eDNA workshop!</strong></p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="phylogenetictree.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Phylogenetic tree</p>
      </div>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reading-parsing-data">2. Reading/Parsing data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#summary-stats">3. Summary stats</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-exploration">4. Data exploration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#phyloseq-functions">4.1 Phyloseq functions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#bar-plots">4.1.1 Bar plots</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#heat-maps">4.1.2 Heat maps</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#richness">4.1.3 Richness</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-diversity">5. Alpha diversity</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#species-accumulation-curves">5.1 Species accumulation curves</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#alpha-diversity-comparison">5.2 Alpha diversity comparison</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#estimate-replication">5.3 Estimate replication</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#phylodiversity">5.4 Phylodiversity</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beta-diversity">6. Beta diversity</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ordination">6.1 Ordination</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#nmds">6.1.1 NMDS</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pcoa">6.1.2 PCoA</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#permanova">6.2 PERMANOVA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#permdisp">6.3 PERMDISP</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#indicator-species">6.4 Indicator species</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Dr Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>