

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Taxonomy assignment &#8212; eDNA workshop HKU</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'taxonomyassignment';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Data pre-processing" href="datapreprocessing.html" />
    <link rel="prev" title="Bioinformatic analysis" href="bioinformaticanalysis.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo3.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo3.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    <no title>
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinformaticanalysis.html">Bioinformatic analysis</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datapreprocessing.html">Data pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogenetictree.html">Phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="statisticalanalysis.html">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Ftaxonomyassignment.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button"
   title="Open an issue"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/taxonomyassignment.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Taxonomy assignment</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-databases">2. Reference databases</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crabs-workflow">2.1 CRABS workflow</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#help-documentation">2.2 Help documentation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-download-sequence-data">2.3 Step 1: download sequence data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-merge-downloaded-sequences">2.4 Step 2: merge downloaded sequences</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-extract-amplicons-through-in-silico-pcr">2.5 Step 3: extract amplicons through <em>in silico</em> PCR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-extract-amplicons-through-pairwise-global-alignments">2.6 Step 4: extract amplicons through pairwise global alignments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-assigning-a-taxonomic-id">2.7 Step 5: assigning a taxonomic ID</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-dereplicating-the-db">2.8 Step 6: dereplicating the DB</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-ref-db-cleanup">2.8 Step 7: ref DB cleanup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-8-export-ref-db">2.9 Step 8: export ref DB</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-coi-reference-db">3. The COI reference DB</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#taxonomic-classification">4. Taxonomic classification</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sintax">4.1 SINTAX</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#idtaxa">4.2 IDTAXA</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#input-formatting">4.2.1 Input formatting</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#classifier-training">4.2.2 Classifier training</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">4.2.3 Classifier training</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#blast">4.3 BLAST</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#online-blast">4.3.1 Online BLAST</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-remote">4.3.2 BLAST+ remote</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-local">4.3.3 BLAST+ local</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output-formatting">5. Output formatting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sintax-formatting">5.1 SINTAX formatting</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#idtaxa-formatting">5.2 IDTAXA formatting</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-formatting">5.3 BLAST formatting</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="taxonomy-assignment">
<h1>Taxonomy assignment<a class="headerlink" href="#taxonomy-assignment" title="Permalink to this heading">#</a></h1>
<section id="introduction">
<h2>1. Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">#</a></h2>
<p>As we have previously discussed, eDNA metabarcoding is the process whereby we amplify a specific gene region of the taxonomic group of interest from a DNA extract obtained from an environmental sample. In our tutorial data set, the COI sequence data originated from settlement plates that were placed in highly impacted and low-impact coastal regions in Hong Kong. After the bioinformatic processing we conducted in the previous section, we ended up with a frequency table (<strong>sequenceData/8-final/zotutable.txt</strong>) and a sequence file (<strong>sequenceData/8-final/zotus.fasta</strong>) containing all the biologically relevant sequences. The next step in the bioinformatic process will be to assign a taxonomic ID to each sequence to determine what species were detected through our eDNA metabarcoding analysis.</p>
<p>There are four basic strategies to taxonomy classification (and an endless number of variations  of those four basic strategies):</p>
<ol class="arabic simple">
<li><p>sequence similarity (SS);</p></li>
<li><p>sequence composition (SC);</p></li>
<li><p>phylogenetic (Ph);</p></li>
<li><p>and probabilistic (Pr).</p></li>
</ol>
<figure class="align-default" id="taxonomy-assignment-categories">
<img alt="_images/taxonomyassignmentcategories.png" src="_images/taxonomyassignmentcategories.png" />
<figcaption>
<p><span class="caption-number">Fig. 19 </span><span class="caption-text">: The four basic strategies to taxonomy classification. Copyright by <a class="reference external" href="https://onlinelibrary.wiley.com/doi/epdf/10.1111/1755-0998.13407">Hleap et al., 2021</a>.</span><a class="headerlink" href="#taxonomy-assignment-categories" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Of the four abovementioned strategies to taxonomy classification, the <strong>Sequency Similarity</strong> and <strong>Sequence Composition</strong> methods are currently the most frequently used in the metabarcoding research community. Multiple comparative experiments have been conducted to determine the most optimal approach to assign a taxonomic ID to metabarcoding data, though it seems to be that the main conclusion of those papers is that their own method is best. One example of a comparative study is the one from <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0470-z">Bokulich et al., 2018</a>.</p>
<figure class="align-default" id="taxonomy-classifier-comparison">
<img alt="_images/taxonomyclassifiercomparison.png" src="_images/taxonomyclassifiercomparison.png" />
<figcaption>
<p><span class="caption-number">Fig. 20 </span><span class="caption-text">: A comparison of taxonomy classifiers. Copyright by <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0470-z">Bokulich et al., 2018</a>.</span><a class="headerlink" href="#taxonomy-classifier-comparison" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Within this tutorial, we will cover three different ways to assign a taxonomic ID to our sequences. However, before doing so, we will need to create a reference database that will be used by the classifiers to our ZOTU list to the reference barcodes.</p>
</section>
<section id="reference-databases">
<h2>2. Reference databases<a class="headerlink" href="#reference-databases" title="Permalink to this heading">#</a></h2>
<p>Several reference databases are available online. The most notable ones being NCBI, EMBL, BOLD, SILVA, and RDP. However, recent research has indicated a need to use custom curated reference databases to increase the accuracy of taxonomy assignment. While certain gene-specific or primer-specific reference databases are available (RDP: 16S microbial; MIDORI: COI eukaryotes; etc.), this essential data source is missing in most instances. We will, therefore, show you how to build your own custom curated reference database using the <a class="reference external" href="https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.13741?af=R">CRABS</a> (Creating Reference databases for Amplicon-Based Sequencing) program. CRABS can be installed via <a class="reference external" href="https://github.com/gjeunen/reference_database_creator">GitHub</a>, Docker, and conda.</p>
<div class="admonition attention">
<p class="admonition-title">Attention</p>
<p>Since the online COI reference material is quite large, it being the standard metazoan barcode gene, we will not build a COI reference database during this tutorial. Instead, we will use a pre-assembled COI database. However, to show you how to build your own database, we will go through the steps to build a 16S Chondrichthyes (sharks and rays) database, as the online reference material is small. Therefore, the steps can be accomplished quite quickly.</p>
</div>
<section id="crabs-workflow">
<h3>2.1 CRABS workflow<a class="headerlink" href="#crabs-workflow" title="Permalink to this heading">#</a></h3>
<p>The CRABS workflow consists of seven parts, including:</p>
<ol class="arabic simple">
<li><p>Downloading and importing data from multiple online repositories, including NCBI, BOLD, EMBL, MitoFish, etc.;</p></li>
<li><p>Retrieving amplicon regions through <em>in silico</em> PCR analysis;</p></li>
<li><p>Retrieving amplicons with missing primer-binding regions through <em>Pairwise Global Alignments</em> with amplicons from the <em>in Silico</em> PCR analysis as seed sequences;</p></li>
<li><p>Generating the taxonomic lineage for amplicons;</p></li>
<li><p>Curating the database via multiple filtering parameters;</p></li>
<li><p>Post-processing functions and visualizations to provide a summary overview of the final reference database; and</p></li>
<li><p>Exporting the custom curated reference database in various formats covering most format requirements of taxonomic assignment tools.</p></li>
</ol>
<figure class="align-default" id="id1">
<img alt="_images/crabsworkflow.png" src="_images/crabsworkflow.png" />
<figcaption>
<p><span class="caption-number">Fig. 21 </span><span class="caption-text">: The CRABS worflow to create your own reference database. Copyright by <a class="reference external" href="https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.13741?af=R">Jeunen et al., 2022</a>.</span><a class="headerlink" href="#id1" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="help-documentation">
<h3>2.2 Help documentation<a class="headerlink" href="#help-documentation" title="Permalink to this heading">#</a></h3>
<p>To bring up the general help documentation for CRABS, we can use the <code class="docutils literal notranslate"><span class="pre">--help</span></code> or <code class="docutils literal notranslate"><span class="pre">-h</span></code> command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--version
crabs<span class="w"> </span>--help
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">base</span><span class="p">)</span> <span class="n">gjeunen</span><span class="nd">@MacBook</span><span class="o">-</span><span class="n">Pro</span><span class="o">-</span><span class="mi">16</span> <span class="n">tutorial</span> <span class="o">%</span> <span class="n">crabs</span> <span class="o">--</span><span class="n">version</span>
<span class="mf">0.1.8</span>


<span class="p">(</span><span class="n">base</span><span class="p">)</span> <span class="n">gjeunen</span><span class="nd">@MacBook</span><span class="o">-</span><span class="n">Pro</span><span class="o">-</span><span class="mi">16</span> <span class="n">tutorial</span> <span class="o">%</span> <span class="n">crabs</span> <span class="o">--</span><span class="n">help</span>
<span class="n">usage</span><span class="p">:</span> <span class="n">crabs</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">version</span><span class="p">]</span> <span class="p">{</span><span class="n">db_download</span><span class="p">,</span><span class="n">db_import</span><span class="p">,</span><span class="n">db_merge</span><span class="p">,</span><span class="n">insilico_pcr</span><span class="p">,</span><span class="n">pga</span><span class="p">,</span><span class="n">assign_tax</span><span class="p">,</span><span class="n">dereplicate</span><span class="p">,</span><span class="n">seq_cleanup</span><span class="p">,</span><span class="n">db_subset</span><span class="p">,</span><span class="n">visualization</span><span class="p">,</span><span class="n">tax_format</span><span class="p">}</span> <span class="o">...</span>

<span class="n">creating</span> <span class="n">a</span> <span class="n">curated</span> <span class="n">reference</span> <span class="n">database</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="p">{</span><span class="n">db_download</span><span class="p">,</span><span class="n">db_import</span><span class="p">,</span><span class="n">db_merge</span><span class="p">,</span><span class="n">insilico_pcr</span><span class="p">,</span><span class="n">pga</span><span class="p">,</span><span class="n">assign_tax</span><span class="p">,</span><span class="n">dereplicate</span><span class="p">,</span><span class="n">seq_cleanup</span><span class="p">,</span><span class="n">db_subset</span><span class="p">,</span><span class="n">visualization</span><span class="p">,</span><span class="n">tax_format</span><span class="p">}</span>

<span class="n">options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">--</span><span class="n">version</span>             <span class="n">show</span> <span class="n">program</span><span class="s1">&#39;s version number and exit</span>
</pre></div>
</div>
</div>
<p>The output from this command shows the functions that are available within CRABS. To get a detailed overview and bring up the documentation of one of the functions, we can use the <code class="docutils literal notranslate"><span class="pre">--help</span></code> or <code class="docutils literal notranslate"><span class="pre">-h</span></code> parameter with the function.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>db_download<span class="w"> </span>-h
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">crabs</span> <span class="n">db_download</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">-</span><span class="n">s</span> <span class="n">SOURCE</span> <span class="p">[</span><span class="o">-</span><span class="n">db</span> <span class="n">DATABASE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUERY</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">SPECIES</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="n">MARKER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">k</span> <span class="n">ORIG</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span> <span class="n">DISCARD</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span> <span class="n">BOLDGAP</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EMAIL</span><span class="p">]</span>
                         <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BATCHSIZE</span><span class="p">]</span>

<span class="n">downloading</span> <span class="n">sequence</span> <span class="n">data</span> <span class="kn">from</span> <span class="nn">online</span> <span class="n">databases</span>

<span class="n">options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
  <span class="o">-</span><span class="n">s</span> <span class="n">SOURCE</span><span class="p">,</span> <span class="o">--</span><span class="n">source</span> <span class="n">SOURCE</span>
                        <span class="n">specify</span> <span class="n">online</span> <span class="n">database</span> <span class="n">used</span> <span class="n">to</span> <span class="n">download</span> <span class="n">sequences</span><span class="o">.</span> <span class="n">Currently</span> <span class="n">supported</span> <span class="n">options</span> <span class="n">are</span><span class="p">:</span> <span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="n">ncbi</span><span class="p">,</span> <span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="n">embl</span><span class="p">,</span> <span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="n">mitofish</span><span class="p">,</span> <span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">bold</span><span class="p">,</span> <span class="p">(</span><span class="mi">5</span><span class="p">)</span>
                        <span class="n">taxonomy</span>
  <span class="o">-</span><span class="n">db</span> <span class="n">DATABASE</span><span class="p">,</span> <span class="o">--</span><span class="n">database</span> <span class="n">DATABASE</span>
                        <span class="n">specific</span> <span class="n">database</span> <span class="n">used</span> <span class="n">to</span> <span class="n">download</span> <span class="n">sequences</span><span class="o">.</span> <span class="n">Example</span> <span class="n">NCBI</span><span class="p">:</span> <span class="n">nucleotide</span><span class="o">.</span> <span class="n">Example</span> <span class="n">EMBL</span><span class="p">:</span> <span class="n">mam</span><span class="o">*.</span> <span class="n">Example</span> <span class="n">BOLD</span><span class="p">:</span> <span class="n">Actinopterygii</span>
  <span class="o">-</span><span class="n">q</span> <span class="n">QUERY</span><span class="p">,</span> <span class="o">--</span><span class="n">query</span> <span class="n">QUERY</span>
                        <span class="n">NCBI</span> <span class="n">query</span> <span class="n">search</span> <span class="n">to</span> <span class="n">limit</span> <span class="n">portion</span> <span class="n">of</span> <span class="n">database</span> <span class="n">to</span> <span class="n">be</span> <span class="n">downloaded</span><span class="o">.</span> <span class="n">Example</span><span class="p">:</span> <span class="s2">&quot;16S[All Fields] AND (&quot;</span><span class="mi">1</span><span class="s2">&quot;[SLEN] : &quot;</span><span class="mi">50000</span><span class="s2">&quot;[SLEN])&quot;</span>
  <span class="o">-</span><span class="n">p</span> <span class="n">SPECIES</span><span class="p">,</span> <span class="o">--</span><span class="n">species</span> <span class="n">SPECIES</span>
                        <span class="n">species</span> <span class="n">to</span> <span class="n">be</span> <span class="n">downloaded</span><span class="p">,</span> <span class="n">either</span> <span class="k">as</span> <span class="n">a</span> <span class="n">string</span> <span class="n">separated</span> <span class="n">by</span> <span class="s2">&quot;+&quot;</span> <span class="ow">or</span> <span class="k">as</span> <span class="n">a</span> <span class="nb">list</span> <span class="ow">in</span> <span class="n">a</span> <span class="o">.</span><span class="n">txt</span> <span class="n">file</span>
  <span class="o">-</span><span class="n">m</span> <span class="n">MARKER</span><span class="p">,</span> <span class="o">--</span><span class="n">marker</span> <span class="n">MARKER</span>
                        <span class="n">genetic</span> <span class="n">marker</span> <span class="n">to</span> <span class="n">download</span> <span class="k">for</span> <span class="n">BOLD</span>
  <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
                        <span class="n">output</span> <span class="n">file</span> <span class="n">name</span>
  <span class="o">-</span><span class="n">k</span> <span class="n">ORIG</span><span class="p">,</span> <span class="o">--</span><span class="n">keep_original</span> <span class="n">ORIG</span>
                        <span class="n">keep</span> <span class="n">original</span> <span class="n">downloaded</span> <span class="n">file</span><span class="p">,</span> <span class="n">default</span> <span class="o">=</span> <span class="s2">&quot;no&quot;</span>
  <span class="o">-</span><span class="n">d</span> <span class="n">DISCARD</span><span class="p">,</span> <span class="o">--</span><span class="n">discard</span> <span class="n">DISCARD</span>
                        <span class="n">output</span> <span class="n">filename</span> <span class="k">for</span> <span class="n">sequences</span> <span class="k">with</span> <span class="n">incorrect</span> <span class="n">formatting</span><span class="p">,</span> <span class="n">default</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">saved</span>
  <span class="o">-</span><span class="n">g</span> <span class="n">BOLDGAP</span><span class="p">,</span> <span class="o">--</span><span class="n">boldgap</span> <span class="n">BOLDGAP</span>
                        <span class="n">incorporate</span> <span class="ow">or</span> <span class="n">discard</span> <span class="n">sequences</span> <span class="k">with</span> <span class="n">gaps</span> <span class="kn">from</span> <span class="nn">BOLD</span> <span class="p">(</span><span class="n">DISCARD</span><span class="o">/</span><span class="n">INCORPORATE</span><span class="p">)</span><span class="o">.</span> <span class="n">Default</span> <span class="o">=</span> <span class="n">DISCARD</span>
  <span class="o">-</span><span class="n">e</span> <span class="n">EMAIL</span><span class="p">,</span> <span class="o">--</span><span class="n">email</span> <span class="n">EMAIL</span>
                        <span class="n">email</span> <span class="n">address</span> <span class="n">to</span> <span class="n">connect</span> <span class="n">to</span> <span class="n">NCBI</span> <span class="n">servers</span>
  <span class="o">-</span><span class="n">b</span> <span class="n">BATCHSIZE</span><span class="p">,</span> <span class="o">--</span><span class="n">batchsize</span> <span class="n">BATCHSIZE</span>
                        <span class="n">number</span> <span class="n">of</span> <span class="n">sequences</span> <span class="n">downloaded</span> <span class="kn">from</span> <span class="nn">NCBI</span> <span class="n">per</span> <span class="n">iteration</span><span class="o">.</span> <span class="n">Default</span> <span class="o">=</span> <span class="mi">5000</span>
</pre></div>
</div>
</div>
</section>
<section id="step-1-download-sequence-data">
<h3>2.3 Step 1: download sequence data<a class="headerlink" href="#step-1-download-sequence-data" title="Permalink to this heading">#</a></h3>
<p>As a first step, we will download the 16S shark sequences from the NCBI online repository, as well as all sequences from the Japanese MitoFish database. Downloading the Japanese MitoFish databases is quite straightforward, as it is a single file. We can use the <code class="docutils literal notranslate"><span class="pre">crabs</span> <span class="pre">db_download</span> <span class="pre">-s</span> <span class="pre">mitofish</span></code> command to accomplish this. The only other parameter we need to provide is the output file name.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>db_download<span class="w"> </span>-s<span class="w"> </span>mitofish<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mitofish.fasta
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>downloading sequences from the MitoFish database
complete_partial_mitogenomes.zip        100%[===============================================================================&gt;]  79.27M  6.05MB/s    in 13s     
formatting mito-all to CRABS format
 93%|██████████████████████████████████████████████████████████████████████████████████████████████▊       | 781202771/840651415 [00:05&lt;00:00, 140055187.99it/s]
found 0 sequences with incorrect accession format
written 825365 sequences to sequenceData/7-refdb/mitofish.fasta
</pre></div>
</div>
</div>
<p>To download data from the NCBI servers, we need to provide a few additional parameters to CRABS. Maybe one of the trickiest ones to understand is the <code class="docutils literal notranslate"><span class="pre">--query</span></code> parameter. We can get the input string for this parameter from the NCBI website, as shown on the screenshot below.</p>
<figure class="align-default" id="db-download-query">
<img alt="_images/dbdownloadquery.png" src="_images/dbdownloadquery.png" />
<figcaption>
<p><span class="caption-number">Fig. 22 </span><span class="caption-text">: A screenshot on how to obtain the query string to download sequence data from the NCBI servers through CRABS.</span><a class="headerlink" href="#db-download-query" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>db_download<span class="w"> </span>-s<span class="w"> </span>ncbi<span class="w"> </span>-db<span class="w"> </span>nucleotide<span class="w"> </span>-q<span class="w"> </span><span class="s1">&#39;16S[All Fields] AND (&quot;Chondrichthyes&quot;[Organism] OR Chondrichthyes[All Fields])&#39;</span><span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/ncbi16Schondrichthyes.fasta<span class="w"> </span>-e<span class="w"> </span>gjeunen@gmail.com
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>downloading sequences from NCBI
&quot;--species&quot; parameter not provided, downloading NCBI data based solely on &quot;--query&quot;: 16S[All Fields] AND (&quot;Chondrichthyes&quot;[Organism] OR Chondrichthyes[All Fields])
looking up the number of sequences that match the query
found 3300 number of sequences matching the query
starting the download
temp_ncbi16Schondrichthyes.fasta_1.fasta           [                                                                                  &lt;=&gt;                ]  28.92M   310KB/s    in 33s     
formatting the downloaded sequencing file to CRABS format
 97%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋   | 29553742/30322979 [00:00&lt;00:00, 275397805.14it/s]
found 0 sequences with incorrect accession format
written 3300 sequences to ncbi16Schondrichthyes.fasta
</pre></div>
</div>
</div>
<p>When downloading data from online repositories, CRABS will immediately reformat the data to a simple 2-line fasta format whereby the header information only contains the unique identifier (accession number or a CRABS number if the accession number is not available). If you have generated your own reference barcodes, you can import them into CRABS format using the <code class="docutils literal notranslate"><span class="pre">db_import</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">2</span><span class="w"> </span>sequenceData/7-refdb/ncbi16Schondrichthyes.fasta
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">OR575552</span>
<span class="n">GCTAGCGTAGCTTATCGAAAGCATGGCACTGAAGATGCTAAGATGAAAAATAAAAATTTCCGCGGGCATGAGAGGCTTGGTCCTGGCCTTAGTGTTAATTGTAATTAGGATTATACATGCAAGTCTCAGCACACCTGTGAGAATGCCCTCATTAGTCTATTAACTAATCAGGAGCTGGTATCAGGCACACGCACGTAGCCCAAGACACCTTGCTAAGCCACACCCCCAAGGGATTTCAGCAGTAATAAACATTGATTTATAAGCGCAAGCTTGAATCAGTTAGGATTAACTGGGTTGGTAAATCTCGTGCCAGCCACCGCGGTTATACGAGTAACCCTTATTAACACTCTACCGGCGTAAAGAGTGGTTTAAGAATTTTTCCCTATGAACTAAAGTTAAGACCCCATTAAGCTGTTATACGCCCTCATGGGTGGAATCACCAACAACGAAAGTAACTTTACAGCACCAGAAAACCTTGACCCCACGACAGTTGAGCCCCAAACTAGGATTAGATACCCTACTATGCTTAACCCCAAACTTAGATAATAACCCACCTATTATCCGCCAGAGTACTACAAGCGCTAGCTTAAAACCCAAAGGACTTGGCGGTGCCTCAGACCCACCTAGAGGAGCCTGTTCTATAATCGATACTCCCCGTTTAACCTCACCACATCTTGCCAATACCGCCTATATACCGCCGTCGTCAGCCCACCCTGTGAGGGTTAAGAAGTAAGCAAAAAGAATTAACTTCCACACGTCAGGTCAAGGTGTAGCGAATGAAGTGGAAAGAAATGGGCTACATTTTCTATTAAGAAAACACGAACAGTAAACTGAAAAATTACTCGAAGGTGGATTTAGCAGTAAGAAAAATTAGAGTATTTTTCTGAAGCCGGCTCTGAAGCGCGCACACACCGCCCGTCACTCTCCTTAAAAAAATACTAATTTATACTTAAACAAAATTATTTTACAAGAGGAGGCAAGTCGTAACATGGTAAGTGTACTGGAAAGTGCACTTGGAATCAAAATGTAGCTAATTTAGTAAAGCACCTCCCTTACACCGAGGAAAAACCCGTGCAATTCGAGTCATTCTGAACATTAAAGCTAGCCTAATTATCATATAAACACAACCCCATTAATTACACACCATTTTTACCAGAACTAAAACATTTTACACTTTTTAGTATAGGCGATAGAACGATAAACCAGCGCAATAGATTAAGTACCGCAAGGGAAAACTGAAAAAGAAATGAAACAAATCATTAAAGTAATAAAAAGCAGAGATTTAACCTCGTACCTTTTGCATCATGATTTAGCTAGATCAATTAGGCAAAAAGATCTTAAGTCTACCTTCCCGAAACTAAACGAGCTACTCCGAAGCAGCAAATTAGAGCCAACCCGTCTCTGTGGCAAAAGAGTGGGAAGACTTCCGAGTAGTGGTGACAGACCTATCGAGTTTAGTGATAGCTGGTTACCCAAAAAAAGAACTTAAATTCTGCATTAATTTTTTACCCACCAATAAGAACATCTCCACAAGATGGTAAATAAGAATTAATAGTTATTTAGAAGAGGTACAGCCCTTCTAAACCAAGATACAACTTTTAAAGGCGGGTAATGATCACATTTTACTAAGGTTTACCCACCAGTAGGCCTGAAAGCAGCCATCTGTAAAGTAAGCGTCACAGCTCCAGCTCGACATAAACCTATAATCTAGATACCACCTCACAACCCCCTACAAAATACTGGGTTATTTTATACAACTATAAAAGAACTTATGCTAAAATGAGTAATAAGAGACCAAACCTCTCCCAACACAAGTGTACATCAGAAAGAATTAAATCACTGATAATTAAACGAATCCAAACTGAGGGTATAATATTAACTCACCACCAACTAGAAAACCTTATTTAACCATTCGTTAATCCTACACAGGAGTGTCCTCAAGGAAAGATTAAAAGAAAATAAAGGAACTCGGCAAACACAAACTCCGCCTGTTTACCAAAAACATCGCCTCTTGCTTACTATAAGAGGTCCCGCCTGCCCTGTGACAATGTTCAACGGCCGCGGTATTTTGACCGTGCGAAGGTAGCGTAATCACTTGTCTTTTAAATGAAGACTTGTATGAAAGGCATCACGAGAGTTTAACTGTCTCTATTTTCTAATCAATGAAATTGATCTTCTCGTGCAGAAGCGAGAATAACACCATTAGACGAGAAGACCCTATGGAGCTTCAAACACTTAAATTAATTATGTAAAATTTTTAACTCTCCGGAGAAAAACCACCCATATAATACCCCTAATTTAACTGTTTTTGGTTGGGGTGACCAGGGGGAAAAAATTATCCCCCCCATCGATTGAGTACTCAGTACTTAAAAATTAGAATGACAACTCTAATTAATAAAACATTTACCGAAACATGACCCAGAATTTATTCTGATCAATGAACCAAGTTACCCTAGGGATAACAGCGCAATCCTTTCCCAGAGTCCCTATCGACGAAAGGGTTTACGACCTCGATGTTGGATCAGGACATCCTAATGGTGCAACCGCTATTAAGGGCTCGTTTGTTCAACGATTAATAGTCCTACGTGATCTGAGTTCAGACCGGAGAAATCCAGGTCAGTTTCTATCTATGAATTAATTTTTCCTAGTACGAAAGGACCGGAAAGATGGAGCCAATACCATGAGCACGCTCCATTTTCACCTACTGAAATAAACTAAATTAGGTAAGAAAACATTACCTAAAACCCGAGACAAGGGTTATTGGGGTGGCAGAGCCTGGTAAATGCAAAAGACCTAAATCCTTTAATCCAGAGGTTCAAATCCTCTCCCCAATTATGCTCCAATACATTTTACTGTACCTCGTAAATCCTCTTGCCTACATCATCCCAGTTCTTTTAGCAACAGCCTTCCTAACCTTGGTCGAACGAAAAATCCTAGGCTACATACAATTCCGCAAGGGCCCCAACATCGTAGGACCATACGGCCTCCTTCAACCAATTGCAGACGGCGTAAAACTTTTTATTAAAGAACCAGTCCGCCCATCAACATCCTCACCCCTCTTGTTCCTAGCAACCCCAACCCTAGCACTAACCCTAGCCCTCCTTTTATGAATACCTCTCCCCCTACCACACTCAATCATCAACTTAAACTTAGGTCTACTATTCATCCTAGCCATCTCAAGCCTAACTGTTTACACAATCCTAGGCTCAGGATGAGCATCCAACTCAAAATATGCCCTCATAGGAGCACTACGAGCCGTAGCACAAACTATTTCCTATGAAGTAACTCTCGGCCTCATCCTCCTATCGATAATCATCTTCACGGGTGGATTCACACTCCACACATTCAACCTAACTCAGGAGGCAATTTGATTGTTTATCCCAGGCTGACCACTTGCTATAATATGGTATATCTCAACCCTGGCAGAAACCAACCGAGCACCGTTCGACTTAACCGAGGGAGAATCAGAACTGGTCTCAGGGTTTAATGTTGAATATGCAGCTGGCTCATTTGCCTTATTCTTCTTAGCCGAATACACAAACATCCTAATAATAAATACACTCTCAGTGATTCTATTTTTAGGCTCCTCCTATAACCCCTCTATACCACAAATCTCCACATTTTACTTAATGATAAAAGCAACACTACTCACCCTAATTTTTTTATGGGTGCGAGCATCATACCCCCGCTTCCGCTATGACCAACTTATACACCTAGTGTGAAAAAATTTCCTGCCACTCACCCTAGCCCTCATCCTATGACACGCAACCCTACCGCTCGCTGTAGTGAGCCTCCCCCCCGCAACCTAAACGGAAACGTGCCTGAATTAAAGGACCACTTTGATAGAGTGGGTCATGAGAGTTAAAATCCCTCCGCTTCCTCCTAGAAAAACAGGACTTGAACCTGCACCCGAGAGATCAAAACTCCCGGTACTTCCTATTATACTATCTCCTAAGTAAAGTCAGCTAATTAAGCTTTTGGGCCCATACCCCAACGATGTTGGTTAAAATCCTTCCTCTACTAATGAACCCTCTAGTACTAACCACTATTATCTCAAGCCTTGGCCTAGGAACCACAATAACATTCATCGGATCCCATTGACTTCTAGTTTGAATAGGTCTTGAAATTAACACCTTAGCCATTATTCCCCTAATAATCTACCACCACCACCCGCGAGCAGTAGAAGCAACCACAAAATATTTCCTAACACAAGCAACTGCCTCCGCCCTACTTTTATTTGCTAGCATCACAAACGCCTGAGACTTGGGCGAATGAAATTTAATTGAATTAATTAACCCAACCTCTGCCACACTAGCAACCATCGCATTAGCTTTAAAAATCGGACTTGCACCACTACACTTTTGACTACCCGAGGTCCTTCAGGGACTAAACCTAACCACAGGCCTGGTCCTAGCCACGTGACAAAAACTCGCCCCATTTGCCATCCTCCTACAACTCCACCCCATCCTTGACTCAAACTTACTGTTATTCCTAGGCATTTTATCAATCATTGTCGGAGGTTGGGGGGGACTCAACCAAACCCAACTACGAAAAATTCTGGCTTACTCATCAATCGCCCACTTGGGATGAATAATTACAATCCTACACTACTCCCCAAACCTCACCAAGCTCAACCTCACTTTATATATTATTATGACCCTCACAACCTTCCTCCTATTTAAACTATTCGACTCCACCAAAATCAACTCCATCGCATCCTCCTCAATCAAATCCCCCCTCCTTTCTGTTATTACATTAATAACCCTCCTCTCCCTAGCCGGCCTGCCACCCCTCTCCGGATTTATACCAAAATGATTAATTTTACAAGAACTAGTTAAACAAAACCTAGCCGCCCCTGCCACAATCATAGCCCTCGCAACACTACTAAGCTTGTTCTTCTACCTACGCCTATCTTACTCAACAACCCTAACAATATCCCCAAATTCTATCTACTTATCATCATCTTGACGAACAAAATCAATACAACAAAATCTTCCCCTAATAATAACCGCCTCTATTACCATCCTACTCCTGCCACTAACCCCAGCTATCTTTATATTAACCCTCTAAGAAATTTAGGCTAAACAGACCAAAAGCCTTCAAAGCTTTCAGTAGGAGTGGAAATCCCCTAATTTCTGCTAAGATTTGCAAGACTTTATCTCACATCTTCTGAATGCAACCCAGATACTTTCATTAAGCTAAAACCTTCTAGATAAATAGGCCTTGATCCTACAAAATCTCAGTTAACAGCTAAGCGTTCAATCCAGCGAACTTTTATCTAGGCTTCCTCCCGCCGTCAATACGGGGAGGCGGGATGAAGCCCCGGGAGGAGGAAACCTCCGGTTTTGGATTTGCAATCCAACGTAACGTCTACTGCAGGACTTGGTAAGAAGAGGAATTTAACCCCTGTTTACGGGGTTACAACCCGCCACTTAGTTCTCAGCCATCTTACCTGTGGCAATTAACCGTTGATTCTTTTCTACAAATCATAAAGATATCGGCACCCTATACTTGATCTTTGGTGCATGGGCAGGAATAGTAGGCACAGCCCTTAGCCTACTAATTCGAGCTGAATTAAGTCAGCCAGGATCCCTTCTTGGTGATGATCAAATCTACAATGTAATTGTGACTGCTCACGCTTTCGTAATAATCTTTTTTATAGTTATACCTGTAATAATTGGTGGGTTTGGAAATTGACTAGTGCCTTTAATAATTGGCGCACCAGATATAGCCTTCCCACGAATAAATAACATAAGCTTTTGATTACTACCCCCGTCGCTCCTTCTACTCTTAGCCTCCGCCGGCGTAGAAGCAGGAGCTGGAACTGGTTGAACGGTCTACCCACCACTCGCCAGTAATATAGCCCACGCGGGGGCATCTGTAGACTTAGCAATTTTTTCACTCCATCTAGCTGGTATCTCCTCAATTCTAGCCTCTATTAACTTTATCACAACCATCATCAACATGAAACCACCCGCCATTTCTCAATACCAAACACCACTTTTTGTTTGATCCATCTTAGTGACAACTGTCCTGCTTCTCCTGGCCCTCCCAGTCCTTGCAGCTGCAATTACAATACTGCTAACTGACCGTAATTTAAATACAACATTTTTTGACCCGTCAGGGGGAGGAGACCCCATCCTATATCAACACTTATTTTGATTCTTTGGTCACCCAGAAGTTTACATTTTAATTTTACCAGGCTTCGGAATAATTTCACACGTGGTAGCCTACTACTCAGGGAAAAAAGAACCTTTTGGCTACATGGGAATAGTCTGAGCAATAATAGCAATCGGTTTACTAGGCTTTATTGTCTGAGCCCATCATATATTTACGGTTGGGATAGACGTCGACACCCGCGCCTACTTCACATCAGCAACAATAATTATTGCTATCCCAACAGGCGTGAAAGTTTTTAGTTGACTAGCGACCCTGCATGGGGGGTCAATTAAATGAGAAACCCCACTCCTTTGGGCTTTAGGTTTCATTTTCTTATTCACAGTTGGCGGCCTCACTGGAATTGTTTTAGCCAACTCCTCTTTAGACATCGTTCTCCATGACACTTATTATGTAGTAGCCCACTTCCACTATGTTTTATCAATAGGGGCAGTATTTGCAATTATAGCGGGCTTCATCCACTGATTCCCTCTATTCTCGGGGTACACCCTTCACTCAACATGAACAAAAACCCAATTCCTAGTAATATTTATTGGGGTAAATTTAACTTTCTTCCCCCAACACTTCTTAGGCTTAGCCGGCATACCTCGACGGTACTCAGACTACCCAGATGCATACGCCCTTTGAAACACGGTTTCATCAATTGGCTCATTAATTTCATTAGTTGCCGTAATTATGTTCCTATTTATTATTTGAGAAGCATTTGCCGCTAAACGCGAAGTTCTATCTGTAGAATTACCGCACACAAACGTGGAGTGACTACACGGCTGCCCGCCACCCTATCACACCTACGAAGAGCCAGCATTTGTACAAATCCAACGAACCCATTTCTAACAAGGAAGGGAGGAATTGAACCCCCGTATATTGATTTCAAGTCAATCACATCACCACTCTGTCACTTTCTTTAAGATTCTAGTAAAATATATTACACTGCCTTGTCAGGGCATAGTCATGAGTTTAAACCTCGTGTATCTTATTTTACAATGGCACACCCAGCACAATTAGGATTTCAAGACGCAGCCTCCCCAGTTATGGAAGAACTTCTACATTTTCATGACCACACACTAATAATCGTATTTTTAATTAGTACTTTGGTCCTTTATATTATCTTAGCAATGGTATCCACTAAACTCACAAACAAATACATTTTAGATTCACAAGAAATCGAAATAGTGTGAACTATTCTCCCTGCTGTGATTCTTATTATAATCGCCCTTCCATCATTACGAATTTTATACCTCATAGATGAGATCAACGACCCACACCTAACCATTAAAGCCATGGGCCACCAATGATATTGAAGCTATGAATACACAGACTACGAAGACCTGGGGTTCGACTCCTACATAATCCAAACTCAAGATTTAGCCCCCGGCCAATTTCGCCTGCTAGAAACTGACCATCGGATAGTTGTACCAATAGAATCCCCAATTCGTGTCTTAGTATCCGCAGAAGATGTACTTCACTCATGAACAGTTCCAGCCCTGGGAGTCAAAATAGATGCAGTACCGGGACGACTAAATCAAACAGCCTTTATTGTCTCCCGCCCAGGTGTGTACTACGGTCAATGCTCGGAAATTTGCGGCGCCAATCATAGCTTTATACCCATCGTAGTTGAAGCAGTCCCACTAGAACACTTTGAATCCTGATCTTCATTAATACTAGAAGAAGCCTCACTAAGAAGCTATACTGGGTTTTAGCATTAGCCTTTTAAGCTAAAAATTGGTGACTCCCCACCACCCTTAGTGACATGCCCCAGTTAAACCCAAATCCATGATTTGCCATTTTCATACTCTCATGAATTTTCTTCCTAGCTATTTTACCAAAAAAAGTGATAACCTATCTATTTAATAATAACCCAACACCAAAAAGCGCTGAAAAACCTAAACCTGAACCCTGAAACTGACCATGAACCTAAGCTTTTTTGACCAATTCTTGAGCCCATCACTTCTTGGAGTCCCACTGATTGCCCTAGCAATCATAATCCCATGGTTAATCTTCCCAGCCCCAACAAACCGCTGACTTAATAACCGCCTCGTGACAATCCAATCATGATTCATCAACCGCTTTACACATCAGCTAATACAACCTATAAATTTTGGAGGTCACAAGTGAGCCACCATTTTAACAGCCCTCATACTATTCTTGATTACTGTTAACCTTCTTGGCCTACTCCCATACACCTTCACACCCACAACCCAATTATCACTTAACATAGCATTTGCCATCCCACTTTGATTAACAACCGTCCTCATCGGAATACTTAACCAACCAACAATTGCCCTTGGACACCTTCTCCCAGAAGGCACTCCGACCCCACTAATTCCAATCCTAATTATCATCGAAACAATTAGCTTGTTCATCCGACCCCTAGCACTAGGCGTCCGACTTACTGCAAACCTGACAGCCGGCCACCTCCTCATACAACTGATTGCCACCGCAGCCTTCGTACTAATTACCATCATGCCAGCCGTAGCCATACTAACCTCATTAATCTTATTCCTATTAACAATTTTAGAAGTAGCCGTGGCAATAATTCAAGCTTACGTATTTGTCCTATTACTAAGTCTCTATCTACAAGAAAACGTCTAATGGCTCACCAAGCACATGCATATCATATAGTCGACCCCAGCCCATGACCCCTAACAGGAGCTGTTGCTGCCTTACTAATAACATCAGGCCTAGCCGTCTGATTTCACTTCCACTCTTTATACCTCCTCTACCTCGGATTAATTCTTCTATTCTTGACCATGATTCAATGGTGGCGAGATGTCATTCGAGAAGGAACATTTCAAGGCCACCACACACCACCTGTTCAGAAAGGACTTCGCTATGGAATAATTTTATTTATTACATCAGAAGTTTTCTTTTTCCTTGGCTTCTTCTGAGCCTTCTACCACTCAAGCCTTGCCCCCACACCAGAACTAGGTGGATGCTGACCACCCACAGGAATCTACCCACTAGACCCCTTCGAAGTCCCCTTATTAAACACTGCAGTCCTTCTTGCATCCGGGGTCACAGTAACCTGAGCCCACCACAGCTTAATAGAAGGTAACCGAAAAGAAGCAATCCAAGCCCTAACTCTAACCGTATTATTAGGATTTTATTTTACAGCTCTTCAAGCCATAGAATACTACGAAGCTCCATTCACCATTGCTGATGGGGTTTATGGTTCAACATTCTTTGTTGCCACAGGTTTTCATGGTCTCCATGTTATTATTGGTTCAACATTTTTAATAGTTTGTCTAATACGACAAATCCAATATCATTTTACATCAGAACATCACTTTGGCTTTGAAGCCGCAGCATGATACTGACATTTTGTCGATGTAGTATGATTATTCCTTTATGTTTCCATCTACTGATGAGGTTCATAATTACTTTTCTAGTAAAAACTAGTACAAGTGATTTCCAATTACTTAATCTTGGTTAAAATCCAAGGGAAAGTAATGAACCTCATCATGTCATCAGTCGCAACCACGGCCCTCATTTCCCTAATCCTCGCTTTCATAGCATTTTGACTCCCATCACTAAACCCAGACAACGAAAAACTATCACCCTACGAATGCGGGTTTGACCCACTAGGGAGTGCACGCCTCCCGTTTTCATTACGTTTTTTCCTCGTGGCAATTTTATTTCTCCTGTTCGACCTCGAAATCGCCCTCCTCCTCCCACTGCCCTGAGGGAATCAACTATTAACACCACCCACTTCACTTCTATGAGCAACAAGCATCATTATTCTATTAACCCTGGGCCTCGTCTACGAGTGGTTTCAAGGTGGCCTTGAATGAGCAGAGTGGGTATTTAGTCCAAACAAGACCACTAATTTCGGCTTAGTAAATTATGGCGAAACACCATAAATACCTTATGTCCCCCATCTACTTCAGCTTTAGCTCAGCATTCACCATAGGTTTAATAGGCCTAGCATTCAACCGATCACACCTCTTATCAGCCTTACTTTGTTTAGAGGGAATAATATTATCCTTGTTTATTGCTATTGCCCTCTGATCAATAACACTAAACTCAGCATCATGCTCAATTACCCCGATAATCATCCTCACTTTCTCTGCCTGTGAAGCAAGCGCAGGTTTAGCCATATTGGTGGCAACCACACGAACCCACGGTTCCGACCACTTACAAAACCTGAGTCTTCTACAATGCTAAAAATTTTAATTCCAACAATCATACTATTCCCCACCACATGACTACTACCCATAAAATGACTGTGACCCGTTATAACCTCCCACACCCTTCTAATCGCACTACTTAGCCTACTATGATTTAAATGAAGCACGGACATTGGGTGAGACTTCTCCAATCACTACATAGGTGTAGACCCACTCTCAGCCCCCCTACTTGTCCTCACCTGTTGGCTCCTCCCATTAATAATCCTGGCCAGTCAAAACCACATATCCATAGAGCCGATTATTCGCCAACGAACTTACATCACACTTCTAATTTCACTACAAACCTTCCTCATCCTAGCATTCAGTGCAACAGAATTAATCATATTTTACATCATATTTGAAGCCACACTTATCCCCACCCTTATCATCATCACTCGATGGGGAAACCAAACCGAACGATTAAATGCAGGAATTTACTTTCTATTCTACACCCTAATTGGCTCCCTTCCTTTACTAGTTGCCCTTCTACTCATACAAAACAACTTGGGCTCATTGTCAATAATTATTATTCAACACCCACAACCACTCAACCTCAACACATGAGCCAACAAACTCTGATGGGCCGCCTGCCTTATCGCCTTCTTAGTAAAAATACCCCTCTATGGTGTTCACCTTTGACTCCCAAAAGCCCATGTTGAAGCCCCCATTGCAGGGTCAATAATCTTAGCCGCCGTACTACTAAAACTTGGTGGTTACGGAATAATACGAATTATTGTTATACTTGACCCCACCACCAAAGAAATAGCATACCCCTTTTTAATCTTAGCCGTCTGAGGGGTAATTATAACAAGCTCAATTTGCCTTCGACAAACTGACCTAAAATCCCTAATCGCTTACTCATCAGTAAGCCACATAGGACTAGTCGCAGCAGCAATTATAATCCAAACACCATGAAGCTTCGCAGGAGCAATTACACTAATAATTGCACATGGGTTAGTCTCGTCCGCTCTCTTCTGCCTAGCAAATACCAATTATGAACGCATCCACAGCCGAACCCTACTCTTAGCACGTGGCATTCAAATTATCCTCCCACTAATGGCAACCTGGTGATTCATTACTAACCTTGCCAACTTAGCTCTACCACCAACCCCTAATTTAATAGGAGAACTAATCATCATCTCCTCATTATTCAATTGATCCAACTGAACAATTTTACTAACGGGACTTGGAGTACTAATTACAGCATTATACTCCCTATACATGTTCTTAATAACCCAACGAGGCCCCACACCTAAACACCTGCTATCCTTAAACCCATCCCACACACGAGAGCACCTCCTCCTCAACCTTCACCTCACCCCCATATTACTATTAATCCTCAAACCCGACCTTATCTGAGGATGAACCTTTTGCCACTATAGTTTAACCAAAACATTAGATTGTGGTTCTAAAAACAAAAGTTAAAACCTTTTTAGTAGCCGAGAGTGGTCTGGGACACAAAGAACTGCTAATTCTTTCTACCACGGTCCGAATCCGTGGCTCACTCAGCCCTTGAAAGATAACAGCCATCTATTGGTCTTAGGAACCAAAAACTCTTAGTGCAACTCTAAGCAAGAGCTATGAGCACAACTATTTTTAACTCAACCTTTCTTCTTATCTTCATCATCCTACTGTACCCACTCATTTCATCACTCTCACCTAAAGAGCTCAACCCAAACTGATCATCTTCACATGTAAAAACCGCTGTAAAAATCTCTTTCTTCATCAGCCTTATCCCACTACTCATTCATCTAGATCAAGGCCTAGAATCAACCACAACCAACTGAAACTGAATCAATATTGGACCTTTTGATATCAACATAAGTTTTATATTTGACTCATACTCAATTATCTTCACCCCAGTAGCCCTTTACGTTACTTGATCAATCCTTGAATTTGCCATCTGATACATACACTCAGATCCAAACATTAACCGCTTCTTCAAATATCTTCTCCTGTTTTTAATTTCCATGATTATCTTAGTAACAGCCAATAATATATTTCAACTATTTATTGGCTGAGAGGGGGTCGGAATCATATCTTTCCTTTTAATTGGTTGGTGGTTCAGTCGAGCAGACGCAAATACTGCGGCCCTACAGGCCGTGATTTACAACCGTGTGGGGGATATTGGACTCATCTTAAGCATAGCCTGATTAGCCACAAACCTAAACTCATGAGAAATTCAACAGCTCTTTATCCTATCAAAAGATAAAGACCTCACCCTGCCTCTTATTGGCTTAGTACTAGCCGCAGCGGGAAAATCTGCACAGTTTGGACTTCACCCGTGACTACCTTCCGCCATAGAAGGGCCCACACCAGTCTCTGCCCTACTCCACTCCAGCACAATAGTGGTAGCCGGCATTTTTCTACTAATTCGCCTTCACCCATTAATCCAGGACAACCAACTAATCTTAACAACATGTTTATGTTTGGGGGCCCTAACCACCCTATTTACTGCCGCATGCGCACTAACACAAAACGACATTAAGAAAATTGTTGCCTTCTCAACATCAAGCCAACTAGGCCTAATAATAGTCACAATCGGACTAAACCAACCACAGTTGGCGTTTCTTCATATCTGCACACATGCTTTCTTTAAAGCTATACTTTTCCTCTGTTCCGGTTCGATTATCCACAGCTTAAATGATGAACAAGACATCCGAAAAATAGGCGGCCTTCATAAACTTCTACCGTTCACCTCTTCATCACTAACTGTGGGCAGCCTTGCCCTCACAGGGATACCATTCCTGTCAGGATTTTTTTCAAAAGATGCCATTATTGAATCCATAAACACCTCCCATTTAAACGCCTGGGCCCTAATTTTAACTCTTATTGCAACATCATTCACCGCCATTTATAGCCTCCGCCTTATCTTTTTCGCCCTAATAAAATTTCCGCGATTCTCCACCCTATCCCCAATCAATGAAAATAACACCTTAGTTATAAACCCCGTCAAACGCCTCGCATACGGAAGCATCTTTGCTGGGTTAATCATTACCTTTAACCTACCGCCAGCAAAAACTCAAATCATAACTATAACCCCACTATTAAAACTTTCTGCCCTCCTGGTAACAATTATCGGTTTACTTCTAGCACTAGAACTAACCAACTTAACTAACACCCAACTAAAAATTACCCCCACCACATCTACCCACCACTTCTCCAACATGCTTGGATATTTTCCAATAGTTGTTCACCGACTCACACCAAAAATTGACCTAAAATGAGCTCAACACATCTCAACACACCTTATTGACCTCACTTGAACTGAAAAAATTGGACCAAAAAGCACCATTATTCAACAAACACCATTAATTAAATTATCAACCCAACCACAACAAGGAATAATCAAAACCTACCTGATACTGCTCTTTTTAACACTAACACTAGCCACCCTAATCACCACCGCTTAAACAACACGCAAACTCCCATATGATAAACCACGAGTTAATTCTAAAACCACAAACAACGTTAAAAGCAGAACTCACCCGCCCAACACTAATATCCAACCACCGTGGGAGTATAACAAAGCTACGCCACCCAGATCTCCTCGAATTATCTCCAGAGTACTTATTTCCTCAACCCCAGCCCAACTTAATCCCAACCATTTATTAAAAAAATATTGACCTACAAAAAAAAGGCCTAATAAATACACACCAACATACAATAACACAGACCAATCGGCCCACGTCTCTGGGTATGGTTCGGCAGCAAGTGCTGCTGTATAAGCAAACACGACCAATATCCCCCCCAAATAAATTAGAAATAAGACTAAAGATAAAAAAGAACCCCCATGTCCCACTAATAAACCACACCCCACTCCAGCAACCACCACCAAACCCAAAGCCGCAAAATATGGGGAGGGGTTTGACGCTACTGCTATCAATCCTAAAATTAAACCCACTAATATTACAAACATAAAATAAATCATTATTCTCACCTGGGCTCTAACCAAGACCGACAACTTGAAAAACTATCGTTGTTCATTCAACTACAAGAATAATAATGACCACCAATATCCGAAAAACCCACCCACTCATTAAAATCGCCAACCACGCTTTAGTCGATTTACCCACACCATCCAACATCTCGATCTGATGGAATTTTGGTTCGCTACTTGGACTTTGCTTATTTATCCAAATTCTCACTGGATTATTTTTAGCCATACACTACACCGCAGACATTTCCACAGCCTTCTCATCAGTGATTCACATCTGTCGTGATGTAAACTATGGCTGACTAATTCGTAACATTCACTCTAACGGAGCCTCAATATTCTTCATTTGTGTTTACTTACACATTGCCCGAGGACTATACTACGGCTCCTACCTCTTTAAAGAAGTGTGAAACATCGGAGTTATTCTACTGTTCCTACTTATAGCAACAGCCTTTGTGGGCTACGTCCTACCATGAGGGCAAATATCCTTCTGAGGTGCCACAGTCATCACCAACCTCCTGTCCGCCCTCCCTTATATCGGGGACACATTAGTACAGTGGGTTTGGGGGGGTTTTTCAATTGATAACGCCACCCTAACACGCTTCTTCGCATTTCACTTCCTCCTACCATTTTTAATCCTAGGTTTAACCCTAATTCATCTACTATTCCTACACGAAACTGGCTCCAACAACCCCCTAGGCCTAAACTCTGACATAGATAAAATTCCATTCCACCCCTACTTCACATACAAAGATCTCTTGGGTTTCCTTATAATAACCATTTTATTAACCACATTGGCTTTATTCTCACCCAACCTTTTAGGTGACGCCGAAAATTTCATTCCCGCAAACCCACTCGTCACACCACCACACATCAAACCAGAGTGGTATTTCCTATTCGCCTACGCCATTCTTCGATCAATTCCTAACAAACTAGGTGGGGTATTAGCCTTAGCATTCTCAATTTTCATCCTAATACTAGTCCCCATACTTCACACCTCAAAACAACGAAGTAGCAGCTTCCGGCCCCTTACACAAATTCTCCTTTGAACACTCGTAGCTAACGCGATCATTCTAACATGAATTGGAGGACAACCCGTTGAACAACCATTTATCCTTGTTGGCCAAATTGCTTCCACTACTTATTTTTCCTTGTTTCTCATTATCATTCCGGCAACTAGCTGGTATGAAAACAAAATGCTCAACTTAAATTAATGTTTTGGTGGCTAAAATAAAAAGCGTCGGCCTTGTAAGTCGAAGACCGGAGGTGCAAATCCTCCCCACAACACATCAGGGAAAGGAGGGTTAAACTCCTGCCATTGGCTCCCAAAGCCAAGATTCTGCCTAAACTGCCCCCTGATCGCTAAGATCCCGCGAAAGCGGGGTTTTTTTGGAGTAAGTCAGTTGTACATATAAGTGATATAGCCCCCATACCCTAATATACCACATACGACATACATGTATAATCACCATATATAGACTTTCCCCATCTATATTACATCTATATGTTTAATACTCATTCATCTATAATCCTCACCCTCATTACATATATGATTAACCCACATTTCACTAAGATCAATAAATTCATTACATTAAATTTTCATTAACCCACATTTCCTTAAGATCACCGGATCCATTACATAAAATCCTTAACCCTCATAAACATAAGATCAATTATTTAACTAGGGAATATAAGAACCGTACAACGATCTCTTAGGTAAATATAGTGTACGGTTTGTGGTACGGTTCTCGAGATATCCCCTATAGGTGATCAAATGCTGGCATTTGGCACCCTTAAAGGGCTGACAGTTCAGACGTATCATGATCTCCAGACCGCTAGTTCCCTTAGTTGTCATTCAACTCTTAACCGTCTCAAGATTTACTGTCCGCCCTGTTTTTTTTTGGGGGGATGCGAGCAATCGCTAAGACCCGGAAGGCCCCAGGCATGGTCATACAGGTAAACCCGAACTCATCTAGACCTTACTCTCTTATTTAACAATACTGAATTCTTGTAATAAGATTGTCAAGTTGACCATATAGACAAAGGCATGGAGATCCTAATATTATAAGTGACACAATTTGGTTTCGTGCGTAATTAAACTTTCGTTCAATATGCACATATCATTTATCCTCATTTCTGTTTGGTAATCGATATCTATTAATGAGAATTAACATTCAACCTATTTGGTACTAACGTCCCTATTCGAGCATAGTTGATAGTAATTGCGCAATATTTCTGTAATATTACCCCGGGTGACGTAACGGAAGACGACGAATTATTAACTTTGGTTTTCTGGGAAAAACCCCCCCTCCCCCCTAATATACAAACTGACTTTTCTCGAAAAACCCCAAAAACGAGGGCCGAATGCATATTCCCATTTTTTAGCGTGTGAAAATATTTGCTATACATTGTTGCAACATGTTTTGCACAATGTGAC</span>
</pre></div>
</div>
</div>
</section>
<section id="step-2-merge-downloaded-sequences">
<h3>2.4 Step 2: merge downloaded sequences<a class="headerlink" href="#step-2-merge-downloaded-sequences" title="Permalink to this heading">#</a></h3>
<p>Once we have downloaded the sequences from the MitoFish and NCBI repositories, we can combine both files using the <code class="docutils literal notranslate"><span class="pre">db_merge</span></code> command. Since there will be overlap in the data we downloaded between the online repositories, we can specify the <code class="docutils literal notranslate"><span class="pre">--uniq</span></code> parameter to only keep unique accession numbers.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>db_merge<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDB.fasta<span class="w"> </span>-u<span class="w"> </span>yes<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/ncbi16Schondrichthyes.fasta<span class="w"> </span>sequenceData/7-refdb/mitofish.fasta
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">merging</span> <span class="nb">all</span> <span class="n">fasta</span> <span class="n">files</span> <span class="ow">and</span> <span class="n">discarding</span> <span class="n">duplicate</span> <span class="n">sequence</span> <span class="n">headers</span>
<span class="n">found</span> <span class="mi">3300</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">7</span><span class="o">-</span><span class="n">refdb</span><span class="o">/</span><span class="n">ncbi16Schondrichthyes</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">found</span> <span class="mi">825365</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">7</span><span class="o">-</span><span class="n">refdb</span><span class="o">/</span><span class="n">mitofish</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">written</span> <span class="mi">825709</span> <span class="n">sequences</span> <span class="n">to</span> <span class="n">mergedDB</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
</div>
</section>
<section id="step-3-extract-amplicons-through-in-silico-pcr">
<h3>2.5 Step 3: extract amplicons through <em>in silico</em> PCR<a class="headerlink" href="#step-3-extract-amplicons-through-in-silico-pcr" title="Permalink to this heading">#</a></h3>
<p>From the merged file, we can start with the first curation step of the reference database, i.e., extracting the amplicon sequences. We will be extracting the amplicon region from each sequence through an <em>in silico</em> PCR analysis. With this analysis, we will locate the forward and reverse primer-binding regions and extract the sequence in between. This will significantly reduce file sizes when larger data sets are downloaded, while also making sure that the necessary information is kept.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>insilico_pcr<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/mergedDB.fasta<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilico.fasta<span class="w"> </span>-f<span class="w"> </span>GACCCTATGGAGCTTTAGAC<span class="w"> </span>-r<span class="w"> </span>CGCTGTTATCCCTADRGTAACT
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>reading sequenceData/7-refdb/mergedDB.fasta into memory
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 794191686/794191686 [00:00&lt;00:00, 1114359519.98it/s]
found 825709 sequences in sequenceData/7-refdb/mergedDB.fasta
running in silico PCR on fasta file containing 825709 sequences
counting the number of sequences found by in silico PCR
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 11323145/11323145 [00:00&lt;00:00, 319601025.34it/s]
found primers in 52758 sequences
reading sequences without primer-binding regions into memory
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 559830454/559830454 [00:00&lt;00:00, 903207100.93it/s]
reverse complementing 772951 untrimmed sequences
running in silico PCR on 772951 reverse complemented untrimmed sequences
counting the number of sequences found by in silico PCR
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 78281/78281 [00:00&lt;00:00, 199231985.09it/s]
found primers in 367 sequences
</pre></div>
</div>
</div>
</section>
<section id="step-4-extract-amplicons-through-pairwise-global-alignments">
<h3>2.6 Step 4: extract amplicons through pairwise global alignments<a class="headerlink" href="#step-4-extract-amplicons-through-pairwise-global-alignments" title="Permalink to this heading">#</a></h3>
<p>Amplicons in the originally downloaded sequences might be missed during the <em>in silico</em> PCR analysis when one or both primer-binding regions are not incorporated in the online deposited sequence. This can happen when the reference barcode is generated with the same primer set or if the deposited sequence is incomplete in the primer-binding regions (denoted as “N” in the sequence). To retrieve those missed amplicons, we can use the already-retrieved amplicons as seed sequences in a Pairwise Global Alignment (PGA) analysis.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>pga<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/mergedDB.fasta<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGA.fasta<span class="w"> </span>-db<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilico.fasta<span class="w"> </span>-f<span class="w"> </span>GACCCTATGGAGCTTTAGAC<span class="w"> </span>-r<span class="w"> </span>CGCTGTTATCCCTADRGTAACT
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>reading sequenceData/7-refdb/mergedDBinsilico.fasta into memory
 95%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏      | 10817435/11401426 [00:00&lt;00:00, 100683726.05it/s]
found 53125 number of sequences in sequenceData/7-refdb/mergedDBinsilico.fasta
reading sequenceData/7-refdb/mergedDB.fasta into memory
 99%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌ | 785113607/794191686 [00:05&lt;00:00, 155879553.73it/s]
found 825709 number of sequences in sequenceData/7-refdb/mergedDB.fasta
analysing 772584 number of sequences for pairwise global alignment
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 558392140/558392140 [00:01&lt;00:00, 321625912.48it/s]
subsetting sequenceData/7-refdb/mergedDB.fasta by discarding sequences longer than 10,000 bp
found 772374 (99.97%) number of sequences in sequenceData/7-refdb/mergedDB.fasta shorter than 10,000 bp
running pairwise global alignment on 772374 number of sequences. This may take a while!
vsearch v2.23.0_macos_aarch64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading file sequenceData/7-refdb/mergedDBinsilico.fasta 100%  
10817412 nt in 53124 seqs, min 73, max 1274, avg 204
minseqlength 32: 1 sequence discarded.
Masking 100% 
Counting k-mers 100% 
Creating k-mer index 100% 
Searching 100%  
Matching unique query sequences: 2515 of 772374 (0.33%)
filtering alignments based on parameter settings
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 140747/140747 [00:00&lt;00:00, 11348463.16it/s]
2515 out of 772374 number of sequences aligned to sequenceData/7-refdb/mergedDBinsilico.fasta
1864 number of sequences achieved an alignment that passed filtering thresholds
written 54989 sequences to sequenceData/7-refdb/mergedDBinsilicoPGA.fasta
</pre></div>
</div>
</div>
</section>
<section id="step-5-assigning-a-taxonomic-id">
<h3>2.7 Step 5: assigning a taxonomic ID<a class="headerlink" href="#step-5-assigning-a-taxonomic-id" title="Permalink to this heading">#</a></h3>
<p>The next curating steps for our reference database require a taxonomic ID for each sequence. Currently, sequences are only identified by a unique accession or CRABS number. Therefore, before we continue curating the reference database, we will need to assign a taxonomic lineage to each sequence. The taxonomic ID is based on the NCBI taxonomy, for which we need to download the files. We can do this through the <code class="docutils literal notranslate"><span class="pre">db_download</span></code> function.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>sequenceData/7-refdb<span class="w"> </span>
crabs<span class="w"> </span>db_download<span class="w"> </span>-s<span class="w"> </span>taxonomy
<span class="nb">cd</span><span class="w"> </span>../../
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">downloading</span> <span class="n">taxonomy</span> <span class="n">information</span>
<span class="n">nucl_gb</span><span class="o">.</span><span class="n">accession2taxid</span><span class="o">.</span><span class="n">gz</span>                     <span class="mi">100</span><span class="o">%</span><span class="p">[</span><span class="o">====================================================================================================&gt;</span><span class="p">]</span>   <span class="mf">2.18</span><span class="n">G</span>  <span class="mf">1.32</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>    <span class="ow">in</span> <span class="mi">10</span><span class="n">m</span> <span class="mi">54</span><span class="n">s</span> 
<span class="n">unzipping</span> <span class="n">nucl_gb</span><span class="o">.</span><span class="n">accession2taxid</span><span class="o">.</span><span class="n">gz</span><span class="o">...</span>
<span class="n">taxdump</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>                                 <span class="mi">100</span><span class="o">%</span><span class="p">[</span><span class="o">====================================================================================================&gt;</span><span class="p">]</span>  <span class="mf">60.37</span><span class="n">M</span>  <span class="mf">2.19</span><span class="n">MB</span><span class="o">/</span><span class="n">s</span>    <span class="ow">in</span> <span class="mi">24</span><span class="n">s</span>     
<span class="n">removing</span> <span class="n">intermediary</span> <span class="n">files</span>
</pre></div>
</div>
</div>
<p>Once the taxonomy files are downloaded, we can assign a taxonomic ID for each amplicon using the <code class="docutils literal notranslate"><span class="pre">assign_tax</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>assign_tax<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGA.fasta<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtax.tsv<span class="w"> </span>-a<span class="w"> </span>sequenceData/7-refdb/nucl_gb.accession2taxid<span class="w"> </span>-t<span class="w"> </span>sequenceData/7-refdb/nodes.dmp<span class="w"> </span>-n<span class="w"> </span>sequenceData/7-refdb/names.dmp<span class="w"> </span>-w<span class="w"> </span>yes
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>retrieving accession numbers from sequenceData/7-refdb/mergedDBinsilicoPGA.fasta
 95%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏      | 11110715/11715190 [00:00&lt;00:00, 103506466.40it/s]
found 54989 accession numbers in sequenceData/7-refdb/mergedDBinsilicoPGA.fasta
reading sequenceData/7-refdb/nucl_gb.accession2taxid into memory
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 12080791485/12080791485 [01:55&lt;00:00, 104443260.06it/s]
reading sequenceData/7-refdb/nodes.dmp into memory
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 191357208/191357208 [00:03&lt;00:00, 48500398.70it/s]
reading sequenceData/7-refdb/names.dmp into memory
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 237051517/237051517 [00:03&lt;00:00, 70933792.18it/s]
processed 54961 entries in sequenceData/7-refdb/nucl_gb.accession2taxid
processed 2530548 entries in sequenceData/7-refdb/nodes.dmp
processed 2530548 entries in sequenceData/7-refdb/names.dmp
assigning a tax ID number to 54989 accession numbers from sequenceData/7-refdb/mergedDBinsilicoPGA.fasta
did not find 28 accession numbers in sequenceData/7-refdb/nucl_gb.accession2taxid, retrieving information through a web search
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 28/28 [00:45&lt;00:00,  1.63s/it]
could not find a taxonomic ID for 0 entries
54989 accession numbers resulted in 14376 unique tax ID numbers
generating taxonomic lineages for 14376 tax ID numbers
assigning a taxonomic lineage to 54989 accession numbers
written 54989 entries to sequenceData/7-refdb/mergedDBinsilicoPGAtax.tsv
</pre></div>
</div>
</div>
<p>The output file of the <code class="docutils literal notranslate"><span class="pre">assign_tax</span></code> function has been altered from a 2-line fasta structure to a tab-delimited file whereby all information of a sequence is provided on a single line. Let’s use the <code class="docutils literal notranslate"><span class="pre">head</span></code> command to investigate the format of the file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">4</span><span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtax.tsv
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">OR575552</span>	<span class="mi">305515</span>	<span class="n">Eukaryota</span>	<span class="n">Chordata</span>	<span class="n">Chondrichthyes</span>	<span class="n">Squaliformes</span>	<span class="n">Dalatiidae</span>	<span class="n">Euprotomicrus</span>	<span class="n">Euprotomicrus_bispinatus</span>	<span class="n">ACTTAAATTAATTATGTAAAATTTTTAACTCTCCGGAGAAAAACCACCCATATAATACCCCTAATTTAACTGTTTTTGGTTGGGGTGACCAGGGGGAAAAAATTATCCCCCCCATCGATTGAGTACTCAGTACTTAAAAATTAGAATGACAACTCTAATTAATAAAACATTTACCGAAACATGACCCAGAATTTATTCTGATCAATGAACCA</span>
<span class="n">OR582714</span>	<span class="mi">862652</span>	<span class="n">Eukaryota</span>	<span class="n">Chordata</span>	<span class="n">Chondrichthyes</span>	<span class="n">Squaliformes</span>	<span class="n">Etmopteridae</span>	<span class="n">Etmopterus</span>	<span class="n">Etmopterus_bigelowi</span>	<span class="n">ACTTAAATTAATCATATAAACTATTAACCCACGGGAATAAATTACATATATACCCCTAATTTAACTGTTTTTGGTTGGGGCGACCAAGGGGGAGAAAAAATCCCCCTCATCGATTGAGTACTTAGTACTTAAAAATTAGAACGACAGTTCTTATTAATGAAATATTTAACGAAAAATGACCCAGTTTTTCTGATCAATGAACCA</span>
<span class="n">OR582700</span>	<span class="mi">263691</span>	<span class="n">Eukaryota</span>	<span class="n">Chordata</span>	<span class="n">Chondrichthyes</span>	<span class="n">Carcharhiniformes</span>	<span class="n">Scyliorhinidae</span>	<span class="n">Apristurus</span>	<span class="n">Apristurus_kampae</span>	<span class="n">ACTTAGACTAATTATGTAATTTTTTTCCGCCTGTGGGTAAAAACAAAAATATAATATTTCTAGTTTAATTGTTTTTGGTTGGGGTGACCAAGGGGAAAAACAAATCCCCCTTATCGACCAAGTACTCAGTACTTAAAAATTAGAGCGACAGCTCTAATCAATAAAACATTTATCGAAAAATGACCCAGGATTTCCTGATCAATGAACCA</span>
<span class="n">OR582686</span>	<span class="mi">671160</span>	<span class="n">Eukaryota</span>	<span class="n">Chordata</span>	<span class="n">Chondrichthyes</span>	<span class="n">Squaliformes</span>	<span class="n">Etmopteridae</span>	<span class="n">Etmopterus</span>	<span class="n">Etmopterus_gracilispinis</span>	<span class="n">ACTTAAATTAATTATGTAAAACTACTAATCCACGGAAATAAACTATTTATATAATATTTCTAATTTAACTGTTTTTGGTTGGGGTGACCGAGGGGAAAAGAAAATCCCCCTCATCGATTGAGTACTTAGTACTTAAAAATTAGAACGACAGTTCTTATTAATAAAATATTTAACGAAAAATGACCCAGTTTTTCTGATCAATGAACCA</span>
</pre></div>
</div>
</div>
</section>
<section id="step-6-dereplicating-the-db">
<h3>2.8 Step 6: dereplicating the DB<a class="headerlink" href="#step-6-dereplicating-the-db" title="Permalink to this heading">#</a></h3>
<p>Now that we have a taxonomic ID for each amplicon sequence in our reference database, we can remove duplicate sequences per species, i.e., dereplicating the reference database.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>dereplicate<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtax.tsv<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtaxderep.tsv<span class="w"> </span>-m<span class="w"> </span>uniq_species
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>dereplicating sequenceData/7-refdb/mergedDBinsilicoPGAtax.tsv, keeping all unique sequences per species
species information found at position 7, starting dereplication process
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 16890510/16890510 [00:00&lt;00:00, 201942731.88it/s]
found 54989 sequences in sequenceData/7-refdb/mergedDBinsilicoPGAtax.tsv
written 25135 sequences to sequenceData/7-refdb/mergedDBinsilicoPGAtaxderep.tsv
</pre></div>
</div>
</div>
</section>
<section id="step-7-ref-db-cleanup">
<h3>2.8 Step 7: ref DB cleanup<a class="headerlink" href="#step-7-ref-db-cleanup" title="Permalink to this heading">#</a></h3>
<p>The final curation step for our custom reference database is to clean up the database using a variety of parameters, including minimum length, maximum length, maximum number of ambiguous base calls, environmental sequences, sequences for which the species name is not provided, and sequences with unspecified taxonomic levels.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>seq_cleanup<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtaxderep.tsv<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtaxderepclean.tsv<span class="w"> </span>-e<span class="w"> </span>yes<span class="w"> </span>-s<span class="w"> </span>yes<span class="w"> </span>-na<span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>CRABS v0.1.8
https://github.com/gjeunen/reference_database_creator


Filter INPUT: 100.0%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|00:00&lt;00:00

SUMMARY STATISTICS:
Number of sequences analysed: 25135
Sequences kept: 22121 (88.01%)
Sequences removed: 3014 (11.99%)
Too short sequences: 326
Too long sequences: 12
Sequences with too many &quot;N&quot;: 622
Environmental sequences: 0
Sequences without proper species ID: 2099
Sequences with too many undefined taxonomic levels: 3
</pre></div>
</div>
</div>
</section>
<section id="step-8-export-ref-db">
<h3>2.9 Step 8: export ref DB<a class="headerlink" href="#step-8-export-ref-db" title="Permalink to this heading">#</a></h3>
<p>Now that we have created a custom curated reference database containing 22,121 16S rRNA gene sequences of fish and sharks, we can export the reference database to be used by different taxonomy classifiers. Multiple formats are implemented in CRABS. Today, we will export the reference database in SINTAX format, as this is one of the classification methods we will be using later on with our COI data set.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>tax_format<span class="w"> </span>-i<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtaxderepclean.tsv<span class="w"> </span>-o<span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtaxderepcleansintax.fasta<span class="w"> </span>-f<span class="w"> </span>sintax
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">4</span><span class="w"> </span>sequenceData/7-refdb/mergedDBinsilicoPGAtaxderepcleansintax.fasta
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">OR575552</span><span class="p">;</span><span class="n">tax</span><span class="o">=</span><span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Chordata</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Chondrichthyes</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Squaliformes</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Dalatiidae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Euprotomicrus</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">Euprotomicrus_bispinatus</span>
<span class="n">ACTTAAATTAATTATGTAAAATTTTTAACTCTCCGGAGAAAAACCACCCATATAATACCCCTAATTTAACTGTTTTTGGTTGGGGTGACCAGGGGGAAAAAATTATCCCCCCCATCGATTGAGTACTCAGTACTTAAAAATTAGAATGACAACTCTAATTAATAAAACATTTACCGAAACATGACCCAGAATTTATTCTGATCAATGAACCA</span>
<span class="o">&gt;</span><span class="n">OR582714</span><span class="p">;</span><span class="n">tax</span><span class="o">=</span><span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Chordata</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Chondrichthyes</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Squaliformes</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Etmopteridae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Etmopterus</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">Etmopterus_bigelowi</span>
<span class="n">ACTTAAATTAATCATATAAACTATTAACCCACGGGAATAAATTACATATATACCCCTAATTTAACTGTTTTTGGTTGGGGCGACCAAGGGGGAGAAAAAATCCCCCTCATCGATTGAGTACTTAGTACTTAAAAATTAGAACGACAGTTCTTATTAATGAAATATTTAACGAAAAATGACCCAGTTTTTCTGATCAATGAACCA</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="the-coi-reference-db">
<h2>3. The COI reference DB<a class="headerlink" href="#the-coi-reference-db" title="Permalink to this heading">#</a></h2>
<p>For our tutorial data set, we will be using the COI reference database: <strong>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta</strong>, which I made through CRABS, supplemented with local barcodes provided by HKU, and subsetted to reduce the file size and speed up classification by excluding non-target taxonomic groups such as land plants and insects for example. This reference database is formatted according to SINTAX specifications and holds 213,937 high quality, unique, COI reference barcodes.</p>
<figure class="align-default" id="coi-reference-db-screenshot">
<img alt="_images/coidbscreenshot2.png" src="_images/coidbscreenshot2.png" />
<figcaption>
<p><span class="caption-number">Fig. 23 </span><span class="caption-text">: A screenshot of the COI reference database</span><a class="headerlink" href="#coi-reference-db-screenshot" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="taxonomic-classification">
<h2>4. Taxonomic classification<a class="headerlink" href="#taxonomic-classification" title="Permalink to this heading">#</a></h2>
<section id="sintax">
<h3>4.1 SINTAX<a class="headerlink" href="#sintax" title="Permalink to this heading">#</a></h3>
<p>Since the COI reference database <strong>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta</strong> is already formatted according to SINTAX specifications, let us first explore the SINTAX algorithm. The SINTAX algorithm predicts taxonomy by using k-mer similarity to identify the top hit in a reference database and provides bootstrap confidence for all ranks in the prediction. Unlike the RDP classifier, SINTAX does not rely on Bayesian posteriors. Hence, training the data set is not necessary and results can be obtained very quickly. SINTAX was developed for USEARCH and is also incorporated into <a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a> through the <code class="docutils literal notranslate"><span class="pre">--sintax</span></code> command. More information about the SINTAX algorithm can be found on the <a class="reference external" href="https://www.drive5.com/usearch/manual/sintax_algo.html">USEARCH website</a>.</p>
<p>The input file for the <code class="docutils literal notranslate"><span class="pre">--sintax</span></code> command is our ZOTU sequence list <strong>sequenceData/8-final/zotus.fasta</strong>. The database to match our sequences can be specified using the <code class="docutils literal notranslate"><span class="pre">--db</span></code> parameter. Finally, the output file is a simple tab-delimited text file, which can be specified using the <code class="docutils literal notranslate"><span class="pre">--tabbedout</span></code> parameter. Although not essential, you can provide a confidence value cutoff level through the <code class="docutils literal notranslate"><span class="pre">--sintax_cutoff</span></code> parameter. For the tutorial data set, let’s set the similarity threshold (<code class="docutils literal notranslate"><span class="pre">--sintax_cutoff</span></code>) at 60%.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vsearch<span class="w"> </span>--sintax<span class="w"> </span>sequenceData/8-final/zotus.fasta<span class="w"> </span>--db<span class="w"> </span>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta<span class="w"> </span>--tabbedout<span class="w"> </span>sequenceData/8-final/sintaxTaxonomy.txt<span class="w"> </span>--sintax_cutoff<span class="w"> </span><span class="m">0</span>.60
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vsearch</span> <span class="n">v2</span><span class="mf">.23.0</span><span class="n">_macos_aarch64</span><span class="p">,</span> <span class="mf">16.0</span><span class="n">GB</span> <span class="n">RAM</span><span class="p">,</span> <span class="mi">8</span> <span class="n">cores</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">torognes</span><span class="o">/</span><span class="n">vsearch</span>

<span class="n">Reading</span> <span class="n">file</span> <span class="n">sequenceData</span><span class="o">/</span><span class="mi">7</span><span class="o">-</span><span class="n">refdb</span><span class="o">/</span><span class="n">newDBcrabsCOIsintax</span><span class="o">.</span><span class="n">fasta</span> <span class="mi">100</span><span class="o">%</span>  
<span class="mi">87166478</span> <span class="n">nt</span> <span class="ow">in</span> <span class="mi">213930</span> <span class="n">seqs</span><span class="p">,</span> <span class="nb">min</span> <span class="mi">101</span><span class="p">,</span> <span class="nb">max</span> <span class="mi">47038</span><span class="p">,</span> <span class="n">avg</span> <span class="mi">407</span>
<span class="n">maxseqlength</span> <span class="mi">50000</span><span class="p">:</span> <span class="mi">7</span> <span class="n">sequences</span> <span class="n">discarded</span><span class="o">.</span>
<span class="n">Counting</span> <span class="n">k</span><span class="o">-</span><span class="n">mers</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Creating</span> <span class="n">k</span><span class="o">-</span><span class="n">mer</span> <span class="n">index</span> <span class="mi">100</span><span class="o">%</span> 
<span class="n">Classifying</span> <span class="n">sequences</span> <span class="mi">100</span><span class="o">%</span>  
<span class="n">Classified</span> <span class="mi">1345</span> <span class="n">of</span> <span class="mi">1345</span> <span class="n">sequences</span> <span class="p">(</span><span class="mf">100.00</span><span class="o">%</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The output shows that SINTAX managed to classify all 1,345 ZOTU sequences. Using the <code class="docutils literal notranslate"><span class="pre">head</span></code> command, we can inspect the first couple of lines in our document.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>sequenceData/8-final/sintaxTaxonomy.txt
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zotu</span><span class="mf">.6</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">0.98</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Bacillariophyta</span><span class="p">(</span><span class="mf">0.38</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Fragilariophyceae</span><span class="p">(</span><span class="mf">0.07</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Licmophorales</span><span class="p">(</span><span class="mf">0.07</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Ulnariaceae</span><span class="p">(</span><span class="mf">0.07</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Ulnaria</span><span class="p">(</span><span class="mf">0.07</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Ulnaria_acus</span><span class="p">(</span><span class="mf">0.07</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Arthropoda</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Malacostraca</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Decapoda</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Alpheidae</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Athanas</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">parvus</span><span class="p">(</span><span class="mf">1.00</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Arthropoda</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Malacostraca</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Decapoda</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Alpheidae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Athanas</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">parvus</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">0.98</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Bacillariophyta</span><span class="p">(</span><span class="mf">0.17</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Fragilariophyceae</span><span class="p">(</span><span class="mf">0.05</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Licmophorales</span><span class="p">(</span><span class="mf">0.05</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Ulnariaceae</span><span class="p">(</span><span class="mf">0.05</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Ulnaria</span><span class="p">(</span><span class="mf">0.05</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Ulnaria_acus</span><span class="p">(</span><span class="mf">0.05</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span>
<span class="n">zotu</span><span class="mf">.8</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Annelida</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Polychaeta</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Phyllodocida</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Nereididae</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Platynereis</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">HK01</span><span class="p">(</span><span class="mf">1.00</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Annelida</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Polychaeta</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Phyllodocida</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Nereididae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Platynereis</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">HK01</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Bacillariophyta</span><span class="p">(</span><span class="mf">0.13</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Coscinodiscophyceae</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Thalassiosirales</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Thalassiosiraceae</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Thalassiosira</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Thalassiosira_nordenskioeldii</span><span class="p">(</span><span class="mf">0.02</span><span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span>
<span class="n">zotu</span><span class="mf">.11</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Bacillariophyta</span><span class="p">(</span><span class="mf">0.33</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Bacillariophyceae</span><span class="p">(</span><span class="mf">0.21</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Bacillariales</span><span class="p">(</span><span class="mf">0.17</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Bacillariaceae</span><span class="p">(</span><span class="mf">0.17</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Cylindrotheca</span><span class="p">(</span><span class="mf">0.07</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Cylindrotheca_closterium</span><span class="p">(</span><span class="mf">0.07</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Arthropoda</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Malacostraca</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Decapoda</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Decapoda_nan</span><span class="p">(</span><span class="mf">0.96</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Decapoda_nan_nan</span><span class="p">(</span><span class="mf">0.96</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Caridea_sp</span><span class="o">.</span><span class="p">(</span><span class="mf">0.96</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Arthropoda</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Malacostraca</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Decapoda</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Decapoda_nan</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Decapoda_nan_nan</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">Caridea_sp</span><span class="o">.</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">0.99</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Chordata</span><span class="p">(</span><span class="mf">0.17</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Actinopteri</span><span class="p">(</span><span class="mf">0.08</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Perciformes</span><span class="p">(</span><span class="mf">0.04</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Serranidae</span><span class="p">(</span><span class="mf">0.01</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Aporops</span><span class="p">(</span><span class="mf">0.01</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Aporops_bilinearis</span><span class="p">(</span><span class="mf">0.01</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span>
<span class="n">zotu</span><span class="mf">.10</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Arthropoda</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Malacostraca</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Decapoda</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Rhynchocinetidae</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Rhynchocinetes</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">brucei</span><span class="p">(</span><span class="mf">1.00</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Arthropoda</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Malacostraca</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Decapoda</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Rhynchocinetidae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Rhynchocinetes</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">brucei</span>
<span class="n">zotu</span><span class="mf">.7</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span><span class="p">(</span><span class="mf">1.00</span><span class="p">),</span><span class="n">p</span><span class="p">:</span><span class="n">Bacillariophyta</span><span class="p">(</span><span class="mf">0.22</span><span class="p">),</span><span class="n">c</span><span class="p">:</span><span class="n">Mediophyceae</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">o</span><span class="p">:</span><span class="n">Cymatosirales</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">f</span><span class="p">:</span><span class="n">Cymatosiraceae</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">g</span><span class="p">:</span><span class="n">Minutocellus</span><span class="p">(</span><span class="mf">0.02</span><span class="p">),</span><span class="n">s</span><span class="p">:</span><span class="n">Minutocellus_polymorphus</span><span class="p">(</span><span class="mf">0.02</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">:</span><span class="n">Eukaryota</span>
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">head</span></code> command output shows that our file is structured in 4 columns, the first containing the ZOTU sequence name, the second providing the classification information without the cut off threshold, the third specifying on which strand the match was made, and the fourth column providing the classification information with the cut off value of 60% applied. From the head command, we can also see that the ZOTU sequence list is not ordered based on sequence number. Something you might want to keep in mind when you want to parse the document.</p>
</section>
<section id="idtaxa">
<h3>4.2 IDTAXA<a class="headerlink" href="#idtaxa" title="Permalink to this heading">#</a></h3>
<p>The second method we will explore is the IDTAXA machine learning algorithm. The IDTAXA algorithm is a machine learning algorithm that is split into two phases: a “training” phase where the classifier learns attributes of the training set, and a “testing” phase where sequences with unknown taxonomic assignments are classified. The IDTAXA algorithm is incorporated in the <a class="reference external" href="http://www2.decipher.codes/Classification.html">DECIPHER R package</a>.</p>
<section id="input-formatting">
<h4>4.2.1 Input formatting<a class="headerlink" href="#input-formatting" title="Permalink to this heading">#</a></h4>
<p>IDTAXA requires three files to run, including (i) a reference database containing sequences with known taxonomic IDs, (ii) a taxonomic ID file containing information about taxonomic ranks, and (iii) the ZOTU sequence file which we created during the bioinformatic processing of our sequence data. We already have access to the third ZOTU file (<strong>sequenceData/8-final/zotus.fasta</strong>) and we can create the first tow files from the SINTAX reference database file (<strong>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta</strong>). The IDTAXA reference database is a fasta file with an altered header structure compared to the SINTAX format we used above. The taxonomic ID file is a simple text file containing five columns separated with an asterisk (’*’). Column 1 refers to the unique index number of to taxon name, column 2 refers to the taxon name, column 3 reference to the unique index number of the parent, column 4 represents the numerical value of the taxonomic level, and column 5 refers to the name of the taxonomic rank.</p>
<p>It should be noted that the IDTAXA algorithm, as implemented in DECIPHER, cannot resolve homonyms, i.e., identical taxon names. IDTAXA will error out when encountering homonyms. To solve this issue, we will create the reference database file and taxonomic ID file using names of two ranks pasted together to generate unique names. For example, the fish family name of Chilodontidae will be transformed to the order name + the family name: CharaciformesChilodontidae, while the gastropod family name of Chilodontidae will be transformed to SeguenziidaChilodontidae. This will be done for every single name, no matter if they are unique or a homonym exists. This “hack” is also the reason why the taxonomic names in the output file will follow this structure. Don’t worry about the formatting of the output file. We will correct the taxonomic names during the file parsing process in section <strong>5.2 IDTAXA formatting</strong>.</p>
<p>To format the sequenceData/7-refdb/newDBcrabsCOIsintax.fasta file to IDTAXA format, we will execute the python script below. This script takes in three user arguments, including the SINTAX reference database, the output file name for the IDTAXA reference database, and the output file name for the IDTAXA taxonomic ID text file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/idtaxaInitFormat
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="kn">import</span> <span class="nn">sys</span>

<span class="n">inFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">dbOut</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">taxOut</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

<span class="n">indexNameDict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Root&#39;</span> <span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
<span class="n">parentDict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Root&#39;</span> <span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">}</span>
<span class="n">levelDict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Root&#39;</span> <span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
<span class="n">rankDict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Root&#39;</span> <span class="p">:</span> <span class="s1">&#39;rootrank&#39;</span><span class="p">}</span>
<span class="n">indexNumber</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">ranks</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s1">&#39;rootrank&#39;</span><span class="p">,</span> <span class="mi">1</span> <span class="p">:</span> <span class="s1">&#39;domain&#39;</span><span class="p">,</span> <span class="mi">2</span> <span class="p">:</span> <span class="s1">&#39;phylum&#39;</span><span class="p">,</span> <span class="mi">3</span> <span class="p">:</span> <span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="mi">4</span> <span class="p">:</span> <span class="s1">&#39;order&#39;</span><span class="p">,</span> <span class="mi">5</span><span class="p">:</span> <span class="s1">&#39;family&#39;</span><span class="p">,</span> <span class="mi">6</span> <span class="p">:</span> <span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="mi">7</span> <span class="p">:</span> <span class="s1">&#39;species&#39;</span><span class="p">}</span>

<span class="n">idTaxaSeqDict</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">inFile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
        <span class="n">previousName</span> <span class="o">=</span> <span class="s1">&#39;Root&#39;</span>
        <span class="n">previousTaxonName</span> <span class="o">=</span> <span class="s1">&#39;Root&#39;</span>
        <span class="n">oldName</span> <span class="o">=</span> <span class="s1">&#39;start&#39;</span>
        <span class="n">levelTax</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;CRABS&#39;</span><span class="p">):</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">seqID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">headerStart</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">seqID</span><span class="si">}</span><span class="s1">;Root&#39;</span>
            <span class="n">lineage</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">:</span><span class="mi">9</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">lineage</span><span class="p">:</span>
                <span class="n">levelTax</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">taxonName</span> <span class="o">=</span> <span class="n">taxon</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">taxonName</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
                    <span class="n">taxonName</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">previousName</span><span class="si">}</span><span class="s1">_insertaeCedis&#39;</span>
                <span class="n">uniqueTaxonName</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">previousTaxonName</span><span class="si">}{</span><span class="n">taxonName</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="n">headerStart</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">headerStart</span><span class="si">}</span><span class="s1">;</span><span class="si">{</span><span class="n">uniqueTaxonName</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="k">if</span> <span class="n">uniqueTaxonName</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">indexNameDict</span><span class="p">:</span>
                    <span class="n">indexNumber</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">indexNameDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">indexNumber</span>
                    <span class="n">parentDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">indexNameDict</span><span class="p">[</span><span class="n">previousName</span><span class="p">]</span>
                    <span class="n">levelDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">levelTax</span>
                    <span class="n">rankDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">ranks</span><span class="p">[</span><span class="n">levelTax</span><span class="p">]</span>
                <span class="n">taxonNumber</span> <span class="o">=</span> <span class="n">indexNameDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span>
                <span class="n">oldName</span> <span class="o">=</span> <span class="n">previousName</span>
                <span class="n">previousName</span> <span class="o">=</span> <span class="n">uniqueTaxonName</span>
                <span class="n">previousTaxonName</span> <span class="o">=</span> <span class="n">taxonName</span>
        <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">seqID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">headerStart</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">seqID</span><span class="si">}</span><span class="s1">;Root&#39;</span>
            <span class="n">lineage</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">lineage</span><span class="p">:</span>
                <span class="n">levelTax</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">taxonName</span> <span class="o">=</span> <span class="n">taxon</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">taxonName</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
                    <span class="n">taxonName</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">previousName</span><span class="si">}</span><span class="s1">_insertaeCedis&#39;</span>
                <span class="n">uniqueTaxonName</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">previousTaxonName</span><span class="si">}{</span><span class="n">taxonName</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="n">headerStart</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">headerStart</span><span class="si">}</span><span class="s1">;</span><span class="si">{</span><span class="n">uniqueTaxonName</span><span class="si">}</span><span class="s1">&#39;</span>
                <span class="k">if</span> <span class="n">uniqueTaxonName</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">indexNameDict</span><span class="p">:</span>
                    <span class="n">indexNumber</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">indexNameDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">indexNumber</span>
                    <span class="n">parentDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">indexNameDict</span><span class="p">[</span><span class="n">previousName</span><span class="p">]</span>
                    <span class="n">levelDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">levelTax</span>
                    <span class="n">rankDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span> <span class="o">=</span> <span class="n">ranks</span><span class="p">[</span><span class="n">levelTax</span><span class="p">]</span>
                <span class="n">taxonNumber</span> <span class="o">=</span> <span class="n">indexNameDict</span><span class="p">[</span><span class="n">uniqueTaxonName</span><span class="p">]</span>
                <span class="n">oldName</span> <span class="o">=</span> <span class="n">previousName</span>
                <span class="n">previousName</span> <span class="o">=</span> <span class="n">uniqueTaxonName</span>
                <span class="n">previousTaxonName</span> <span class="o">=</span> <span class="n">taxonName</span>          
        <span class="k">else</span><span class="p">:</span>
            <span class="n">idTaxaSeqDict</span><span class="p">[</span><span class="n">headerStart</span><span class="p">]</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>                

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">taxOut</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">indexName</span> <span class="ow">in</span> <span class="n">indexNameDict</span><span class="p">:</span>
        <span class="n">printLine</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">indexNameDict</span><span class="p">[</span><span class="n">indexName</span><span class="p">]</span><span class="si">}</span><span class="s1">*</span><span class="si">{</span><span class="n">indexName</span><span class="si">}</span><span class="s1">*</span><span class="si">{</span><span class="n">parentDict</span><span class="p">[</span><span class="n">indexName</span><span class="p">]</span><span class="si">}</span><span class="s1">*</span><span class="si">{</span><span class="n">levelDict</span><span class="p">[</span><span class="n">indexName</span><span class="p">]</span><span class="si">}</span><span class="s1">*</span><span class="si">{</span><span class="n">rankDict</span><span class="p">[</span><span class="n">indexName</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">printLine</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">dbOut</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">seqFile</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">idTaxaSeqDict</span><span class="p">:</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">seqFile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">idTaxaSeqDict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/idtaxaInitFormat
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/idtaxaInitFormat<span class="w"> </span>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta<span class="w"> </span>sequenceData/7-refdb/idtaxa.fasta<span class="w"> </span>sequenceData/7-refdb/idtaxa.txt
</pre></div>
</div>
<p>Let’s check if the two IDTAXA files have been created in the <strong>sequenceData/7-refdb/</strong> subdirectory and explore their structure using the <code class="docutils literal notranslate"><span class="pre">head</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-ltr<span class="w"> </span>sequenceData/7-refdb
head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>sequenceData/7-refdb/idtaxa.fasta
head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>sequenceData/7-refdb/idtaxa.txt
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>-rwxr-xr-x@ 1 gjeunen  staff  240602642  6 Oct 20:28 newDBcrabsCOIsintax.fasta
-rw-r--r--  1 gjeunen  staff    7098352  7 Oct 20:42 idtaxa.txt
-rw-r--r--  1 gjeunen  staff  268731720  7 Oct 20:42 idtaxa.fasta

&gt;AB010933;Root;RootEukaryota;EukaryotaArthropoda;ArthropodaInsecta;InsectaHymenoptera;HymenopteraFormicidae;FormicidaeCataglyphis;CataglyphisCataglyphis_rosenhaueri
GTTCCTTTAATACTAGGTTCCCCAGATATAGCTTATCCACGAATAGACAATATAAGGTTTTGATTACTACCCCCTTCTATTACCTTACTAATTTTAAGAAATTTTATTAATGATGGAACAGGAACAGGATGAACTGTTTATCCCCCCTTAGCATCTAATATTTTTCATAATGGCCCTTCTGTTGATCTTACTATTTTTTCTCTCCACATTGCAGGAATATCTTCAATTTTAGGAGCAATCAATTTTATTTCTACAATTTTAAATATACACCATAATAATATTTCTATAGATAAAGTTCCTCTTCTTGTTTGATCAATTTTAATTACAGCAATTCTTCTTCTCCTATCCCTTCCTGTTTTAGCAGGAGCAATTACCATATTACTTACAGACCGAAATCTTAACACTTCATTCTTTGACCCTTCAGGAGGAGGAGACCCTATTCTTTACCAACATCTATTCTGATTCTTTGGCCATCCTGAAGTTTATATTTTAATTCTACCAGGATTTGGATTAATTTCTCATATTATTATAAATGAAAGAGGTAAAAAAGAAACTTTTGGGGCTTTAGGAATATTTTATGCTCTTATAGCAATTGGATTTTTAGGATTTGTGGTTTGAGCACATCATATATTCACAATTGGCTTAGATGTTGATACACGAGCCTATTTTACCTCTGCAACTATAATTATTGCTATTCCTACTGGAATTAAAATTTTTAGATGAATTACAACTTTGCATGGAACAAAAATCAATAATAATTCATCCTTATGATGAACTATAGGATTTATTTTTTTATTTACAAGAGGAGGTTTAACAGGAATAATACTTTCAAATTCATCTAATGATATTATCCTTCATGATACATATTATGTTGTAGCTCACTTCCATTTTGTCATATCTATAGGAACTGTATTTGCTAATAGTGCTAGATTTATTCACTGATTCCCATTAATAACTGGATTCTCATTAAATAA
&gt;AB015863;Root;RootEukaryota;EukaryotaArthropoda;ArthropodaInsecta;InsectaLepidoptera;LepidopteraSaturniidae;SaturniidaeAntheraea;AntheraeaAntheraea_pernyi
CGAAAATGACTTTATTCTACAAATCATAAAGATATTGGAACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACCTCATTAAGACTTCTAATTCGAGCAGAATTAGGAACCCCAGGGTCTTTAATTGGAGATGACCAAATCTATAATACCATTGTAACAGCTCATGCTTTTATTATAATTTTTTTCATAGTTATACCTATCATAATTGGAGGATTTGGAAATTGATTAATTCCATTAATATTAGGAGCCCCTGATATAGCTTTCCCACGAATAAATAATATAAGTTTTTGACTATTACCCCCCTCTTTAACCCTATTAATCTCCAGAAGAATTGTAGAAAATGGAGCTGGAACTGGATGAACAGTTTACCCCCCTCTCTCTTCAAATATTGCTCATGGAGGATCTTCAGTAGATCTTGCTATTTTTTCCCTTCATCTTGCAGGTATTTCATCAATTTTAGGAGCAATTAATTTTATTACTACAATTATTAATATACGAATAAATAATTTATCATTTGATCAAATACCTTTATTTGTCTGAGCTGTTGGAATTACAGCTTTCTTACTTCTTTTATCATTACCTGTTTTAGCTGGAGCTATTACTATACTTTTAACAGATCGAAACTTAAATACTTCTTTTTTTGATCCTGCTGGTGGAGGAGATCCAATTTTATATCAACATTTATTTTGATTTTTTGGTCATCCA
&gt;AB015869;Root;RootEukaryota;EukaryotaArthropoda;ArthropodaInsecta;InsectaLepidoptera;LepidopteraSaturniidae;SaturniidaeSaturnia;SaturniaSaturnia_japonica
CGAAAATGACTTTATTCAACTAATCATAAAGATATTGGAACTTTATACTTTATTTTTGGAATTTGAGCAGGAATAGTAGGTACTTCTTTAAGATTACTAATTCGAGCTGAATTAGGAACCCCCGGATCTTTAATTGGAGATGATCAAATTTATAATACTATTGTAACAGCTCACGCTTTTATTATAATTTTTTTCATAGTTATACCTATTATAATTGGAGGATTTGGAAATTGATTAATCCCTTTAATATTAGGAGCCCCTGACATAGCTTTTCCTCGAATAAATAATATGAGCTTTTGATTATTGCCTCCTTCTTTAACTCTTTTAATCTCCAGAAGAATTGTAGAAAATGGAGCAGGTACAGGATGAACAGTTTATCCTCCTTTATCTTCTAATATTGCTCACAGAGGAACTTCAGTAGATTTAGCTATTTTTTCCCTTCATCTTGCTGGAATTTCTTCTATTTTAGGGGCTATTAATTTTATTACGACAATTATTAATATACGAATAAATAATATAACATTCGATCAAATACCTTTATTTGTATGAGCTGTTGGAATTACAGCTTTTCTTCTTTTATTATCTCTTCCTGTTTTAGCGGGAGCTATTACTATATTATTAACAGATCGAAATTTAAATACCTCTTTTTTTGACCCTGCAGGAGGAGGTGACCCAATTCTTTACCAACATCTTTTTTGATTTTTTGGGCACCCA
&gt;AB020225;Root;RootEukaryota;EukaryotaBacillariophyta;BacillariophytaBacillariophyceae;BacillariophyceaeBacillariales;BacillarialesBacillariaceae;BacillariaceaeNitzschia;NitzschiaNitzschia_frustulum
TTTATCAGGTATTATTGCTCACTCTGGTGGTTCTGTGGATTTAGCGATTTTCAGTCTTCACTTATCTGGAGCTGCCTCAATTTTAGGAGCAATTAATTTTATTTGTACCATAGTAAATATGAGAACTGAAAGTTTACCGTTTCATAAATTACCTCTATTCGTATGGGCGGTTTTTATTACAGCTATATTACTATTATTATCACTTCCTGTTTTAGCAGGTGCGATTACAATGTTATTAACAGATCGAAATTTTAATACTACTTTTTTTGACCCCGCAGGTGGAGGAGATCCCGTTTTGTTTCAACACTTATTT
&gt;AB020228;Root;RootEukaryota;EukaryotaBacillariophyta;BacillariophytaFragilariophyceae;FragilariophyceaeThalassionemales;ThalassionemalesThalassionemataceae;ThalassionemataceaeThalassionema;ThalassionemaThalassionema_nitzschioides
ACTATCAAGTATAATCGCTCATTCTGGCGGTTCGGTAGATTTAGCAATTTTTAGTTTACATGTTTCTGGAGCGGCTTCAATTTTAGGAGCAATTAATTTTATTTGTACTATTTTTAACATGCGTGTAAAAAGTTTATCATTTCATAAATTACCATTATTTGTTTGGGCTATTTTAATAACTGCCGTTTTACTTCTTTTATCTTTACCTGTTTTAGCAGGGGCAATTACAATGTTATTAACTGATAGAAATTTTAATACAACTTTTTTTGATCCAGCAGGAGGAGGTGATCCGGTGTTATATCAACATTTATTT

0*Root*-1*0*rootrank
1*RootEukaryota*0*1*domain
2*EukaryotaArthropoda*1*2*phylum
3*ArthropodaInsecta*2*3*class
4*InsectaHymenoptera*3*4*order
5*HymenopteraFormicidae*4*5*family
6*FormicidaeCataglyphis*5*6*genus
7*CataglyphisCataglyphis_rosenhaueri*6*7*species
8*InsectaLepidoptera*3*4*order
9*LepidopteraSaturniidae*8*5*family
</pre></div>
</div>
</div>
<p>The output of the <code class="docutils literal notranslate"><span class="pre">head</span></code> commands shows the structure of both files. Note the slight difference in structure between the IDTAXA and SINTAX reference database format. We can also see the “hack” applied to <strong>sequenceData/7-refdb-idtaxa.txt</strong>, whereby the names are pasted together, e.g., <em>RootEukaryota</em>, rather than <em>Eukaryota</em>.</p>
</section>
<section id="classifier-training">
<h4>4.2.2 Classifier training<a class="headerlink" href="#classifier-training" title="Permalink to this heading">#</a></h4>
<p>Once the reference database has been formatted to IDTAXA format, we need to train the classifier on the training set, i.e., the formatted reference database. Since IDTAXA is incorporated into the R package DECIPHER, the following code needs to be run in an R environment, such as <a class="reference external" href="https://posit.co/download/rstudio-desktop/">RStudio</a>. I suggest opening a new R script to copy-paste all the remaining code for this section and save it as <strong>sequenceData/1-scripts/idtaxa.R</strong>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">DECIPHER</span><span class="p">)</span>
<span class="nf">setwd</span><span class="p">(</span><span class="s">&#39;/Users/gjeunen/Documents/work/lectures/HKU2023/tutorial/sequenceData/7-refdb&#39;</span><span class="p">)</span>

<span class="nf">set.seed</span><span class="p">(</span><span class="m">1989</span><span class="p">)</span>

<span class="n">seqIDTAXA</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readDNAStringSet</span><span class="p">(</span><span class="s">&#39;idtaxa.fasta&#39;</span><span class="p">)</span>
<span class="n">rankIDTAXA</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="s">&#39;idtaxa.txt&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;Index&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Name&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Parent&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Level&#39;</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;Rank&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;*&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>

<span class="n">groups</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">seqIDTAXA</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>
<span class="n">groups</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;(.*)(Root;)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\\2&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">groups</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>
<span class="n">groupCounts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">table</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>
<span class="n">uniqueGroups</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">groupCounts</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">uniqueGroups</span><span class="p">)</span>

<span class="c1"># count number of sequences per group and, optionally, select only a subset of sequences if the group is deemed too large (10 seqs)</span>
<span class="n">maxGroupSize</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="c1"># max sequences per label (&gt;= 1)</span>
<span class="n">remove</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">logical</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">seqIDTAXA</span><span class="p">))</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="n">groupCounts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">maxGroupSize</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="n">groups</span><span class="o">==</span><span class="n">uniqueGroups</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="w">  </span><span class="n">keep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">index</span><span class="p">),</span>
<span class="w">                 </span><span class="n">maxGroupSize</span><span class="p">)</span>
<span class="w">  </span><span class="n">remove</span><span class="p">[</span><span class="n">index</span><span class="p">[</span><span class="o">-</span><span class="n">keep</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">TRUE</span>
<span class="p">}</span>
<span class="nf">sum</span><span class="p">(</span><span class="n">remove</span><span class="p">)</span><span class="w"> </span><span class="c1"># number of sequences eliminated</span>

<span class="c1"># train the classifier. Set &#39;maxIterations&#39; to number of times the classifier will be trained.</span>
<span class="n">maxIterations</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="c1"># must be &gt;= 1</span>
<span class="n">allowGroupRemoval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">FALSE</span>
<span class="n">probSeqsPrev</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">integer</span><span class="p">()</span><span class="w"> </span><span class="c1"># suspected problem sequences from prior iteration</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">seq_len</span><span class="p">(</span><span class="n">maxIterations</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Training iteration: &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="c1"># train the classifier</span>
<span class="w">  </span><span class="n">trainingSet</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">LearnTaxa</span><span class="p">(</span><span class="n">seqIDTAXA</span><span class="p">[</span><span class="o">!</span><span class="n">remove</span><span class="p">],</span>
<span class="w">                           </span><span class="nf">names</span><span class="p">(</span><span class="n">seqIDTAXA</span><span class="p">)[</span><span class="o">!</span><span class="n">remove</span><span class="p">],</span>
<span class="w">                           </span><span class="n">rankIDTAXA</span><span class="p">,</span><span class="w"> </span><span class="n">maxChildren</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span>
<span class="w">  </span><span class="c1"># look for problem sequences</span>
<span class="w">  </span><span class="n">probSeqs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">trainingSet</span><span class="o">$</span><span class="n">problemSequences</span><span class="o">$</span><span class="n">Index</span>
<span class="w">  </span><span class="nf">if </span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">probSeqs</span><span class="p">)</span><span class="o">==</span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;No problem sequences remaining.\n&quot;</span><span class="p">)</span>
<span class="w">    </span><span class="n">break</span>
<span class="w">  </span><span class="p">}</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="nf">if </span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">probSeqs</span><span class="p">)</span><span class="o">==</span><span class="nf">length</span><span class="p">(</span><span class="n">probSeqsPrev</span><span class="p">)</span><span class="w"> </span><span class="o">&amp;&amp;</span>
<span class="w">             </span><span class="nf">all</span><span class="p">(</span><span class="n">probSeqsPrev</span><span class="o">==</span><span class="n">probSeqs</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Iterations converged.\n&quot;</span><span class="p">)</span>
<span class="w">    </span><span class="n">break</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">if </span><span class="p">(</span><span class="n">i</span><span class="o">==</span><span class="n">maxIterations</span><span class="p">)</span>
<span class="w">    </span><span class="n">break</span>
<span class="w">  </span><span class="n">probSeqsPrev</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">probSeqs</span>
<span class="w">  </span><span class="c1"># remove any problem sequences</span>
<span class="w">  </span><span class="n">index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="o">!</span><span class="n">remove</span><span class="p">)[</span><span class="n">probSeqs</span><span class="p">]</span>
<span class="w">  </span><span class="n">remove</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">TRUE</span><span class="w"> </span><span class="c1"># remove all problem sequences</span>
<span class="w">  </span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="n">allowGroupRemoval</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># replace any removed groups</span>
<span class="w">    </span><span class="n">missing</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="o">!</span><span class="p">(</span><span class="n">uniqueGroups</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">groups</span><span class="p">[</span><span class="o">!</span><span class="n">remove</span><span class="p">])</span>
<span class="w">    </span><span class="n">missing</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">uniqueGroups</span><span class="p">[</span><span class="n">missing</span><span class="p">]</span>
<span class="w">    </span><span class="nf">if </span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="n">index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">index</span><span class="p">[</span><span class="n">groups</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">missing</span><span class="p">]</span>
<span class="w">      </span><span class="n">remove</span><span class="p">[</span><span class="n">index</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">FALSE</span><span class="w"> </span><span class="c1"># don&#39;t remove</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">}</span>
<span class="nf">sum</span><span class="p">(</span><span class="n">remove</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">probSeqs</span><span class="p">)</span>
<span class="n">trainingSet</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">trainingSet</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; trainingSet
  A training set of class &#39;Taxa&#39;
   * K-mer size: 9
   * Number of rank levels: 9
   * Total number of sequences: 134730
   * Number of groups: 44133
   * Number of problem groups: 0
   * Number of problem sequences: 8
</pre></div>
</div>
</div>
<figure class="align-default" id="idtaxa-training-set">
<img alt="_images/idtaxatrainingset2.png" src="_images/idtaxatrainingset2.png" />
<figcaption>
<p><span class="caption-number">Fig. 24 </span><span class="caption-text">: The plot of the IDTAXA training set.</span><a class="headerlink" href="#idtaxa-training-set" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The output of the <strong>trainingSet</strong> shows that the reference database was trained on a k-mer size of 9 bp, we have 9 ranks incorporated in the taxonomic lineage, a total number of 134,730 sequences, and a total number of 44,133 taxonomic groups. This sequence number is lower than previously mentioned. During the training stage of the classifier, only 10 reads per species were kept (as recommended by the developers), even though CRABS identified them as being unique for that species. For this trained classifier, we iterated over the reference database 3 times. After three iterations, 8 sequences are identified as problematic, i.e., potentially misclassified. Those are represented as magenta lines in the taxonomic tree figure. The figures below provide additional information on the number of sequences with a species ID and a bar graph showing the number of sequences per species.</p>
</section>
<section id="id2">
<h4>4.2.3 Classifier training<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h4>
<p>Now that the classifier is trained on the reference barcodes, we can use the training set to assign taxonomic IDs to new sequences, i.e., our ZOTU sequences. So, let’s load our ZOTU fasta file and assign taxonomic IDs using the IDTAXA algorithm. Again, the code below should be run in an R environment, such as RStudio. Copy-paste the R code below after the R code used to load in the training set.</p>
<div class="note admonition">
<p class="admonition-title">Set seed for reproducibility</p>
<p>The IDTAXA algorithm uses bootstrapping, which involves random sampling to obtain a confidence score. For this reason, the classifications are expected to change slightly if the classification process is repeated with the same inputs. To allow for the replicability of the results, we can set a random seed number before classification using the command <code class="docutils literal notranslate"><span class="pre">set.seed(1989)</span></code>. You can find this line of code in the R-script above, when we loaded in the training set. In case this is undesirable for your intent, remove this specific line from the R code to allow for randomness of subsampling during bootstrapping.</p>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># read in the ZOTU fasta file</span>
<span class="n">fastaIDTAXA</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readDNAStringSet</span><span class="p">(</span><span class="s">&#39;../8-final/zotus.fasta&#39;</span><span class="p">)</span>

<span class="c1"># classify sequences</span>
<span class="n">ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">IdTaxa</span><span class="p">(</span><span class="n">fastaIDTAXA</span><span class="p">,</span><span class="w"> </span><span class="n">trainingSet</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;extended&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;top&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">40</span><span class="p">,</span><span class="w"> </span><span class="n">processors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">)</span>
<span class="n">ids</span>

<span class="c1"># plot results</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="w"> </span><span class="n">trainingSet</span><span class="p">)</span>

<span class="c1"># export output</span>
<span class="n">output</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span>
<span class="w">                 </span><span class="nf">function</span><span class="p">(</span><span class="n">id</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">                   </span><span class="nf">paste</span><span class="p">(</span><span class="n">id</span><span class="o">$</span><span class="n">taxon</span><span class="p">,</span>
<span class="w">                         </span><span class="s">&quot; (&quot;</span><span class="p">,</span>
<span class="w">                         </span><span class="nf">round</span><span class="p">(</span><span class="n">id</span><span class="o">$</span><span class="n">confidence</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span>
<span class="w">                         </span><span class="s">&quot;%)&quot;</span><span class="p">,</span>
<span class="w">                         </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span>
<span class="w">                         </span><span class="n">collapse</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;: &quot;</span><span class="p">)</span>
<span class="w">                 </span><span class="p">})</span>
<span class="c1"># Create an empty character vector to store the output</span>
<span class="n">outputFile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">character</span><span class="p">(</span><span class="m">0</span><span class="p">)</span>
<span class="c1"># Combine the zotu ID and information and add them to the output vector</span>
<span class="nf">for </span><span class="p">(</span><span class="n">zotu_id</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">output</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">entry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="n">zotu_id</span><span class="p">,</span><span class="w"> </span><span class="n">output</span><span class="p">[</span><span class="n">zotu_id</span><span class="p">],</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">)</span><span class="w">  </span><span class="c1"># Use tab as a separator</span>
<span class="w">  </span><span class="n">outputFile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">outputFile</span><span class="p">,</span><span class="w"> </span><span class="n">entry</span><span class="p">)</span>
<span class="p">}</span>
<span class="c1"># Write output to file</span>
<span class="nf">writeLines</span><span class="p">(</span><span class="n">outputFile</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;../8-final/idtaxaTaxonomy.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; ids &lt;- IdTaxa(fastaIDTAXA, trainingSet, type = &#39;extended&#39;, strand = &#39;top&#39;, threshold = 40, processors = NULL)
  |===============================================================================================================| 100%

Time difference of 507.63 secs

&gt; ids
  A test set of class &#39;Taxa&#39; with length 1345
       confidence name                 taxon
   [1]        41% zotu.1               Root; RootEukaryota; unclassified_RootEukaryota...                                
   [2]        43% zotu.2               Root; RootEukaryota; unclassified_RootEukaryota...                                
   [3]        95% zotu.3               Root; RootEukaryota; EukaryotaArthropoda; ArthropodaMalacostraca; MalacostracaD...
   [4]        51% zotu.4               Root; RootEukaryota; EukaryotaArthropoda; ArthropodaMalacostraca; MalacostracaD...
   [5]        47% zotu.5               Root; RootEukaryota; unclassified_RootEukaryota...                                
   ...        ... ...                  ...
[1341]        47% zotu.1341            Root; RootEukaryota; unclassified_RootEukaryota                                   
[1342]        43% zotu.1342            Root; RootEukaryota; unclassified_RootEukaryota                                   
[1343]        56% zotu.1343            Root; RootEukaryota; unclassified_RootEukaryota                                   
[1344]        53% zotu.1344            Root; RootEukaryota; unclassified_RootEukaryota                                   
[1345]        55% zotu.1345            Root; RootEukaryota; unclassified_RootEukaryota   
</pre></div>
</div>
</div>
<figure class="align-default" id="idtaxa-result">
<img alt="_images/idtaxaresults2.png" src="_images/idtaxaresults2.png" />
<figcaption>
<p><span class="caption-number">Fig. 25 </span><span class="caption-text">: The plot of the IDTAXA results.</span><a class="headerlink" href="#idtaxa-result" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Plotting the IDTAXA results shows a pie graph with all the different taxonomic groups coloured differently, as well as a taxonomic tree where coloured branches (same as the pie graph colours) represent the reference barcodes that were matched to the query sequences. Note the difference in time it took between SINTAX and IDTAXA. Also note that IDTAXA classifies a large proportion to Eukaryota, as the authors describe IDTAXA as a more conserved classifier compared to other implementations.</p>
<p>We also exported the results to <strong>sequenceData/8-final/idtaxaTaxonomy.txt</strong>. Let us look at the structure of this file using the <code class="docutils literal notranslate"><span class="pre">head</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>sequenceData/8-final/idtaxaTaxonomy.txt
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zotu</span><span class="mf">.1</span> <span class="n">Root</span> <span class="p">(</span><span class="mf">41.2</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mf">41.2</span><span class="o">%</span><span class="p">):</span> <span class="n">unclassified_RootEukaryota</span> <span class="p">(</span><span class="mf">41.2</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">Root</span> <span class="p">(</span><span class="mf">43.9</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mf">43.9</span><span class="o">%</span><span class="p">):</span> <span class="n">unclassified_RootEukaryota</span> <span class="p">(</span><span class="mf">43.9</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">Root</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">EukaryotaArthropoda</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">ArthropodaMalacostraca</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">MalacostracaDecapoda</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">DecapodaAlpheidae</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">AlpheidaeAthanas</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">):</span> <span class="n">Athanasparvus</span> <span class="p">(</span><span class="mf">95.2</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">Root</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">EukaryotaArthropoda</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">ArthropodaMalacostraca</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">MalacostracaDecapoda</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">DecapodaHippolytidae</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">HippolytidaeThor</span> <span class="p">(</span><span class="mf">95.7</span><span class="o">%</span><span class="p">):</span> <span class="n">ThorHK01</span> <span class="p">(</span><span class="mf">51.5</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">Root</span> <span class="p">(</span><span class="mf">47.9</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mf">47.6</span><span class="o">%</span><span class="p">):</span> <span class="n">unclassified_RootEukaryota</span> <span class="p">(</span><span class="mf">47.6</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.6</span> <span class="n">Root</span> <span class="p">(</span><span class="mf">44.1</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mf">43.7</span><span class="o">%</span><span class="p">):</span> <span class="n">unclassified_RootEukaryota</span> <span class="p">(</span><span class="mf">43.7</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.7</span> <span class="n">Root</span> <span class="p">(</span><span class="mf">42.6</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mf">42.6</span><span class="o">%</span><span class="p">):</span> <span class="n">unclassified_RootEukaryota</span> <span class="p">(</span><span class="mf">42.6</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.8</span> <span class="n">Root</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">EukaryotaAnnelida</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">AnnelidaPolychaeta</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">PolychaetaPhyllodocida</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">PhyllodocidaNereididae</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">NereididaePlatynereis</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">PlatynereisHK01</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.9</span> <span class="n">Root</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">EukaryotaAnnelida</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">AnnelidaPolychaeta</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">PolychaetaPhyllodocida</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">PhyllodocidaNereididae</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">NereididaePlatynereis</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">PlatynereisHK01</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">)</span>
<span class="n">zotu</span><span class="mf">.10</span> <span class="n">Root</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">RootEukaryota</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">EukaryotaArthropoda</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">ArthropodaMalacostraca</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">MalacostracaDecapoda</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">DecapodaRhynchocinetidae</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">RhynchocinetidaeRhynchocinetes</span> <span class="p">(</span><span class="mi">100</span><span class="o">%</span><span class="p">):</span> <span class="n">Rhynchocinetesbrucei</span> <span class="p">(</span><span class="mf">58.3</span><span class="o">%</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">head</span></code> output shows that the <strong>sequenceData/8-final/idtaxaTaxonomy.txt</strong> file is structured in two columns, with the first column specifying the ZOTU sequence name and the second column providing the taxonomic ID. Hence, a very similar structure to SINTAX, though the third column indicating the strand the match was made is missing, as well as the fourth column where SINTAX provided the taxonomic ID with the similarity threshold applied.</p>
</section>
</section>
<section id="blast">
<h3>4.3 BLAST<a class="headerlink" href="#blast" title="Permalink to this heading">#</a></h3>
<section id="online-blast">
<h4>4.3.1 Online BLAST<a class="headerlink" href="#online-blast" title="Permalink to this heading">#</a></h4>
<p>The third and final method we will explore is BLAST (Basic Local Alignment Search Tool), probably the most well known manner to assign a taxonomic ID to sequences. As the name suggests, BLAST assigns a taxonomic ID to a sequence through best hits from local alignments against a reference database.</p>
<p>The popularity of BLAST likely stems from its direct implementation into the <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">NCBI online reference database repository</a>. To check out the website, we will copy-paste the first 5 sequences in our ZOTU file (<strong>sequenceData/8-final/zotus.fasta</strong>) and perform a Nucleotide BLAST, as shown in the screenshot below.</p>
<figure class="align-default" id="blast-website">
<img alt="_images/blastwebsite.png" src="_images/blastwebsite.png" />
<figcaption>
<p><span class="caption-number">Fig. 26 </span><span class="caption-text">: A screenshot of the BLAST website</span><a class="headerlink" href="#blast-website" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Below are the top BLAST hit results for the first 5 sequences, followed by a screenshot of the website for the first ZOTU sequence (zotu.1).</p>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<p>Results for:    Description:           Qcov:   e-val:    Pident:<br />
zotu.1 (313 bp) Gammarus mucronatus    42%     4e-32     87.22%
zotu.2 (313 bp) Mitostoma chrysomelas  89%     2e-31     76.60%
zotu.3 (313 bp) Caridea sp.            99%     7e-90     86.58%
zotu.4 (313 bp) Caridea sp.            100%    2e-160    100.00%
zotu.5 (313 bp) Metaphire tosaensis    96%     9e-64     82.12%</p>
</div>
<figure class="align-default" id="blast-website-results">
<img alt="_images/blastwebsiteresults.png" src="_images/blastwebsiteresults.png" />
<figcaption>
<p><span class="caption-number">Fig. 27 </span><span class="caption-text">: A screenshot of the results from the BLAST website for sequence ID zotu.1</span><a class="headerlink" href="#blast-website-results" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>We will explore the results on the website more in depth during the workshop, but the most important numbers are provided in the output above, including the ZOTU sequence name (Results for:), the taxonomic ID the sequence best matches to (Description:), the percentage of how much the reference sequence provides coverage to the query sequence (Qcov:), the expect value is a parameter that describes the number of hits one can “expect” to see by chance when searching a database of a particular size and describes the random background noise (e-val:), and the percent identity value of how well the query and reference match over the covered region (Pident:).</p>
</section>
<section id="blast-remote">
<h4>4.3.2 BLAST+ remote<a class="headerlink" href="#blast-remote" title="Permalink to this heading">#</a></h4>
<p>While the website is extremely handy, it will be time-consuming if we have to manually check all 1,345 ZOTU sequences in our <strong>sequenceData/8-final/zotus.fasta</strong> file. Luckily, BLAST is also available via the Command Line Interface, called <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html">BLAST+</a>. When we conduct BLAST via the command line (<code class="docutils literal notranslate"><span class="pre">blastn</span></code>), we can specify the parameter <code class="docutils literal notranslate"><span class="pre">-remote</span></code>, which will tell the program to access the NCBI servers and match your query sequences to the online NCBI database. To specify we want to search against the nucleotide database, we can provide the <code class="docutils literal notranslate"><span class="pre">-db</span> <span class="pre">nt</span></code> parameter. Our query sequences are provided through the <code class="docutils literal notranslate"><span class="pre">-query</span></code> parameter. We can also set the maximum number of target sequences we get back, the minimum percent identity threshold, and the minimum query coverage threshold through the <code class="docutils literal notranslate"><span class="pre">-max_target_seqs</span></code>, <code class="docutils literal notranslate"><span class="pre">perc_identity</span></code>, and <code class="docutils literal notranslate"><span class="pre">-qcov_hsp_perc</span></code> parameters. Finally, the output file (parameter: <code class="docutils literal notranslate"><span class="pre">-out</span></code>) can take on different formats, but the most common is a tab-delimited structure, which is invoked by <code class="docutils literal notranslate"><span class="pre">-outfmt</span> <span class="pre">&quot;6&quot;</span></code>. Table C1 on this <a class="reference external" href="https://scicomp.ethz.ch/public/manual/BLAST/BLAST.pdf">website</a> provides an overview of all the different options to the BLAST+ search application.</p>
<p>To show you how you can run remote BLASTs in the Terminal, let’s extract the first five sequences from our <strong>sequenceData/8-final/zotus.fasta</strong> file and BLAST them remotely against the nucleotide NCBI database. To do all of this in the Terminal, rather than making a new document with the first five sequences, we can use the <code class="docutils literal notranslate"><span class="pre">head</span></code> command to extract the first five sequences and pipe the results to the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> command.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If you would like the scientific names provided in the output file, we first need to download and extract the taxonomic database from the NCBI website. We can do this using the following two commands</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz
tar<span class="w"> </span>-xvf<span class="w"> </span>taxdb.tar.gz
</pre></div>
</div>
<p>If you do not download the taxonomic database, the output file column <code class="docutils literal notranslate"><span class="pre">sscinames</span></code> in the code below will default to <code class="docutils literal notranslate"><span class="pre">N/A</span></code>.</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">25</span><span class="w"> </span>sequenceData/8-final/zotus.fasta<span class="w"> </span><span class="p">|</span><span class="w"> </span>blastn<span class="w"> </span>-db<span class="w"> </span>nt<span class="w"> </span>-query<span class="w"> </span>-<span class="w"> </span>-out<span class="w"> </span>sequenceData/8-final/BLASTremote.txt<span class="w"> </span>-max_target_seqs<span class="w"> </span><span class="m">10</span><span class="w"> </span>-perc_identity<span class="w"> </span><span class="m">80</span><span class="w"> </span>-qcov_hsp_perc<span class="w"> </span><span class="m">80</span><span class="w"> </span>-remote<span class="w"> </span>-outfmt<span class="w"> </span><span class="s2">&quot;6 qaccver saccver staxid sscinames length pident mismatch qcovs evalue bitscore qstart qend sstart send gapopen&quot;</span>
</pre></div>
</div>
<p>To display the output results of the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> command in the Terminal window, we can use the <code class="docutils literal notranslate"><span class="pre">cat</span></code> command.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>We can use the <code class="docutils literal notranslate"><span class="pre">cat</span></code> command to display the contents of a file to the Terminal window. However, it is not recommended to do this for large files, as it will all be read into memory. For large files, it is best to use a command such as <code class="docutils literal notranslate"><span class="pre">less</span></code>.</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>sequenceData/8-final/BLASTremote.txt
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zotu</span><span class="mf">.1</span> <span class="n">MH087519</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">500</span> <span class="mi">632</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KU906005</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">485</span> <span class="mi">617</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP255185</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">519</span> <span class="mi">651</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP255165</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">492</span> <span class="mi">624</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP254955</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">519</span> <span class="mi">651</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP254688</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">518</span> <span class="mi">650</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP254656</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">519</span> <span class="mi">651</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP254605</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">518</span> <span class="mi">650</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP254404</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">518</span> <span class="mi">650</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KP254360</span><span class="mf">.1</span> <span class="mi">315708</span> <span class="n">Gammarus</span> <span class="n">mucronatus</span> <span class="mi">133</span> <span class="mf">87.218</span> <span class="mi">17</span> <span class="mi">42</span> <span class="mf">4.49e-32</span> <span class="mi">152</span> <span class="mi">174</span> <span class="mi">306</span> <span class="mi">519</span> <span class="mi">651</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">AB500136</span><span class="mf">.1</span> <span class="mi">322050</span> <span class="n">Neoseiulus</span> <span class="n">womersleyi</span> <span class="mi">167</span> <span class="mf">82.036</span> <span class="mi">29</span> <span class="mi">53</span> <span class="mf">9.72e-29</span> <span class="mi">141</span> <span class="mi">147</span> <span class="mi">312</span> <span class="mi">187</span> <span class="mi">353</span> <span class="mi">1</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">HM465067</span><span class="mf">.1</span> <span class="mi">874932</span> <span class="n">Caridea</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LPdivOTU347</span> <span class="mi">313</span> <span class="mf">86.581</span> <span class="mi">40</span> <span class="mi">99</span> <span class="mf">6.90e-90</span> <span class="mi">344</span> <span class="mi">2</span> <span class="mi">313</span> <span class="mi">294</span> <span class="mi">605</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">HM466580</span><span class="mf">.1</span> <span class="mi">875001</span> <span class="n">Caridea</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LPdivOTU96</span> <span class="mi">313</span> <span class="mf">100.000</span> <span class="mi">0</span> <span class="mi">100</span> <span class="mf">1.74e-160</span> <span class="mi">579</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ560339</span><span class="mf">.1</span> <span class="mi">2858060</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466310</span> <span class="mi">313</span> <span class="mf">98.403</span> <span class="mi">5</span> <span class="mi">100</span> <span class="mf">3.79e-152</span> <span class="mi">551</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ560299</span><span class="mf">.1</span> <span class="mi">2858059</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466309</span> <span class="mi">313</span> <span class="mf">98.403</span> <span class="mi">5</span> <span class="mi">100</span> <span class="mf">3.79e-152</span> <span class="mi">551</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">GQ260974</span><span class="mf">.1</span> <span class="mi">653025</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LP</span><span class="o">-</span><span class="mi">2009</span><span class="p">;</span><span class="n">Caridea</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LPdivOTU96</span> <span class="mi">313</span> <span class="mf">98.403</span> <span class="mi">5</span> <span class="mi">100</span> <span class="mf">3.79e-152</span> <span class="mi">551</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ560028</span><span class="mf">.1</span> <span class="mi">2858073</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466326</span> <span class="mi">313</span> <span class="mf">98.083</span> <span class="mi">6</span> <span class="mi">100</span> <span class="mf">1.76e-150</span> <span class="mi">545</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ559959</span><span class="mf">.1</span> <span class="mi">2858071</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466317</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466318</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466319</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466320</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466321</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466322</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466323</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466324</span> <span class="mi">313</span> <span class="mf">98.083</span> <span class="mi">6</span> <span class="mi">100</span> <span class="mf">1.76e-150</span> <span class="mi">545</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ559957</span><span class="mf">.1</span> <span class="mi">2858061</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466313</span> <span class="mi">313</span> <span class="mf">98.083</span> <span class="mi">6</span> <span class="mi">100</span> <span class="mf">1.76e-150</span> <span class="mi">545</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ560029</span><span class="mf">.1</span> <span class="mi">2858074</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466328</span> <span class="mi">311</span> <span class="mf">98.071</span> <span class="mi">6</span> <span class="mi">99</span> <span class="mf">2.28e-149</span> <span class="mi">542</span> <span class="mi">3</span> <span class="mi">313</span> <span class="mi">348</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ560087</span><span class="mf">.1</span> <span class="mi">2858063</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466315</span> <span class="mi">313</span> <span class="mf">97.764</span> <span class="mi">7</span> <span class="mi">100</span> <span class="mf">8.20e-149</span> <span class="mi">540</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span> <span class="n">MZ559958</span><span class="mf">.1</span> <span class="mi">2858064</span> <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466316</span> <span class="mi">313</span> <span class="mf">97.764</span> <span class="mi">7</span> <span class="mi">100</span> <span class="mf">8.20e-149</span> <span class="mi">540</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">AB542683</span><span class="mf">.1</span> <span class="mi">506673</span> <span class="n">Metaphire</span> <span class="n">tosaensis</span> <span class="mi">302</span> <span class="mf">82.119</span> <span class="mi">52</span> <span class="mi">96</span> <span class="mf">9.25e-64</span> <span class="mi">257</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">332</span> <span class="mi">632</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">DQ224174</span><span class="mf">.1</span> <span class="mi">351234</span> <span class="n">Amynthas</span> <span class="n">lini</span> <span class="mi">302</span> <span class="mf">81.788</span> <span class="mi">53</span> <span class="mi">96</span> <span class="mf">4.30e-62</span> <span class="mi">252</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">351</span> <span class="mi">651</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">KM187649</span><span class="mf">.1</span> <span class="mi">1544308</span> <span class="n">Metabonellia</span> <span class="n">haswelli</span> <span class="mi">302</span> <span class="mf">81.457</span> <span class="mi">54</span> <span class="mi">96</span> <span class="mf">2.00e-60</span> <span class="mi">246</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">351</span> <span class="mi">651</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">HM219172</span><span class="mf">.2</span> <span class="mi">857735</span> <span class="n">Perionyx</span> <span class="n">sp</span><span class="o">.</span> <span class="mi">1</span> <span class="n">WL</span><span class="o">-</span><span class="mi">2010</span> <span class="mi">301</span> <span class="mf">81.395</span> <span class="mi">56</span> <span class="mi">96</span> <span class="mf">2.00e-60</span> <span class="mi">246</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">350</span> <span class="mi">650</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">MT472583</span><span class="mf">.1</span> <span class="mi">2736579</span> <span class="n">Perionyx</span> <span class="n">sp</span><span class="o">.</span> <span class="n">E1R2</span> <span class="mi">301</span> <span class="mf">81.063</span> <span class="mi">57</span> <span class="mi">96</span> <span class="mf">9.32e-59</span> <span class="mi">241</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">352</span> <span class="mi">652</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">KT716821</span><span class="mf">.1</span> <span class="mi">1737888</span> <span class="n">Perionyx</span> <span class="n">sp</span><span class="o">.</span> <span class="n">DHS</span><span class="o">-</span><span class="mi">2015</span> <span class="mi">301</span> <span class="mf">81.063</span> <span class="mi">57</span> <span class="mi">96</span> <span class="mf">9.32e-59</span> <span class="mi">241</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">362</span> <span class="mi">662</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">LC703190</span><span class="mf">.1</span> <span class="mi">585715</span> <span class="n">Metaphire</span> <span class="n">megascolidioides</span> <span class="mi">302</span> <span class="mf">81.126</span> <span class="mi">55</span> <span class="mi">96</span> <span class="mf">9.32e-59</span> <span class="mi">241</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">263</span> <span class="mi">563</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">JF509718</span><span class="mf">.1</span> <span class="mi">6431</span> <span class="n">Urechis</span> <span class="n">caupo</span> <span class="mi">302</span> <span class="mf">81.126</span> <span class="mi">55</span> <span class="mi">96</span> <span class="mf">9.32e-59</span> <span class="mi">241</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">350</span> <span class="mi">650</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">AB542648</span><span class="mf">.1</span> <span class="mi">585715</span> <span class="n">Metaphire</span> <span class="n">megascolidioides</span> <span class="mi">302</span> <span class="mf">81.126</span> <span class="mi">55</span> <span class="mi">96</span> <span class="mf">9.32e-59</span> <span class="mi">241</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">332</span> <span class="mi">632</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span> <span class="n">AB536863</span><span class="mf">.1</span> <span class="mi">585715</span> <span class="n">Metaphire</span> <span class="n">megascolidioides</span> <span class="mi">302</span> <span class="mf">81.126</span> <span class="mi">55</span> <span class="mi">96</span> <span class="mf">9.32e-59</span> <span class="mi">241</span> <span class="mi">6</span> <span class="mi">306</span> <span class="mi">332</span> <span class="mi">632</span> <span class="mi">2</span>
</pre></div>
</div>
</div>
<p>The output shows similar results to the results we obtained from the website. However, now we immediately have the results in a tab-delimited text file that we can easily parse using a python script for example.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>It should be noted that if you ran BLAST via the website, you can use the BLAST+ command line software to download a similar table to your computer using the <code class="docutils literal notranslate"><span class="pre">blast_formatter</span></code> function. To accomplish this, you would need to provide the RID tag with the <code class="docutils literal notranslate"><span class="pre">-rid</span></code> parameter. Below is a screenshot on where you can locate this RID tag on the website and the code to download the BLAST hits of the 5 sequences we checked on the website. Online BLAST results typically stay available through the RID tag for 24 - 48h. While you can set the maximum number of sequences per entry to be downloaded <code class="docutils literal notranslate"><span class="pre">-max_target_seqs</span></code>, this method does not allow the user to set a percent identity or query coverage threshold.</p>
<figure class="align-default" id="blast-rid-number">
<img alt="_images/blastRID.png" src="_images/blastRID.png" />
<figcaption>
<p><span class="caption-number">Fig. 28 </span><span class="caption-text">: A screenshot of where to locate the RID tag on the NCBI website to download online BLAST results</span><a class="headerlink" href="#blast-rid-number" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>blast_formatter<span class="w"> </span>-rid<span class="w"> </span>J3FG1AXN013<span class="w"> </span>-out<span class="w"> </span>sequenceData/8-final/blastonline.txt<span class="w"> </span>-max_target_seqs<span class="w"> </span><span class="m">10</span><span class="w"> </span>-outfmt<span class="w"> </span><span class="s2">&quot;6 qaccver saccver staxid sscinames length pident mismatch qcovs evalue bitscore qstart qend sstart send gapopen&quot;</span>
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>sequenceData/8-final/BLASTonline.txt
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zotu</span><span class="mf">.1</span>  <span class="n">MH087519</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">500</span>  <span class="mi">632</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KU906005</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">485</span>  <span class="mi">617</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP255185</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">519</span>  <span class="mi">651</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP255165</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">492</span>  <span class="mi">624</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP254955</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">519</span>  <span class="mi">651</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP254688</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">518</span>  <span class="mi">650</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP254656</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">519</span>  <span class="mi">651</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP254605</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">518</span>  <span class="mi">650</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP254404</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">518</span>  <span class="mi">650</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span>  <span class="n">KP254360</span><span class="mf">.1</span>  <span class="mi">315708</span>  <span class="n">Gammarus</span> <span class="n">mucronatus</span>  <span class="mi">133</span>  <span class="mf">87.218</span>  <span class="mi">17</span>  <span class="mi">42</span>  <span class="mf">4.49e-32</span>  <span class="mi">152</span>  <span class="mi">174</span>  <span class="mi">306</span>  <span class="mi">519</span>  <span class="mi">651</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">KY268691</span><span class="mf">.1</span>  <span class="mi">673248</span>  <span class="n">Mitostoma</span> <span class="n">chrysomelas</span>  <span class="mi">282</span>  <span class="mf">76.596</span>  <span class="mi">60</span>  <span class="mi">89</span>  <span class="mf">1.62e-31</span>  <span class="mi">150</span>  <span class="mi">29</span>  <span class="mi">307</span>  <span class="mi">374</span>  <span class="mi">652</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">KY268314</span><span class="mf">.1</span>  <span class="mi">673248</span>  <span class="n">Mitostoma</span> <span class="n">chrysomelas</span>  <span class="mi">282</span>  <span class="mf">76.241</span>  <span class="mi">61</span>  <span class="mi">89</span>  <span class="mf">2.09e-30</span>  <span class="mi">147</span>  <span class="mi">29</span>  <span class="mi">307</span>  <span class="mi">374</span>  <span class="mi">652</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">OK072767</span><span class="mf">.1</span>  <span class="mi">2899957</span>  <span class="n">Paraleptamphopidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">NZGDW0096</span>  <span class="mi">261</span>  <span class="mf">77.011</span>  <span class="mi">56</span>  <span class="mi">82</span>  <span class="mf">2.09e-30</span>  <span class="mi">147</span>  <span class="mi">50</span>  <span class="mi">307</span>  <span class="mi">376</span>  <span class="mi">635</span>  <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">OK072738</span><span class="mf">.1</span>  <span class="mi">2899963</span>  <span class="n">Paraleptamphopidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">NZGDW0110</span>  <span class="mi">261</span>  <span class="mf">77.011</span>  <span class="mi">56</span>  <span class="mi">82</span>  <span class="mf">2.09e-30</span>  <span class="mi">147</span>  <span class="mi">50</span>  <span class="mi">307</span>  <span class="mi">343</span>  <span class="mi">602</span>  <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">JF444969</span><span class="mf">.1</span>  <span class="mi">38668</span>  <span class="n">Onychomys</span> <span class="n">leucogaster</span>  <span class="mi">273</span>  <span class="mf">76.923</span>  <span class="mi">51</span>  <span class="mi">85</span>  <span class="mf">7.51e-30</span>  <span class="mi">145</span>  <span class="mi">29</span>  <span class="mi">295</span>  <span class="mi">373</span>  <span class="mi">639</span>  <span class="mi">10</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">OK072800</span><span class="mf">.1</span>  <span class="mi">2899962</span>  <span class="n">Paraleptamphopidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">NZGDW0107</span>  <span class="mi">261</span>  <span class="mf">76.628</span>  <span class="mi">57</span>  <span class="mi">82</span>  <span class="mf">2.70e-29</span>  <span class="mi">143</span>  <span class="mi">50</span>  <span class="mi">307</span>  <span class="mi">343</span>  <span class="mi">602</span>  <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">AB500136</span><span class="mf">.1</span>  <span class="mi">322050</span>  <span class="n">Neoseiulus</span> <span class="n">womersleyi</span>  <span class="mi">167</span>  <span class="mf">82.036</span>  <span class="mi">29</span>  <span class="mi">53</span>  <span class="mf">9.72e-29</span>  <span class="mi">141</span>  <span class="mi">147</span>  <span class="mi">312</span>  <span class="mi">187</span>  <span class="mi">353</span>  <span class="mi">1</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">KY269411</span><span class="mf">.1</span>  <span class="mi">673248</span>  <span class="n">Mitostoma</span> <span class="n">chrysomelas</span>  <span class="mi">282</span>  <span class="mf">75.887</span>  <span class="mi">62</span>  <span class="mi">89</span>  <span class="mf">3.50e-28</span>  <span class="mi">139</span>  <span class="mi">29</span>  <span class="mi">307</span>  <span class="mi">374</span>  <span class="mi">652</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">KY269355</span><span class="mf">.1</span>  <span class="mi">673248</span>  <span class="n">Mitostoma</span> <span class="n">chrysomelas</span>  <span class="mi">282</span>  <span class="mf">75.887</span>  <span class="mi">62</span>  <span class="mi">89</span>  <span class="mf">3.50e-28</span>  <span class="mi">139</span>  <span class="mi">29</span>  <span class="mi">307</span>  <span class="mi">374</span>  <span class="mi">652</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span>  <span class="n">KY268900</span><span class="mf">.1</span>  <span class="mi">673248</span>  <span class="n">Mitostoma</span> <span class="n">chrysomelas</span>  <span class="mi">282</span>  <span class="mf">75.887</span>  <span class="mi">62</span>  <span class="mi">89</span>  <span class="mf">3.50e-28</span>  <span class="mi">139</span>  <span class="mi">29</span>  <span class="mi">307</span>  <span class="mi">374</span>  <span class="mi">652</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">HM465067</span><span class="mf">.1</span>  <span class="mi">874932</span>  <span class="n">Caridea</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LPdivOTU347</span>  <span class="mi">313</span>  <span class="mf">86.581</span>  <span class="mi">40</span>  <span class="mi">99</span>  <span class="mf">6.90e-90</span>  <span class="mi">344</span>  <span class="mi">2</span>  <span class="mi">313</span>  <span class="mi">294</span>  <span class="mi">605</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">MG093570</span><span class="mf">.1</span>  <span class="mi">2362403</span>  <span class="n">Mycetophila</span> <span class="n">sp</span><span class="o">.</span> <span class="n">BIOUG21226</span><span class="o">-</span><span class="n">D02</span>  <span class="mi">302</span>  <span class="mf">79.470</span>  <span class="mi">58</span>  <span class="mi">95</span>  <span class="mf">7.30e-50</span>  <span class="mi">211</span>  <span class="mi">13</span>  <span class="mi">312</span>  <span class="mi">357</span>  <span class="mi">656</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">KR651281</span><span class="mf">.1</span>  <span class="mi">1881779</span>  <span class="n">Mycetophila</span> <span class="n">sp</span><span class="o">.</span> <span class="n">BOLD</span><span class="o">-</span><span class="mi">2016</span>  <span class="mi">302</span>  <span class="mf">79.139</span>  <span class="mi">59</span>  <span class="mi">95</span>  <span class="mf">3.40e-48</span>  <span class="mi">206</span>  <span class="mi">13</span>  <span class="mi">312</span>  <span class="mi">358</span>  <span class="mi">657</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">KM960518</span><span class="mf">.1</span>  <span class="mi">1750023</span>  <span class="n">Mycetophilidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">BOLD</span><span class="p">:</span><span class="n">ACG6335</span>  <span class="mi">302</span>  <span class="mf">78.808</span>  <span class="mi">60</span>  <span class="mi">95</span>  <span class="mf">1.58e-46</span>  <span class="mi">200</span>  <span class="mi">13</span>  <span class="mi">312</span>  <span class="mi">358</span>  <span class="mi">657</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">MN868958</span><span class="mf">.1</span>  <span class="mi">2690821</span>  <span class="n">Rhamphomyia</span> <span class="n">dudai</span>  <span class="mi">300</span>  <span class="mf">78.333</span>  <span class="mi">63</span>  <span class="mi">95</span>  <span class="mf">2.65e-44</span>  <span class="mi">193</span>  <span class="mi">14</span>  <span class="mi">312</span>  <span class="mi">359</span>  <span class="mi">657</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">MT826940</span><span class="mf">.1</span>  <span class="mi">149068</span>  <span class="n">Nematoscelis</span> <span class="n">megalops</span>  <span class="mi">299</span>  <span class="mf">78.261</span>  <span class="mi">65</span>  <span class="mi">95</span>  <span class="mf">2.65e-44</span>  <span class="mi">193</span>  <span class="mi">14</span>  <span class="mi">312</span>  <span class="mi">373</span>  <span class="mi">671</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">KM978363</span><span class="mf">.1</span>  <span class="mi">1302337</span>  <span class="n">Orchesellides</span> <span class="n">sinensis</span>  <span class="mi">300</span>  <span class="mf">78.000</span>  <span class="mi">66</span>  <span class="mi">95</span>  <span class="mf">3.42e-43</span>  <span class="mi">189</span>  <span class="mi">14</span>  <span class="mi">313</span>  <span class="mi">359</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">KR039462</span><span class="mf">.1</span>  <span class="mi">30126</span>  <span class="n">Graminella</span> <span class="n">mohri</span>  <span class="mi">304</span>  <span class="mf">77.961</span>  <span class="mi">63</span>  <span class="mi">96</span>  <span class="mf">1.23e-42</span>  <span class="mi">187</span>  <span class="mi">11</span>  <span class="mi">312</span>  <span class="mi">362</span>  <span class="mi">663</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">KM023138</span><span class="mf">.1</span>  <span class="mi">82746</span>  <span class="n">Orius</span> <span class="n">niger</span>  <span class="mi">305</span>  <span class="mf">78.033</span>  <span class="mi">61</span>  <span class="mi">96</span>  <span class="mf">1.23e-42</span>  <span class="mi">187</span>  <span class="mi">11</span>  <span class="mi">312</span>  <span class="mi">356</span>  <span class="mi">657</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.3</span>  <span class="n">KM022126</span><span class="mf">.1</span>  <span class="mi">82746</span>  <span class="n">Orius</span> <span class="n">niger</span>  <span class="mi">305</span>  <span class="mf">78.033</span>  <span class="mi">61</span>  <span class="mi">96</span>  <span class="mf">1.23e-42</span>  <span class="mi">187</span>  <span class="mi">11</span>  <span class="mi">312</span>  <span class="mi">101</span>  <span class="mi">402</span>  <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">HM466580</span><span class="mf">.1</span>  <span class="mi">875001</span>  <span class="n">Caridea</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LPdivOTU96</span>  <span class="mi">313</span>  <span class="mf">100.000</span>  <span class="mi">0</span>  <span class="mi">100</span>  <span class="mf">1.74e-160</span>  <span class="mi">579</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ560339</span><span class="mf">.1</span>  <span class="mi">2858060</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466310</span>  <span class="mi">313</span>  <span class="mf">98.403</span>  <span class="mi">5</span>  <span class="mi">100</span>  <span class="mf">3.79e-152</span>  <span class="mi">551</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ560299</span><span class="mf">.1</span>  <span class="mi">2858059</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466309</span>  <span class="mi">313</span>  <span class="mf">98.403</span>  <span class="mi">5</span>  <span class="mi">100</span>  <span class="mf">3.79e-152</span>  <span class="mi">551</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">GQ260974</span><span class="mf">.1</span>  <span class="mi">653025</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LP</span><span class="o">-</span><span class="mi">2009</span><span class="p">;</span><span class="n">Caridea</span> <span class="n">sp</span><span class="o">.</span> <span class="n">LPdivOTU96</span>  <span class="mi">313</span>  <span class="mf">98.403</span>  <span class="mi">5</span>  <span class="mi">100</span>  <span class="mf">3.79e-152</span>  <span class="mi">551</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ560028</span><span class="mf">.1</span>  <span class="mi">2858073</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466326</span>  <span class="mi">313</span>  <span class="mf">98.083</span>  <span class="mi">6</span>  <span class="mi">100</span>  <span class="mf">1.76e-150</span>  <span class="mi">545</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ559959</span><span class="mf">.1</span>  <span class="mi">2858071</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466317</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466318</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466319</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466320</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466321</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466322</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466323</span><span class="p">;</span><span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466324</span>  <span class="mi">313</span>  <span class="mf">98.083</span>  <span class="mi">6</span>  <span class="mi">100</span>  <span class="mf">1.76e-150</span>  <span class="mi">545</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ559957</span><span class="mf">.1</span>  <span class="mi">2858061</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466313</span>  <span class="mi">313</span>  <span class="mf">98.083</span>  <span class="mi">6</span>  <span class="mi">100</span>  <span class="mf">1.76e-150</span>  <span class="mi">545</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ560029</span><span class="mf">.1</span>  <span class="mi">2858074</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466328</span>  <span class="mi">311</span>  <span class="mf">98.071</span>  <span class="mi">6</span>  <span class="mi">99</span>  <span class="mf">2.28e-149</span>  <span class="mi">542</span>  <span class="mi">3</span>  <span class="mi">313</span>  <span class="mi">348</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ560087</span><span class="mf">.1</span>  <span class="mi">2858063</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466315</span>  <span class="mi">313</span>  <span class="mf">97.764</span>  <span class="mi">7</span>  <span class="mi">100</span>  <span class="mf">8.20e-149</span>  <span class="mi">540</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.4</span>  <span class="n">MZ559958</span><span class="mf">.1</span>  <span class="mi">2858064</span>  <span class="n">Hippolytidae</span> <span class="n">sp</span><span class="o">.</span> <span class="n">USNM</span> <span class="mi">1466316</span>  <span class="mi">313</span>  <span class="mf">97.764</span>  <span class="mi">7</span>  <span class="mi">100</span>  <span class="mf">8.20e-149</span>  <span class="mi">540</span>  <span class="mi">1</span>  <span class="mi">313</span>  <span class="mi">346</span>  <span class="mi">658</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">AB542683</span><span class="mf">.1</span>  <span class="mi">506673</span>  <span class="n">Metaphire</span> <span class="n">tosaensis</span>  <span class="mi">302</span>  <span class="mf">82.119</span>  <span class="mi">52</span>  <span class="mi">96</span>  <span class="mf">9.25e-64</span>  <span class="mi">257</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">332</span>  <span class="mi">632</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">DQ224174</span><span class="mf">.1</span>  <span class="mi">351234</span>  <span class="n">Amynthas</span> <span class="n">lini</span>  <span class="mi">302</span>  <span class="mf">81.788</span>  <span class="mi">53</span>  <span class="mi">96</span>  <span class="mf">4.30e-62</span>  <span class="mi">252</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">351</span>  <span class="mi">651</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">KM187649</span><span class="mf">.1</span>  <span class="mi">1544308</span>  <span class="n">Metabonellia</span> <span class="n">haswelli</span>  <span class="mi">302</span>  <span class="mf">81.457</span>  <span class="mi">54</span>  <span class="mi">96</span>  <span class="mf">2.00e-60</span>  <span class="mi">246</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">351</span>  <span class="mi">651</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">HM219172</span><span class="mf">.2</span>  <span class="mi">857735</span>  <span class="n">Perionyx</span> <span class="n">sp</span><span class="o">.</span> <span class="mi">1</span> <span class="n">WL</span><span class="o">-</span><span class="mi">2010</span>  <span class="mi">301</span>  <span class="mf">81.395</span>  <span class="mi">56</span>  <span class="mi">96</span>  <span class="mf">2.00e-60</span>  <span class="mi">246</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">350</span>  <span class="mi">650</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">MT472583</span><span class="mf">.1</span>  <span class="mi">2736579</span>  <span class="n">Perionyx</span> <span class="n">sp</span><span class="o">.</span> <span class="n">E1R2</span>  <span class="mi">301</span>  <span class="mf">81.063</span>  <span class="mi">57</span>  <span class="mi">96</span>  <span class="mf">9.32e-59</span>  <span class="mi">241</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">352</span>  <span class="mi">652</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">KT716821</span><span class="mf">.1</span>  <span class="mi">1737888</span>  <span class="n">Perionyx</span> <span class="n">sp</span><span class="o">.</span> <span class="n">DHS</span><span class="o">-</span><span class="mi">2015</span>  <span class="mi">301</span>  <span class="mf">81.063</span>  <span class="mi">57</span>  <span class="mi">96</span>  <span class="mf">9.32e-59</span>  <span class="mi">241</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">362</span>  <span class="mi">662</span>  <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">LC703190</span><span class="mf">.1</span>  <span class="mi">585715</span>  <span class="n">Metaphire</span> <span class="n">megascolidioides</span>  <span class="mi">302</span>  <span class="mf">81.126</span>  <span class="mi">55</span>  <span class="mi">96</span>  <span class="mf">9.32e-59</span>  <span class="mi">241</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">263</span>  <span class="mi">563</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">JF509718</span><span class="mf">.1</span>  <span class="mi">6431</span>  <span class="n">Urechis</span> <span class="n">caupo</span>  <span class="mi">302</span>  <span class="mf">81.126</span>  <span class="mi">55</span>  <span class="mi">96</span>  <span class="mf">9.32e-59</span>  <span class="mi">241</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">350</span>  <span class="mi">650</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">AB542648</span><span class="mf">.1</span>  <span class="mi">585715</span>  <span class="n">Metaphire</span> <span class="n">megascolidioides</span>  <span class="mi">302</span>  <span class="mf">81.126</span>  <span class="mi">55</span>  <span class="mi">96</span>  <span class="mf">9.32e-59</span>  <span class="mi">241</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">332</span>  <span class="mi">632</span>  <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.5</span>  <span class="n">AB536863</span><span class="mf">.1</span>  <span class="mi">585715</span>  <span class="n">Metaphire</span> <span class="n">megascolidioides</span>  <span class="mi">302</span>  <span class="mf">81.126</span>  <span class="mi">55</span>  <span class="mi">96</span>  <span class="mf">9.32e-59</span>  <span class="mi">241</span>  <span class="mi">6</span>  <span class="mi">306</span>  <span class="mi">332</span>  <span class="mi">632</span>  <span class="mi">2</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="blast-local">
<h4>4.3.3 BLAST+ local<a class="headerlink" href="#blast-local" title="Permalink to this heading">#</a></h4>
<p>While running BLAST+ remotely is very handy, it is frowned upon to run remote BLAST searches for thousands of sequences and should be restricted to a couple of sequences at a time. The reason for this is that we don’t want to abuse our priviledge to take up too much of the traffic on the NCBI servers. One way to circumvent the issue is to set up a script that iterates over our 1,345 ZOTU sequences to run remote BLAST searches separately. Similarly, posting a file with thousands of sequences for an online BLAST will encounter time-out errors.</p>
<p>Hence, when dealing with large numbers of sequences, it is best to run BLAST+ on a local database. If you are working on a supercomputer, they might have a locally installed and up-to-date version of the full NCBI database available to you. This option, however, is not available to everyone. Besides BLASTing against the full NCBI database (which is a couple of TB in size), we could also use the database we created with CRABS to run BLAST locally. Since running BLAST on a curated reference database you have built yourself is more accessible to everyone than a local version of the full NCBI database on a supercomputer, we will use the curated database option for this tutorial.</p>
<p>The first step before running BLAST, is to reformat the initial database, similarly to what we had to do before we could run the IDTAXA machine learning algorithm. To generate a local BLAST database, we first need to remove the SINTAX-specific formatting of our database. We can do this using the following python script, which will remove all the header information and only keep the accession numbers as sequence headers. This python script takes in two user arguments, including the input file and an output file name for the BLAST database.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/blastInitFormat.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="kn">import</span> <span class="nn">sys</span>

<span class="n">inputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">outputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="n">newDict</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">inputFile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
      <span class="n">seqID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
      <span class="k">if</span> <span class="s1">&#39;CRABS_&#39;</span> <span class="ow">in</span> <span class="n">seqID</span><span class="p">:</span>
        <span class="n">seqID</span> <span class="o">=</span> <span class="n">seqID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">newDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">]</span> <span class="o">=</span> <span class="n">line</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">newDict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/blastInitFormat.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/blastInitFormat.py<span class="w"> </span>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta<span class="w"> </span>sequenceData/7-refdb/blast.fasta
</pre></div>
</div>
<p>Once we have generated a fasta file with the correct format, we can create a BLAST database using the <code class="docutils literal notranslate"><span class="pre">makeblastdb</span></code> command in the BLAST+ suite.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>makeblastdb<span class="w"> </span>-in<span class="w"> </span>sequenceData/7-refdb/blast.fasta<span class="w"> </span>-dbtype<span class="w"> </span>nucl<span class="w"> </span>-parse_seqids<span class="w"> </span>-out<span class="w"> </span>sequenceData/7-refdb/blastDBCOI
</pre></div>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Building</span> <span class="n">a</span> <span class="n">new</span> <span class="n">DB</span><span class="p">,</span> <span class="n">current</span> <span class="n">time</span><span class="p">:</span> <span class="mi">10</span><span class="o">/</span><span class="mi">10</span><span class="o">/</span><span class="mi">2023</span> <span class="mi">10</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">06</span>
<span class="n">New</span> <span class="n">DB</span> <span class="n">name</span><span class="p">:</span>   <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">gjeunen</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">lectures</span><span class="o">/</span><span class="n">HKU2023</span><span class="o">/</span><span class="n">tutorial</span><span class="o">/</span><span class="n">sequenceData</span><span class="o">/</span><span class="mi">7</span><span class="o">-</span><span class="n">refdb</span><span class="o">/</span><span class="n">blastDBCOI</span>
<span class="n">New</span> <span class="n">DB</span> <span class="n">title</span><span class="p">:</span>  <span class="n">sequenceData</span><span class="o">/</span><span class="mi">7</span><span class="o">-</span><span class="n">refdb</span><span class="o">/</span><span class="n">blast</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">Sequence</span> <span class="nb">type</span><span class="p">:</span> <span class="n">Nucleotide</span>
<span class="n">Deleted</span> <span class="n">existing</span> <span class="n">Nucleotide</span> <span class="n">BLAST</span> <span class="n">database</span> <span class="n">named</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">gjeunen</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">work</span><span class="o">/</span><span class="n">lectures</span><span class="o">/</span><span class="n">HKU2023</span><span class="o">/</span><span class="n">tutorial</span><span class="o">/</span><span class="n">sequenceData</span><span class="o">/</span><span class="mi">7</span><span class="o">-</span><span class="n">refdb</span><span class="o">/</span><span class="n">blastDBCOI</span>
<span class="n">Keep</span> <span class="n">MBits</span><span class="p">:</span> <span class="n">T</span>
<span class="n">Maximum</span> <span class="n">file</span> <span class="n">size</span><span class="p">:</span> <span class="mi">1000000000</span><span class="n">B</span>
<span class="n">Adding</span> <span class="n">sequences</span> <span class="kn">from</span> <span class="nn">FASTA</span><span class="p">;</span> <span class="n">added</span> <span class="mi">208189</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="mf">3.29866</span> <span class="n">seconds</span><span class="o">.</span>
</pre></div>
</div>
</div>
<p>We can now use <code class="docutils literal notranslate"><span class="pre">blastn</span></code> locally to assign a taxonomic ID to each ZOTU sequence using the BLAST algorithm.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>blastn<span class="w"> </span>-query<span class="w"> </span>sequenceData/8-final/zotus.fasta<span class="w"> </span>-db<span class="w"> </span>sequenceData/7-refdb/blastDBCOI<span class="w"> </span>-max_target_seqs<span class="w"> </span><span class="m">10</span><span class="w"> </span>-perc_identity<span class="w"> </span><span class="m">50</span><span class="w"> </span>-qcov_hsp_perc<span class="w"> </span><span class="m">50</span><span class="w"> </span>-outfmt<span class="w"> </span><span class="s2">&quot;6 qaccver saccver staxid sscinames length pident mismatch qcovs evalue bitscore qstart qend sstart send gapopen&quot;</span><span class="w"> </span>-out<span class="w"> </span>sequenceData/8-final/blastTaxonomy.txt
</pre></div>
</div>
<p>By using the <code class="docutils literal notranslate"><span class="pre">head</span></code> command, we can inspect the first couple of lines of the output document.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">20</span><span class="w"> </span>sequenceData/8-final/blastTaxonomy.txt
</pre></div>
</div>
<div class="tip admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">zotu</span><span class="mf">.1</span> <span class="n">KT708260</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">297</span> <span class="mf">75.421</span> <span class="mi">73</span> <span class="mi">95</span> <span class="mf">4.53e-34</span> <span class="mi">145</span> <span class="mi">11</span> <span class="mi">307</span> <span class="mi">356</span> <span class="mi">652</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">KT707216</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">297</span> <span class="mf">75.421</span> <span class="mi">73</span> <span class="mi">95</span> <span class="mf">4.53e-34</span> <span class="mi">145</span> <span class="mi">11</span> <span class="mi">307</span> <span class="mi">356</span> <span class="mi">652</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.1</span> <span class="n">MF744717</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">252</span> <span class="mf">76.190</span> <span class="mi">58</span> <span class="mi">80</span> <span class="mf">3.53e-30</span> <span class="mi">132</span> <span class="mi">57</span> <span class="mi">307</span> <span class="mi">402</span> <span class="mi">652</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">KY268691</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">282</span> <span class="mf">76.596</span> <span class="mi">60</span> <span class="mi">89</span> <span class="mf">9.74e-36</span> <span class="mi">150</span> <span class="mi">29</span> <span class="mi">307</span> <span class="mi">374</span> <span class="mi">652</span> <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">OK072767</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">261</span> <span class="mf">77.011</span> <span class="mi">56</span> <span class="mi">82</span> <span class="mf">1.26e-34</span> <span class="mi">147</span> <span class="mi">50</span> <span class="mi">307</span> <span class="mi">376</span> <span class="mi">635</span> <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">OK072738</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">261</span> <span class="mf">77.011</span> <span class="mi">56</span> <span class="mi">82</span> <span class="mf">1.26e-34</span> <span class="mi">147</span> <span class="mi">50</span> <span class="mi">307</span> <span class="mi">343</span> <span class="mi">602</span> <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">KY268314</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">282</span> <span class="mf">76.241</span> <span class="mi">61</span> <span class="mi">89</span> <span class="mf">1.26e-34</span> <span class="mi">147</span> <span class="mi">29</span> <span class="mi">307</span> <span class="mi">374</span> <span class="mi">652</span> <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">OK072800</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">261</span> <span class="mf">76.628</span> <span class="mi">57</span> <span class="mi">82</span> <span class="mf">1.63e-33</span> <span class="mi">143</span> <span class="mi">50</span> <span class="mi">307</span> <span class="mi">343</span> <span class="mi">602</span> <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">KY269355</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">282</span> <span class="mf">75.887</span> <span class="mi">62</span> <span class="mi">89</span> <span class="mf">2.11e-32</span> <span class="mi">139</span> <span class="mi">29</span> <span class="mi">307</span> <span class="mi">374</span> <span class="mi">652</span> <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">KY268900</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">282</span> <span class="mf">75.887</span> <span class="mi">62</span> <span class="mi">89</span> <span class="mf">2.11e-32</span> <span class="mi">139</span> <span class="mi">29</span> <span class="mi">307</span> <span class="mi">374</span> <span class="mi">652</span> <span class="mi">5</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">OK072832</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">261</span> <span class="mf">76.245</span> <span class="mi">58</span> <span class="mi">82</span> <span class="mf">2.73e-31</span> <span class="mi">135</span> <span class="mi">50</span> <span class="mi">307</span> <span class="mi">355</span> <span class="mi">614</span> <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">OK072819</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">261</span> <span class="mf">76.245</span> <span class="mi">58</span> <span class="mi">82</span> <span class="mf">2.73e-31</span> <span class="mi">135</span> <span class="mi">50</span> <span class="mi">307</span> <span class="mi">348</span> <span class="mi">607</span> <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.2</span> <span class="n">OK072786</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">261</span> <span class="mf">76.245</span> <span class="mi">58</span> <span class="mi">82</span> <span class="mf">2.73e-31</span> <span class="mi">135</span> <span class="mi">50</span> <span class="mi">307</span> <span class="mi">376</span> <span class="mi">635</span> <span class="mi">4</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">BHKG</span><span class="o">-</span><span class="mi">0179</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">313</span> <span class="mf">99.681</span> <span class="mi">1</span> <span class="mi">100</span> <span class="mf">4.87e-163</span> <span class="mi">573</span> <span class="mi">1</span> <span class="mi">313</span> <span class="mi">346</span> <span class="mi">658</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">HM465067</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">313</span> <span class="mf">86.581</span> <span class="mi">40</span> <span class="mi">99</span> <span class="mf">4.16e-94</span> <span class="mi">344</span> <span class="mi">2</span> <span class="mi">313</span> <span class="mi">294</span> <span class="mi">605</span> <span class="mi">2</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">MT826940</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">299</span> <span class="mf">78.261</span> <span class="mi">65</span> <span class="mi">96</span> <span class="mf">1.59e-48</span> <span class="mi">193</span> <span class="mi">14</span> <span class="mi">312</span> <span class="mi">373</span> <span class="mi">671</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">MT826939</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">299</span> <span class="mf">77.926</span> <span class="mi">66</span> <span class="mi">96</span> <span class="mf">7.42e-47</span> <span class="mi">187</span> <span class="mi">14</span> <span class="mi">312</span> <span class="mi">373</span> <span class="mi">671</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">MT826938</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">299</span> <span class="mf">77.926</span> <span class="mi">66</span> <span class="mi">96</span> <span class="mf">7.42e-47</span> <span class="mi">187</span> <span class="mi">14</span> <span class="mi">312</span> <span class="mi">373</span> <span class="mi">671</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">MT826937</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">293</span> <span class="mf">78.157</span> <span class="mi">64</span> <span class="mi">94</span> <span class="mf">7.42e-47</span> <span class="mi">187</span> <span class="mi">14</span> <span class="mi">306</span> <span class="mi">373</span> <span class="mi">665</span> <span class="mi">0</span>
<span class="n">zotu</span><span class="mf">.3</span> <span class="n">MT826936</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="n">N</span><span class="o">/</span><span class="n">A</span> <span class="mi">299</span> <span class="mf">77.926</span> <span class="mi">66</span> <span class="mi">96</span> <span class="mf">7.42e-47</span> <span class="mi">187</span> <span class="mi">14</span> <span class="mi">312</span> <span class="mi">373</span> <span class="mi">671</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
<p>From the <code class="docutils literal notranslate"><span class="pre">head</span></code> output, we can see that the local BLAST output file follows the same structure as the remote BLAST output file. However, note the missing values for the taxonomic ID and species names! We have excluded this information when we generated the database, as errors occur when using custom barcodes for which no accession numbers exist. No worries about this though, as we will generate the taxonomic ID, species name, and taxonomic lineage for each BLAST hit when parsing the output file in section <strong>5.3 BLAST formatting</strong>.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>For people who’s reference database solely exists out of NCBI sequences with valid accession numbers, you can also generate a reference database that includes the taxonomic ID and species name information. When the reference database is built in such a manner the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> output will contain the taxonomic ID and species name values, rather than defaulting to <code class="docutils literal notranslate"><span class="pre">N/A</span></code>. Provided below is the code to build such a database.</p>
<p>If we would like to add the taxonomic ID and scientific names in the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> output file, as in the <code class="docutils literal notranslate"><span class="pre">blastn</span> <span class="pre">-remote</span></code> option, we will need to run some additional lines of code before using the <code class="docutils literal notranslate"><span class="pre">makeblastdb</span></code> command and ensure only valid accession numbers are in the reference database and incorporated in the <code class="docutils literal notranslate"><span class="pre">nucl_gb.accession2taxid</span></code> (which we downloaded when building a reference database with CRABS in section <strong>2. Reference databases</strong>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you would like to have the taxonomic ID and scientific name outputs, but your reference database includes invalid accession number, i.e., non-NCBI sequences, you can add your sequence ID’s to the <code class="docutils literal notranslate"><span class="pre">nucl_gb.accession2taxid</span></code> file, which is a simple tab-delimited text file that matches accession numbers to NCBI taxIDs. This will need to be done manually, though.</p>
</div>
<p><strong>First</strong>, we need to generate an accession number to taxonomic ID mapping file. This can be done by removing the second and fourth column from the <code class="docutils literal notranslate"><span class="pre">nucl_gb.accession2taxid</span></code> file we downloaded when building a reference database with CRABS in section <strong>2. Reference databases</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tail<span class="w"> </span>-n<span class="w"> </span>+2<span class="w"> </span>sequenceData/7-refdb/nucl_gb.accession2taxid<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{print $1, $3}&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>sequenceData/7-refdb/taxIDmap.txt
</pre></div>
</div>
<p><strong>Second</strong>, if you haven’t downloaded the taxonomic NCBI database during section <strong>4.3.2 BLAST+ remote</strong>, you need to run the following two commands (this can be skipped if you already downloaded these files).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz
tar<span class="w"> </span>-xvf<span class="w"> </span>taxdb.tar.gz
</pre></div>
</div>
<p><strong>Third</strong>, we can run an altered version of the <code class="docutils literal notranslate"><span class="pre">makeblastdb</span></code> command, where we specify the <strong>sequenceData/7-refdb/taxIDmap.txt</strong> file using the <code class="docutils literal notranslate"><span class="pre">-taxid_map</span></code> parameter.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>makeblastdb<span class="w"> </span>-in<span class="w"> </span>sequenceData/7-refdb/blast.fasta<span class="w"> </span>-dbtype<span class="w"> </span>nucl<span class="w"> </span>-parse_seqids<span class="w"> </span>-out<span class="w"> </span>sequenceData/7-refdb/blastDBCOI<span class="w"> </span>-taxid_map<span class="w"> </span>sequenceData/7-refdb/taxIDmap.txt
</pre></div>
</div>
<p>Once this command is executed, you can run the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> command with an output that specifies the taxonomic ID and species name values for each BLAST hit.</p>
</div>
</section>
</section>
</section>
<section id="output-formatting">
<h2>5. Output formatting<a class="headerlink" href="#output-formatting" title="Permalink to this heading">#</a></h2>
<p>At this stage, we have explored three taxonomy assignment algorithms and created multiple files containing the taxonomic ID for each ZOTU sequence, including <strong>sequenceData/8-final/sintaxTaxonomy.txt</strong> for the SINTAX classifier, <strong>sequenceData/8-final/idtaxaTaxonomy.txt</strong> for the IDTAXA classifier, and <strong>sequenceData/8-final/blastTaxonomy.txt</strong> for the BLAST classifier. Each of these files has a unique, but similar structure. For the data exploration and statistical analyses in R, it would be easiest if all have the same structure, so they can be read in using the same code. Therefore, let’s reformat all files to a unique format, so they can be easily compared and read into R. We will be using some simple python parsing scripts to accomplish this goal.</p>
<section id="sintax-formatting">
<h3>5.1 SINTAX formatting<a class="headerlink" href="#sintax-formatting" title="Permalink to this heading">#</a></h3>
<p>The SINTAX output is a tab-delimited text file, where the first column contains the ZOTU identifiers. These identifiers are not ordered based on number, though all ZOTU numbers are present in the document, even when no taxonomic ID could be assigned. The second column contains the taxonomic ID information, including lineage and confidence values. The third column denotes the DNA strand the assignment was made. The fourth column and last column contains the taxonomic lineage up to the point where the confidence value is larger than the cut off value, 70% in our case. We can use the python script below to parse the SINTAX document. This python script takes five user arguments, including the SINTAX input file, an output file name for the new lineage document, an output file name for the new score document, the number of ZOTUs, and the similarity threshold value.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/sintaxOutputFormat.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">collections</span>

<span class="n">sintaxInputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sintaxLineageOutputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">sintaxScoreOutputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">zotuNumber</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
<span class="n">cutOffValue</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">5</span><span class="p">])</span>

<span class="n">headerRanks</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;kingdom&#39;</span><span class="p">,</span> <span class="s1">&#39;phylum&#39;</span><span class="p">,</span> <span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="s1">&#39;order&#39;</span><span class="p">,</span> <span class="s1">&#39;family&#39;</span><span class="p">,</span> <span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="s1">&#39;species&#39;</span><span class="p">]</span>
<span class="n">idLineageDict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
<span class="n">idScoreDict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">sintaxInputFile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="p">:</span>
    <span class="n">seqID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">lineage</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">lineageItems</span> <span class="o">=</span> <span class="n">lineage</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">lineageItems</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">headerRanks</span><span class="p">:</span>
        <span class="n">idLineageDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
        <span class="n">idScoreDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lineageItems</span><span class="p">)):</span>
        <span class="n">rankID</span> <span class="o">=</span> <span class="n">lineageItems</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">scoreID</span> <span class="o">=</span> <span class="n">lineageItems</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">float</span><span class="p">(</span><span class="n">scoreID</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">cutOffValue</span><span class="p">:</span>
          <span class="n">idLineageDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="n">rankID</span>
          <span class="n">idScoreDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="n">scoreID</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="n">idLineageDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
          <span class="n">idScoreDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="n">scoreID</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">sintaxLineageOutputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
  <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">headerRanks</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">zotuNumber</span><span class="p">):</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">idLineageDict</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">v</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">sintaxScoreOutputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
  <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">headerRanks</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">zotuNumber</span><span class="p">):</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">idScoreDict</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">v</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/sintaxOutputFormat.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/sintaxOutputFormat.py<span class="w"> </span>sequenceData/8-final/sintaxTaxonomy.txt<span class="w"> </span>sequenceData/8-final/sintaxLineage.txt<span class="w"> </span>sequenceData/8-final/sintaxScore.txt<span class="w"> </span><span class="m">1345</span><span class="w"> </span><span class="m">0</span>.7
</pre></div>
</div>
<p>Opening both documents shows that we have created two tab-delimited text files, with <strong>sequenceData/8-final/sintaxLineage.txt</strong> containing the lineage information of the taxonomic ID and <strong>sequenceData/8-final/sintaxScore.txt</strong> containing the confidence score of the taxonomic ID for each level.</p>
<figure class="align-default" id="sintax-results-after-parsing-the-output-document">
<img alt="_images/sintaxparsing2.png" src="_images/sintaxparsing2.png" />
<figcaption>
<p><span class="caption-number">Fig. 29 </span><span class="caption-text">: A screenshot of the SINTAX results after parsing the output document.</span><a class="headerlink" href="#sintax-results-after-parsing-the-output-document" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that we set a manual confidence threshold of 70% and have set the taxonomic ID to <code class="docutils literal notranslate"><span class="pre">NA</span></code> if this value wasn’t reached. Changing the value in the python script allows you to change the similarity threshold after SINTAX classification to see how different thresholds impact the final taxonomic ID.</p>
</div>
</section>
<section id="idtaxa-formatting">
<h3>5.2 IDTAXA formatting<a class="headerlink" href="#idtaxa-formatting" title="Permalink to this heading">#</a></h3>
<p>The IDTAXA output is very similarly structured as the SINTAX output, except that the first column is ordered on ZOTU identifiers, the second column is slightly differently formatted but contains the same information, and the third and fourth columns are missing. We can use the python script below to format the IDTAXA file, which takes three user arguments, including the input file, an output file name for the new lineage document, and an output file name for the new score document. As the ZOTU sequences are ordered and ZOTU sequences are included even when no taxonomic ID could be achieved, we do not need to provide the total number of sequences in this python script.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This python document also reverts the “hack” used with homonym names.</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/idtaxaOutputFormat.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">collections</span>

<span class="k">def</span> <span class="nf">removeBeforeSecondUppercase</span><span class="p">(</span><span class="n">input_string</span><span class="p">):</span>
    <span class="n">uppercase_count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">result</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">input_string</span><span class="p">:</span>
        <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">char</span><span class="o">.</span><span class="n">isupper</span><span class="p">():</span>
            <span class="n">uppercase_count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">uppercase_count</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">input_string</span><span class="p">[</span><span class="n">count</span> <span class="o">-</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">break</span>
    <span class="k">return</span> <span class="n">result</span>

<span class="n">idTaxaInputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">idTaxaLineageOutputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">idTaxaScoreOutputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

<span class="n">headerRanks</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;root&#39;</span><span class="p">,</span> <span class="s1">&#39;kingdom&#39;</span><span class="p">,</span> <span class="s1">&#39;phylum&#39;</span><span class="p">,</span> <span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="s1">&#39;order&#39;</span><span class="p">,</span> <span class="s1">&#39;family&#39;</span><span class="p">,</span> <span class="s1">&#39;genus&#39;</span><span class="p">,</span> <span class="s1">&#39;species&#39;</span><span class="p">]</span>
<span class="n">idLineageDict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
<span class="n">idScoreDict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">idTaxaInputFile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputFile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inputFile</span><span class="p">:</span>
    <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">seqID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">lineage</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">lineageItems</span> <span class="o">=</span> <span class="n">lineage</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;: &#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">headerRanks</span><span class="p">)):</span>
      <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;unclassified&#39;</span> <span class="ow">in</span> <span class="n">lineageItems</span><span class="p">[</span><span class="n">item</span><span class="p">]:</span>
          <span class="n">idLineageDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
          <span class="n">idScoreDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
        <span class="k">else</span><span class="p">:</span>
          <span class="n">lineageString</span> <span class="o">=</span> <span class="n">lineageItems</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
          <span class="k">if</span> <span class="n">lineageString</span> <span class="o">!=</span> <span class="s1">&#39;Root&#39;</span><span class="p">:</span>
            <span class="n">lineageString</span> <span class="o">=</span> <span class="n">removeBeforeSecondUppercase</span><span class="p">(</span><span class="n">lineageString</span><span class="p">)</span>
          <span class="n">scoreString</span> <span class="o">=</span> <span class="n">lineageItems</span><span class="p">[</span><span class="n">item</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)</span>
          <span class="n">idScoreDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="n">scoreString</span>
          <span class="n">idLineageDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="n">lineageString</span>
      <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
        <span class="n">idLineageDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
        <span class="n">idScoreDict</span><span class="p">[</span><span class="n">seqID</span><span class="p">][</span><span class="n">headerRanks</span><span class="p">[</span><span class="n">item</span><span class="p">]]</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">idTaxaLineageOutputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
  <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">headerRanks</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">idLineageDict</span><span class="p">:</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">idLineageDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">g</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">idTaxaScoreOutputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
  <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">headerRanks</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">idScoreDict</span><span class="p">:</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">idScoreDict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">g</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/idtaxaOutputFormat.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/idtaxaOutputFormat.py<span class="w"> </span>sequenceData/8-final/idtaxaTaxonomy.txt<span class="w"> </span>sequenceData/8-final/idtaxaLineage.txt<span class="w"> </span>sequenceData/8-final/idtaxaScore.txt
</pre></div>
</div>
<p>Similarly to the SINTAX parsing, opening both documents shows that we have created two tab-delimited text files, with <strong>sequenceData/8-final/idtaxaLineage.txt</strong> containing the lineage information of the taxonomic ID and <strong>sequenceData/8-final/idtaxaScore.txt</strong> containing the confidence score of the taxonomic ID for each level.</p>
<figure class="align-default" id="idtaxa-results-after-parsing-the-output-document">
<img alt="_images/idtaxaparsing2.png" src="_images/idtaxaparsing2.png" />
<figcaption>
<p><span class="caption-number">Fig. 30 </span><span class="caption-text">: A screenshot of the IDTAXA results after parsing the output document.</span><a class="headerlink" href="#idtaxa-results-after-parsing-the-output-document" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="blast-formatting">
<h3>5.3 BLAST formatting<a class="headerlink" href="#blast-formatting" title="Permalink to this heading">#</a></h3>
<p>The BLAST output is a tab-delimited text file, whereby the first column is an ordered ZOTU identifier list. However, ZOTUs for which no match was found are missing. The second column provides the accession number for the best match. The fourth column should contain the species name, but is defaulted to <code class="docutils literal notranslate"><span class="pre">N/A</span></code> as we used in house barcodes (see section <strong>4.3.3 BLAST+ local</strong>) and the following columns report the statistics for the BLAST match. We can use the python script below to parse the document. It takes four user arguments, including the BLAST input file, the file name for the output document, the number of ZOTUs, and the original sintax-formatted reference database (<strong>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta</strong>) to retrieve the species name and taxonomic lineage for each accession number.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The following python code only takes into account the top BLAST hit to provide an example of how to parse BLAST documents. For your own research project, I do not recommend relying solely on top BLAST hits, but rather determine the MRCA (Most Recent Common Ancestor) from the BLAST output first. Generating an MRCA is outside the scope of this tutorial. For more information on how to generate MRCA taxonomic IDs from BLAST output files, please have a look at this <a class="reference external" href="https://github.com/gjeunen/LCA_BLAST_calculator">GitHub repo</a> on MRCA calculation I have written a while ago. It contains a full working software program to calculate MRCAs from BLAST output files.</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nano<span class="w"> </span>sequenceData/1-scripts/blastOutputFormat.py
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="ch">#! /usr/bin/env python3</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">subprocess</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">import</span> <span class="nn">collections</span>

<span class="n">blastInputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">BLASTOutputFile</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">numberOfZotus</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
<span class="n">refDB</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
<span class="n">zotuIDblastDict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
<span class="n">zotuDict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">accessionDict</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">refDB</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">refFile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">refFile</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
      <span class="n">accessionNumber</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
      <span class="k">if</span> <span class="n">accessionNumber</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;CRABS_&#39;</span><span class="p">):</span>
        <span class="n">accessionNumber</span> <span class="o">=</span> <span class="n">accessionNumber</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
      <span class="n">tax</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
      <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tax</span><span class="p">)):</span>
        <span class="n">tax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">tax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
      <span class="n">taxLineage</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tax</span><span class="p">)</span>
      <span class="n">accessionDict</span><span class="p">[</span><span class="n">accessionNumber</span><span class="p">]</span> <span class="o">=</span> <span class="n">taxLineage</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">blastInputFile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">:</span>
  <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
    <span class="n">zotuID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">zotuID</span> <span class="ow">in</span> <span class="n">zotuDict</span><span class="p">:</span>
      <span class="k">continue</span>
    <span class="n">zotuDict</span><span class="p">[</span><span class="n">zotuID</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">pident</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">5</span><span class="p">]</span>
    <span class="n">qcov</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">7</span><span class="p">]</span>
    <span class="n">acc</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">zotuIDblastDict</span><span class="p">[</span><span class="n">zotuID</span><span class="p">][</span><span class="s1">&#39;pident&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pident</span>
    <span class="n">zotuIDblastDict</span><span class="p">[</span><span class="n">zotuID</span><span class="p">][</span><span class="s1">&#39;qcov&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">qcov</span>
    <span class="n">zotuIDblastDict</span><span class="p">[</span><span class="n">zotuID</span><span class="p">][</span><span class="s1">&#39;accession&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">acc</span>

    
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">BLASTOutputFile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
  <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;kingdom&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;phylum&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;class&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;order&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;family&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;genus&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;species&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;pident&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;qcov&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">numberOfZotus</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">zotuIDblastDict</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
      <span class="n">accessionN</span> <span class="o">=</span> <span class="n">zotuIDblastDict</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">][</span><span class="s1">&#39;accession&#39;</span><span class="p">]</span>
      <span class="n">pidentN</span> <span class="o">=</span> <span class="n">zotuIDblastDict</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">][</span><span class="s1">&#39;pident&#39;</span><span class="p">]</span>
      <span class="n">qcovN</span> <span class="o">=</span> <span class="n">zotuIDblastDict</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;zotu.</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">][</span><span class="s1">&#39;qcov&#39;</span><span class="p">]</span>
      <span class="n">_</span> <span class="o">=</span> <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\t</span><span class="si">{</span><span class="n">accessionDict</span><span class="p">[</span><span class="n">accessionN</span><span class="p">]</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">pidentN</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">qcovN</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Press <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">x</span></code> to exit out of the editor, followed by <code class="docutils literal notranslate"><span class="pre">y</span></code> and <code class="docutils literal notranslate"><span class="pre">return</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chmod<span class="w"> </span>+x<span class="w"> </span>sequenceData/1-scripts/blastOutputFormat.py
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./sequenceData/1-scripts/blastOutputFormat.py<span class="w"> </span>sequenceData/8-final/blastTaxonomy.txt<span class="w"> </span>sequenceData/8-final/blastLineage.txt<span class="w"> </span><span class="m">1345</span><span class="w"> </span>sequenceData/7-refdb/newDBcrabsCOIsintax.fasta
</pre></div>
</div>
<p>Opening the parsed BLAST document shows that we have created a tab-delimited text file, containing the ZOTU sequence ID in the first column, species ID in the second column, percent identity value in the third column, and query coverage value in the fourth column. Note the difference in sequences that have a BLAST hit compared to the SINTAX and IDTAXA classifiers, even though we used the same reference database! This is not due to BLAST being a better classifier, rather than our low thresholds for allowing a BLAST hit to be accepted. Increasing the percent identity and query coverage value to SINTAX or IDTAXA standard will result in a similar drop out of sequences without a taxonomic ID.</p>
<figure class="align-default" id="blast-results-after-parsing-the-output-document">
<img alt="_images/blastparsing2.png" src="_images/blastparsing2.png" />
<figcaption>
<p><span class="caption-number">Fig. 31 </span><span class="caption-text">: A screenshot of the BLAST results after parsing the output document.</span><a class="headerlink" href="#blast-results-after-parsing-the-output-document" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p><strong>That’s it for today, see you tomorrow when we will explore the data and conduct statistical analyses to determine differences between samples in our tutorial data set!</strong></p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="bioinformaticanalysis.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Bioinformatic analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="datapreprocessing.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Data pre-processing</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-databases">2. Reference databases</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crabs-workflow">2.1 CRABS workflow</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#help-documentation">2.2 Help documentation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-download-sequence-data">2.3 Step 1: download sequence data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-merge-downloaded-sequences">2.4 Step 2: merge downloaded sequences</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-extract-amplicons-through-in-silico-pcr">2.5 Step 3: extract amplicons through <em>in silico</em> PCR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-extract-amplicons-through-pairwise-global-alignments">2.6 Step 4: extract amplicons through pairwise global alignments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-assigning-a-taxonomic-id">2.7 Step 5: assigning a taxonomic ID</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-dereplicating-the-db">2.8 Step 6: dereplicating the DB</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-ref-db-cleanup">2.8 Step 7: ref DB cleanup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-8-export-ref-db">2.9 Step 8: export ref DB</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-coi-reference-db">3. The COI reference DB</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#taxonomic-classification">4. Taxonomic classification</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sintax">4.1 SINTAX</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#idtaxa">4.2 IDTAXA</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#input-formatting">4.2.1 Input formatting</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#classifier-training">4.2.2 Classifier training</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">4.2.3 Classifier training</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#blast">4.3 BLAST</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#online-blast">4.3.1 Online BLAST</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-remote">4.3.2 BLAST+ remote</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-local">4.3.3 BLAST+ local</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output-formatting">5. Output formatting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sintax-formatting">5.1 SINTAX formatting</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#idtaxa-formatting">5.2 IDTAXA formatting</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-formatting">5.3 BLAST formatting</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Dr Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>